// ontimize-web-ngx/esm2015/lib/components/user-info/user-info-configuration-item/o-user-info-configuration-item.directive.js
import * as tslib_1 from "tslib";
import { Directive, Injector } from "@angular/core";
import { Router } from "@angular/router";
import { InputConverter } from "../../../decorators";
import { DialogService } from "../../../services";
export const DEFAULT_INPUTS_O_USER_INFO_MENU_ITEM = [
    'name',
    'route',
    'icon',
    'action',
    'confirm',
    'confirmText: confirm-text',
    'tooltip',
    'class'
];
export class OUserInfoConfigurationItemDirective {
    constructor(injector) {
        this.injector = injector;
        this.toolTip = false;
        this.confirm = false;
        this.router = this.injector.get(Router);
        this.dialogService = this.injector.get(DialogService);
    }
    triggerOnClick(e) {
        if (this.route) {
            this.navigate();
        }
        else {
            this.executeItemAction();
        }
    }
    navigate() {
        if (this.router.url !== this.route) {
            this.router.navigate([this.route]);
        }
    }
    executeItemAction() {
        if (this.confirm) {
            this.dialogService.confirm('CONFIRM', this.confirmText || 'MESSAGES.CONFIRM_ACTION').then(result => result ? this.action() : null);
        }
        else {
            this.action();
        }
    }
}
OUserInfoConfigurationItemDirective.decorators = [
    { type: Directive, args: [{
                selector: 'o-user-info-configuration-item',
                inputs: DEFAULT_INPUTS_O_USER_INFO_MENU_ITEM
            },] }
];
OUserInfoConfigurationItemDirective.ctorParameters = () => [
    { type: Injector }
];
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OUserInfoConfigurationItemDirective.prototype, "toolTip", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OUserInfoConfigurationItemDirective.prototype, "confirm", void 0);
