// ontimize-web-ngx/esm2015/lib/directives/o-file-drag-and-drop.directive.js
import { Directive, EventEmitter, HostBinding, HostListener, Output } from '@angular/core';
export class OFileDragAndDropDirective {
    constructor() {
        this.onFileDropped = new EventEmitter();
    }
    onDragOver(evt) {
        evt.preventDefault();
        evt.stopPropagation();
        this.fileDragging = true;
    }
    onDragLeave(evt) {
        evt.preventDefault();
        evt.stopPropagation();
        this.fileDragging = false;
    }
    onDrop(evt) {
        evt.preventDefault();
        evt.stopPropagation();
        this.fileDragging = false;
        const files = evt.dataTransfer.files;
        if (files.length > 0) {
            this.onFileDropped.emit(files);
        }
    }
}
OFileDragAndDropDirective.decorators = [
    { type: Directive, args: [{
                selector: '[oFileDragAndDrop]'
            },] }
];
OFileDragAndDropDirective.propDecorators = {
    onFileDropped: [{ type: Output }],
    fileDragging: [{ type: HostBinding, args: ['class.o-file-dragging',] }],
    onDragOver: [{ type: HostListener, args: ['dragover', ['$event'],] }],
    onDragLeave: [{ type: HostListener, args: ['dragleave', ['$event'],] }],
    onDrop: [{ type: HostListener, args: ['drop', ['$event'],] }]
};
