// ontimize-web-ngx/esm5/lib/services/state/o-component-state.service.js
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { Util } from '../../util';
import { LocalStorageService } from '../local-storage.service';
import { DefaultComponentStateClass, DefaultServiceComponentStateClass } from './o-component-state.class';
var AbstractComponentStateService = (function () {
    function AbstractComponentStateService(injector) {
        this.injector = injector;
        this.localStorageService = injector.get(LocalStorageService);
    }
    AbstractComponentStateService.prototype.initialize = function (comp) {
        this.component = comp;
        if (Util.isDefined(this.state)) {
            this.initializeState(this.state);
        }
    };
    AbstractComponentStateService.prototype.initializeState = function (state) {
        if (Util.isDefined(this.state) &&
            ((Util.isDefined(this.component.storeState) && this.component.storeState
                ||
                    !Util.isDefined(this.component.storeState)))) {
            state.setData(this.localStorageService.getComponentStorage(this.component, this.component.getRouteKey()));
        }
    };
    AbstractComponentStateService.decorators = [
        { type: Injectable }
    ];
    AbstractComponentStateService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return AbstractComponentStateService;
}());
export { AbstractComponentStateService };
var DefaultComponentStateService = (function (_super) {
    tslib_1.__extends(DefaultComponentStateService, _super);
    function DefaultComponentStateService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DefaultComponentStateService.prototype.initialize = function (comp) {
        this.state = new DefaultComponentStateClass();
        _super.prototype.initialize.call(this, comp);
    };
    DefaultComponentStateService.decorators = [
        { type: Injectable }
    ];
    return DefaultComponentStateService;
}(AbstractComponentStateService));
export { DefaultComponentStateService };
var DefaultServiceComponentStateService = (function (_super) {
    tslib_1.__extends(DefaultServiceComponentStateService, _super);
    function DefaultServiceComponentStateService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DefaultServiceComponentStateService.prototype.initialize = function (comp) {
        this.state = new DefaultServiceComponentStateClass();
        _super.prototype.initialize.call(this, comp);
    };
    DefaultServiceComponentStateService.decorators = [
        { type: Injectable }
    ];
    return DefaultServiceComponentStateService;
}(AbstractComponentStateService));
export { DefaultServiceComponentStateService };
