// ontimize-web-ngx/esm5/lib/services/state/o-component-state.class.js
import * as tslib_1 from "tslib";
var AbstractComponentStateClass = (function () {
    function AbstractComponentStateClass() {
    }
    return AbstractComponentStateClass;
}());
export { AbstractComponentStateClass };
var AbstractServiceComponentStateClass = (function (_super) {
    tslib_1.__extends(AbstractServiceComponentStateClass, _super);
    function AbstractServiceComponentStateClass() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AbstractServiceComponentStateClass;
}(AbstractComponentStateClass));
export { AbstractServiceComponentStateClass };
var DefaultComponentStateClass = (function (_super) {
    tslib_1.__extends(DefaultComponentStateClass, _super);
    function DefaultComponentStateClass() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DefaultComponentStateClass.prototype.setData = function (data) {
        Object.assign(this, data);
    };
    return DefaultComponentStateClass;
}(AbstractComponentStateClass));
export { DefaultComponentStateClass };
var DefaultServiceComponentStateClass = (function (_super) {
    tslib_1.__extends(DefaultServiceComponentStateClass, _super);
    function DefaultServiceComponentStateClass() {
        return _super.call(this) || this;
    }
    Object.defineProperty(DefaultServiceComponentStateClass.prototype, "filterCaseSensitive", {
        get: function () {
            return this['filter-case-sensitive'];
        },
        set: function (value) {
            this['filter-case-sensitive'] = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DefaultServiceComponentStateClass.prototype, "quickFilterValue", {
        get: function () {
            return this['filterValue'];
        },
        set: function (value) {
            this['filterValue'] = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DefaultServiceComponentStateClass.prototype, "queryRows", {
        get: function () {
            return this['query-rows'];
        },
        set: function (value) {
            this['query-rows'] = value;
        },
        enumerable: true,
        configurable: true
    });
    DefaultServiceComponentStateClass.prototype.setData = function (data) {
        Object.assign(this, data);
    };
    return DefaultServiceComponentStateClass;
}(AbstractServiceComponentStateClass));
export { DefaultServiceComponentStateClass };
