// ontimize-web-ngx/esm5/lib/services/state/o-filter-builder-component-state.class.js
import * as tslib_1 from "tslib";
import { Util } from '../../util';
import { DefaultComponentStateClass } from './o-component-state.class';
var OFilterBuilderComponentStateClass = (function (_super) {
    tslib_1.__extends(OFilterBuilderComponentStateClass, _super);
    function OFilterBuilderComponentStateClass() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(OFilterBuilderComponentStateClass.prototype, "storedFilterBuilders", {
        get: function () {
            return this['stored-filter-builders'] || [];
        },
        set: function (value) {
            this['stored-filter-builders'] = value;
        },
        enumerable: true,
        configurable: true
    });
    OFilterBuilderComponentStateClass.prototype.addStoredFilter = function (filter) {
        if (!Util.isDefined(this['stored-filter-builders'])) {
            this['stored-filter-builders'] = [];
        }
        this.storedFilterBuilders.push(filter);
    };
    OFilterBuilderComponentStateClass.prototype.deleteStoredFilter = function (filterName) {
        var index = this.storedFilterBuilders.findIndex(function (item) { return item.name === filterName; });
        if (index >= 0) {
            this.storedFilterBuilders.splice(index, 1);
        }
    };
    OFilterBuilderComponentStateClass.prototype.applyFilter = function (filterName) {
        var filter = this.getStoredFilter(filterName);
        if (filter) {
            this.filterBuilderValues = filter;
        }
    };
    OFilterBuilderComponentStateClass.prototype.getStoredFilter = function (filterName) {
        var result = [];
        var filter = this.storedFilterBuilders.find(function (item) { return item.name === filterName; });
        if (Util.isDefined(filter)) {
            result = filter['filter-builder-values'];
        }
        return result;
    };
    return OFilterBuilderComponentStateClass;
}(DefaultComponentStateClass));
export { OFilterBuilderComponentStateClass };
