// ontimize-web-ngx/esm5/lib/services/state/o-grid-component-state.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Codes } from '../../util/codes';
import { Util } from '../../util/util';
import { AbstractComponentStateService } from './o-component-state.service';
import { OGridComponentStateClass } from './o-grid-component-state.class';
var OGridComponentStateService = (function (_super) {
    tslib_1.__extends(OGridComponentStateService, _super);
    function OGridComponentStateService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OGridComponentStateService.prototype.initialize = function (component) {
        this.state = new OGridComponentStateClass();
        _super.prototype.initialize.call(this, component);
    };
    OGridComponentStateService.prototype.initializeState = function (state) {
        _super.prototype.initializeState.call(this, state);
    };
    OGridComponentStateService.prototype.getDataToStore = function () {
        var dataToStore = Object.assign({}, this.state);
        dataToStore['query-rows'] = this.component.queryRows;
        dataToStore['currentPage'] = this.component.currentPage;
        if (this.component.storePaginationState) {
            dataToStore['queryRecordOffset'] = Math.max((this.state.queryRecordOffset - this.component.dataArray.length), (this.state.queryRecordOffset - this.component.queryRows));
        }
        else {
            delete dataToStore['queryRecordOffset'];
        }
        if (Util.isDefined(this.component.sortColumnOrder)) {
            dataToStore['sort-column'] = this.component.sortColumnOrder.columnName + Codes.COLUMNS_ALIAS_SEPARATOR +
                (this.component.sortColumnOrder.ascendent ? Codes.ASC_SORT : Codes.DESC_SORT);
        }
        dataToStore['filter-case-sensitive'] = this.component.isFilterCaseSensitive();
        if (this.component.quickFilter && Util.isDefined(this.component.quickFilterComponent)) {
            dataToStore['quickFilterActiveColumns'] = this.component.quickFilterComponent.getActiveColumns().join(Codes.ARRAY_INPUT_SEPARATOR);
        }
        return dataToStore;
    };
    OGridComponentStateService.decorators = [
        { type: Injectable }
    ];
    return OGridComponentStateService;
}(AbstractComponentStateService));
export { OGridComponentStateService };
