// ontimize-web-ngx/esm5/lib/services/state/o-list-component-state.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Codes } from '../../util/codes';
import { Util } from '../../util/util';
import { AbstractComponentStateService } from './o-component-state.service';
import { OListComponentStateClass } from './o-list-component-state.class';
var OListComponentStateService = (function (_super) {
    tslib_1.__extends(OListComponentStateService, _super);
    function OListComponentStateService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OListComponentStateService.prototype.initialize = function (component) {
        this.state = new OListComponentStateClass();
        _super.prototype.initialize.call(this, component);
    };
    OListComponentStateService.prototype.initializeState = function (state) {
        _super.prototype.initializeState.call(this, state);
    };
    OListComponentStateService.prototype.getDataToStore = function () {
        var dataToStore = Object.assign({}, this.state);
        dataToStore['query-rows'] = this.component.queryRows;
        if (!this.component.storePaginationState) {
            delete dataToStore['queryRecordOffset'];
        }
        if (this.component.quickFilter && Util.isDefined(this.component.quickFilterComponent)) {
            dataToStore['quickFilterActiveColumns'] = this.component.quickFilterComponent.getActiveColumns().join(Codes.ARRAY_INPUT_SEPARATOR);
        }
        dataToStore['filter-case-sensitive'] = this.component.isFilterCaseSensitive();
        dataToStore.selection = this.state.selection;
        return dataToStore;
    };
    OListComponentStateService.prototype.refreshSelection = function () {
        this.state.selection = this.getSelectionState();
    };
    OListComponentStateService.prototype.getSelectionState = function () {
        var selection = [];
        if (this.component) {
            var keys_1 = this.component.getKeys();
            this.component.getSelectedItems().forEach(function (item) {
                var data = {};
                keys_1.forEach(function (key) {
                    data[key] = item[key];
                });
                selection.push(data);
            });
        }
        return selection;
    };
    OListComponentStateService.decorators = [
        { type: Injectable }
    ];
    return OListComponentStateService;
}(AbstractComponentStateService));
export { OListComponentStateService };
