// ontimize-web-ngx-report/esm2015/lib/components/report-on-demand/style-dialog/style-dialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/report-on-demand/style-dialog/style-dialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject, Optional } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
import { Util } from 'ontimize-web-ngx';
import { Constants } from '../../../util/constants';
import { Utils } from '../../../util/utils';
export class StyleDialogComponent {
    /**
     * @param {?} dialogo
     * @param {?} data
     */
    constructor(dialogo, data) {
        this.dialogo = dialogo;
        this.data = data;
        this.dataAlignment = [
            { "name": "left", "icon": "format_align_left" },
            { "name": "center", "icon": "format_align_center" },
            { "name": "right", "icon": "format_align_right" }
        ];
        this.reportColumn = Utils.cloneObject(data);
        if (this.reportColumn.columnStyle && !Util.isDefined(this.reportColumn.columnStyle.width)) {
            this.reportColumn.columnStyle.width = Constants.DEFAULT_WIDTH_COLUMN_STYLE;
        }
        if (this.reportColumn.columnStyle && !Util.isDefined(this.reportColumn.columnStyle.alignment)) {
            this.reportColumn.columnStyle.alignment = Constants.DEFAULT_ALIGNMENT_COLUMN_STYLE;
        }
        if (!this.reportColumn.columnStyle) {
            this.reportColumn.columnStyle = { width: Constants.DEFAULT_WIDTH_COLUMN_STYLE, alignment: Constants.DEFAULT_ALIGNMENT_COLUMN_STYLE };
        }
    }
    /**
     * @return {?}
     */
    confirm() {
        this.dialogo.close(this.reportColumn);
    }
}
StyleDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-style-dialog',
                template: "<span mat-dialog-title>{{ 'COLUMN_CONFIGURATION' | oReportsTranslate }}</span>\n<mat-dialog-content>\n  <div mat-subheader>{{ 'SAVE_COLUMN_CONFIGURATION' | oReportsTranslate }}</div>\n  <div fxLayout=\"column\" fxLayoutAlign=\"start start\" fxLayoutGap=\"12px\">\n    <form #form>\n      <mat-form-field fxFill>\n        <mat-label>{{ 'COLUMN_NAME' | oReportsTranslate }}</mat-label>\n        <input id=\"columnNameInput\" name=\"columnNameInput\" matInput placeholder=\"{{ 'COLUMN_NAME' | oReportsTranslate }}\" required\n          [(ngModel)]=\"reportColumn.name\" #columnNameInput=\"ngModel\" [ngModelOptions]=\"{standalone: true}\">\n      </mat-form-field>\n      <mat-form-field fxFill>\n        <input id=\"widthColumnInput\" name=\"widthColumnInput\" matInput type=\"number\" placeholder=\"{{ 'COLUMN_WIDTH' | oReportsTranslate }}\" required\n          [(ngModel)]=\"reportColumn.columnStyle.width\" #widthColumnInput=\"ngModel\" [ngModelOptions]=\"{standalone: true}\">\n        <span matSuffix>px</span>\n        <mat-hint></mat-hint>\n      </mat-form-field>\n      <mat-radio-group [(ngModel)]=\"reportColumn.columnStyle.alignment\" fxFill fxLayout=\"row\" fxLayoutAlign=\"space-between center\"\n        [ngModelOptions]=\"{standalone: true}\">\n        <mat-radio-button *ngFor=\"let aligmentItem of dataAlignment\" [value]=\"aligmentItem.name\" [checked]=\"reportColumn.columnStyle.alignment === aligmentItem\">\n          <mat-icon fxFlex>\n            {{ aligmentItem.icon }}\n          </mat-icon>\n        </mat-radio-button>\n      </mat-radio-group>\n    </form>\n  </div>\n</mat-dialog-content>\n<mat-dialog-actions align=\"end\">\n  <button type=\"button\" mat-stroked-button class=\"mat-primary\" [mat-dialog-close]=\"false\" class=\"o-button-default\">\n    {{ 'CANCEL' | oTranslate }}</button>\n  <button type=\"button\" mat-stroked-button class=\"mat-primary\" (click)=\"confirm()\" [mat-dialog-close]=\"true\"\n    [disabled]=\"columnNameInput.invalid || widthColumnInput.invalid\">\n    {{ 'SAVE' | oTranslate }}</button>\n</mat-dialog-actions>\n"
            }] }
];
/** @nocollapse */
StyleDialogComponent.ctorParameters = () => [
    { type: MatDialogRef, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DIALOG_DATA,] }] }
];
if (false) {
    /** @type {?} */
    StyleDialogComponent.prototype.reportColumn;
    /** @type {?} */
    StyleDialogComponent.prototype.dataAlignment;
    /** @type {?} */
    StyleDialogComponent.prototype.dialogo;
    /** @type {?} */
    StyleDialogComponent.prototype.data;
}
