// ontimize-web-ngx/esm5/lib/components/input/combo/combo-renderer/date/o-combo-renderer-date.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { OMomentPipe } from '../../../../../pipes/o-moment.pipe';
import { DEFAULT_INPUTS_O_COMBO_RENDERER, OComboCustomRenderer } from '../o-combo-renderer.class';
export var DEFAULT_INPUTS_O_COMBO_RENDERER_DATE = tslib_1.__spread(DEFAULT_INPUTS_O_COMBO_RENDERER, [
    'format'
]);
var OComboRendererDateComponent = (function (_super) {
    tslib_1.__extends(OComboRendererDateComponent, _super);
    function OComboRendererDateComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.setComponentPipe();
        return _this;
    }
    OComboRendererDateComponent.prototype.setComponentPipe = function () {
        this.componentPipe = new OMomentPipe(this.injector);
    };
    OComboRendererDateComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        this.pipeArguments = {
            format: this.format
        };
    };
    OComboRendererDateComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-combo-renderer-date',
                    template: "<ng-template #templateref let-value=\"value\">\n  {{getComboData(value)}}\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_COMBO_RENDERER_DATE
                }] }
    ];
    OComboRendererDateComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OComboRendererDateComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
    };
    return OComboRendererDateComponent;
}(OComboCustomRenderer));
export { OComboRendererDateComponent };
