// ontimize-web-ngx/esm5/lib/components/table/extensions/dialog/group-by-columns/o-table-group-by-columns-dialog.component.js
import { ChangeDetectionStrategy, Component, Inject, ViewChild, ViewEncapsulation } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
import { Util } from '../../../../../util/util';
import { ODualListSelectorComponent } from '../../../../dual-list-selector/o-dual-list-selector.component';
var OTableGroupByColumnsDialogComponent = (function () {
    function OTableGroupByColumnsDialogComponent(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.columns = [];
        this.groupedColumns = [];
        this.groupedColumnTypes = [];
        if (Util.isArray(data.groupedColumns) && !Util.isArrayEmpty(data.groupedColumns)) {
            this.groupedColumns = data.groupedColumns.map(function (attr) {
                var indexCol = data.columnsData.findIndex(function (col) { return col.attr === attr; });
                if (indexCol > -1) {
                    return data.columnsData[indexCol];
                }
            });
        }
        var self = this;
        if (Util.isArray(data.columnsData)) {
            this.columns = data.columnsData.filter(function (oCol) { return oCol.visible && oCol.groupable && self.groupedColumns.findIndex(function (x) { return x.attr === oCol.attr; }) === -1; });
        }
        if (Util.isDefined(data.groupedColumnTypes)) {
            this.groupedColumnTypes = data.groupedColumnTypes;
        }
    }
    OTableGroupByColumnsDialogComponent.prototype.getGroupedColumns = function () {
        return this.dualListSelector.getSelectedItems().map(function (oCol) { return oCol.attr; });
    };
    OTableGroupByColumnsDialogComponent.prototype.getGroupedColumnTypes = function () {
        return this.dualListSelector.getGroupedColumnTypes();
    };
    OTableGroupByColumnsDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table-group-by-columns-dialog',
                    template: "<span mat-dialog-title>{{ 'TABLE.BUTTONS.GROUPBY' | oTranslate }}</span>\n\n<mat-dialog-content>\n  <o-dual-list-selector #dualListSelector [data-source]=\"columns\" [data-destination]=\"groupedColumns\" [grouped-column-types]=\"groupedColumnTypes\"\n    key=\"attr\" display=\"title\" title-list-data-source=\"TABLE_GROUP_BY_COLUMNS_DIALOG.ALL_COLUMNS\"\n    title-list-data-destination=\"TABLE_GROUP_BY_COLUMNS_DIALOG.GROUPED_COLUMNS\" description=\"TABLE_GROUP_BY_COLUMNS_DIALOG.DESCRIPTION\" fxFlexFill>\n  </o-dual-list-selector>\n</mat-dialog-content>\n\n<mat-dialog-actions fxLayoutAlign=\"end center\">\n  <button type=\"button\" mat-stroked-button [mat-dialog-close]=\"false\" class=\"o-button-default cancel\">{{ 'CANCEL' | oTranslate }}</button>\n  <button type=\"button\" mat-stroked-button [mat-dialog-close]=\"true\" class=\"o-button-primary\">{{ 'ACCEPT' | oTranslate }}</button>\n</mat-dialog-actions>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.o-table-group-by-column-dialog]': 'true'
                    },
                    styles: [".o-table-group-by-column-dialog{display:flex;flex-direction:column;height:100%}.o-table-group-by-column-dialog .mat-dialog-content{flex:1}"]
                }] }
    ];
    OTableGroupByColumnsDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    OTableGroupByColumnsDialogComponent.propDecorators = {
        dualListSelector: [{ type: ViewChild, args: ['dualListSelector', { static: false },] }]
    };
    return OTableGroupByColumnsDialogComponent;
}());
export { OTableGroupByColumnsDialogComponent };
