// ontimize-web-ngx/esm5/lib/components/table/extensions/default-o-table.datasource.js
import * as tslib_1 from "tslib";
import { DataSource } from '@angular/cdk/collections';
import { EventEmitter } from '@angular/core';
import { BehaviorSubject, merge, Subject, Subscription } from 'rxjs';
import { distinctUntilChanged, map } from 'rxjs/operators';
import { ColumnValueFilterOperator } from '../../../types/table/o-column-value-filter.type';
import { Codes } from '../../../util';
import { Util } from '../../../util/util';
import { OTableGroupedRow } from './row/o-table-row-group.class';
var OnRangeChangeVirtualScroll = (function () {
    function OnRangeChangeVirtualScroll(data) {
        this.range = data;
    }
    return OnRangeChangeVirtualScroll;
}());
export { OnRangeChangeVirtualScroll };
var DefaultOTableDataSource = (function (_super) {
    tslib_1.__extends(DefaultOTableDataSource, _super);
    function DefaultOTableDataSource(table) {
        var _this = _super.call(this) || this;
        _this.table = table;
        _this.dataTotalsChange = new BehaviorSubject([]);
        _this._virtualPageChange = new BehaviorSubject(new OnRangeChangeVirtualScroll({ start: 0, end: 0 }));
        _this._quickFilterChange = new BehaviorSubject('');
        _this._columnValueFilterChange = new BehaviorSubject(null);
        _this.groupByColumnChange = new Subject();
        _this.filteredData = [];
        _this.aggregateData = {};
        _this.onRenderedDataChange = new EventEmitter();
        _this._renderedData = [];
        _this.resultsLength = 0;
        _this.columnValueFilters = [];
        _this.groupedRowState = [];
        _this.activeAggregates = {};
        _this.groupedRowsSubscription = new Subscription();
        _this.levelsExpansionState = {};
        _this._database = table.daoTable;
        if (_this._database) {
            _this.resultsLength = _this._database.data.length;
        }
        if (table.paginator) {
            _this._paginator = table.matpaginator;
        }
        if (_this.table.virtualScrollViewport) {
            _this.table.virtualScrollViewport.renderedRangeStream
                .pipe(distinctUntilChanged())
                .subscribe(function (value) {
                _this._virtualPageChange.next(new OnRangeChangeVirtualScroll(value));
            });
        }
        _this._tableOptions = table.oTableOptions;
        _this._sort = table.sort;
        return _this;
    }
    Object.defineProperty(DefaultOTableDataSource.prototype, "data", {
        get: function () { return this.dataTotalsChange.value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DefaultOTableDataSource.prototype, "quickFilter", {
        get: function () { return this._quickFilterChange.value || ''; },
        set: function (filter) {
            this._quickFilterChange.next(filter);
        },
        enumerable: true,
        configurable: true
    });
    DefaultOTableDataSource.prototype.sortFunction = function (a, b) {
        return this._sort.sortFunction(a, b);
    };
    Object.defineProperty(DefaultOTableDataSource.prototype, "renderedData", {
        get: function () {
            return this._renderedData;
        },
        set: function (arg) {
            this._renderedData = arg;
            this.onRenderedDataChange.emit();
        },
        enumerable: true,
        configurable: true
    });
    DefaultOTableDataSource.prototype.connect = function () {
        var _this = this;
        var displayDataChanges = [
            this._database.dataChange
        ];
        if (!this.table.pageable) {
            if (this._sort) {
                displayDataChanges.push(this._sort.oSortChange);
            }
            if (this._tableOptions.filter) {
                displayDataChanges.push(this._quickFilterChange);
            }
            if (this._paginator) {
                displayDataChanges.push(this._paginator.page);
            }
        }
        if (this.table.virtualScrollViewport) {
            displayDataChanges.push(this._virtualPageChange);
        }
        displayDataChanges.push(this._columnValueFilterChange);
        if (this.table.groupable) {
            displayDataChanges.push(this.groupByColumnChange);
        }
        return merge.apply(void 0, tslib_1.__spread(displayDataChanges)).pipe(map(function (x) {
            var data = Object.assign([], _this._database.data);
            if (x instanceof OnRangeChangeVirtualScroll) {
                data = _this.getVirtualScrollData(_this.renderedData, x);
            }
            else {
                if (_this.existsAnyCalculatedColumn()) {
                    data = _this.getColumnCalculatedData(data);
                }
                if (!_this.table.pageable) {
                    data = _this.getColumnValueFilterData(data);
                    data = _this.getQuickFilterData(data);
                    data = _this.getSortedData(data);
                }
                _this.filteredData = Object.assign([], data);
                if (_this.table.pageable) {
                    var totalRecordsNumber = _this.table.getTotalRecordsNumber();
                    _this.resultsLength = totalRecordsNumber !== undefined ? totalRecordsNumber : data.length;
                }
                else {
                    _this.resultsLength = data.length;
                    data = _this.getPaginationData(data);
                }
                if (_this.table.groupable && !Util.isArrayEmpty(_this.table.groupedColumnsArray) && data.length > 0) {
                    data = _this.getGroupedData(data);
                }
                _this.renderedData = data;
                if (_this.table.virtualScrollViewport && !_this._paginator) {
                    data = _this.getVirtualScrollData(data, new OnRangeChangeVirtualScroll({ start: 0, end: Codes.LIMIT_SCROLLVIRTUAL }));
                }
                _this.aggregateData = _this.getAggregatesData(_this.renderedData);
            }
            return data;
        }));
    };
    DefaultOTableDataSource.prototype.getGroupedData = function (data) {
        data = this.getSubGroupsOfGroupedRow(data);
        data = this.filterCollapsedRowGroup(data);
        return data;
    };
    DefaultOTableDataSource.prototype.getSubGroupsOfGroupedRow = function (data) {
        return data = this.getSublevel(data, 0);
    };
    DefaultOTableDataSource.prototype.getAggregatesData = function (data) {
        var _this = this;
        var obj = {};
        if (typeof this._tableOptions === 'undefined') {
            return obj;
        }
        this._tableOptions.columns.forEach(function (column) {
            var totalValue = '';
            if (column.aggregate && column.visible) {
                totalValue = _this.calculateAggregate(data, column.attr, column.aggregate.operator);
            }
            var key = column.attr;
            obj[key] = totalValue;
        });
        return obj;
    };
    DefaultOTableDataSource.prototype.getColumnCalculatedData = function (data) {
        var self = this;
        var calculatedCols = this._tableOptions.columns.filter(function (oCol) { return oCol.visible && oCol.calculate !== undefined; });
        return data.map(function (row) {
            calculatedCols.forEach(function (oColumn) {
                var value;
                if (typeof oColumn.calculate === 'string') {
                    value = self.transformFormula(oColumn.calculate, row);
                }
                else if (typeof oColumn.calculate === 'function') {
                    value = oColumn.calculate(row);
                }
                row[oColumn.attr] = isNaN(value) ? 0 : value;
            });
            return row;
        });
    };
    DefaultOTableDataSource.prototype.transformFormula = function (formulaArg, row) {
        var formula = formulaArg;
        var columnsAttr = this._tableOptions.columns.map(function (oCol) { return oCol.attr; });
        columnsAttr.forEach(function (column) {
            formula = formula.replace(column, row[column]);
        });
        var resultFormula = '';
        try {
            resultFormula = (new Function('return ' + formula))();
        }
        catch (e) {
            console.error('Operation defined in the calculated column is incorrect ');
        }
        return resultFormula;
    };
    DefaultOTableDataSource.prototype.getQuickFilterData = function (data) {
        var _this = this;
        if (Util.isDefined(this.quickFilter) && this.quickFilter.length > 0) {
            var filterData_1 = !this._tableOptions.filterCaseSensitive ? this.quickFilter.toLowerCase() : this.quickFilter;
            return data.filter(function (item) {
                var passCustomFilter = _this.fulfillsCustomFilterFunctions(filterData_1, item);
                var passSearchString = _this.fulfillsQuickfilter(filterData_1, item);
                return passCustomFilter || passSearchString;
            });
        }
        else {
            return data;
        }
    };
    DefaultOTableDataSource.prototype.getPaginationData = function (data) {
        if (!this._paginator || isNaN(this._paginator.pageSize)) {
            return data;
        }
        var startIndex = isNaN(this._paginator.pageSize) ? 0 : this._paginator.pageIndex * this._paginator.pageSize;
        if (data.length > 0 && data.length < startIndex) {
            startIndex = 0;
            this._paginator.pageIndex = 0;
        }
        return data.splice(startIndex, this._paginator.pageSize);
    };
    DefaultOTableDataSource.prototype.getVirtualScrollData = function (data, x) {
        return data.slice(x.range.start, x.range.end);
    };
    DefaultOTableDataSource.prototype.disconnect = function () {
        this.dataTotalsChange.complete();
        this._quickFilterChange.complete();
        this._columnValueFilterChange.complete();
        this.groupByColumnChange.complete();
        this._virtualPageChange.complete();
    };
    DefaultOTableDataSource.prototype.fulfillsCustomFilterFunctions = function (filter, item) {
        var customFilterCols = this.table.oTableOptions.columns.filter(function (oCol) { return oCol.useCustomFilterFunction(); });
        return customFilterCols.some(function (oCol) { return oCol.renderer.filterFunction(item[oCol.attr], item, filter); });
    };
    DefaultOTableDataSource.prototype.fulfillsQuickfilter = function (filter, item) {
        var columns = this._tableOptions.columns.filter(function (oCol) { return oCol.useQuickfilterFunction(); });
        var searchStr = columns.map(function (oCol) { return oCol.getFilterValue(item[oCol.attr], item).join(' '); }).join(' ');
        if (!this._tableOptions.filterCaseSensitive) {
            searchStr = searchStr.toLowerCase();
        }
        return searchStr.indexOf(filter) !== -1;
    };
    DefaultOTableDataSource.prototype.getSortedData = function (data) {
        return this._sort.getSortedData(data);
    };
    DefaultOTableDataSource.prototype.getTableData = function () {
        return this._database.data;
    };
    DefaultOTableDataSource.prototype.getCurrentData = function () {
        return this.getData();
    };
    DefaultOTableDataSource.prototype.getCurrentAllData = function () {
        return this.getAllData(false, false);
    };
    DefaultOTableDataSource.prototype.getCurrentRendererData = function () {
        return this.getRenderedData(this.renderedData);
    };
    DefaultOTableDataSource.prototype.getAllRendererData = function () {
        return this.getAllData(true, true);
    };
    Object.defineProperty(DefaultOTableDataSource.prototype, "sqlTypes", {
        get: function () {
            return this._database.sqlTypes;
        },
        enumerable: true,
        configurable: true
    });
    DefaultOTableDataSource.prototype.getData = function () {
        return this.renderedData;
    };
    DefaultOTableDataSource.prototype.getRenderedData = function (data) {
        var visibleColumns = this._tableOptions.columns.filter(function (oCol) { return oCol.visible; });
        return data.map(function (row) {
            var obj = {};
            visibleColumns.forEach(function (oCol) {
                var useRenderer = oCol.renderer && oCol.renderer.getCellData;
                obj[oCol.attr] = useRenderer ? oCol.renderer.getCellData(row[oCol.attr], row) : row[oCol.attr];
            });
            return obj;
        });
    };
    DefaultOTableDataSource.prototype.getAllData = function (usingRendererers, onlyVisibleColumns) {
        var tableColumns = this._tableOptions.columns;
        if (onlyVisibleColumns) {
            tableColumns = this._tableOptions.columns.filter(function (oCol) { return oCol.visible; });
        }
        return this.filteredData.map(function (row) {
            var obj = {};
            tableColumns.forEach(function (oCol) {
                var useRenderer = usingRendererers && oCol.renderer && oCol.renderer.getCellData;
                obj[oCol.attr] = useRenderer ? oCol.renderer.getCellData(row[oCol.attr], row) : row[oCol.attr];
            });
            return obj;
        });
    };
    DefaultOTableDataSource.prototype.getRenderersData = function (data, tableColumns) {
        return data.map(function (row) {
            var obj = Object.assign({}, row);
            tableColumns.forEach(function (oCol) {
                obj[oCol.attr] = oCol.renderer.getCellData(row[oCol.attr], row);
            });
            return obj;
        });
    };
    DefaultOTableDataSource.prototype.getColumnData = function (ocolumn) {
        return this.renderedData.map(function (row) {
            var obj = {};
            if (ocolumn) {
                obj[ocolumn] = row[ocolumn];
            }
            return obj;
        });
    };
    DefaultOTableDataSource.prototype.initializeColumnsFilters = function (filters) {
        var _this = this;
        this.columnValueFilters = [];
        filters.forEach(function (filter) {
            _this.columnValueFilters.push(filter);
        });
        if (!this.table.pageable) {
            this._columnValueFilterChange.next(null);
        }
    };
    DefaultOTableDataSource.prototype.isColumnValueFilterActive = function () {
        return this.columnValueFilters.length !== 0;
    };
    DefaultOTableDataSource.prototype.getColumnValueFilters = function () {
        return this.columnValueFilters;
    };
    DefaultOTableDataSource.prototype.getColumnValueFilterByAttr = function (attr) {
        return this.columnValueFilters.find(function (item) { return item.attr === attr; });
    };
    DefaultOTableDataSource.prototype.clearColumnFilters = function (trigger, columnsAttr) {
        if (trigger === void 0) { trigger = true; }
        if (Util.isDefined(columnsAttr)) {
            this.columnValueFilters = this.columnValueFilters.filter(function (x) { return !columnsAttr.includes(x.attr); });
        }
        else {
            this.columnValueFilters = [];
        }
        if (trigger) {
            this._columnValueFilterChange.next(null);
        }
    };
    DefaultOTableDataSource.prototype.clearColumnFilter = function (attr, trigger) {
        if (trigger === void 0) { trigger = true; }
        this.columnValueFilters = this.columnValueFilters.filter(function (x) { return x.attr !== attr; });
        if (trigger) {
            this._columnValueFilterChange.next(null);
        }
    };
    DefaultOTableDataSource.prototype.addColumnFilter = function (filter) {
        var existingFilterIndex = this.columnValueFilters.findIndex(function (item) { return item.attr === filter.attr; });
        if (existingFilterIndex > -1) {
            this.columnValueFilters.splice(existingFilterIndex, 1, filter);
        }
        else {
            var validFilter = Util.isDefined(filter.values);
            if (validFilter) {
                switch (filter.operator) {
                    case ColumnValueFilterOperator.IN:
                        validFilter = filter.values.length > 0;
                        break;
                    case ColumnValueFilterOperator.BETWEEN:
                        validFilter = filter.values.length === 2;
                        break;
                    case ColumnValueFilterOperator.EQUAL:
                    case ColumnValueFilterOperator.LESS_EQUAL:
                    case ColumnValueFilterOperator.MORE_EQUAL:
                        validFilter = true;
                        break;
                    default:
                        validFilter = false;
                }
                if (validFilter) {
                    this.columnValueFilters.push(filter);
                }
            }
        }
        if (!this.table.pageable) {
            this._columnValueFilterChange.next(null);
        }
    };
    DefaultOTableDataSource.prototype.getColumnValueFilterData = function (data) {
        var _this = this;
        this.columnValueFilters.forEach(function (filter) {
            var filterColumn = _this.table.oTableOptions.columns.find(function (col) { return col.attr === filter.attr; });
            if (filterColumn) {
                switch (filter.operator) {
                    case ColumnValueFilterOperator.IN:
                        var filterValues_1 = (filter.values || []).reduce(function (previous, current) {
                            return previous.concat(filterColumn.getFilterValue(current).map(function (f) { return Util.normalizeString(f); }));
                        }, []);
                        data = data.filter(function (item) {
                            if (filterColumn.renderer && filterColumn.renderer.filterFunction) {
                                return filterColumn.renderer.filterFunction(item[filter.attr], item);
                            }
                            else {
                                var colValues_1 = filterColumn.getFilterValue(item[filter.attr], item).map(function (f) { return Util.normalizeString(f); });
                                return filterValues_1.some(function (value) { return colValues_1.indexOf(value) !== -1; });
                            }
                        });
                        break;
                    case ColumnValueFilterOperator.EQUAL:
                        var normalizedValue_1 = Util.normalizeString(filter.values);
                        data = data.filter(function (item) {
                            var colValues = filterColumn.getFilterValue(item[filter.attr], item).map(function (f) { return Util.normalizeString(f); });
                            var regExp;
                            if (normalizedValue_1.includes('*')) {
                                regExp = new RegExp('^' + normalizedValue_1.split('*').join('.*') + '$');
                            }
                            return colValues.some(function (colValue) { return regExp ? regExp.test(colValue) : colValue.includes(normalizedValue_1); });
                        });
                        break;
                    case ColumnValueFilterOperator.BETWEEN:
                        data = data.filter(function (item) { return item[filter.attr] >= filter.values[0] && item[filter.attr] <= filter.values[1]; });
                        break;
                    case ColumnValueFilterOperator.MORE_EQUAL:
                        data = data.filter(function (item) { return item[filter.attr] >= filter.values; });
                        break;
                    case ColumnValueFilterOperator.LESS_EQUAL:
                        data = data.filter(function (item) { return item[filter.attr] <= filter.values; });
                        break;
                }
            }
        });
        return data;
    };
    DefaultOTableDataSource.prototype.getAggregateData = function (column) {
        var obj = {};
        var totalValue = '';
        if (typeof this._tableOptions === 'undefined') {
            return new Array(obj);
        }
        totalValue = this.aggregateData[column.attr];
        return totalValue;
    };
    DefaultOTableDataSource.prototype.calculateAggregate = function (data, columnAttr, operator) {
        var resultAggregate;
        if (typeof operator === 'string') {
            switch (operator.toLowerCase()) {
                case 'count':
                    resultAggregate = this.count(columnAttr, data);
                    break;
                case 'min':
                    resultAggregate = this.min(columnAttr, data);
                    break;
                case 'max':
                    resultAggregate = this.max(columnAttr, data);
                    break;
                case 'avg':
                    resultAggregate = this.avg(columnAttr, data);
                    break;
                default:
                    resultAggregate = this.sum(columnAttr, data);
                    break;
            }
        }
        else {
            var columnData = this.getColumnData(columnAttr);
            if (typeof operator === 'function') {
                resultAggregate = operator(columnData);
            }
        }
        return resultAggregate;
    };
    DefaultOTableDataSource.prototype.sum = function (column, data) {
        var value = 0;
        if (data) {
            if (data[0] instanceof OTableGroupedRow) {
                this.getDataFromFirstLevelTableGroupRow(data).forEach(function (x) {
                    value = x.getColumnAggregateData(column).reduce(function (acumulator, currentValue) {
                        return acumulator + (isNaN(currentValue[column]) ? 0 : currentValue[column]);
                    }, value);
                });
            }
            else {
                value = data.reduce(function (acumulator, currentValue) {
                    return acumulator + (isNaN(currentValue[column]) ? 0 : currentValue[column]);
                }, value);
            }
        }
        return +(value).toFixed(2);
    };
    DefaultOTableDataSource.prototype.count = function (column, data) {
        var value = 0;
        if (data) {
            if (data[0] instanceof OTableGroupedRow) {
                this.getDataFromFirstLevelTableGroupRow(data).forEach(function (x) {
                    value = x.getColumnAggregateData(column).reduce(function (acumulator) {
                        return acumulator + 1;
                    }, value);
                });
            }
            else {
                value = data.reduce(function (acumulator) {
                    return acumulator + 1;
                }, 0);
            }
        }
        return value;
    };
    DefaultOTableDataSource.prototype.avg = function (column, data) {
        var totalSum = this.sum(column, data);
        var totalCount = this.count(column, data);
        return +((totalSum === 0 || totalCount === 0) ? 0 : (totalSum / totalCount)).toFixed(2);
    };
    DefaultOTableDataSource.prototype.min = function (column, data) {
        var tempMin = [];
        if (data[0] instanceof OTableGroupedRow) {
            tempMin = this.getDataFromFirstLevelTableGroupRow(data).map(function (x) {
                return Math.min.apply(Math, tslib_1.__spread(x.getColumnAggregateData(column).map(function (x) { return x[column]; })));
            });
        }
        else {
            tempMin = data.map(function (x) { return x[column]; });
        }
        return tempMin.length > 0 ? Math.min.apply(Math, tslib_1.__spread(tempMin)) : 0;
    };
    DefaultOTableDataSource.prototype.max = function (column, data) {
        var tempMax = [];
        if (data[0] instanceof OTableGroupedRow) {
            tempMax = this.getDataFromFirstLevelTableGroupRow(data).map(function (x) {
                return Math.max.apply(Math, tslib_1.__spread(x.getColumnAggregateData(column).map(function (x) { return x[column]; })));
            });
        }
        else {
            tempMax = data.map(function (x) { return x[column]; });
        }
        return tempMax.length > 0 ? Math.max.apply(Math, tslib_1.__spread(tempMax)) : 0;
    };
    DefaultOTableDataSource.prototype.isFirstLevelTableGroupRow = function (tableRowGroupData) {
        return tableRowGroupData instanceof OTableGroupedRow && tableRowGroupData.level === 1;
    };
    DefaultOTableDataSource.prototype.getDataFromFirstLevelTableGroupRow = function (data) {
        var _this = this;
        return data.filter(function (x) { return _this.isFirstLevelTableGroupRow(x); });
    };
    DefaultOTableDataSource.prototype.existsAnyCalculatedColumn = function () {
        return this._tableOptions.columns.find(function (oCol) { return oCol.calculate !== undefined; }) !== undefined;
    };
    DefaultOTableDataSource.prototype.updateRenderedRowData = function (rowData) {
        var tableKeys = this.table.getKeys();
        var record = this.renderedData.find(function (data) {
            var found = true;
            for (var i = 0, len = tableKeys.length; i < len; i++) {
                var key = tableKeys[i];
                if (data[key] !== rowData[key]) {
                    found = false;
                    break;
                }
            }
            return found;
        });
        if (Util.isDefined(record)) {
            Object.assign(record, rowData);
        }
    };
    DefaultOTableDataSource.prototype.getDataInformationByGroup = function (data, level) {
        var _this = this;
        var recordHash = {};
        data.forEach(function (row, i) {
            var keys = {};
            for (var i_1 = 0; i_1 <= level; i_1++) {
                keys[_this.table.groupedColumnsArray[i_1]] = _this.table.getColumnDataByAttr(_this.table.groupedColumnsArray[i_1], row);
            }
            var recordKey = JSON.stringify(keys);
            if (recordHash.hasOwnProperty(recordKey)) {
                recordHash[recordKey].push(i);
            }
            else {
                recordHash[recordKey] = [i];
            }
        });
        return recordHash;
    };
    DefaultOTableDataSource.prototype.recalculateColumnAggregate = function (columnAttr, row) {
        var aggregateConf = row.getActiveColumnAggregateConfiguration(columnAttr);
        var data = row.getColumnAggregateData(columnAttr);
        var value = this.calculateAggregate(data, aggregateConf.attr, aggregateConf.aggregateFunction || aggregateConf.aggregate);
        row.setColumnAggregateValue(columnAttr, value);
    };
    DefaultOTableDataSource.prototype.getSublevel = function (data, level, parent) {
        var _this = this;
        if (level >= this.table.groupedColumnsArray.length) {
            return data;
        }
        var recordHash = this.getDataInformationByGroup(data, level);
        var result = [];
        Object.keys(recordHash).forEach(function (recordKey) {
            var row = new OTableGroupedRow({
                column: _this.table.groupedColumnsArray[level],
                keysAsString: recordKey,
                level: level + 1,
                parent: parent
            });
            _this.groupedRowsSubscription.add(row.aggregateFunctionChange.subscribe(function (arg) {
                if (arg.changeAllGroupedRows) {
                    _this.activeAggregates[arg.columnAttr] = arg.activeAggregate;
                    _this.renderedData.filter(function (row) { return row instanceof OTableGroupedRow; }).forEach(function (row) {
                        row.setColumnActiveAggregateFunction(arg.columnAttr, arg.activeAggregate, false);
                        _this.recalculateColumnAggregate(arg.columnAttr, row);
                    });
                }
                else {
                    _this.recalculateColumnAggregate(arg.columnAttr, arg.row);
                }
            }));
            var affectedIndexes = recordHash[row.keysAsString];
            var groupData = data.filter(function (row, index) { return affectedIndexes.includes(index); });
            _this.table.visibleColArray.forEach(function (columnAttr, i) {
                if (i === 0) {
                    row.title = _this.getTextGroupRow(row, affectedIndexes.length);
                }
                var useColumnAggregate = _this.table.useColumnGroupingAggregate(columnAttr);
                if (useColumnAggregate) {
                    row.initializeColumnAggregate(columnAttr, _this.table.getColumnGroupingComponent(columnAttr));
                    if (Util.isDefined(_this.activeAggregates[columnAttr])) {
                        row.setColumnActiveAggregateFunction(columnAttr, _this.activeAggregates[columnAttr], false);
                    }
                    if (row.hasActiveAggregate(columnAttr)) {
                        var aggregateData = groupData.map(function (x) { var obj = {}; obj[columnAttr] = x[columnAttr]; return obj; });
                        row.setColumnAggregateData(columnAttr, aggregateData);
                        var aggregateConf = row.getActiveColumnAggregateConfiguration(columnAttr);
                        var value = _this.calculateAggregate(aggregateData, aggregateConf.attr, aggregateConf.aggregateFunction || aggregateConf.aggregate);
                        row.setColumnAggregateValue(columnAttr, value);
                    }
                }
            });
            row.expanded = _this.getExpansionState(row);
            var subGroup = _this.getSublevel(groupData, level + 1, row);
            subGroup.unshift(row);
            result = result.concat(subGroup);
        });
        return result;
    };
    DefaultOTableDataSource.prototype.filterCollapsedRowGroup = function (data) {
        var _this = this;
        return data.filter(function (row) { return (row instanceof OTableGroupedRow) ? row.visible : _this.belongsToExpandedGroupedRow(data, row); });
    };
    DefaultOTableDataSource.prototype.belongsToExpandedGroupedRow = function (data, row) {
        var parent;
        var match = false;
        for (var index = 0; index < data.length && !match; index++) {
            if (data[index] instanceof OTableGroupedRow) {
                parent = data[index];
            }
            else if (Util.isEquivalent(data[index], row)) {
                match = true;
            }
        }
        return Util.isDefined(parent) ? (parent.visible && parent.expanded) : true;
    };
    DefaultOTableDataSource.prototype.updateGroupedColumns = function () {
        this.groupByColumnChange.next();
    };
    DefaultOTableDataSource.prototype.toggleGroupByColumn = function (rowGroup) {
        if (rowGroup.expandSameLevel(this.table.expandGroupsSameLevel)) {
            this.levelsExpansionState[rowGroup.level] = !rowGroup.expanded;
        }
        else {
            this.updateStateRowGrouped(rowGroup);
        }
        this.groupByColumnChange.next();
    };
    DefaultOTableDataSource.prototype.setRowGroupLevelExpansion = function (rowGroup, value) {
        this.levelsExpansionState[rowGroup.level] = value;
        this.groupByColumnChange.next();
    };
    DefaultOTableDataSource.prototype.updateStateRowGrouped = function (rowGroup) {
        var stateRowGrouped = this.groupedRowState.find(function (row) { return rowGroup.keysAsString === row.keysAsString && JSON.stringify(rowGroup.parent) === JSON.stringify(row.parent); });
        if (Util.isDefined(stateRowGrouped)) {
            stateRowGrouped.expanded = !stateRowGrouped.expanded;
        }
        else {
            rowGroup.expanded = !rowGroup.expanded;
            this.groupedRowState.push(rowGroup);
        }
    };
    DefaultOTableDataSource.prototype.getTextGroupRow = function (group, totalCounts) {
        var _this = this;
        var field = this.table.groupedColumnsArray[group.level - 1];
        var value = JSON.parse(group.keysAsString)[this.table.groupedColumnsArray[group.level - 1]];
        var oCol = this.table.getOColumn(field);
        if (!value && Util.isDefined(oCol.renderer) && this.table.isInstanceOfOTableCellRendererServiceComponent(oCol.renderer)) {
            value = ' - ';
            if (!this.table.onDataLoadedCellRendererSubscription) {
                this.table.onDataLoadedCellRendererSubscription = oCol.renderer.onDataLoaded.subscribe(function (x) {
                    _this.updateGroupedColumns();
                });
            }
        }
        return this.table.translateService.get(oCol.title) + ': ' + value + ' (' + totalCounts + ')';
    };
    DefaultOTableDataSource.prototype.getExpansionState = function (row) {
        var expansionState = !parent || !this.table.collapseGroupedColumns;
        if (row.expandSameLevel(this.table.expandGroupsSameLevel)) {
            expansionState = this.levelsExpansionState.hasOwnProperty(row.level) ? this.levelsExpansionState[row.level] : expansionState;
        }
        else {
            var rowGroup = this.groupedRowState.find(function (x) { return x.keysAsString === row.keysAsString && JSON.stringify(x.parent) === JSON.stringify(row.parent); });
            expansionState = rowGroup ? rowGroup.expanded : expansionState;
        }
        return expansionState;
    };
    return DefaultOTableDataSource;
}(DataSource));
export { DefaultOTableDataSource };
