// ontimize-web-ngx-report/esm5/lib/components/report/o-report-new/o-report-new.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/report/o-report-new/o-report-new.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, ViewEncapsulation } from '@angular/core';
import { DialogService, OFileInputComponent, OFormComponent, OTextInputComponent } from 'ontimize-web-ngx';
import { OAlertService } from '../../../services/o-alert.service';
var OReportNewComponent = /** @class */ (function () {
    function OReportNewComponent(alertService, dialogService) {
        this.alertService = alertService;
        this.dialogService = dialogService;
        this.loading = false;
    }
    /**
     * @return {?}
     */
    OReportNewComponent.prototype.getFileData = /**
     * @return {?}
     */
    function () {
        return {
            'name': this.name.getValue(),
            'type': this.type.getValue(),
            'description': this.description.getValue(),
        };
    };
    /**
     * @return {?}
     */
    OReportNewComponent.prototype.onUploadFile = /**
     * @return {?}
     */
    function () {
        this.loading = false;
        this.confirm();
        this.form._clearAndCloseFormAfterInsert();
    };
    /**
     * @return {?}
     */
    OReportNewComponent.prototype.onError = /**
     * @return {?}
     */
    function () {
        if (this.dialogService) {
            this.dialogService.error('ERROR', 'SERVER_ERROR_MESSAGE');
            this.loading = false;
        }
    };
    /**
     * @return {?}
     */
    OReportNewComponent.prototype.confirm = /**
     * @return {?}
     */
    function () {
        this.alertService.announceAlert('alert');
    };
    /**
     * @param {?} e
     * @return {?}
     */
    OReportNewComponent.prototype.onClickSave = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var _this = this;
        Object.keys(this.form.formGroup.controls).forEach((/**
         * @param {?} control
         * @return {?}
         */
        function (control) {
            _this.form.formGroup.controls[control].markAsTouched();
        }));
        if (!this.form.formGroup.valid) {
            this.dialogService.alert('ERROR', 'MESSAGES.FORM_VALIDATION_ERROR');
            return;
        }
        this.loading = true;
        this.file.onClickUpload(e);
    };
    OReportNewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-report-new',
                    template: "<o-form #form service-type=\"report\" entity=\"addReport\" keys=\"UUID\" header-actions=\"R;I;U;D\" show-header-navigation=\"no\">\n  <div o-form-toolbar-buttons>\n    <button type=\"button\" class=\"o-form-toolbar-button\" mat-stroked-button (click)=\"onClickSave($event)\" attr=\"fillreport\">\n      <mat-icon>save</mat-icon>\n      <span>{{ 'SAVE'| oReportsTranslate}}</span>\n    </button>\n  </div>\n  <o-column attr=\"row1\" title=\"REPORT_DATA\" icon=\"description\" layout-gap=\"14px\" fxFill>\n    <div fxLayout=\"row wrap\" fxLayoutGap=\"14px\" fxLayoutAlign=\"start start\" fxFlex>\n      <o-text-input #name class=\"input-padding\" attr=\"NAME\" label=\"REPORT_NAME\" fxFlex=\"50\" required=\"yes\"></o-text-input>\n      <o-text-input #type class=\"input-padding\" attr=\"REPORT_TYPE\" label=\"REPORT_TYPE\" fxFlex required=\"yes\"></o-text-input>\n    </div>\n    <o-textarea-input #description attr=\"DESCRIPTION\" label=\"REPORT_DESCRIPTION\" fxFlex required=\"yes\"></o-textarea-input>\n    <o-file-input #file attr=\"file\" fxFlex=\"50\" service=\"report\" service-type=\"report\" entity=\"addReport\" label=\"REPORT_ZIPFOLDER\" show-info=\"yes\"\n      workspace-key=\"id\" required=\"yes\" (onUploadFile)=\"onUploadFile()\" accept-file-type=\".zip\" [additional-data]=\"getFileData()\"\n      (onErrorFile)=\"onError()\">\n    </o-file-input>\n  </o-column>\n</o-form>\n<div class=\"loading\" *ngIf=\"loading\">\n  <div class=\"spinner\">\n    <mat-spinner mode=\"indeterminate\"></mat-spinner>\n  </div>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.app-report-store-new]': 'true'
                    },
                    styles: [".o-form-toolbar-button.o-form-toolbar-button-insert{display:none}.loading{position:fixed;top:0;left:0;right:0;bottom:0;width:100%;height:100%;background-color:#fff;opacity:.5;z-index:1000}.spinner{position:absolute;top:50%;left:50%;margin-top:-20px;margin-left:-25px;width:50px;height:40px;text-align:center;font-size:10px;opacity:1}"]
                }] }
    ];
    /** @nocollapse */
    OReportNewComponent.ctorParameters = function () { return [
        { type: OAlertService },
        { type: DialogService }
    ]; };
    OReportNewComponent.propDecorators = {
        name: [{ type: ViewChild, args: ['name', { static: true },] }],
        type: [{ type: ViewChild, args: ['type', { static: true },] }],
        description: [{ type: ViewChild, args: ['description', { static: true },] }],
        form: [{ type: ViewChild, args: ['form', { static: true },] }],
        file: [{ type: ViewChild, args: ['file', { static: true },] }]
    };
    return OReportNewComponent;
}());
export { OReportNewComponent };
if (false) {
    /** @type {?} */
    OReportNewComponent.prototype.name;
    /** @type {?} */
    OReportNewComponent.prototype.type;
    /** @type {?} */
    OReportNewComponent.prototype.description;
    /** @type {?} */
    OReportNewComponent.prototype.form;
    /** @type {?} */
    OReportNewComponent.prototype.file;
    /** @type {?} */
    OReportNewComponent.prototype.subscription;
    /** @type {?} */
    OReportNewComponent.prototype.loading;
    /**
     * @type {?}
     * @private
     */
    OReportNewComponent.prototype.alertService;
    /**
     * @type {?}
     * @private
     */
    OReportNewComponent.prototype.dialogService;
}
