// ontimize-web-ngx/esm5/lib/components/table/extensions/dialog/o-table-base-dialog.class.js
import { QueryList, ViewChildren } from '@angular/core';
import { OMatErrorDirective } from '../../../../directives/o-mat-error.directive';
import { O_MAT_ERROR_OPTIONS } from '../../../../services/factories';
import { Codes } from '../../../../util/codes';
import { Util } from '../../../../util/util';
var OTableBaseDialogClass = (function () {
    function OTableBaseDialogClass(injector) {
        this.injector = injector;
        try {
            this.errorOptions = this.injector.get(O_MAT_ERROR_OPTIONS) || {};
        }
        catch (e) {
            this.errorOptions = {};
        }
    }
    OTableBaseDialogClass.prototype.setFormControl = function (formControl) {
        this.formControl = formControl;
    };
    Object.defineProperty(OTableBaseDialogClass.prototype, "tooltipClass", {
        get: function () {
            var result;
            var liteError = this.errorOptions.type === Codes.O_MAT_ERROR_LITE;
            if (liteError && Util.isDefined(this.formControl) && this.formControlHasErrors()) {
                result = "o-tooltip o-mat-error";
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableBaseDialogClass.prototype, "tooltipText", {
        get: function () {
            var result;
            var liteError = this.errorOptions.type === Codes.O_MAT_ERROR_LITE;
            if (liteError && this.formControlHasErrors() && this.oMatErrorChildren && this.oMatErrorChildren.length > 0) {
                result = this.oMatErrorChildren
                    .filter(function (oMatError) { return Util.isDefined(oMatError.text); })
                    .map(function (oMatError) { return oMatError.text; })
                    .join('\n');
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    OTableBaseDialogClass.prototype.formControlHasErrors = function () {
        return Util.isDefined(this.formControl) && this.formControl.touched && Util.isDefined(this.formControl.errors);
    };
    OTableBaseDialogClass.prototype.hasError = function (fControl, error) {
        return fControl && fControl.touched && fControl.hasError(error);
    };
    OTableBaseDialogClass.propDecorators = {
        oMatErrorChildren: [{ type: ViewChildren, args: [OMatErrorDirective,] }]
    };
    return OTableBaseDialogClass;
}());
export { OTableBaseDialogClass };
