// ontimize-web-ngx-report/esm2015/lib/components/report-on-demand/apply-configuration/apply-configuration-dialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/report-on-demand/apply-configuration/apply-configuration-dialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SelectionModel } from '@angular/cdk/collections';
import { Component, Inject, Optional, ViewChild } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef, MatSelectionList } from '@angular/material';
import { DialogService, Util } from 'ontimize-web-ngx';
import { OReportService } from '../../../services/o-report.service';
export class ApplyConfigurationDialogComponent {
    /**
     * @param {?} dialogService
     * @param {?} reportsService
     * @param {?} dialogo
     * @param {?} data
     */
    constructor(dialogService, reportsService, dialogo, data) {
        this.dialogService = dialogService;
        this.reportsService = reportsService;
        this.dialogo = dialogo;
        this.data = data;
        this.configurationListData = [];
        this.getConfigurations();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.configurationList.selectedOptions = new SelectionModel(false);
    }
    /**
     * @return {?}
     */
    loadConfiguration() {
        this.dialogo.close(this.selectedConfiguration);
    }
    /**
     * @return {?}
     */
    remove() {
        this.dialogService.confirm('CONFIRM', 'TABLE.DIALOG.CONFIRM_REMOVE_CONFIGURATION').then((/**
         * @param {?} result
         * @return {?}
         */
        result => {
            if (result) {
                this.reportsService.deletePreferences(this.selectedConfiguration.ID).subscribe((/**
                 * @param {?} res
                 * @return {?}
                 */
                res => {
                    if (res && res.code === 0) {
                        this.getConfigurations();
                    }
                }));
            }
        }));
    }
    /**
     * @param {?} option
     * @return {?}
     */
    onClickConfigurationList(option) {
        this.selectedConfiguration = option;
    }
    /**
     * @return {?}
     */
    getConfigurations() {
        this.reportsService.getPreferences(this.data.entity, this.data.service).subscribe((/**
         * @param {?} resp
         * @return {?}
         */
        resp => {
            if (resp.isSuccessful()) {
                this.setDataArray(resp.data);
            }
        }));
    }
    /**
     * @param {?} data
     * @return {?}
     */
    setDataArray(data) {
        if (Util.isArray(data)) {
            this.configurationListData = data;
        }
        else {
            this.configurationListData = [];
        }
    }
}
ApplyConfigurationDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'apply-configuration-dialog',
                template: "<span mat-dialog-title>{{ 'TABLE.BUTTONS.APPLY_CONFIGURATION' | oTranslate }}</span>\n<mat-dialog-content fxLayout=\"column\">\n  <div mat-subheader>{{ 'TABLE.DIALOG.APPLY_CONFIGURATION' | oTranslate }}</div>\n  <mat-selection-list #configurationList dense class=\"apply-configuration-dialog-list\">\n    <mat-list-option (click)=\"onClickConfigurationList(configuration)\" checkboxPosition=\"before\"\n      *ngFor=\"let configuration of configurationListData\" [value]=\"configuration\">\n      <span matLine>{{ configuration.NAME }}</span>\n      <span matLine>{{ configuration.DESCRIPTION }}</span>\n    </mat-list-option>\n  </mat-selection-list>\n</mat-dialog-content>\n\n<mat-dialog-actions align=\"end\">\n  <button type=\"button\" (click)=\"remove()\" mat-stroked-button [disabled]=\"configurationList.selectedOptions.selected.length!==1\"\n    class=\"o-button-danger\">\n    {{ 'DELETE' | oTranslate }}\n  </button>\n  <span fxFlex></span>\n  <button type=\"button\" mat-stroked-button [mat-dialog-close]=\"false\" class=\"o-button-default\">{{ 'CANCEL' | oTranslate }}</button>\n  <button type=\"button\" (click)=\"loadConfiguration()\" mat-stroked-button [mat-dialog-close]=\"true\"\n    [disabled]=\"configurationList.selectedOptions.selected.length!==1\" class=\"o-button-primary\">{{ 'OK' | oTranslate}}</button>\n</mat-dialog-actions>\n"
            }] }
];
/** @nocollapse */
ApplyConfigurationDialogComponent.ctorParameters = () => [
    { type: DialogService },
    { type: OReportService },
    { type: MatDialogRef },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DIALOG_DATA,] }] }
];
ApplyConfigurationDialogComponent.propDecorators = {
    configurationList: [{ type: ViewChild, args: [MatSelectionList, { static: true },] }]
};
if (false) {
    /**
     * @type {?}
     * @protected
     */
    ApplyConfigurationDialogComponent.prototype.configurationList;
    /** @type {?} */
    ApplyConfigurationDialogComponent.prototype.configurationListData;
    /** @type {?} */
    ApplyConfigurationDialogComponent.prototype.selectedConfiguration;
    /**
     * @type {?}
     * @private
     */
    ApplyConfigurationDialogComponent.prototype.dialogService;
    /**
     * @type {?}
     * @private
     */
    ApplyConfigurationDialogComponent.prototype.reportsService;
    /** @type {?} */
    ApplyConfigurationDialogComponent.prototype.dialogo;
    /** @type {?} */
    ApplyConfigurationDialogComponent.prototype.data;
}
