// ontimize-web-ngx/esm5/lib/components/table/column/cell-renderer/o-base-table-cell-renderer.class.js
import { FilterExpressionUtils } from '../../../../util/filter-expression.utils';
import { Util } from '../../../../util/util';
import { OTableColumnComponent } from '../o-table-column.component';
export var DEFAULT_INPUTS_O_BASE_TABLE_CELL_RENDERER = [
    'filterSource: filter-source',
    'filterFunction: filter-function'
];
var OBaseTableCellRenderer = (function () {
    function OBaseTableCellRenderer(injector) {
        this.injector = injector;
        this._filterSource = 'render';
        this.tableColumn = this.injector.get(OTableColumnComponent);
    }
    OBaseTableCellRenderer.prototype.ngOnInit = function () {
        this.initialize();
    };
    OBaseTableCellRenderer.prototype.initialize = function () {
    };
    OBaseTableCellRenderer.prototype.ngAfterContentInit = function () {
        if (typeof this.filterFunction !== 'function') {
            this.filterFunction = undefined;
        }
        this.registerRenderer();
    };
    Object.defineProperty(OBaseTableCellRenderer.prototype, "table", {
        get: function () {
            return this.tableColumn.table;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OBaseTableCellRenderer.prototype, "column", {
        get: function () {
            return this.tableColumn.attr;
        },
        enumerable: true,
        configurable: true
    });
    OBaseTableCellRenderer.prototype.registerRenderer = function () {
        this.tableColumn.registerRenderer(this);
        if (!Util.isDefined(this.type) && Util.isDefined(this.tableColumn.type)) {
            this.type = this.tableColumn.type;
        }
    };
    OBaseTableCellRenderer.prototype.getCellData = function (cellvalue, rowvalue) {
        var parsedValue;
        if (this.componentPipe && this.pipeArguments !== undefined && cellvalue !== undefined) {
            parsedValue = this.componentPipe.transform(cellvalue, this.pipeArguments);
        }
        else {
            parsedValue = cellvalue;
        }
        return parsedValue;
    };
    OBaseTableCellRenderer.prototype.getTooltip = function (cellValue, rowValue) {
        return this.getCellData(cellValue, rowValue);
    };
    Object.defineProperty(OBaseTableCellRenderer.prototype, "filterSource", {
        get: function () {
            return this._filterSource;
        },
        set: function (val) {
            var lowerVal = (val || '').toLowerCase();
            this._filterSource = (lowerVal === 'render' || lowerVal === 'data' || lowerVal === 'both') ? lowerVal : 'render';
        },
        enumerable: true,
        configurable: true
    });
    OBaseTableCellRenderer.prototype.getFilter = function (cellValue, rowValue) {
        var result;
        switch (this.filterSource) {
            case 'render':
                result = [this.getCellData(cellValue, rowValue)];
                break;
            case 'data':
                result = [cellValue];
                break;
            case 'both':
                result = [cellValue, this.getCellData(cellValue, rowValue)];
                break;
        }
        return result;
    };
    OBaseTableCellRenderer.prototype.getFilterExpression = function (quickFilter) {
        return FilterExpressionUtils.buildExpressionLike(this.tableColumn.attr, quickFilter);
    };
    return OBaseTableCellRenderer;
}());
export { OBaseTableCellRenderer };
