// ontimize-web-ngx/esm2015/lib/components/table/extensions/o-table.dao.js
import { BehaviorSubject, merge, of } from 'rxjs';
export class OTableDao {
    constructor(dataService, entity, methods) {
        this.dataService = dataService;
        this.entity = entity;
        this.methods = methods;
        this.usingStaticData = false;
        this._isLoadingResults = false;
        this.dataChange = new BehaviorSubject([]);
        this.sqlTypesChange = new BehaviorSubject({});
    }
    get data() { return this.dataChange.value; }
    get sqlTypes() { return this.sqlTypesChange.value; }
    getQuery(queryArgs) {
        this.isLoadingResults = true;
        return this.dataService[this.methods.query].apply(this.dataService, queryArgs);
    }
    removeQuery(filters, sqlTypes) {
        return merge(...filters.map((kv => this.dataService[this.methods.delete](kv, this.entity, sqlTypes))));
    }
    insertQuery(av, sqlTypes) {
        if (this.usingStaticData) {
            this.data.push(av);
            return of(this.data);
        }
        else {
            return this.dataService[this.methods.insert](av, this.entity, sqlTypes);
        }
    }
    updateQuery(kv, av, sqlTypes) {
        if (this.usingStaticData) {
            return of([]);
        }
        else {
            return this.dataService[this.methods.update](kv, av, this.entity, sqlTypes);
        }
    }
    setDataArray(data) {
        this.dataChange.next(data);
        this.isLoadingResults = false;
        return of(data);
    }
    setAsynchronousColumn(value, rowData) {
        let index = null;
        for (let i = 0; i < this.data.length; i++) {
            if (this.data[i] === rowData) {
                index = i;
                break;
            }
        }
        if (index !== null) {
            Object.assign(this.data[index], value);
        }
    }
    get isLoadingResults() {
        return this._isLoadingResults;
    }
    set isLoadingResults(val) {
        if (val) {
            this.cleanTimer();
            this.loadingTimer = setTimeout(() => {
                this._isLoadingResults = val;
            }, 500);
        }
        else {
            this.cleanTimer();
            this._isLoadingResults = val;
        }
    }
    cleanTimer() {
        if (this.loadingTimer) {
            clearTimeout(this.loadingTimer);
        }
    }
}
