// ontimize-web-ngx/esm2015/lib/components/list/list-item/o-list-item.component.js
import { ChangeDetectorRef, Component, ContentChild, ContentChildren, ElementRef, forwardRef, Inject, Injector, Optional, QueryList, Renderer2, ViewChild, ViewEncapsulation } from '@angular/core';
import { MatLine, MatListAvatarCssMatStyler, MatListItem } from '@angular/material';
import { Util } from '../../../util/util';
import { OListComponent } from '../o-list.component';
export class OListItemComponent {
    constructor(elRef, _renderer, _injector, cd, _list) {
        this.elRef = elRef;
        this._renderer = _renderer;
        this._injector = _injector;
        this.cd = cd;
        this._list = _list;
    }
    set _hasAvatar(avatar) {
        const listItemNativeEl = this.elRef.nativeElement.getElementsByTagName('mat-list-item');
        if (listItemNativeEl && listItemNativeEl.length === 1) {
            if ((avatar !== null && avatar !== undefined)) {
                this._renderer.addClass(listItemNativeEl[0], 'mat-list-avatar');
            }
            else {
                this._renderer.removeClass(listItemNativeEl[0], 'mat-list-avatar');
            }
        }
    }
    ngAfterContentInit() {
        const matLinesRef = this._lines;
        const ngAfterContentInitOriginal = this._innerListItem.ngAfterContentInit;
        this._innerListItem.ngAfterContentInit = function () {
            const emptyDiv = this._element.nativeElement.querySelector('.mat-list-text:empty');
            if (emptyDiv) {
                emptyDiv.remove();
            }
            this._lines = matLinesRef;
            ngAfterContentInitOriginal.apply(this);
        };
    }
    onDetailIconClicked(e) {
        if (Util.isDefined(e)) {
            e.stopPropagation();
        }
        this._list.viewDetail(this.modelData);
    }
    onEditIconClicked(e) {
        if (Util.isDefined(e)) {
            e.stopPropagation();
        }
        this._list.editDetail(this.modelData);
    }
    setItemData(data) {
        if (!Util.isDefined(this.modelData)) {
            this.modelData = data;
            this.cd.detectChanges();
        }
    }
    getItemData() {
        return this.modelData;
    }
    onCheckboxChange() {
        if (this._list.selectable && Util.isDefined(this.modelData)) {
            this._list.setSelected(this.modelData);
        }
    }
    onCheckboxClicked(event) {
        event.stopPropagation();
    }
    get isSelected() {
        return this._list.isItemSelected(this.modelData);
    }
}
OListItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-list-item',
                template: "<mat-list-item #innerListItem>\n  <div class=\"o-list-row-action o-list-item-selection-check\" *ngIf=\"_list && _list.selectable\">\n    <mat-checkbox [checked]=\"isSelected\" (change)=\"onCheckboxChange()\" (click)=\"onCheckboxClicked($event)\"></mat-checkbox>\n  </div>\n  <ng-content select=\"[o-list-item-avatar], [matListAvatar], [matListIcon]\"> </ng-content>\n  <div class=\"mat-list-text\">\n    <ng-content select=\"[matLine]\"></ng-content>\n  </div>\n  <ng-content></ng-content>\n  <div fxLayout=\"row\" class=\"row-buttons-container\">\n    <div class=\"o-list-row-action o-list-item-icon\" *ngIf=\"_list && _list.editButtonInRow\" (click)=\"onEditIconClicked($event)\">\n      <mat-icon class=\"material-icons\">{{ _list.editButtonInRowIcon }}</mat-icon>\n    </div>\n    <div class=\"o-list-row-action o-list-item-icon\" *ngIf=\"_list && _list.detailButtonInRow\" (click)=\"onDetailIconClicked($event)\">\n      <mat-icon class=\"material-icons\">{{ _list.detailButtonInRowIcon }}</mat-icon>\n    </div>\n  </div>\n</mat-list-item>",
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.o-list-item]': 'true'
                },
                styles: [".o-list-item{display:flex}.o-list-item .mat-list-item{width:100%}.o-list-item .mat-list-item .o-list-item-selection-check{padding-right:16px}.o-list-item .mat-list-item .row-buttons-container .o-list-row-action{height:24px;text-align:center;cursor:pointer}.mat-list[dense] .o-list-item .o-list-row-action .mat-checkbox-inner-container{height:16px;width:16px}"]
            }] }
];
OListItemComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: Injector },
    { type: ChangeDetectorRef },
    { type: OListComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(() => OListComponent),] }] }
];
OListItemComponent.propDecorators = {
    _lines: [{ type: ContentChildren, args: [MatLine,] }],
    _innerListItem: [{ type: ViewChild, args: ['innerListItem', { static: true },] }],
    _hasAvatar: [{ type: ContentChild, args: [MatListAvatarCssMatStyler, { static: false },] }]
};
