// ontimize-web-ngx/esm5/lib/components/input/date-input/o-date-input.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, forwardRef, Inject, Injector, Optional, ViewChild, ViewEncapsulation } from '@angular/core';
import { MediaObserver } from '@angular/flex-layout';
import { DateAdapter, MAT_DATE_LOCALE, MatDatepicker, MatDatepickerInput } from '@angular/material';
import moment from 'moment';
import { InputConverter } from '../../../decorators/input-converter';
import { MomentService } from '../../../services/moment.service';
import { OntimizeMomentDateAdapter } from '../../../shared/material/date/ontimize-moment-date-adapter';
import { SQLTypes } from '../../../util/sqltypes';
import { Util } from '../../../util/util';
import { OFormValue } from '../../form/o-form-value';
import { OFormComponent } from '../../form/o-form.component';
import { OFormDataComponent } from '../../o-form-data-component.class';
import { OValueChangeEvent } from '../../o-value-change-event.class';
import { DEFAULT_INPUTS_O_TEXT_INPUT, DEFAULT_OUTPUTS_O_TEXT_INPUT } from '../text-input/o-text-input.component';
export var DEFAULT_OUTPUTS_O_DATE_INPUT = tslib_1.__spread(DEFAULT_OUTPUTS_O_TEXT_INPUT);
export var DEFAULT_INPUTS_O_DATE_INPUT = tslib_1.__spread([
    'valueType: value-type',
    'oformat: format',
    'olocale: locale',
    'oStartView: start-view',
    'oMinDate: min',
    'oMaxDate: max',
    'oTouchUi: touch-ui',
    'oStartAt: start-at',
    'filterDate: filter-date',
    'textInputEnabled: text-input-enabled',
    'dateClass: date-class'
], DEFAULT_INPUTS_O_TEXT_INPUT);
var ODateInputComponent = (function (_super) {
    tslib_1.__extends(ODateInputComponent, _super);
    function ODateInputComponent(form, dateAdapter, elRef, injector) {
        var _this = _super.call(this, form, elRef, injector) || this;
        _this.textInputEnabled = true;
        _this._oformat = 'L';
        _this.updateLocaleOnChange = false;
        _this.oStartView = 'month';
        _this._valueType = 'timestamp';
        _this.momentDateAdapter = dateAdapter;
        _this._defaultSQLTypeKey = 'DATE';
        _this.momentSrv = _this.injector.get(MomentService);
        _this.media = _this.injector.get(MediaObserver);
        return _this;
    }
    Object.defineProperty(ODateInputComponent.prototype, "oMinDate", {
        set: function (value) {
            if (value) {
                var momentD = this.getValueAsMoment(value);
                if (Util.isDefined(momentD)) {
                    this.datepickerInput.min = momentD.toDate();
                    this.minDateString = momentD.format(this.oformat);
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ODateInputComponent.prototype, "oMaxDate", {
        set: function (value) {
            if (value) {
                var momentD = this.getValueAsMoment(value);
                if (Util.isDefined(momentD)) {
                    this.datepickerInput.max = momentD.toDate();
                    this.maxDateString = momentD.format(this.oformat);
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    ODateInputComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.initialize();
        if (!this.olocale) {
            this.updateLocaleOnChange = true;
            this.olocale = this.momentSrv.getLocale();
        }
        if (this.oformat) {
            this.momentDateAdapter.oFormat = this.oformat;
        }
        this.momentDateAdapter.setLocale(this.olocale);
        if (this.oStartView) {
            this.datepicker.startView = this.oStartView;
        }
        if (this.oStartAt) {
            this.datepicker.startAt = new Date(this.oStartAt);
        }
        if (this.updateLocaleOnChange) {
            this.onLanguageChangeSubscription = this.translateService.onLanguageChanged.subscribe(function () {
                _this.momentDateAdapter.setLocale(_this.translateService.getCurrentLang());
                _this.setValue(_this.getValue());
            });
        }
        this.subscribeToMediaChanges();
    };
    ODateInputComponent.prototype.subscribeToMediaChanges = function () {
        var _this = this;
        this.mediaSubscription = this.media.asObservable().subscribe(function (change) {
            if (['xs', 'sm'].indexOf(change[0].mqAlias) !== -1) {
                _this.touchUi = Util.isDefined(_this.oTouchUi) ? _this.oTouchUi : true;
            }
            if (['md', 'lg', 'xl'].indexOf(change[0].mqAlias) !== -1) {
                _this.touchUi = Util.isDefined(_this.oTouchUi) ? _this.oTouchUi : false;
            }
        });
    };
    ODateInputComponent.prototype.ngOnDestroy = function () {
        _super.prototype.ngOnDestroy.call(this);
        if (this.mediaSubscription) {
            this.mediaSubscription.unsubscribe();
        }
        if (this.onLanguageChangeSubscription) {
            this.onLanguageChangeSubscription.unsubscribe();
        }
    };
    ODateInputComponent.prototype.getValueAsDate = function () {
        return this.dateValue;
    };
    ODateInputComponent.prototype.getValue = function () {
        var timestampValue = _super.prototype.getValue.call(this);
        if (timestampValue && timestampValue instanceof Date) {
            timestampValue = timestampValue.getTime();
        }
        return timestampValue;
    };
    Object.defineProperty(ODateInputComponent.prototype, "showClearButton", {
        get: function () {
            return this.clearButton && !this.isReadOnly && this.enabled && this.matInputRef.nativeElement.value;
        },
        enumerable: true,
        configurable: true
    });
    ODateInputComponent.prototype.open = function () {
        if (!this.isReadOnly && this.enabled) {
            this.datepicker.open();
        }
    };
    ODateInputComponent.prototype.onChangeEvent = function (event) {
        var isValid = event.value && event.value.isValid && event.value.isValid();
        var val = isValid ? event.value.valueOf() : event.value;
        var parsedVal = Util.parseByValueType(val, this.valueType, this.oformat);
        this.setValue(parsedVal, {
            changeType: OValueChangeEvent.USER_CHANGE,
            emitEvent: false,
            emitModelToViewChange: false
        });
    };
    ODateInputComponent.prototype.onClickInput = function (e) {
        if (!this.textInputEnabled) {
            this.open();
        }
    };
    Object.defineProperty(ODateInputComponent.prototype, "filterDate", {
        get: function () {
            return this._filterDate;
        },
        set: function (val) {
            this._filterDate = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ODateInputComponent.prototype, "dateClass", {
        get: function () {
            return this._dateClass;
        },
        set: function (val) {
            this._dateClass = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ODateInputComponent.prototype, "oformat", {
        get: function () {
            return this._oformat;
        },
        set: function (val) {
            this._oformat = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ODateInputComponent.prototype, "minDateString", {
        get: function () {
            return this._minDateString;
        },
        set: function (val) {
            this._minDateString = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ODateInputComponent.prototype, "maxDateString", {
        get: function () {
            return this._maxDateString;
        },
        set: function (val) {
            this._maxDateString = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ODateInputComponent.prototype, "touchUi", {
        get: function () {
            return this.oTouchUi || false;
        },
        set: function (val) {
            this.oTouchUi = val;
            this.datepicker.touchUi = this.touchUi;
        },
        enumerable: true,
        configurable: true
    });
    ODateInputComponent.prototype.ensureODateValueType = function (val) {
        if (!Util.isDefined(val)) {
            return val;
        }
        var result = val;
        switch (this.valueType) {
            case 'string':
                if (typeof val === 'string') {
                    var m = moment(val, this.oformat);
                    if (m.isValid()) {
                        this.dateValue = new Date(m.valueOf());
                    }
                }
                else {
                    result = undefined;
                }
                break;
            case 'date':
                if ((val instanceof Date)) {
                    this.dateValue = val;
                }
                else {
                    result = undefined;
                }
                break;
            case 'timestamp':
                if (typeof val === 'number') {
                    this.dateValue = new Date(val);
                }
                else {
                    result = undefined;
                }
                break;
            case 'iso-8601':
                if (typeof val !== 'string') {
                    var acceptTimestamp = typeof val === 'number' && this.getSQLType() === SQLTypes.TIMESTAMP;
                    if (acceptTimestamp) {
                        this.dateValue = new Date(val);
                    }
                    else {
                        result = undefined;
                    }
                }
                else {
                    var m = moment(val);
                    if (m.isValid()) {
                        this.dateValue = new Date(m.valueOf());
                    }
                    else {
                        result = undefined;
                    }
                }
                break;
            default:
                break;
        }
        if (!Util.isDefined(result)) {
            console.warn("ODateInputComponent value (" + val + ") is not consistent with value-type (" + this.valueType + ")");
        }
        return result;
    };
    ODateInputComponent.prototype.setFormValue = function (val, options, setDirty) {
        if (setDirty === void 0) { setDirty = false; }
        var value = val;
        if (val instanceof OFormValue) {
            value = val.value;
        }
        this.ensureODateValueType(value);
        _super.prototype.setFormValue.call(this, value, options, setDirty);
    };
    Object.defineProperty(ODateInputComponent.prototype, "valueType", {
        get: function () {
            return this._valueType;
        },
        set: function (val) {
            this._valueType = Util.convertToODateValueType(val);
        },
        enumerable: true,
        configurable: true
    });
    ODateInputComponent.prototype.createFormControl = function (cfg, validators) {
        this._fControl = _super.prototype.createFormControl.call(this, cfg, validators);
        if (!this.isEmpty() && !this.form.isInInsertMode()) {
            this._fControl.markAsTouched();
        }
        return this._fControl;
    };
    ODateInputComponent.prototype.getValueAsMoment = function (val) {
        if (!Util.isDefined(val)) {
            return val;
        }
        var result;
        switch (true) {
            case this.valueType === 'string' && typeof val === 'string':
                result = moment(val, this.oformat);
                break;
            case this.valueType === 'date' && val instanceof Date:
            case this.valueType === 'timestamp' && typeof val === 'number':
            case this.valueType === 'iso-8601' && typeof val === 'string':
                result = moment(val);
                break;
            case this.valueType === 'iso-8601':
                if (typeof val !== 'string') {
                    var acceptTimestamp = typeof val === 'number' && this.getSQLType() === SQLTypes.TIMESTAMP;
                    if (acceptTimestamp) {
                        result = moment(val);
                    }
                }
                break;
            default:
                break;
        }
        return Util.isDefined(result) && result.isValid() ? result : undefined;
    };
    ODateInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-date-input',
                    template: "<div fxLayout=\"row\" fxLayoutAlign=\"space-between center\" [formGroup]=\"getFormGroup()\" [matTooltip]=\"tooltip\"\n  [matTooltipClass]=\"tooltipClass\" [matTooltipPosition]=\"tooltipPosition\" [matTooltipShowDelay]=\"tooltipShowDelay\"\n  [matTooltipHideDelay]=\"tooltipHideDelay\" (click)=\"onClickInput($event)\">\n  <mat-form-field [appearance]=\"appearance\" [floatLabel]=\"floatLabel\" [hideRequiredMarker]=\"hideRequiredMarker\"\n    [class.custom-width]=\"hasCustomWidth\" class=\"icon-field\" fxFlexFill>\n    <mat-label *ngIf=\"labelVisible\">{{ olabel | oTranslate }}</mat-label>\n    <input matInput #matInputRef [matDatepicker]=\"picker\" [id]=\"getAttribute()\" [formControlName]=\"getAttribute()\"\n      [placeholder]=\"placeHolder\" [readonly]=\"isReadOnly || !textInputEnabled\" (focus)=\"innerOnFocus($event)\"\n      (blur)=\"innerOnBlur($event)\" [matDatepickerFilter]=\"filterDate\" (dateChange)=\"onChangeEvent($event)\"\n      [required]=\"isRequired\" [value]=\"getValueAsDate()\">\n\n    <button type=\"button\" *ngIf=\"showClearButton\" matSuffix mat-icon-button (click)=\"onClickClearValue($event)\">\n      <mat-icon svgIcon=\"ontimize:close\"></mat-icon>\n    </button>\n    <mat-datepicker-toggle matSuffix [disabled]=\"isReadOnly || !enabled\" [class.read-only]=\"isReadOnly\" [for]=\"picker\">\n      <mat-icon matDatepickerToggleIcon>today</mat-icon>\n    </mat-datepicker-toggle>\n    <mat-datepicker #picker [disabled]=\"isReadOnly || !enabled\" [dateClass]=\"dateClass\"></mat-datepicker>\n    <mat-error *oMatError=\"hasError('required')\">\n      {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n    </mat-error>\n    <mat-error *oMatError=\"hasError('matDatepickerParse')\">\n      {{ 'FORM_VALIDATION.DATE_PARSE' | oTranslate }} {{ oformat }}\n    </mat-error>\n    <mat-error *oMatError=\"hasError('matDatepickerFilter')\">\n      {{ 'FORM_VALIDATION.DATE_FILTER' | oTranslate }}\n    </mat-error>\n    <mat-error *oMatError=\"hasError('matDatepickerMin')\">\n      {{ 'FORM_VALIDATION.DATE_MIN' | oTranslate }} {{ minDateString }}\n    </mat-error>\n    <mat-error *oMatError=\"hasError('matDatepickerMax')\">\n      {{ 'FORM_VALIDATION.DATE_MAX' | oTranslate }} {{ maxDateString }}\n    </mat-error>\n    <mat-error *ngFor=\"let oError of getActiveOErrors()\">\n      {{ oError.text | oTranslate }}\n    </mat-error>\n  </mat-form-field>\n</div>",
                    outputs: DEFAULT_OUTPUTS_O_DATE_INPUT,
                    inputs: DEFAULT_INPUTS_O_DATE_INPUT,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        { provide: DateAdapter, useClass: OntimizeMomentDateAdapter, deps: [MAT_DATE_LOCALE] }
                    ]
                }] }
    ];
    ODateInputComponent.ctorParameters = function () { return [
        { type: OFormComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OFormComponent; }),] }] },
        { type: DateAdapter },
        { type: ElementRef },
        { type: Injector }
    ]; };
    ODateInputComponent.propDecorators = {
        datepicker: [{ type: ViewChild, args: ['picker', { static: true },] }],
        datepickerInput: [{ type: ViewChild, args: [MatDatepickerInput, { static: true },] }],
        matInputRef: [{ type: ViewChild, args: ['matInputRef', { read: ElementRef, static: true },] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], ODateInputComponent.prototype, "textInputEnabled", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], ODateInputComponent.prototype, "oTouchUi", void 0);
    return ODateInputComponent;
}(OFormDataComponent));
export { ODateInputComponent };
