// ontimize-web-ngx/esm2015/lib/components/form/services/o-form-message.service.js
import { Injectable } from '@angular/core';
export class OFormMessageService {
    getQueryErrorMessage() {
        return 'MESSAGES.ERROR_QUERY';
    }
    getUpdateSuccessMessage() {
        return 'MESSAGES.SAVED';
    }
    getUpdateErrorMessage() {
        return 'MESSAGES.ERROR_UPDATE';
    }
    getDeleteSuccessMessage() {
        return 'MESSAGES.DELETED';
    }
    getDeleteErrorMessage() {
        return 'MESSAGES.ERROR_DELETE';
    }
    getDeleteConfirmationMessage() {
        return 'MESSAGES.CONFIRM_DELETE';
    }
    getDeleteConfirmationDialogTitle() {
        return 'CONFIRM';
    }
    getInsertSuccessMessage() {
        return 'MESSAGES.INSERTED';
    }
    getInsertErrorMessage() {
        return 'MESSAGES.ERROR_INSERT';
    }
    getValidationError() {
        return 'MESSAGES.FORM_VALIDATION_ERROR';
    }
    getValidationErrorDialogTitle() {
        return 'ERROR';
    }
    getNothingToUpdateMessage() {
        return 'MESSAGES.FORM_NOTHING_TO_UPDATE_INFO';
    }
    getActionPermissionNotEnabledMessage() {
        return 'MESSAGES.OPERATION_NOT_ALLOWED_PERMISSION';
    }
    getDiscardChangesConfirmationMessage() {
        return 'MESSAGES.FORM_CHANGES_WILL_BE_LOST';
    }
    getDiscardChangesConfirmationDialogTitle() {
        return 'CONFIRM';
    }
}
OFormMessageService.decorators = [
    { type: Injectable }
];
