// ontimize-web-ngx/esm5/lib/layouts/app-layout/o-app-layout.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, EventEmitter, ViewChild, ViewEncapsulation } from '@angular/core';
import { OAppHeaderComponent } from '../../components/app-header/o-app-header.component';
import { OAppSidenavComponent } from '../../components/app-sidenav/o-app-sidenav.component';
import { OUserInfoConfigurationDirective } from '../../components/user-info/user-info-configuration/o-user-info-configuration.directive';
import { InputConverter } from '../../decorators/input-converter';
import { Codes } from '../../util/codes';
import { Util } from '../../util/util';
export var DEFAULT_INPUTS_O_APP_LAYOUT = [
    'mode',
    'sidenavMode: sidenav-mode',
    'sidenavOpened: sidenav-opened',
    '_showHeader: show-header',
    'showUserInfo: show-user-info',
    'showLanguageSelector: show-language-selector',
    'useFlagIcons: use-flag-icons',
    'openedSidenavImg: opened-sidenav-image',
    'closedSidenavImg: closed-sidenav-image',
    'headerColor: header-color',
    'headerHeight: header-height',
    'showTitle: show-title',
    'staticTitle: static-title',
    'showStaticTitle: show-static-title'
];
export var DEFAULT_OUTPUTS_O_APP_LAYOUT = [
    'beforeOpenSidenav',
    'afterOpenSidenav',
    'beforeCloseSidenav',
    'afterCloseSidenav'
];
var OAppLayoutComponent = (function () {
    function OAppLayoutComponent() {
        this.sidenavOpened = true;
        this.showUserInfo = true;
        this.showLanguageSelector = true;
        this.useFlagIcons = false;
        this._showHeader = true;
        this.showTitle = false;
        this.showStaticTitle = false;
        this.headerHeight = Codes.DEFAULT_ROW_HEIGHT;
        this.beforeOpenSidenav = new EventEmitter();
        this.afterOpenSidenav = new EventEmitter();
        this.beforeCloseSidenav = new EventEmitter();
        this.afterCloseSidenav = new EventEmitter();
    }
    Object.defineProperty(OAppLayoutComponent.prototype, "showHeader", {
        get: function () {
            return this._showHeader;
        },
        set: function (val) {
            this._showHeader = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OAppLayoutComponent.prototype, "mode", {
        get: function () {
            return this._mode;
        },
        set: function (val) {
            var m = Codes.OAppLayoutModes.find(function (e) { return e === val; });
            if (Util.isDefined(m)) {
                this._mode = m;
                if (this._mode === 'mobile' && !Util.isDefined(this.showHeader)) {
                    this.showHeader = true;
                }
            }
            else {
                console.error('Invalid `o-app-layout` mode (' + val + ')');
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OAppLayoutComponent.prototype, "sidenavMode", {
        get: function () {
            return this._sidenavMode;
        },
        set: function (val) {
            var m = Codes.OSidenavModes.find(function (e) { return e === val; });
            if (Util.isDefined(m)) {
                this._sidenavMode = m;
            }
            else {
                console.error('Invalid `o-app-layout` sidenav-mode (' + val + ')');
            }
        },
        enumerable: true,
        configurable: true
    });
    OAppLayoutComponent.prototype.sidenavToggle = function (opened) {
        opened ? this.beforeOpenSidenav.emit() : this.beforeCloseSidenav.emit();
    };
    OAppLayoutComponent.prototype.afterToggle = function (opened) {
        opened ? this.afterOpenSidenav.emit() : this.afterCloseSidenav.emit();
    };
    OAppLayoutComponent.prototype.ngAfterViewInit = function () {
        if (this.appHeader && this.appHeader.userInfo) {
            this.appHeader.userInfo.registerUserInfoConfiguration(this.userInfoConfiguration);
        }
    };
    OAppLayoutComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-app-layout',
                    inputs: DEFAULT_INPUTS_O_APP_LAYOUT,
                    outputs: DEFAULT_OUTPUTS_O_APP_LAYOUT,
                    template: "<o-app-sidenav #appSidenav [sidenav-mode]=\"sidenavMode\" [opened]=\"sidenavOpened\" [show-user-info]=\"showUserInfo\" [show-toggle-button]=\"!showHeader\"\n  [opened-sidenav-image]=\"openedSidenavImg\" [closed-sidenav-image]=\"closedSidenavImg\" [layout-mode]=\"mode\" [class.header-layout]=\"showHeader\"\n  (onSidenavToggle)=\"sidenavToggle($event)\" (afterSidenavToggle)=\"afterToggle($event)\">\n\n  <ng-content select=\"o-app-layout-sidenav[position=start]\" ngProjectAs=\"o-app-layout-sidenav-projection-start\">\n  </ng-content>\n  <ng-content select=\"o-app-layout-sidenav[position=end]\" ngProjectAs=\"o-app-layout-sidenav-projection-end\">\n  </ng-content>\n  <ng-content select=\"o-app-layout-sidenav\" ngProjectAs=\"o-app-layout-sidenav-projection-end\"></ng-content>\n\n  <o-app-header #appHeader *ngIf=\"showHeader\" [show-title]=\"showTitle\" [show-static-title]=\"showStaticTitle\" [static-title]=\"staticTitle\"\n    [show-user-info]=\"showUserInfo\" [show-language-selector]=\"showLanguageSelector\" [use-flag-icons]=\"useFlagIcons\"\n    (onSidenavToggle)=\"appSidenav.toggleSidenav()\" [color]=\"headerColor\" [header-height]=\"headerHeight\">\n    <mat-icon class=\"sidenav-toggle\" svgIcon=\"ontimize:menu\" (click)=\"appSidenav.toggleSidenav()\" *ngIf=\"appSidenav.isMobileMode()\"></mat-icon>\n    <ng-content select=\"o-app-layout-header[position=start]\" ngProjectAs=\"o-app-layout-header-projection-start\">\n    </ng-content>\n    <ng-content select=\"o-app-layout-header[position=end]\" ngProjectAs=\"o-app-layout-header-projection-end\">\n    </ng-content>\n    <ng-content select=\"o-app-layout-header\" ngProjectAs=\"o-app-layout-header-projection-start\"></ng-content>\n  </o-app-header>\n\n  <div class=\"application-layout-content-wrapper\" [class.header-layout]=\"showHeader\">\n    <ng-content></ng-content>\n  </div>\n</o-app-sidenav>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".application-layout-content-wrapper{z-index:1;position:absolute;top:0;bottom:0;left:0;right:0;overflow:auto;padding:16px}"]
                }] }
    ];
    OAppLayoutComponent.propDecorators = {
        appSidenav: [{ type: ViewChild, args: ['appSidenav', { static: false },] }],
        appHeader: [{ type: ViewChild, args: ['appHeader', { static: false },] }],
        userInfoConfiguration: [{ type: ContentChild, args: [OUserInfoConfigurationDirective, { static: false },] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "sidenavOpened", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "showUserInfo", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "showLanguageSelector", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "useFlagIcons", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "_showHeader", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "showTitle", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", String)
    ], OAppLayoutComponent.prototype, "staticTitle", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "showStaticTitle", void 0);
    return OAppLayoutComponent;
}());
export { OAppLayoutComponent };
