// ontimize-web-ngx/esm5/lib/components/button/o-button.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, ViewEncapsulation } from '@angular/core';
import { InputConverter } from '../../decorators/input-converter';
import { Codes } from '../../util';
export var DEFAULT_INPUTS_O_BUTTON = [
    'oattr: attr',
    'olabel: label',
    'otype: type',
    'icon',
    'svgIcon : svg-icon',
    'iconPosition: icon-position',
    'image',
    'enabled',
    'color'
];
export var DEFAULT_OUTPUTS_O_BUTTON = [
    'onClick',
    'click'
];
var OButtonComponent = (function () {
    function OButtonComponent() {
        this.iconPosition = Codes.ICON_POSITION_LEFT;
        this.enabled = true;
        this.onClick = new EventEmitter();
        this.click = new EventEmitter();
        this.otype = OButtonComponent.DEFAULT_TYPE;
    }
    OButtonComponent.prototype.ngOnInit = function () {
        if (this.otype) {
            this.otype = this.otype.toUpperCase();
        }
    };
    OButtonComponent.prototype.onButtonClick = function (event) {
        event.stopPropagation();
        if (this.enabled) {
            this.click.emit(event);
            this.onClick.emit(event);
        }
    };
    Object.defineProperty(OButtonComponent.prototype, "needsIconButtonClass", {
        get: function () {
            return (this.icon !== undefined || this.svgIcon !== undefined) && (this.olabel === undefined || this.olabel === '');
        },
        enumerable: true,
        configurable: true
    });
    OButtonComponent.prototype.isFab = function () {
        return this.otype === 'FAB';
    };
    OButtonComponent.prototype.isRaised = function () {
        return this.otype === 'RAISED';
    };
    OButtonComponent.prototype.isFlat = function () {
        return this.otype === 'FLAT';
    };
    OButtonComponent.prototype.isStroked = function () {
        return (this.otype === 'STROKED' || !this.otype);
    };
    OButtonComponent.prototype.isBasic = function () {
        return this.otype === 'BASIC';
    };
    OButtonComponent.prototype.isMiniFab = function () {
        return this.otype === 'FAB-MINI';
    };
    OButtonComponent.prototype.isIconButton = function () {
        return this.otype === 'ICON';
    };
    OButtonComponent.DEFAULT_TYPE = 'STROKED';
    OButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-button',
                    inputs: DEFAULT_INPUTS_O_BUTTON,
                    outputs: DEFAULT_OUTPUTS_O_BUTTON,
                    template: "<button type=\"button\" (click)=\"onButtonClick($event)\" *ngIf=\"isBasic()\" mat-button [class.mat-icon-button]=\"needsIconButtonClass\"\n  [disabled]=\"!enabled\" [color]=\"color\">\n  <ng-container *ngTemplateOutlet=\"buttonTemplate\"></ng-container>\n</button>\n\n<button type=\"button\" (click)=\"onButtonClick($event)\" *ngIf=\"isRaised()\" mat-raised-button class=\"mat-raised-button\"\n  [class.mat-icon-button]=\"needsIconButtonClass\" [disabled]=\"!enabled\" [color]=\"color\">\n  <ng-container *ngTemplateOutlet=\"buttonTemplate\"></ng-container>\n</button>\n\n<button type=\"button\" (click)=\"onButtonClick($event)\" *ngIf=\"isStroked()\" mat-stroked-button [class.mat-icon-button]=\"needsIconButtonClass\"\n  [disabled]=\"!enabled\" [color]=\"color\">\n  <ng-container *ngTemplateOutlet=\"buttonTemplate\"></ng-container>\n</button>\n\n<button type=\"button\" (click)=\"onButtonClick($event)\" *ngIf=\"isFlat()\" mat-flat-button [class.mat-icon-button]=\"needsIconButtonClass\"\n  [disabled]=\"!enabled\" [color]=\"color\">\n  <ng-container *ngTemplateOutlet=\"buttonTemplate\"></ng-container>\n</button>\n\n<button type=\"button\" (click)=\"onButtonClick($event)\" *ngIf=\"isIconButton()\" mat-icon-button [class.mat-icon-button]=\"needsIconButtonClass\"\n  [disabled]=\"!enabled\" [color]=\"color\">\n  <ng-container *ngTemplateOutlet=\"buttonTemplate\"></ng-container>\n</button>\n\n<button type=\"button\" (click)=\"onButtonClick($event)\" *ngIf=\"isFab()\" mat-fab [disabled]=\"!enabled\" [color]=\"color\">\n  <ng-container *ngTemplateOutlet=\"buttonTemplate\"></ng-container>\n</button>\n\n<button type=\"button\" (click)=\"onButtonClick($event)\" *ngIf=\"isMiniFab()\" mat-mini-fab [disabled]=\"!enabled\" [color]=\"color\">\n  <ng-container *ngTemplateOutlet=\"buttonTemplate\"></ng-container>\n</button>\n\n\n<ng-template #iconTemplate>\n  <img src=\"{{ image }}\" class=\"o-button-image\" *ngIf=\"image!=undefined\" alt=\"button image\" />\n  <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}\n  </mat-icon>\n  <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\" [svgIcon]=\"svgIcon\"></mat-icon>\n</ng-template>\n\n<ng-template #buttonTemplate>\n\n  <ng-container *ngIf=\"iconPosition==='left' || iconPosition==='top' \">\n    <ng-container *ngTemplateOutlet=\"iconTemplate\"></ng-container>\n  </ng-container>\n\n  <br *ngIf=\"iconPosition==='top'\" />\n\n  {{ olabel | oTranslate }}\n\n  <br *ngIf=\"iconPosition==='bottom'\" />\n\n  <ng-container *ngIf=\"iconPosition==='right' || iconPosition==='bottom'\">\n    <ng-container *ngTemplateOutlet=\"iconTemplate\"></ng-container>\n  </ng-container>\n</ng-template>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-button]': 'true',
                        '[class.o-button-icon-position-top]': 'iconPosition==="top"',
                        '[class.o-button-icon-position-bottom]': 'iconPosition==="bottom"'
                    },
                    styles: [".o-button{display:inline-block;pointer-events:none}.o-button.o-button-icon-position-bottom .mat-button,.o-button.o-button-icon-position-bottom .mat-flat-button,.o-button.o-button-icon-position-bottom .mat-raised-button,.o-button.o-button-icon-position-bottom .mat-stroked-button,.o-button.o-button-icon-position-top .mat-button,.o-button.o-button-icon-position-top .mat-flat-button,.o-button.o-button-icon-position-top .mat-raised-button,.o-button.o-button-icon-position-top .mat-stroked-button{height:auto;line-height:normal}.o-button button{pointer-events:auto}.o-button button.mat-button,.o-button button.mat-flat-button,.o-button button.mat-raised-button,.o-button button.mat-stroked-button{width:100%}.o-button button *{pointer-events:none}.o-button button .o-button-image{max-height:24px}.o-button .mat-icon-button .mat-button-ripple.mat-ripple{border-radius:50%}"]
                }] }
    ];
    OButtonComponent.ctorParameters = function () { return []; };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OButtonComponent.prototype, "enabled", void 0);
    return OButtonComponent;
}());
export { OButtonComponent };
