// ontimize-web-ngx/esm5/lib/directives/o-file-drag-and-drop.directive.js
import { Directive, EventEmitter, HostBinding, HostListener, Output } from '@angular/core';
var OFileDragAndDropDirective = (function () {
    function OFileDragAndDropDirective() {
        this.onFileDropped = new EventEmitter();
    }
    OFileDragAndDropDirective.prototype.onDragOver = function (evt) {
        evt.preventDefault();
        evt.stopPropagation();
        this.fileDragging = true;
    };
    OFileDragAndDropDirective.prototype.onDragLeave = function (evt) {
        evt.preventDefault();
        evt.stopPropagation();
        this.fileDragging = false;
    };
    OFileDragAndDropDirective.prototype.onDrop = function (evt) {
        evt.preventDefault();
        evt.stopPropagation();
        this.fileDragging = false;
        var files = evt.dataTransfer.files;
        if (files.length > 0) {
            this.onFileDropped.emit(files);
        }
    };
    OFileDragAndDropDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[oFileDragAndDrop]'
                },] }
    ];
    OFileDragAndDropDirective.propDecorators = {
        onFileDropped: [{ type: Output }],
        fileDragging: [{ type: HostBinding, args: ['class.o-file-dragging',] }],
        onDragOver: [{ type: HostListener, args: ['dragover', ['$event'],] }],
        onDragLeave: [{ type: HostListener, args: ['dragleave', ['$event'],] }],
        onDrop: [{ type: HostListener, args: ['drop', ['$event'],] }]
    };
    return OFileDragAndDropDirective;
}());
export { OFileDragAndDropDirective };
