// ontimize-web-ngx/esm5/lib/directives/o-mat-error.directive.js
import { Directive, ElementRef, Injector, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { O_MAT_ERROR_OPTIONS } from '../services/factories';
import { Codes } from '../util/codes';
import { Util } from '../util/util';
var OMatErrorDirective = (function () {
    function OMatErrorDirective(element, templateRef, viewContainer, injector) {
        this.element = element;
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.injector = injector;
        try {
            this.errorOptions = this.injector.get(O_MAT_ERROR_OPTIONS) || {};
        }
        catch (e) {
            this.errorOptions = {};
        }
        if (!Util.isDefined(this.errorOptions.type)) {
            this.errorOptions.type = Codes.O_MAT_ERROR_STANDARD;
        }
    }
    Object.defineProperty(OMatErrorDirective.prototype, "oMatError", {
        set: function (val) {
            var _this = this;
            if (val) {
                this.viewContainer.createEmbeddedView(this.templateRef);
                if (this.errorOptions.type === Codes.O_MAT_ERROR_LITE) {
                    setTimeout(function () {
                        try {
                            _this.text = _this.element.nativeElement.parentElement.getElementsByTagName('mat-error')[0].textContent;
                        }
                        catch (e) {
                            _this.text = undefined;
                        }
                        _this.viewContainer.clear();
                    }, 0);
                }
            }
            else {
                this.text = undefined;
                this.viewContainer.clear();
            }
        },
        enumerable: true,
        configurable: true
    });
    OMatErrorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[oMatError]'
                },] }
    ];
    OMatErrorDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: TemplateRef },
        { type: ViewContainerRef },
        { type: Injector }
    ]; };
    OMatErrorDirective.propDecorators = {
        oMatError: [{ type: Input }]
    };
    return OMatErrorDirective;
}());
export { OMatErrorDirective };
