// ontimize-web-ngx/esm2015/lib/components/table/extensions/header/table-option/o-table-option.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, Injector, ViewEncapsulation } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
export const DEFAULT_INPUTS_O_TABLE_OPTION = [
    'oattr: attr',
    'enabled',
    'icon',
    'showCheckboxOption : show-checkbox-option',
    'olabel: label',
    'active'
];
export const DEFAULT_OUTPUTS_O_TABLE_OPTION = [
    'onClick'
];
export class OTableOptionComponent {
    constructor(injector, elRef) {
        this.injector = injector;
        this.elRef = elRef;
        this.onClick = new EventEmitter();
        this.enabled = true;
        this.showCheckboxOption = false;
        this.active = false;
        try {
            this.cd = this.injector.get(ChangeDetectorRef);
        }
        catch (e) {
        }
    }
    innerOnClick() {
        this.onClick.emit();
        this.setActive(!this.active);
    }
    get activeCheckboxOption() {
        return this.showCheckboxOption && this.active;
    }
    setActive(val) {
        this.active = val;
        this.cd.detectChanges();
    }
}
OTableOptionComponent.O_TABLE_OPTION_ACTIVE_CLASS = 'o-table-option-active';
OTableOptionComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-option',
                template: "<ng-container *ngIf=\"showCheckboxOption; then checkboxTemplate; else buttonTemplate\"></ng-container>\n\n<ng-template #checkboxTemplate>\n  <span mat-menu-item>\n    <mat-checkbox [disabled]=\"!enabled\" [ngModel]=\"activeCheckboxOption\" (change)=\"innerOnClick()\" disableRipple=\"true\"> \n      {{ olabel | oTranslate }}\n      <mat-icon *ngIf=\"icon\"> {{ icon }}</mat-icon>\n    </mat-checkbox>\n  </span>\n</ng-template>\n\n<ng-template #buttonTemplate>\n  <button [disabled]=\"!enabled\" type=\"button\" (click)=\"innerOnClick()\" mat-menu-item fxLayout=\"row\"\n    fxLayoutAlign=\"start center\">\n    {{ olabel | oTranslate }}\n    <mat-icon *ngIf=\"icon\"> {{ icon }}</mat-icon>\n  </button>\n</ng-template>",
                inputs: DEFAULT_INPUTS_O_TABLE_OPTION,
                outputs: DEFAULT_OUTPUTS_O_TABLE_OPTION,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[class.o-table-option]': 'true',
                },
                styles: [""]
            }] }
];
OTableOptionComponent.ctorParameters = () => [
    { type: Injector },
    { type: ElementRef }
];
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTableOptionComponent.prototype, "enabled", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTableOptionComponent.prototype, "showCheckboxOption", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTableOptionComponent.prototype, "active", void 0);
