// ontimize-web-ngx/esm2015/lib/components/input/combo/combo-renderer/percentage/o-combo-renderer-percentage.component.js
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { OPercentPipe } from '../../../../../pipes/o-percentage.pipe';
import { NumberService } from '../../../../../services/number.service';
import { DEFAULT_INPUTS_O_COMBO_RENDERER_REAL, OComboRendererRealComponent } from '../real/o-combo-renderer-real.component';
export const DEFAULT_INPUTS_O_COMBO_RENDERER_PERCENTAGE = [
    ...DEFAULT_INPUTS_O_COMBO_RENDERER_REAL,
    'valueBase: value-base'
];
export class OComboRendererPercentageComponent extends OComboRendererRealComponent {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.decimalSeparator = '.';
        this.minDecimalDigits = 0;
        this.maxDecimalDigits = 0;
        this.valueBase = 1;
        this.numberService = this.injector.get(NumberService);
        this.setComponentPipe();
    }
    setComponentPipe() {
        this.componentPipe = new OPercentPipe(this.injector);
    }
    initialize() {
        super.initialize();
        this.pipeArguments = {
            minDecimalDigits: this.minDecimalDigits,
            maxDecimalDigits: this.maxDecimalDigits,
            decimalSeparator: this.decimalSeparator,
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator,
            valueBase: this.valueBase
        };
    }
}
OComboRendererPercentageComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-combo-renderer-percentage',
                template: "<ng-template #templateref let-value=\"value\">\n  {{getComboData(value)}}\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_COMBO_RENDERER_PERCENTAGE
            }] }
];
OComboRendererPercentageComponent.ctorParameters = () => [
    { type: Injector }
];
OComboRendererPercentageComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
};
