// ontimize-web-ngx/esm5/lib/layouts/form-layout/dialog/options/o-form-layout-dialog-options.directive.js
import { Directive, Input } from '@angular/core';
import { BooleanConverter } from '../../../../decorators/input-converter';
import { OFormLayoutManagerComponent } from '../../o-form-layout-manager.component';
var OFormLayoutDialogOptionsDirective = (function () {
    function OFormLayoutDialogOptionsDirective(formLayoutManager) {
        this.formLayoutManager = formLayoutManager;
        this.width = '';
        this.height = '';
        this.class = '';
        this._closeOnNavigation = true;
        this._disableClose = true;
    }
    OFormLayoutDialogOptionsDirective.prototype.ngOnChanges = function () {
        if (this.formLayoutManager) {
            this.formLayoutManager.addDialogOptions(this.getOptions());
        }
    };
    Object.defineProperty(OFormLayoutDialogOptionsDirective.prototype, "closeOnNavigation", {
        set: function (value) {
            this._closeOnNavigation = BooleanConverter(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OFormLayoutDialogOptionsDirective.prototype, "disableClose", {
        set: function (value) {
            this._disableClose = BooleanConverter(value);
        },
        enumerable: true,
        configurable: true
    });
    OFormLayoutDialogOptionsDirective.prototype.getOptions = function () {
        var result = {
            width: this.width,
            minWidth: this.minWidth,
            maxWidth: this.maxWidth,
            height: this.height,
            minHeight: this.minHeight,
            maxHeight: this.maxHeight,
            class: this.class,
            position: this.position,
            backdropClass: this.backdropClass,
            disableClose: this._disableClose,
            closeOnNavigation: this._closeOnNavigation,
            title: this.title,
            labelColumns: this.labelColumns,
            separator: this.separator
        };
        Object.keys(result).forEach(function (key) { return result[key] == null ? delete result[key] : {}; });
        return result;
    };
    OFormLayoutDialogOptionsDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'o-form-layout-dialog-options, o-form-layout-manager[mode="dialog"]'
                },] }
    ];
    OFormLayoutDialogOptionsDirective.ctorParameters = function () { return [
        { type: OFormLayoutManagerComponent }
    ]; };
    OFormLayoutDialogOptionsDirective.propDecorators = {
        width: [{ type: Input }],
        minWidth: [{ type: Input, args: ['min-width',] }],
        maxWidth: [{ type: Input, args: ['max-width',] }],
        height: [{ type: Input }],
        minHeight: [{ type: Input, args: ['min-height',] }],
        maxHeight: [{ type: Input, args: ['max-height',] }],
        class: [{ type: Input }],
        position: [{ type: Input }],
        backdropClass: [{ type: Input, args: ['backdrop-class',] }],
        closeOnNavigation: [{ type: Input, args: ['close-on-navigation',] }],
        disableClose: [{ type: Input, args: ['disable-close',] }],
        title: [{ type: Input }],
        labelColumns: [{ type: Input, args: ['label-columns',] }],
        separator: [{ type: Input }]
    };
    return OFormLayoutDialogOptionsDirective;
}());
export { OFormLayoutDialogOptionsDirective };
