// ontimize-web-ngx/esm2015/lib/layouts/form-layout/tabgroup/options/o-form-layout-tabgroup-options.directive.js
import { ContentChild, Directive, Input, TemplateRef } from '@angular/core';
import { BooleanConverter, NumberConverter } from '../../../../decorators/input-converter';
import { OFormLayoutManagerComponent } from '../../o-form-layout-manager.component';
export class OFormLayoutTabGroupOptionsDirective {
    constructor(formLayoutManager) {
        this.formLayoutManager = formLayoutManager;
        this._disableAnimation = true;
        this.iconPosition = 'left';
    }
    ngOnChanges() {
        if (this.formLayoutManager) {
            this.formLayoutManager.addTabGroupOptions(this.getOptions());
        }
    }
    set disableAnimation(value) {
        this._disableAnimation = BooleanConverter(value);
    }
    set maxTabs(value) {
        this._maxTabs = NumberConverter(value);
    }
    getOptions() {
        const result = {
            backgroundColor: this.backgroundColor,
            color: this.color,
            headerPosition: this.headerPosition,
            disableAnimation: this._disableAnimation,
            icon: this.icon,
            iconPosition: this.iconPosition,
            titleDataOrigin: this.titleDataOrigin,
            title: this.title,
            labelColumns: this.labelColumns,
            separator: this.separator,
            maxTabs: this._maxTabs
        };
        Object.keys(result).forEach(key => result[key] == null ? delete result[key] : {});
        return result;
    }
}
OFormLayoutTabGroupOptionsDirective.decorators = [
    { type: Directive, args: [{
                selector: 'o-form-layout-tabgroup-options, o-form-layout-manager[mode="tab"]'
            },] }
];
OFormLayoutTabGroupOptionsDirective.ctorParameters = () => [
    { type: OFormLayoutManagerComponent }
];
OFormLayoutTabGroupOptionsDirective.propDecorators = {
    backgroundColor: [{ type: Input, args: ['background-color',] }],
    color: [{ type: Input }],
    disableAnimation: [{ type: Input, args: ['disable-animation',] }],
    headerPosition: [{ type: Input, args: ['header-position',] }],
    icon: [{ type: Input }],
    iconPosition: [{ type: Input, args: ['icon-position',] }],
    titleDataOrigin: [{ type: Input, args: ['title-data-origin',] }],
    title: [{ type: Input }],
    labelColumns: [{ type: Input, args: ['label-columns',] }],
    separator: [{ type: Input }],
    maxTabs: [{ type: Input, args: ['max-tabs',] }],
    templateMatTabLabel: [{ type: ContentChild, args: [TemplateRef, { static: false },] }]
};
