// ontimize-web-ngx/esm5/lib/components/input/listpicker/listpicker-renderer/date/o-list-picker-renderer-date.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { OMomentPipe } from '../../../../../pipes/o-moment.pipe';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER, OListPickerCustomRenderer } from '../o-list-picker-renderer.class';
export var DEFAULT_INPUTS_O_LISTPICKER_RENDERER_DATE = tslib_1.__spread(DEFAULT_INPUTS_O_LISTPICKER_RENDERER, [
    'format'
]);
var OListPickerRendererDateComponent = (function (_super) {
    tslib_1.__extends(OListPickerRendererDateComponent, _super);
    function OListPickerRendererDateComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.setComponentPipe();
        return _this;
    }
    OListPickerRendererDateComponent.prototype.setComponentPipe = function () {
        this.componentPipe = new OMomentPipe(this.injector);
    };
    OListPickerRendererDateComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        this.pipeArguments = {
            format: this.format
        };
    };
    OListPickerRendererDateComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-list-picker-renderer-date',
                    template: "<ng-template #templateref let-value=\"value\">\n  {{getListPickerValue(value)}}\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_LISTPICKER_RENDERER_DATE
                }] }
    ];
    OListPickerRendererDateComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OListPickerRendererDateComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
    };
    return OListPickerRendererDateComponent;
}(OListPickerCustomRenderer));
export { OListPickerRendererDateComponent };
