// ontimize-web-ngx/esm5/lib/components/bar-menu/menu-group/o-bar-menu-group.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, forwardRef, Inject, Injector, ViewEncapsulation } from '@angular/core';
import { Util } from '../../../util/util';
import { OBarMenuComponent } from '../o-bar-menu.component';
import { DEFAULT_INPUTS_O_BASE_MENU_ITEM, OBaseMenuItemClass } from '../o-base-menu-item.class';
export var DEFAULT_INPUTS_O_BAR_MENU_GROUP = tslib_1.__spread(DEFAULT_INPUTS_O_BASE_MENU_ITEM);
var OBarMenuGroupComponent = (function (_super) {
    tslib_1.__extends(OBarMenuGroupComponent, _super);
    function OBarMenuGroupComponent(menu, elRef, injector) {
        var _this = _super.call(this, menu, elRef, injector) || this;
        _this.menu = menu;
        _this.elRef = elRef;
        _this.injector = injector;
        _this.id = 'm_' + String((new Date()).getTime() + Util.randomNumber().toString());
        return _this;
    }
    OBarMenuGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-bar-menu-group',
                    template: "<div class=\"fake-li\" [class.fake-li-hover]=\"isHovered\">\n  <label [attr.for]=\"id\" class=\"toggle mat-list-item\">\n    <mat-icon class=\"o-bar-menu-group-icon mat-24\">{{ icon }}</mat-icon>\n    {{ title | oTranslate }}\n    <mat-icon class=\"o-bar-menu-group-icon-arrow\" *ngIf=\"!checkbox.checked\">chevron_right</mat-icon>\n    <mat-icon class=\"o-bar-menu-group-icon-arrow\" *ngIf=\"checkbox.checked\">expand_more</mat-icon>\n  </label>\n  <a class=\"o-bar-menu-group-title\">\n    <mat-icon class=\"o-bar-menu-group-icon mat-24\">{{ icon }}</mat-icon> {{ title | oTranslate }}\n    <mat-icon class=\"o-bar-menu-group-icon-arrow\">chevron_right</mat-icon>\n  </a>\n  <input [attr.id]=\"id\" type=\"checkbox\" #checkbox />\n  <div *ngIf=\"!disabled\" class=\"fake-ul mat-primary mat-elevation-z4\">\n    <ng-content></ng-content>\n  </div>\n</div>",
                    inputs: DEFAULT_INPUTS_O_BAR_MENU_GROUP,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-bar-menu-group]': 'true',
                        '[attr.disabled]': 'disabled'
                    },
                    styles: [".o-bar-menu-group[disabled=true] .fake-li{opacity:.5}.o-bar-menu-group[disabled=true] .fake-li a.o-bar-menu-group-title{cursor:default}.o-bar-menu-group .fake-li .o-bar-menu-group-icon,.o-bar-menu-group .fake-ul .o-bar-menu-group-icon{display:inline-block;margin:0 10px 0 0}.o-bar-menu-group .fake-li .o-bar-menu-group-icon~.o-bar-menu-group-title,.o-bar-menu-group .fake-ul .o-bar-menu-group-icon~.o-bar-menu-group-title{padding-left:0}.o-bar-menu-group .fake-li .o-bar-menu-group-icon-arrow,.o-bar-menu-group .fake-ul .o-bar-menu-group-icon-arrow{margin-left:auto}"]
                }] }
    ];
    OBarMenuGroupComponent.ctorParameters = function () { return [
        { type: OBarMenuComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OBarMenuComponent; }),] }] },
        { type: ElementRef },
        { type: Injector }
    ]; };
    return OBarMenuGroupComponent;
}(OBaseMenuItemClass));
export { OBarMenuGroupComponent };
