// ontimize-web-ngx/esm5/lib/services/ontimize/ontimize-base-service.class.js
import * as tslib_1 from "tslib";
import { BaseService } from '../base-service.class';
var OntimizeBaseService = (function (_super) {
    tslib_1.__extends(OntimizeBaseService, _super);
    function OntimizeBaseService(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.kv = {};
        _this.av = [];
        _this.sqltypes = {};
        _this.pagesize = 10;
        _this.offset = 0;
        _this.orderby = [];
        _this.totalsize = -1;
        return _this;
    }
    OntimizeBaseService.prototype.configureService = function (config) {
        _super.prototype.configureService.call(this, config);
    };
    OntimizeBaseService.prototype.startsession = function (user, password) {
        return null;
    };
    OntimizeBaseService.prototype.endsession = function (user, sessionId) {
        return null;
    };
    OntimizeBaseService.prototype.hassession = function (user, sessionId) {
        return null;
    };
    OntimizeBaseService.prototype.clientErrorFallback = function (errorCode) {
        if (errorCode === 401) {
            this.authService.logout();
        }
    };
    return OntimizeBaseService;
}(BaseService));
export { OntimizeBaseService };
