// ontimize-web-ngx/esm2015/lib/components/table/extensions/header/table-columns-grouping/o-table-columns-grouping.component.js
import { ChangeDetectionStrategy, Component, ContentChildren, forwardRef, Inject, Injector, QueryList } from '@angular/core';
import { Util } from '../../../../../util/util';
import { OTableComponent } from '../../../o-table.component';
import { OTableColumnsGroupingColumnComponent } from './columns/o-table-columns-grouping-column.component';
export const DEFAULT_INPUTS_O_TABLE_COLUMN_GROUPING = [
    'columns',
    'excludedAggregateColumns: excluded-aggregate-columns'
];
export const DEFAULT_OUTPUTS_O_TABLE_COLUMN_GROUPING = [];
export class OTableColumnsGroupingComponent {
    constructor(injector, table) {
        this.injector = injector;
        this.table = table;
        this._columnsArray = [];
        this._excludedAggregateColumnsArray = [];
    }
    set columns(arg) {
        this._columnsArray = Util.parseArray(arg, true);
    }
    get columnsArray() {
        return this._columnsArray;
    }
    set excludedAggregateColumns(arg) {
        this._excludedAggregateColumnsArray = Util.parseArray(arg, true);
    }
    ngOnInit() {
        if (this._columnsArray.length === 0) {
            this._columnsArray = this.table.groupedColumnsArray;
        }
        this.table.setOTableColumnsGrouping(this);
    }
    useColumnAggregate(columnAttr, hasDefaultAggregate) {
        if (this._excludedAggregateColumnsArray.includes(columnAttr)) {
            return false;
        }
        const columnConf = this.getColumnGrouping(columnAttr);
        return hasDefaultAggregate || Util.isDefined(columnConf);
    }
    getColumnGrouping(columnAttr) {
        return this.groupingColumns.find(col => col.attr === columnAttr);
    }
}
OTableColumnsGroupingComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-columns-grouping',
                template: ' ',
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_TABLE_COLUMN_GROUPING,
                outputs: DEFAULT_OUTPUTS_O_TABLE_COLUMN_GROUPING
            }] }
];
OTableColumnsGroupingComponent.ctorParameters = () => [
    { type: Injector },
    { type: OTableComponent, decorators: [{ type: Inject, args: [forwardRef(() => OTableComponent),] }] }
];
OTableColumnsGroupingComponent.propDecorators = {
    groupingColumns: [{ type: ContentChildren, args: [OTableColumnsGroupingColumnComponent, { descendants: true },] }]
};
