// ontimize-web-ngx/esm2015/lib/components/input/combo/combo-renderer/real/o-combo-renderer-real.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { ORealPipe } from '../../../../../pipes/o-real.pipe';
import { NumberService } from '../../../../../services/number.service';
import { DEFAULT_INPUTS_O_COMBO_RENDERER_INTEGER, OComboRendererIntegerComponent } from '../integer/o-combo-renderer-integer.component';
export const DEFAULT_INPUTS_O_COMBO_RENDERER_REAL = [
    ...DEFAULT_INPUTS_O_COMBO_RENDERER_INTEGER,
    'decimalSeparator: decimal-separator',
    'minDecimalDigits: min-decimal-digits',
    'maxDecimalDigits: max-decimal-digits'
];
export class OComboRendererRealComponent extends OComboRendererIntegerComponent {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.minDecimalDigits = 2;
        this.maxDecimalDigits = 2;
        this.decimalSeparator = '.';
        this.numberService = this.injector.get(NumberService);
        this.setComponentPipe();
    }
    setComponentPipe() {
        this.componentPipe = new ORealPipe(this.injector);
    }
    initialize() {
        super.initialize();
        this.pipeArguments = {
            minDecimalDigits: this.minDecimalDigits,
            maxDecimalDigits: this.maxDecimalDigits,
            decimalSeparator: this.decimalSeparator,
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator
        };
    }
}
OComboRendererRealComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-combo-renderer-real',
                template: "<ng-template #templateref let-value=\"value\">\n  {{ getComboData(value)}}\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_COMBO_RENDERER_REAL
            }] }
];
OComboRendererRealComponent.ctorParameters = () => [
    { type: Injector }
];
OComboRendererRealComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
};
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Number)
], OComboRendererRealComponent.prototype, "minDecimalDigits", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Number)
], OComboRendererRealComponent.prototype, "maxDecimalDigits", void 0);
