// ontimize-web-ngx/esm5/lib/components/contextmenu/context-menu/o-context-menu-content.component.js
import { Component, EventEmitter, HostListener, Injector, ViewChild } from '@angular/core';
import { MatMenuTrigger } from '@angular/material';
import { OContextMenuGroupComponent } from '../context-menu-group/o-context-menu-group.component';
export var DEFAULT_CONTEXT_MENU_CONTENT_INPUTS = [
    'menuItems',
    'externalMenuItems',
    'overlay',
    'data',
    'menuClass'
];
export var DEFAULT_CONTEXT_MENU_CONTENT_OUTPUTS = [
    'execute',
    'close'
];
var OContextMenuContentComponent = (function () {
    function OContextMenuContentComponent(injector) {
        this.injector = injector;
        this.execute = new EventEmitter();
        this.close = new EventEmitter();
    }
    OContextMenuContentComponent.prototype.click = function () {
        this.closeContent();
    };
    OContextMenuContentComponent.prototype.ngOnInit = function () {
        this.initialize();
    };
    OContextMenuContentComponent.prototype.ngAfterViewInit = function () {
        this.trigger.openMenu();
    };
    OContextMenuContentComponent.prototype.initialize = function () {
        var menuItemsArray = this.menuItems ? this.menuItems.toArray() : [];
        var externalItemsArray = this.externalMenuItems ? this.externalMenuItems.toArray() : [];
        this.allMenuItems = menuItemsArray.concat(externalItemsArray);
        this.setData(this.allMenuItems);
    };
    OContextMenuContentComponent.prototype.setData = function (items) {
        var _this = this;
        if (this.data) {
            (items || []).forEach(function (menuItem) {
                menuItem.data = _this.data;
                if (menuItem instanceof OContextMenuGroupComponent) {
                    _this.setData(menuItem.children);
                }
            });
        }
    };
    OContextMenuContentComponent.prototype.onMenuClosed = function (e) {
        this.closeContent();
    };
    OContextMenuContentComponent.prototype.closeContent = function () {
        this.trigger.closeMenu();
        this.close.emit();
    };
    OContextMenuContentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-context-menu-content',
                    template: "<button mat-button [matMenuTriggerFor]=\"menu.childMenu\" (menuClosed)=\"onMenuClosed($event)\"></button>\n<o-wrapper-content-menu #menu [items]=\"allMenuItems\" [class]=\"menuClass\"> </o-wrapper-content-menu>",
                    inputs: DEFAULT_CONTEXT_MENU_CONTENT_INPUTS,
                    outputs: DEFAULT_CONTEXT_MENU_CONTENT_OUTPUTS,
                    host: {
                        '[class.o-context-menu-content]': 'true'
                    }
                }] }
    ];
    OContextMenuContentComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OContextMenuContentComponent.propDecorators = {
        trigger: [{ type: ViewChild, args: [MatMenuTrigger, { static: false },] }],
        click: [{ type: HostListener, args: ['document:click',] }]
    };
    return OContextMenuContentComponent;
}());
export { OContextMenuContentComponent };
