// ontimize-web-ngx/esm5/lib/components/bar-menu/menu-nested/o-bar-menu-nested.component.js
import { ChangeDetectionStrategy, Component, Injector } from '@angular/core';
import { AppMenuService } from '../../../services/app-menu.service';
export var DEFAULT_INPUTS_O_BAR_MENU_NESTED = [
    'items'
];
var OBarMenuNestedComponent = (function () {
    function OBarMenuNestedComponent(injector) {
        this.injector = injector;
        this.appMenuService = this.injector.get(AppMenuService);
    }
    OBarMenuNestedComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-bar-menu-nested',
                    template: "<ng-container *ngFor=\"let item of items\">\n  <ng-container [ngSwitch]=\"appMenuService.getMenuItemType(item)\">\n\n    <o-bar-menu-group *ngSwitchCase=\"'group'\" [title]=\"item.name\" [attr]=\"item.id\" [tooltip]=\"item.tooltip\" [icon]=\"item.icon\" [ngClass]=\"item.class\">\n      <o-bar-menu-nested [items]=\"item.items\"> </o-bar-menu-nested>\n    </o-bar-menu-group>\n\n    <o-locale-bar-menu-item *ngSwitchCase=\"'locale'\" [title]=\"item.name\" [attr]=\"item.id\" [tooltip]=\"item.tooltip\" [icon]=\"item.icon\"\n      [locale]=\"item.locale\" [ngClass]=\"item.class\">\n    </o-locale-bar-menu-item>\n\n    <o-bar-menu-item *ngSwitchDefault [title]=\"item.name\" [attr]=\"item.id\" [tooltip]=\"item.tooltip\" [icon]=\"item.icon\" [route]=\"item.route\"\n      [ngClass]=\"item.class\"></o-bar-menu-item>\n\n  </ng-container>\n</ng-container>\n",
                    inputs: DEFAULT_INPUTS_O_BAR_MENU_NESTED,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    OBarMenuNestedComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OBarMenuNestedComponent;
}());
export { OBarMenuNestedComponent };
