// ontimize-web-ngx-report/esm5/lib/components/report-on-demand/select-function-dialog/select-function-dialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/report-on-demand/select-function-dialog/select-function-dialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject, Optional } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
var SelectFunctionDialogComponent = /** @class */ (function () {
    function SelectFunctionDialogComponent(dialogo, reportFunction) {
        this.dialogo = dialogo;
        this.reportFunction = reportFunction;
        this.functions = [
            { id: 'SUM', value: 'DIALOG.SELECT_FUNCTION.SUM' },
            { id: 'AVERAGE', value: 'DIALOG.SELECT_FUNCTION.AVG' },
            { id: 'MAX', value: 'DIALOG.SELECT_FUNCTION.MAX' },
            { id: 'MIN', value: 'DIALOG.SELECT_FUNCTION.MIN' }
        ];
        this.selectedFunction = reportFunction.type;
    }
    /**
     * @return {?}
     */
    SelectFunctionDialogComponent.prototype.save = /**
     * @return {?}
     */
    function () {
        this.dialogo.close({ columnName: this.reportFunction.columnName, type: this.selectedFunction });
    };
    SelectFunctionDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-select-function-dialog',
                    template: "<span mat-dialog-title> {{ 'DIALOG.SELECT_FUNCTION.TITLE' | oReportsTranslate }}</span>\n\n<mat-dialog-content>\n  <div mat-subheader>{{ 'DIALOG.SELECT_FUNCTION.DESCRIPTION' | oReportsTranslate }}</div>\n\n  <mat-radio-group name=\"Function\" [(ngModel)]=\"selectedFunction\" fxLayout=\"column\" fxLayoutAlign=\"start start\" fxLayoutGap=\"12px\">\n    <mat-radio-button *ngFor=\"let function of functions\" [value]=\"function.id\">\n      {{ function.value | oReportsTranslate }}\n    </mat-radio-button>\n  </mat-radio-group>\n\n</mat-dialog-content>\n\n<mat-dialog-actions align=\"end\">\n  <button type=\"button\" mat-stroked-button class=\"mat-primary\" [mat-dialog-close]=\"false\" class=\"o-button-default\">\n    {{ 'CANCEL' | oTranslate }}</button>\n  <button type=\"button\" mat-stroked-button class=\"mat-primary\" (click)=\"save()\" [mat-dialog-close]=\"true\" [disabled]=\"!selectedFunction\">\n    {{'SAVE' | oTranslate }}</button>\n</mat-dialog-actions>\n",
                    styles: [".mat-radio-group{margin:0 12px}"]
                }] }
    ];
    /** @nocollapse */
    SelectFunctionDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    return SelectFunctionDialogComponent;
}());
export { SelectFunctionDialogComponent };
if (false) {
    /** @type {?} */
    SelectFunctionDialogComponent.prototype.selectedFunction;
    /** @type {?} */
    SelectFunctionDialogComponent.prototype.functions;
    /** @type {?} */
    SelectFunctionDialogComponent.prototype.dialogo;
    /** @type {?} */
    SelectFunctionDialogComponent.prototype.reportFunction;
}
