// ontimize-web-ngx/esm2015/lib/components/table/column/cell-renderer/o-base-table-cell-renderer.class.js
import { FilterExpressionUtils } from '../../../../util/filter-expression.utils';
import { Util } from '../../../../util/util';
import { OTableColumnComponent } from '../o-table-column.component';
export const DEFAULT_INPUTS_O_BASE_TABLE_CELL_RENDERER = [
    'filterSource: filter-source',
    'filterFunction: filter-function'
];
export class OBaseTableCellRenderer {
    constructor(injector) {
        this.injector = injector;
        this._filterSource = 'render';
        this.tableColumn = this.injector.get(OTableColumnComponent);
    }
    ngOnInit() {
        this.initialize();
    }
    initialize() {
    }
    ngAfterContentInit() {
        if (typeof this.filterFunction !== 'function') {
            this.filterFunction = undefined;
        }
        this.registerRenderer();
    }
    get table() {
        return this.tableColumn.table;
    }
    get column() {
        return this.tableColumn.attr;
    }
    registerRenderer() {
        this.tableColumn.registerRenderer(this);
        if (!Util.isDefined(this.type) && Util.isDefined(this.tableColumn.type)) {
            this.type = this.tableColumn.type;
        }
    }
    getCellData(cellvalue, rowvalue) {
        let parsedValue;
        if (this.componentPipe && this.pipeArguments !== undefined && cellvalue !== undefined) {
            parsedValue = this.componentPipe.transform(cellvalue, this.pipeArguments);
        }
        else {
            parsedValue = cellvalue;
        }
        return parsedValue;
    }
    getTooltip(cellValue, rowValue) {
        return this.getCellData(cellValue, rowValue);
    }
    set filterSource(val) {
        const lowerVal = (val || '').toLowerCase();
        this._filterSource = (lowerVal === 'render' || lowerVal === 'data' || lowerVal === 'both') ? lowerVal : 'render';
    }
    get filterSource() {
        return this._filterSource;
    }
    getFilter(cellValue, rowValue) {
        let result;
        switch (this.filterSource) {
            case 'render':
                result = [this.getCellData(cellValue, rowValue)];
                break;
            case 'data':
                result = [cellValue];
                break;
            case 'both':
                result = [cellValue, this.getCellData(cellValue, rowValue)];
                break;
        }
        return result;
    }
    getFilterExpression(quickFilter) {
        return FilterExpressionUtils.buildExpressionLike(this.tableColumn.attr, quickFilter);
    }
}
