// ontimize-web-ngx/esm5/lib/components/input/combo/combo-renderer/real/o-combo-renderer-real.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { ORealPipe } from '../../../../../pipes/o-real.pipe';
import { NumberService } from '../../../../../services/number.service';
import { DEFAULT_INPUTS_O_COMBO_RENDERER_INTEGER, OComboRendererIntegerComponent } from '../integer/o-combo-renderer-integer.component';
export var DEFAULT_INPUTS_O_COMBO_RENDERER_REAL = tslib_1.__spread(DEFAULT_INPUTS_O_COMBO_RENDERER_INTEGER, [
    'decimalSeparator: decimal-separator',
    'minDecimalDigits: min-decimal-digits',
    'maxDecimalDigits: max-decimal-digits'
]);
var OComboRendererRealComponent = (function (_super) {
    tslib_1.__extends(OComboRendererRealComponent, _super);
    function OComboRendererRealComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.minDecimalDigits = 2;
        _this.maxDecimalDigits = 2;
        _this.decimalSeparator = '.';
        _this.numberService = _this.injector.get(NumberService);
        _this.setComponentPipe();
        return _this;
    }
    OComboRendererRealComponent.prototype.setComponentPipe = function () {
        this.componentPipe = new ORealPipe(this.injector);
    };
    OComboRendererRealComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        this.pipeArguments = {
            minDecimalDigits: this.minDecimalDigits,
            maxDecimalDigits: this.maxDecimalDigits,
            decimalSeparator: this.decimalSeparator,
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator
        };
    };
    OComboRendererRealComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-combo-renderer-real',
                    template: "<ng-template #templateref let-value=\"value\">\n  {{ getComboData(value)}}\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_COMBO_RENDERER_REAL
                }] }
    ];
    OComboRendererRealComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OComboRendererRealComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Number)
    ], OComboRendererRealComponent.prototype, "minDecimalDigits", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Number)
    ], OComboRendererRealComponent.prototype, "maxDecimalDigits", void 0);
    return OComboRendererRealComponent;
}(OComboRendererIntegerComponent));
export { OComboRendererRealComponent };
