// ontimize-web-ngx/esm5/lib/components/input/combo/combo-renderer/o-combo-renderer.class.js
import { Util } from '../../../../util';
import { OComboComponent } from '../o-combo.component';
export var DEFAULT_INPUTS_O_COMBO_RENDERER = [];
var OComboCustomRenderer = (function () {
    function OComboCustomRenderer(injector) {
        this.injector = injector;
        this.comboComponent = this.injector.get(OComboComponent);
    }
    OComboCustomRenderer.prototype.ngOnInit = function () {
        this.initialize();
    };
    OComboCustomRenderer.prototype.initialize = function () {
    };
    OComboCustomRenderer.prototype.ngAfterContentInit = function () {
        this.registerRenderer();
    };
    OComboCustomRenderer.prototype.registerRenderer = function () {
        this.comboComponent.registerRenderer(this);
    };
    OComboCustomRenderer.prototype.getComboData = function (record) {
        var _this = this;
        var descTxt = '';
        if (!Util.isDefined(record)) {
            return descTxt;
        }
        this.comboComponent.descriptionColArray.forEach(function (col, index) {
            if (Util.isDefined(record[col])) {
                var value = record[col];
                if (_this.componentPipe && _this.pipeArguments != null) {
                    value = _this.componentPipe.transform(value, _this.pipeArguments);
                }
                if (Util.isDefined(value)) {
                    descTxt += value;
                    if (index < _this.comboComponent.descriptionColArray.length - 1) {
                        descTxt += _this.comboComponent.separator;
                    }
                }
            }
        });
        return descTxt;
    };
    return OComboCustomRenderer;
}());
export { OComboCustomRenderer };
