// ontimize-web-ngx/esm2015/lib/components/button-toggle/o-button-toggle-group/o-button-toggle-group.component.js
import * as tslib_1 from "tslib";
import { Component, ComponentFactoryResolver, ContentChildren, EventEmitter, forwardRef, QueryList, ViewChild, ViewContainerRef, ViewEncapsulation } from '@angular/core';
import { MatButtonToggleGroup } from '@angular/material';
import { InputConverter } from '../../../decorators/input-converter';
import { Util } from '../../../util/util';
import { OButtonToggleComponent } from '../o-button-toggle.component';
export const DEFAULT_INPUTS_O_BUTTON_TOGGLE_GROUP = [
    'oattr: attr',
    'name',
    'enabled',
    'layout',
    'multiple',
    'value'
];
export const DEFAULT_OUTPUTS_O_BUTTON_TOGGLE_GROUP = [
    'onChange'
];
export class OButtonToggleGroupComponent {
    constructor(resolver) {
        this.resolver = resolver;
        this.DEFAULT_INPUTS_O_BUTTON_TOGGLE_GROUP = DEFAULT_INPUTS_O_BUTTON_TOGGLE_GROUP;
        this.DEFAULT_OUTPUTS_O_BUTTON_TOGGLE_GROUP = DEFAULT_OUTPUTS_O_BUTTON_TOGGLE_GROUP;
        this._enabled = true;
        this.layout = 'row';
        this.multiple = false;
        this.onChange = new EventEmitter();
    }
    get enabled() {
        if (this._innerButtonToggleGroup instanceof MatButtonToggleGroup) {
            return !this._innerButtonToggleGroup.disabled;
        }
        return true;
    }
    set enabled(val) {
        if (this._innerButtonToggleGroup instanceof MatButtonToggleGroup) {
            val = Util.parseBoolean(String(val));
            this._innerButtonToggleGroup.disabled = !val;
        }
    }
    ngOnInit() {
        if (!Util.isDefined(this.name)) {
            this.name = this.oattr;
        }
    }
    ngAfterViewInit() {
        this.buildChildren();
        this._children.changes.subscribe(() => this.buildChildren());
    }
    buildChildren() {
        const factory = this.resolver.resolveComponentFactory(OButtonToggleComponent);
        this._viewContainerRef.clear();
        const childList = this._children.map((child) => {
            const componentRef = this._viewContainerRef.createComponent(factory);
            componentRef.instance.oattr = child.oattr;
            componentRef.instance.label = child.label;
            componentRef.instance.icon = child.icon;
            componentRef.instance.iconPosition = child.iconPosition;
            componentRef.instance.checked = child.checked;
            componentRef.instance.enabled = child.enabled;
            componentRef.instance.value = child.value;
            componentRef.instance.name = this.name;
            componentRef.instance.onChange = child.onChange;
            componentRef.changeDetectorRef.detectChanges();
            return componentRef.instance;
        });
        this._innerButtonToggleGroup._buttonToggles.reset(childList.map(c => c._innerButtonToggle));
        this._children.reset(childList);
    }
    getValue() {
        return this._innerButtonToggleGroup ? this._innerButtonToggleGroup.value : void 0;
    }
    setValue(val) {
        this._innerButtonToggleGroup.value = val;
    }
}
OButtonToggleGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-button-toggle-group',
                template: "<mat-button-toggle-group [disabled]=\"!enabled\" [name]=\"name\" [value]=\"value\" [vertical]=\"layout==='column'\" [multiple]=\"multiple\"\n  (change)=\"onChange.emit($event)\">\n  <ng-container #childContainer></ng-container>\n</mat-button-toggle-group>\n",
                inputs: DEFAULT_INPUTS_O_BUTTON_TOGGLE_GROUP,
                outputs: DEFAULT_OUTPUTS_O_BUTTON_TOGGLE_GROUP,
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.o-button-toggle-group]': 'true'
                },
                entryComponents: [OButtonToggleComponent]
            }] }
];
OButtonToggleGroupComponent.ctorParameters = () => [
    { type: ComponentFactoryResolver }
];
OButtonToggleGroupComponent.propDecorators = {
    _innerButtonToggleGroup: [{ type: ViewChild, args: [MatButtonToggleGroup, { static: false },] }],
    _viewContainerRef: [{ type: ViewChild, args: ['childContainer', { read: ViewContainerRef, static: false },] }],
    _children: [{ type: ContentChildren, args: [forwardRef(() => OButtonToggleComponent),] }]
};
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OButtonToggleGroupComponent.prototype, "multiple", void 0);
