// ontimize-web-ngx/esm2015/lib/layouts/form-layout/split-pane/options/o-form-layout-split-pane-options.directive.js
import { Directive, Input } from '@angular/core';
import { OFormLayoutManagerComponent } from '../../o-form-layout-manager.component';
export class OFormLayoutSplitPaneOptionsDirective {
    constructor(formLayoutManager) {
        this.formLayoutManager = formLayoutManager;
    }
    ngOnChanges() {
        if (this.formLayoutManager) {
            this.formLayoutManager.addSplitPaneOptions(this.getOptions());
        }
    }
    getOptions() {
        const result = {
            mainWidth: this.mainWidth,
            mainMaxWidth: this.mainMaxWidth,
            mainMinWidth: this.mainMinWidth,
            detailWidth: this.detailWidth,
            detailMaxWidth: this.detailMaxWidth,
            detailMinWidth: this.detailMinWidth
        };
        Object.keys(result).forEach(key => result[key] == null ? delete result[key] : {});
        return result;
    }
}
OFormLayoutSplitPaneOptionsDirective.decorators = [
    { type: Directive, args: [{
                selector: 'o-form-layout-split-pane-options, o-form-layout-manager[mode="split-pane"]'
            },] }
];
OFormLayoutSplitPaneOptionsDirective.ctorParameters = () => [
    { type: OFormLayoutManagerComponent }
];
OFormLayoutSplitPaneOptionsDirective.propDecorators = {
    mainWidth: [{ type: Input, args: ['main-width',] }],
    mainMaxWidth: [{ type: Input, args: ['main-max-width',] }],
    mainMinWidth: [{ type: Input, args: ['main-min-width',] }],
    detailWidth: [{ type: Input, args: ['detail-width',] }],
    detailMaxWidth: [{ type: Input, args: ['detail-max-width',] }],
    detailMinWidth: [{ type: Input, args: ['detail-min-width',] }]
};
