// ontimize-web-ngx-report/esm2015/lib/services/o-report-store.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/o-report-store.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from "@angular/core";
import { MatDialog } from "@angular/material";
import { DialogService } from "ontimize-web-ngx";
import { OReportViewerComponent } from "../components/report/o-report-viewer/o-report-viewer.component";
import { Utils } from "../util/utils";
import { OReportService } from "./o-report.service";
export class OReportStoreService {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
        this.dialogService = this.injector.get(DialogService);
        this.dialog = this.injector.get(MatDialog);
        this.reportService = this.injector.get(OReportService);
    }
    /**
     * @param {?} reportId
     * @param {?} parametersValues
     * @param {?} filter
     * @return {?}
     */
    openFillReport(reportId, parametersValues, filter) {
        this.reportService.configureService(this.reportService.getDefaultServiceConfiguration());
        this.reportService.configureAdapter();
        /** @type {?} */
        let kv = { 'UUID': reportId };
        this.reportService.query(kv, null, 'getReport', {}).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            if (res && res.data.length && res.code === 0) {
                /** @type {?} */
                let parameters = res.data[0].PARAMETERS;
                /** @type {?} */
                let name = res.data[0].NAME;
                /** @type {?} */
                let av = [];
                if (parameters.length > 0) {
                    av = [reportId];
                    /** @type {?} */
                    let values = Object.values(parametersValues);
                    for (let value of values) {
                        av.push(value);
                    }
                }
                else {
                    av = [reportId];
                }
                /** @type {?} */
                const data = { 'params': av, 'filter': filter, 'name': name };
                Utils.openModalVisor(this.dialog, OReportViewerComponent, data);
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        err => {
            if (this.dialogService) {
                this.dialogService.error('ERROR', 'SERVER_ERROR_MESSAGE');
            }
            console.log(err);
        }));
    }
}
OReportStoreService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
OReportStoreService.ctorParameters = () => [
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    OReportStoreService.prototype.reportService;
    /**
     * @type {?}
     * @protected
     */
    OReportStoreService.prototype.dialogService;
    /**
     * @type {?}
     * @protected
     */
    OReportStoreService.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    OReportStoreService.prototype.injector;
}
