// ontimize-web-ngx-report/esm2015/lib/services/ontimize-report.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ontimize-report.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { MatDialog } from '@angular/material';
import { DialogService } from 'ontimize-web-ngx';
import { ReportOnDemandComponent } from '../components/report-on-demand/report-on-demand/report-on-demand.component';
import { Utils } from '../util/utils';
import * as i0 from "@angular/core";
export class OntimizeReportService {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
        this.dialogService = this.injector.get(DialogService);
        this.dialog = this.injector.get(MatDialog);
    }
    /**
     * @param {?} table
     * @return {?}
     */
    openReportOnDemand(table) {
        Utils.openModalVisor(this.dialog, ReportOnDemandComponent, table);
    }
}
OntimizeReportService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
OntimizeReportService.ctorParameters = () => [
    { type: Injector }
];
/** @nocollapse */ OntimizeReportService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OntimizeReportService_Factory() { return new OntimizeReportService(i0.ɵɵinject(i0.INJECTOR)); }, token: OntimizeReportService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    OntimizeReportService.prototype.dialogService;
    /**
     * @type {?}
     * @protected
     */
    OntimizeReportService.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    OntimizeReportService.prototype.injector;
}
