// ontimize-web-ngx/esm2015/lib/components/dual-list-selector/dual-list-selector-item/date/o-dual-list-selector-date-item.component.js
import { Component, ChangeDetectionStrategy } from "@angular/core";
export const DEFAULT_DUAL_LIST_SELECTOR_DATE_ITEM = [
    'item',
    'groupedDateColumns: grouped-date-columns',
];
export class ODualListSelectorDateItemComponent {
    constructor() {
        this.dateTypes = [
            { value: 'MONTH', viewValue: 'DUAL_LIST_SELECTOR.GROUP_BY_MONTH' },
            { value: 'YEAR_MONTH', viewValue: 'DUAL_LIST_SELECTOR.GROUP_BY_YEAR_MONTH' },
            { value: 'YEAR', viewValue: 'DUAL_LIST_SELECTOR.GROUP_BY_YEAR' },
            { value: 'YEAR_MONTH_DAY', viewValue: 'DUAL_LIST_SELECTOR.GROUP_BY_YEAR_MONTH_DAY' }
        ];
        this.item = "";
    }
    onSelectionChange(event, itemSelected) {
        let value = event.value;
        let attr = itemSelected;
        let index = this.groupedDateColumns.findIndex(column => column.attr == attr);
        if (index != -1) {
            this.groupedDateColumns.splice(index, 1);
        }
        this.groupedDateColumns.push({ "attr": attr, "type": value });
    }
    getSelectValue() {
        let index = this.groupedDateColumns.findIndex(column => column.attr == this.item);
        return index != -1 ? this.groupedDateColumns[index].type : 'YEAR_MONTH_DAY';
    }
    getViewValue() {
        let value = this.getSelectValue();
        const indexFindValue = this.dateTypes.findIndex(type => type.value == value);
        return indexFindValue > -1 ? this.dateTypes[indexFindValue].viewValue : 'DUAL_LIST_SELECTOR.GROUP_BY_YEAR_MONTH_DAY';
    }
}
ODualListSelectorDateItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-dual-list-selector-date-item',
                template: "<mat-select [value]=\"getSelectValue()\" (selectionChange)=\"onSelectionChange($event,item)\" required>\n  <mat-select-trigger>\n    {{item | oTranslate}} ({{getViewValue() | oTranslate}})\n  </mat-select-trigger>\n  <mat-option *ngFor=\"let type of dateTypes\" [value]=\"type.value\">\n    {{type.viewValue | oTranslate}}\n  </mat-option>\n</mat-select>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_DUAL_LIST_SELECTOR_DATE_ITEM
            }] }
];
