// ontimize-web-ngx-report/esm2015/ontimize-web-ngx-report.js
/**
 * @fileoverview added by tsickle
 * Generated from: ontimize-web-ngx-report.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { OReportModule, OReportDetailComponent, OReportHomeComponent, OReportNewComponent, OReportViewerComponent, OReportRoutingModule, ApplyConfigurationDialogComponent, ReportOnDemandComponent, SavePreferencesDialogComponent, SelectFunctionDialogComponent, StyleDialogComponent, OReportService, OntimizeReportService, OReportStoreService, DefaultOReportPreferences, OReportsTranslatePipe } from './public-api';
export { OREPORT_DECLARATION_MODULES as ɵa, OREPORT_ENTRY_COMPONENTS_MODULES as ɵc, OREPORT_IMPORTS_MODULES as ɵb } from './lib/config/o-components';
export { OREPORT_PROVIDERS as ɵf, reportServiceFactory as ɵe } from './lib/config/o-providers';
export { OAlertService as ɵd } from './lib/services/o-alert.service';
