// ontimize-web-ngx/esm5/lib/components/table/o-table.component.js
import * as tslib_1 from "tslib";
import { animate, state, style, transition, trigger } from '@angular/animations';
import { SelectionModel } from '@angular/cdk/collections';
import { moveItemInArray } from '@angular/cdk/drag-drop';
import { DomPortalOutlet, TemplatePortal } from '@angular/cdk/portal';
import { CdkVirtualScrollViewport, VIRTUAL_SCROLL_STRATEGY } from '@angular/cdk/scrolling';
import { ApplicationRef, ChangeDetectionStrategy, Component, ComponentFactoryResolver, ContentChild, ContentChildren, ElementRef, EventEmitter, forwardRef, HostListener, Inject, Injector, NgZone, Optional, QueryList, TemplateRef, ViewChild, ViewContainerRef, ViewEncapsulation } from '@angular/core';
import { MatDialog, MatTab, MatTabGroup } from '@angular/material';
import moment from 'moment';
import { BehaviorSubject, combineLatest, Observable, of } from 'rxjs';
import { debounceTime, distinctUntilChanged, filter, map } from 'rxjs/operators';
import { BooleanConverter, InputConverter } from '../../decorators/input-converter';
import { ComponentStateServiceProvider, OntimizeServiceProvider, O_COMPONENT_STATE_SERVICE } from '../../services/factories';
import { SnackBarService } from '../../services/snackbar.service';
import { OTableComponentStateService } from '../../services/state/o-table-component-state.service';
import { ColumnValueFilterOperator } from '../../types/table/o-column-value-filter.type';
import { TableFilterByColumnDialogResult } from '../../types/table/o-table-filter-by-column-data.type';
import { ObservableWrapper } from '../../util/async';
import { Codes } from '../../util/codes';
import { FilterExpressionUtils } from '../../util/filter-expression.utils';
import { PermissionsUtils } from '../../util/permissions';
import { ServiceUtils } from '../../util/service.utils';
import { SQLTypes } from '../../util/sqltypes';
import { Util } from '../../util/util';
import { OFormComponent } from '../form/o-form.component';
import { AbstractOServiceComponent, DEFAULT_INPUTS_O_SERVICE_COMPONENT, DEFAULT_OUTPUTS_O_SERVICE_COMPONENT } from '../o-service-component.class';
import { OColumn } from './column/o-column.class';
import { OTableContextMenuComponent } from './extensions/contextmenu/o-table-context-menu.component';
import { DefaultOTableOptions } from './extensions/default-o-table-options.class';
import { OTableFilterByColumnDataDialogComponent } from './extensions/dialog/filter-by-column/o-table-filter-by-column-data-dialog.component';
import { OBaseTablePaginator } from './extensions/footer/paginator/o-base-table-paginator.class';
import { OTableButtonComponent } from './extensions/header/table-button/o-table-button.component';
import { OTableColumnsFilterComponent } from './extensions/header/table-columns-filter/o-table-columns-filter.component';
import { OTableOptionComponent } from './extensions/header/table-option/o-table-option.component';
import { OTableDataSourceService } from './extensions/o-table-datasource.service';
import { OTableVirtualScrollStrategy } from './extensions/o-table-strategy.service';
import { OTableDao } from './extensions/o-table.dao';
import { OTableGroupedRow } from './extensions/row/o-table-row-group.class';
import { OTableRowExpandableComponent, OTableRowExpandedChange } from './extensions/row/table-row-expandable/o-table-row-expandable.component';
import { OMatSort } from './extensions/sort/o-mat-sort';
import { O_TABLE_GLOBAL_CONFIG } from './utils/o-table.tokens';
import { OTableColumnSelectAllDirective } from './extensions/header/table-column-select-all/o-table-column-select-all.directive';
export var DEFAULT_INPUTS_O_TABLE = tslib_1.__spread(DEFAULT_INPUTS_O_SERVICE_COMPONENT, [
    'visibleColumns: visible-columns',
    'defaultVisibleColumns: default-visible-columns',
    'sortColumns: sort-columns',
    'quickFilterCallback: quick-filter-function',
    'deleteButton: delete-button',
    'refreshButton: refresh-button',
    'columnsVisibilityButton: columns-visibility-button',
    'exportButton: export-button',
    'showConfigurationOption: show-configuration-option',
    'showButtonsText: show-buttons-text',
    'selectAllCheckbox: select-all-checkbox',
    'paginationControls: pagination-controls',
    'fixedHeader: fixed-header',
    'showTitle: show-title',
    'editionMode: edition-mode',
    'selectionMode: selection-mode',
    'horizontalScroll: horizontal-scroll',
    'showPaginatorFirstLastButtons: show-paginator-first-last-buttons',
    'autoAlignTitles: auto-align-titles',
    'multipleSort: multiple-sort',
    'selectAllCheckboxVisible: select-all-checkbox-visible',
    'orderable',
    'resizable',
    'keepSelectedItems: keep-selected-items',
    'exportMode: export-mode',
    'exportServiceType: export-service-type',
    'autoAdjust: auto-adjust',
    'showFilterOption: show-filter-option',
    'visibleExportDialogButtons: visible-export-dialog-buttons',
    'rowClass: row-class',
    'filterColumnActiveByDefault:filter-column-active-by-default',
    'groupedColumns: grouped-columns',
    'groupable',
    'expandGroupsSameLevel: expand-groups-same-level',
    'collapseGroupedColumns: collapse-grouped-columns',
    'virtualScroll: virtual-scroll',
    'contextMenu: context-menu',
    'showExpandableIconFunction: show-expandable-icon-function',
    'showReportOnDemandOption: show-report-on-demand-option',
    'showChartsOnDemandOption: show-charts-on-demand-option',
    'showResetWidthOption: show-reset-width-option',
    'disableSelectionFunction: disable-selection-function'
]);
export var DEFAULT_OUTPUTS_O_TABLE = tslib_1.__spread(DEFAULT_OUTPUTS_O_SERVICE_COMPONENT, [
    'onRowSelected',
    'onRowDeselected',
    'onRowDeleted'
]);
var stickyHeaderSelector = '.mat-header-row .mat-table-sticky';
var stickyFooterSelector = '.mat-footer-row .mat-table-sticky';
var rowSelector = '.mat-row';
var headerSelector = '.mat-header-row';
var footerSelector = '.mat-footer-row';
var OTableComponent = (function (_super) {
    tslib_1.__extends(OTableComponent, _super);
    function OTableComponent(injector, elRef, dialog, _viewContainerRef, appRef, _componentFactoryResolver, form, scrollStrategy) {
        var _this = _super.call(this, injector, elRef, form) || this;
        _this.injector = injector;
        _this.dialog = dialog;
        _this._viewContainerRef = _viewContainerRef;
        _this.appRef = appRef;
        _this._componentFactoryResolver = _componentFactoryResolver;
        _this.scrollStrategy = scrollStrategy;
        _this.DETAIL_MODE_NONE = Codes.DETAIL_MODE_NONE;
        _this.EDIT_MODE_NONE = Codes.EDITION_MODE_NONE;
        _this.tableHeaders = [];
        _this.portalHost = [];
        _this.selectAllCheckbox = false;
        _this.exportButton = true;
        _this.showConfigurationOption = true;
        _this.columnsVisibilityButton = true;
        _this.showFilterOption = true;
        _this.showReportOnDemandOption = true;
        _this.showChartsOnDemandOption = true;
        _this.showButtonsText = true;
        _this.filterColumnActiveByDefault = true;
        _this.showResetWidthOption = true;
        _this.insertButton = true;
        _this.refreshButton = true;
        _this.deleteButton = true;
        _this.fixedHeader = true;
        _this.showTitle = false;
        _this.editionMode = Codes.EDITION_MODE_NONE;
        _this.selectionMode = Codes.SELECTION_MODE_MULTIPLE;
        _this._horizontalScroll = false;
        _this.showPaginatorFirstLastButtons = true;
        _this.autoAlignTitles = false;
        _this.multipleSort = true;
        _this.orderable = true;
        _this.resizable = true;
        _this.autoAdjust = true;
        _this.groupable = true;
        _this.expandGroupsSameLevel = true;
        _this.collapseGroupedColumns = false;
        _this.virtualScroll = true;
        _this.contextMenu = true;
        _this.keepSelectedItems = true;
        _this.exportMode = Codes.EXPORT_MODE_VISIBLE;
        _this.searcheableColumns = [];
        _this.groupedColumnTypes = [];
        _this._visibleColArray = [];
        _this.sortColArray = [];
        _this.pendingQuery = false;
        _this.pendingQueryFilter = undefined;
        _this.setStaticData = false;
        _this.avoidQueryColumns = [];
        _this.asyncLoadColumns = [];
        _this.asyncLoadSubscriptions = {};
        _this.finishQuerySubscription = false;
        _this.onRowSelected = new EventEmitter();
        _this.onRowDeselected = new EventEmitter();
        _this.onRowDeleted = new EventEmitter();
        _this.onReinitialize = new EventEmitter();
        _this.onContentChange = new EventEmitter();
        _this.onFilterByColumnChange = new EventEmitter();
        _this.showTotalsSubject = new BehaviorSubject(false);
        _this.showTotals = _this.showTotalsSubject.asObservable();
        _this.loadingSortingSubject = new BehaviorSubject(false);
        _this.loadingSorting = _this.loadingSortingSubject.asObservable();
        _this.loadingScrollSubject = new BehaviorSubject(false);
        _this.loadingScroll = _this.loadingScrollSubject.asObservable();
        _this.showLoading = combineLatest([
            _this.loading.pipe(debounceTime(200)),
            _this.loadingSorting,
            _this.loadingScroll
        ]).pipe(distinctUntilChanged(function (prev, curr) { return prev[0] === curr[0] && prev[1] === curr[1] && prev[2] === curr[2]; }), map(function (res) { return res.some(function (r) { return r; }); }));
        _this.showFirstInsertableRow = false;
        _this.showLastInsertableRow = false;
        _this.clickDelay = 200;
        _this.clickPrevent = false;
        _this.onUpdateScrolledState = new EventEmitter();
        _this.storePaginationState = false;
        _this.pageScrollVirtual = 1;
        _this.groupedColumnsArray = [];
        _this._isColumnFiltersActive = false;
        _this.groupingHeadersRows = [];
        _this.triggerSelectionEvents = true;
        _this._oTableOptions = new DefaultOTableOptions();
        try {
            _this.tabGroupContainer = _this.injector.get(MatTabGroup);
            _this.tabContainer = _this.injector.get(MatTab);
        }
        catch (error) {
        }
        _this.snackBarService = _this.injector.get(SnackBarService);
        _this.getGlobalConfig();
        return _this;
    }
    Object.defineProperty(OTableComponent.prototype, "cdkVirtualScrollViewport", {
        set: function (value) {
            var _this = this;
            if (value != this.virtualScrollViewport) {
                this.virtualScrollViewport = value;
                this.updateHeaderAndFooterStickyPositions();
                if (this.checkViewportSizeSubscription) {
                    this.checkViewportSizeSubscription.unsubscribe();
                }
                if (this.virtualScrollViewport) {
                    this.checkViewportSizeSubscription = this.checkViewPortSubject.subscribe(function (x) {
                        if (x) {
                            _this.checkViewportSize();
                        }
                    });
                }
                this.setDatasource();
                this.registerSortListener();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "diameterSpinner", {
        get: function () {
            var minHeight = OTableComponent.DEFAULT_BASE_SIZE_SPINNER;
            var height = 0;
            if (this.spinnerContainer && this.spinnerContainer.nativeElement) {
                height = this.spinnerContainer.nativeElement.offsetHeight;
            }
            if (height > 0 && height <= 100) {
                return Math.floor(height - (height * 0.1));
            }
            else {
                return minHeight;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "oTableOptions", {
        get: function () {
            return this._oTableOptions;
        },
        set: function (value) {
            this._oTableOptions = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "quickFilter", {
        get: function () {
            return this._quickFilter;
        },
        set: function (value) {
            value = Util.parseBoolean(String(value));
            this._quickFilter = value;
            this._oTableOptions.filter = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "filterCaseSensitive", {
        get: function () {
            return this._filterCaseSensitive;
        },
        set: function (value) {
            this._filterCaseSensitive = BooleanConverter(value);
            if (this._oTableOptions) {
                this._oTableOptions.filterCaseSensitive = this._filterCaseSensitive;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "horizontalScroll", {
        get: function () {
            return this._horizontalScroll;
        },
        set: function (value) {
            this._horizontalScroll = BooleanConverter(value);
            this.refreshColumnsWidth();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "enabled", {
        get: function () {
            return this.oenabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "selectAllCheckboxVisible", {
        get: function () {
            return this._selectAllCheckboxVisible;
        },
        set: function (value) {
            this._selectAllCheckboxVisible = BooleanConverter(value);
            if (this.state) {
                this._selectAllCheckboxVisible = BooleanConverter(this.state.selectColumnVisible);
            }
            this._oTableOptions.selectColumn.visible = this._selectAllCheckboxVisible;
            this.initializeCheckboxColumn();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "visibleColArray", {
        get: function () {
            return this._visibleColArray;
        },
        set: function (arg) {
            var permissionsBlocked = this.permissions ? this.permissions.columns.filter(function (col) { return col.visible === false; }).map(function (col) { return col.attr; }) : [];
            var permissionsChecked = arg.filter(function (value) { return permissionsBlocked.indexOf(value) === -1; });
            this._visibleColArray = permissionsChecked;
            if (this._oTableOptions) {
                var containsSelectionCol = this._oTableOptions.visibleColumns.indexOf(Codes.NAME_COLUMN_SELECT) !== -1;
                var containsExpandableCol = this._oTableOptions.visibleColumns.indexOf(Codes.NAME_COLUMN_EXPANDABLE) !== -1;
                if (containsSelectionCol) {
                    this._visibleColArray.unshift(Codes.NAME_COLUMN_SELECT);
                }
                if (containsSelectionCol && containsExpandableCol) {
                    this._visibleColArray = [this._visibleColArray[0]].concat(Codes.NAME_COLUMN_EXPANDABLE, this._visibleColArray.splice(1));
                }
                else {
                    if (containsExpandableCol) {
                        this._visibleColArray.unshift(Codes.NAME_COLUMN_EXPANDABLE);
                    }
                }
                this._oTableOptions.visibleColumns = this._visibleColArray;
                this.groupingHeadersRows = this._oTableOptions.visibleColumns.map(function (visibleCol) { return 'groupHeader-' + visibleCol; });
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "currentPage", {
        get: function () {
            return this._currentPage;
        },
        set: function (val) {
            this._currentPage = val;
            if (this.paginator) {
                this.paginator.pageIndex = val;
                if (this.matpaginator) {
                    this.matpaginator.pageIndex = val;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.updateScrolledState = function () {
        var _this = this;
        if (this.horizontalScroll) {
            setTimeout(function () {
                var bodyWidth = _this.tableBodyEl.nativeElement.clientWidth;
                var scrollWidth = _this.tableBodyEl.nativeElement.scrollWidth;
                var previousState = _this.horizontalScrolled;
                _this.horizontalScrolled = scrollWidth > bodyWidth;
                if (previousState !== _this.horizontalScrolled) {
                    _this.onUpdateScrolledState.emit(_this.horizontalScrolled);
                }
            }, 0);
        }
        this.refreshColumnsWidth();
        this.checkViewportSize();
    };
    Object.defineProperty(OTableComponent.prototype, "isColumnFiltersActive", {
        get: function () {
            return this._isColumnFiltersActive;
        },
        set: function (val) {
            this._isColumnFiltersActive = val;
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.getGlobalConfig = function () {
        try {
            this.oTableGlobalConfig = this.injector.get(O_TABLE_GLOBAL_CONFIG);
            this.autoAdjust = this.oTableGlobalConfig.autoAdjust;
        }
        catch (error) {
        }
    };
    Object.defineProperty(OTableComponent.prototype, "state", {
        get: function () {
            return this.componentStateService.state;
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.ngOnInit = function () {
        this.initialize();
        if (this.oTableButtons && this.tableButtons && this.tableButtons.length > 0) {
            this.oTableButtons.registerButtons(this.tableButtons.toArray());
        }
    };
    OTableComponent.prototype.ngAfterViewInit = function () {
        _super.prototype.afterViewInit.call(this);
        this.initTableAfterViewInit();
        if (this.oTableMenu) {
            this.matMenu = this.oTableMenu.matMenu;
            this.oTableMenu.registerOptions(this.tableOptions.toArray());
        }
        if (this.tableRowExpandable) {
            this.expandableItem = new SelectionModel(this.tableRowExpandable.multiple, []);
            this.createExpandableColumn();
        }
    };
    OTableComponent.prototype.ngAfterContentInit = function () {
        if (this.tableColumnSelectAllContentChild) {
            this.setCustomDefinitionInSelectColumn(this.tableColumnSelectAllContentChild);
        }
    };
    OTableComponent.prototype.setCustomDefinitionInSelectColumn = function (definition) {
        if (definition.title) {
            this._oTableOptions.selectColumn.title = definition.title;
        }
        if (definition.width) {
            this._oTableOptions.selectColumn.width = definition.width;
        }
        if (definition.minWidth) {
            this._oTableOptions.selectColumn.minWidth = definition.minWidth;
        }
        if (definition.maxWidth) {
            this._oTableOptions.selectColumn.maxWidth = definition.maxWidth;
        }
        if (definition.resizable) {
            this._oTableOptions.selectColumn.resizable = definition.resizable;
        }
    };
    OTableComponent.prototype.ngAfterViewChecked = function () {
        this.cd.detectChanges();
    };
    OTableComponent.prototype.ngOnChanges = function (changes) {
        if (Util.isDefined(changes.autoAdjust) && changes.autoAdjust.currentValue !== changes.autoAdjust.previousValue) {
            this.autoAdjust = changes.autoAdjust.currentValue;
        }
    };
    OTableComponent.prototype.updateHeaderAndFooterStickyPositions = function () {
        var _this = this;
        if (this.virtualScrollSubscription) {
            this.virtualScrollSubscription.unsubscribe();
        }
        if (this.virtualScrollViewport) {
            var zone_1 = this.injector.get(NgZone);
            this.virtualScrollSubscription = this.scrollStrategy.stickyChange.pipe(distinctUntilChanged(), filter(function () { return _this.fixedHeader || _this.hasInsertableRow(); })).subscribe(function (x) {
                zone_1.run(function () {
                    _this.elRef.nativeElement.querySelectorAll(stickyHeaderSelector).forEach(function (el) {
                        el.style.top = -x + 'px';
                    });
                    _this.elRef.nativeElement.querySelectorAll(stickyFooterSelector).forEach(function (el) {
                        el.style.bottom = x + 'px';
                    });
                });
            });
        }
    };
    OTableComponent.prototype.createExpandableColumn = function () {
        this._oTableOptions.expandableColumn = new OColumn();
        this._oTableOptions.expandableColumn.visible = this.tableRowExpandable && this.tableRowExpandable.expandableColumnVisible;
        this.updateStateExpandedColumn();
    };
    OTableComponent.prototype.ngOnDestroy = function () {
        if (this.portalHost) {
            this.portalHost.forEach(function (x) { return x.detach(); });
        }
        this.destroy();
    };
    OTableComponent.prototype.getSuffixColumnInsertable = function () {
        return Codes.SUFFIX_COLUMN_INSERTABLE;
    };
    OTableComponent.prototype.getActionsPermissions = function () {
        return this.permissions ? (this.permissions.actions || []) : [];
    };
    OTableComponent.prototype.getMenuPermissions = function () {
        var result = this.permissions ? this.permissions.menu : undefined;
        return result ? result : {
            visible: true,
            enabled: true,
            items: []
        };
    };
    OTableComponent.prototype.getOColumnPermissions = function (attr) {
        var columns = this.permissions ? (this.permissions.columns || []) : [];
        return columns.find(function (comp) { return comp.attr === attr; }) || { attr: attr, enabled: true, visible: true };
    };
    OTableComponent.prototype.getActionPermissions = function (attr) {
        var actionsPerm = this.permissions ? (this.permissions.actions || []) : [];
        var permissions = actionsPerm.find(function (p) { return p.attr === attr; });
        return permissions || {
            attr: attr,
            visible: true,
            enabled: true
        };
    };
    OTableComponent.prototype.checkEnabledActionPermission = function (attr) {
        var actionsPerm = this.permissions ? (this.permissions.actions || []) : [];
        var permissions = actionsPerm.find(function (p) { return p.attr === attr; });
        var enabledPermision = PermissionsUtils.checkEnabledPermission(permissions);
        if (!enabledPermision) {
            this.snackBarService.open('MESSAGES.OPERATION_NOT_ALLOWED_PERMISSION');
        }
        return enabledPermision;
    };
    OTableComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        this._oTableOptions = new DefaultOTableOptions();
        if (this.tabGroupContainer && this.tabContainer) {
            this.registerTabListener();
        }
        this.initializeParams();
        this.initializeDao();
        this.permissions = this.permissionsService.getTablePermissions(this.oattr, this.actRoute);
    };
    OTableComponent.prototype.initializeDao = function () {
        var queryMethodName = this.pageable ? this.paginatedQueryMethod : this.queryMethod;
        var methods = {
            query: queryMethodName,
            update: this.updateMethod,
            delete: this.deleteMethod,
            insert: this.insertMethod
        };
        if (this.staticData) {
            this.queryOnBind = false;
            this.queryOnInit = false;
            this.daoTable = new OTableDao(undefined, this.entity, methods);
            this.setDataArray(this.staticData);
        }
        else {
            this.configureService();
            this.daoTable = new OTableDao(this.dataService, this.entity, methods);
        }
    };
    OTableComponent.prototype.reinitialize = function (options) {
        if (options) {
            var clonedOpts = Object.assign({}, options);
            if (clonedOpts.hasOwnProperty('entity')) {
                this.entity = clonedOpts.entity;
            }
            if (clonedOpts.hasOwnProperty('service')) {
                this.service = clonedOpts.service;
            }
            if (clonedOpts.hasOwnProperty('columns')) {
                this.columns = clonedOpts.columns;
            }
            if (clonedOpts.hasOwnProperty('visibleColumns')) {
                this.visibleColumns = clonedOpts.visibleColumns;
            }
            if (clonedOpts.hasOwnProperty('defaultVisibleColumns')) {
                this.defaultVisibleColumns = clonedOpts.defaultVisibleColumns;
            }
            if (clonedOpts.hasOwnProperty('keys')) {
                this.keys = clonedOpts.keys;
            }
            if (clonedOpts.hasOwnProperty('sortColumns')) {
                this.sortColumns = clonedOpts.sortColumns;
            }
            if (clonedOpts.hasOwnProperty('parentKeys')) {
                this.parentKeys = clonedOpts.parentKeys;
            }
            if (clonedOpts.hasOwnProperty('filterColumns')) {
                if (!this.oTableColumnsFilterComponent) {
                    this.oTableColumnsFilterComponent = new OTableColumnsFilterComponent(this.injector, this);
                    this.oTableMenu.onVisibleFilterOptionChange.next(this.filterColumnActiveByDefault);
                    this.oTableColumnsFilterComponent.columns = clonedOpts.filterColumns;
                }
                else {
                    this.oTableColumnsFilterComponent.columns = this.searcheableColumns.join(";");
                }
            }
        }
        this.destroy();
        this.initialize();
        this.state.reset(this.pageable);
        this.initTableAfterViewInit();
        this.onReinitialize.emit(null);
    };
    OTableComponent.prototype.initTableAfterViewInit = function () {
        this.parseVisibleColumns();
        this.parseSearcheableColumns();
        this.setDatasource();
        this.parseGroupedColumns();
        this.parseGroupedColumnTypes();
        this.parseSortColumns();
        this.registerSortListener();
        this.setFiltersConfiguration();
        this.addDefaultRowButtons();
        if (Util.isDefined(this.oTableColumnsGroupingComponent)) {
            this.setGroupColumns(this.oTableColumnsGroupingComponent.columnsArray);
        }
        if (this.queryOnInit) {
            this.queryData();
        }
    };
    OTableComponent.prototype.parseSearcheableColumns = function () {
        var _this = this;
        this.searcheableColumns = this.visibleColArray.filter(function (col) {
            var oCol = _this.getOColumn(col);
            return oCol && oCol.searchable;
        });
    };
    OTableComponent.prototype.destroy = function () {
        var _this = this;
        _super.prototype.destroy.call(this);
        if (this.tabGroupChangeSubscription) {
            this.tabGroupChangeSubscription.unsubscribe();
        }
        if (this.selectionChangeSubscription) {
            this.selectionChangeSubscription.unsubscribe();
        }
        if (this.sortSubscription) {
            this.sortSubscription.unsubscribe();
        }
        if (this.onRenderedDataChange) {
            this.onRenderedDataChange.unsubscribe();
        }
        if (this.contextMenuSubscription) {
            this.contextMenuSubscription.unsubscribe();
        }
        if (this.virtualScrollSubscription) {
            this.virtualScrollSubscription.unsubscribe();
        }
        if (this.checkViewportSizeSubscription) {
            this.checkViewportSizeSubscription.unsubscribe();
        }
        if (this.scrollStrategy) {
            this.scrollStrategy.destroy();
        }
        Object.keys(this.asyncLoadSubscriptions).forEach(function (idx) {
            if (_this.asyncLoadSubscriptions[idx]) {
                _this.asyncLoadSubscriptions[idx].unsubscribe();
            }
        });
    };
    OTableComponent.prototype.getDataToStore = function () {
        return this.componentStateService.getDataToStore();
    };
    OTableComponent.prototype.registerQuickFilter = function (arg) {
        var _this = this;
        var quickFilter = arg;
        this.quickFilterComponent = undefined;
        this.oTableQuickFilterComponent = quickFilter;
        if (Util.isDefined(this.oTableQuickFilterComponent)) {
            this.oTableQuickFilterComponent.setValue(this.state.quickFilterValue, false);
            this.quickFilterSubscription = this.oTableQuickFilterComponent.onChange.subscribe(function (val) {
                _this.onSearch.emit(val);
            });
        }
    };
    OTableComponent.prototype.registerPagination = function (value) {
        this.paginationControls = true;
        this.paginator = value;
    };
    OTableComponent.prototype.registerContextMenu = function (value) {
        var _this = this;
        this.tableContextMenu = value;
        this.contextMenuSubscription = this.tableContextMenu.onShow.subscribe(function (params) {
            params.class = 'o-table-context-menu ' + _this.rowHeight;
            if (params.data && !_this.isRowSelected(params.data.rowValue)) {
                _this.clearSelection();
                _this.selectedRow(params.data.rowValue);
            }
        });
    };
    OTableComponent.prototype.registerDefaultColumn = function (column) {
        if (Util.isDefined(this.getOColumn(column))) {
            return;
        }
        var colDef = this.createOColumn(column, this);
        this.pushOColumnDefinition(colDef);
    };
    OTableComponent.prototype.registerColumn = function (column) {
        var columnAttr = (typeof column === 'string') ? column : column.attr;
        var columnPermissions = this.getOColumnPermissions(columnAttr);
        if (!columnPermissions.visible) {
            return;
        }
        if (typeof column === 'string') {
            this.registerDefaultColumn(column);
            return;
        }
        var columnDef = this.getOColumn(column.attr);
        if (Util.isDefined(columnDef) && Util.isDefined(columnDef.definition)) {
            return;
        }
        var colDef = this.createOColumn(column.attr, this, column);
        var columnWidth = column.width;
        var storedData = this.state.getColumnDisplay(colDef);
        if (Util.isDefined(storedData) && Util.isDefined(storedData.width)) {
            if (this.state.initialConfiguration.columnsDisplay) {
                var initialStoredData = this.state.initialConfiguration.getColumnDisplay(colDef);
                if (initialStoredData && initialStoredData.width === colDef.definition.originalWidth) {
                    columnWidth = storedData.width;
                }
            }
            else {
                columnWidth = storedData.width;
            }
        }
        if (Util.isDefined(columnWidth)) {
            colDef.width = columnWidth;
        }
        if (column && (column.asyncLoad || column.type === 'action')) {
            this.avoidQueryColumns.push(column.attr);
            if (column.asyncLoad) {
                this.asyncLoadColumns.push(column.attr);
            }
        }
        this.pushOColumnDefinition(colDef);
    };
    OTableComponent.prototype.pushOColumnDefinition = function (colDef) {
        colDef.visible = (this._visibleColArray.indexOf(colDef.attr) !== -1);
        var alreadyExisting = this.getOColumn(colDef.attr);
        if (alreadyExisting !== undefined) {
            var replacingIndex = this._oTableOptions.columns.indexOf(alreadyExisting);
            this._oTableOptions.columns[replacingIndex] = colDef;
        }
        else {
            this._oTableOptions.columns.push(colDef);
        }
        this.ensureColumnsOrder();
        this.refreshEditionModeWarn();
    };
    OTableComponent.prototype.refreshEditionModeWarn = function () {
        var _this = this;
        if (this.editionMode !== Codes.EDITION_MODE_NONE) {
            return;
        }
        var editableColumns = this._oTableOptions.columns.filter(function (col) {
            return Util.isDefined(col.editor);
        });
        setTimeout(function () {
            if (editableColumns.length > 0 && !_this.hasInsertableRow()) {
                console.warn('Using a column with a editor but there is no edition-mode defined');
            }
        }, 100);
    };
    OTableComponent.prototype.registerColumnAggregate = function (column) {
        this.showTotalsSubject.next(true);
        var alreadyExisting = this.getOColumn(column.attr);
        if (alreadyExisting !== undefined) {
            var replacingIndex = this._oTableOptions.columns.indexOf(alreadyExisting);
            this._oTableOptions.columns[replacingIndex].aggregate = column;
        }
    };
    OTableComponent.prototype.registerTableHeaders = function (tableHeader) {
        if (this.tableHeaders.findIndex(function (header) { return header.column.attr === tableHeader.column.attr; }) === -1) {
            this.tableHeaders.push(tableHeader);
        }
    };
    OTableComponent.prototype.parseVisibleColumns = function (defaultConfiguration) {
        var _this = this;
        if (defaultConfiguration === void 0) { defaultConfiguration = false; }
        if (this.state.columnsDisplay) {
            var stateCols_1 = [];
            this.state.columnsDisplay.forEach(function (oCol, index) {
                var isVisibleColInColumns = _this._oTableOptions.columns.find(function (col) { return col.attr === oCol.attr; }) !== undefined;
                if (isVisibleColInColumns) {
                    stateCols_1.push(oCol);
                }
                else {
                    console.warn('Unable to load the column ' + oCol.attr + ' from the localstorage');
                }
            });
            if (defaultConfiguration) {
                stateCols_1 = this.state.initialConfiguration.columnsDisplay;
            }
            else {
                stateCols_1 = this.checkChangesVisibleColummnsInInitialConfiguration(stateCols_1);
            }
            this._oTableOptions.columns.sort(function (a, b) {
                var indexA = stateCols_1.findIndex(function (col) { return col.attr === a.attr; });
                var indexB = stateCols_1.findIndex(function (col) { return col.attr === b.attr; });
                return indexA - indexB;
            });
            this.visibleColArray = stateCols_1.filter(function (item) { return item.visible; }).map(function (item) { return item.attr; });
        }
        else {
            this.visibleColArray = Util.parseArray(this.defaultVisibleColumns ? this.defaultVisibleColumns : this.visibleColumns, true);
            this._oTableOptions.columns.sort(function (a, b) { return _this.visibleColArray.indexOf(a.attr) - _this.visibleColArray.indexOf(b.attr); });
        }
    };
    OTableComponent.prototype.checkChangesVisibleColummnsInInitialConfiguration = function (stateCols) {
        if (this.state.initialConfiguration.columnsDisplay) {
            var originalVisibleColArray = this.state.initialConfiguration.columnsDisplay.filter(function (x) { return x.visible; }).map(function (x) { return x.attr; });
            var visibleColArray_1 = Util.parseArray(this.visibleColumns, true);
            var colToAddInVisibleCol = Util.differenceArrays(visibleColArray_1, originalVisibleColArray);
            colToAddInVisibleCol.forEach(function (colAdd) {
                var indexCol = stateCols.findIndex(function (col) { return col.attr === colAdd; });
                if (indexCol > -1) {
                    stateCols[indexCol].visible = true;
                }
                stateCols.sort(function (a, b) { return visibleColArray_1.indexOf(a.attr) - visibleColArray_1.indexOf(b.attr); });
            });
            var colToDeleteInVisibleCol_1 = Util.differenceArrays(originalVisibleColArray, visibleColArray_1);
            if (colToDeleteInVisibleCol_1.length > 0) {
                stateCols = stateCols.filter(function (col) { return colToDeleteInVisibleCol_1.indexOf(col.attr) === -1; });
            }
            var changeSortVisibleColumns = JSON.stringify(visibleColArray_1) !== JSON.stringify(originalVisibleColArray);
            if (changeSortVisibleColumns && visibleColArray_1.length === originalVisibleColArray.length) {
                visibleColArray_1.forEach(function (col, toIndex) {
                    var fromIndexToChange = stateCols.findIndex(function (stateCol) { return stateCol.attr === col; });
                    moveItemInArray(stateCols, fromIndexToChange, toIndex);
                });
            }
        }
        return stateCols;
    };
    OTableComponent.prototype.parseSortColumns = function () {
        var _this = this;
        var sortColumnsParam = this.state.sortColumns || this.sortColumns;
        this.sortColArray = ServiceUtils.parseSortColumns(sortColumnsParam);
        if (this.state.sortColumns && this.state.initialConfiguration.sortColumns) {
            var initialConfigSortColumnsArray = ServiceUtils.parseSortColumns(this.state.initialConfiguration.sortColumns);
            var originalSortColumnsArray = ServiceUtils.parseSortColumns(this.sortColumns);
            var colToAddInVisibleCol = Util.differenceArrays(originalSortColumnsArray, initialConfigSortColumnsArray);
            if (colToAddInVisibleCol.length > 0) {
                colToAddInVisibleCol.forEach(function (colAdd) {
                    _this.sortColArray.push(colAdd);
                });
            }
            var colToDelInVisibleCol = Util.differenceArrays(initialConfigSortColumnsArray, originalSortColumnsArray);
            if (colToDelInVisibleCol.length > 0) {
                colToDelInVisibleCol.forEach(function (colDel) {
                    _this.sortColArray.forEach(function (col, i) {
                        if (col.columnName === colDel.columnName) {
                            _this.sortColArray.splice(i, 1);
                        }
                    });
                });
            }
        }
        for (var i = this.sortColArray.length - 1; i >= 0; i--) {
            var colName = this.sortColArray[i].columnName;
            var oCol = this.getOColumn(colName);
            if (!Util.isDefined(oCol) || !oCol.orderable) {
                this.sortColArray.splice(i, 1);
            }
        }
    };
    OTableComponent.prototype.ensureColumnsOrder = function () {
        var _this = this;
        var columnsOrder = [];
        if (this.state.columnsDisplay) {
            columnsOrder = this.state.columnsDisplay.map(function (item) { return item.attr; });
        }
        else {
            columnsOrder = this.colArray.filter(function (attr) { return _this.visibleColArray.indexOf(attr) === -1; });
            columnsOrder.push.apply(columnsOrder, tslib_1.__spread(this.visibleColArray));
        }
        this._oTableOptions.columns.sort(function (a, b) {
            if (columnsOrder.indexOf(a.attr) === -1) {
                return 0;
            }
            else {
                return columnsOrder.indexOf(a.attr) - columnsOrder.indexOf(b.attr);
            }
        });
    };
    OTableComponent.prototype.initializeParams = function () {
        var _this = this;
        if (!this.visibleColumns) {
            this.visibleColumns = this.columns;
        }
        if (this.colArray.length) {
            this.colArray.forEach(function (x) { return _this.registerColumn(x); });
            this.ensureColumnsOrder();
        }
        this._oTableOptions.filter = this.quickFilter;
        if (this.state.currentPage) {
            this.currentPage = this.state.currentPage;
        }
        if (!this.paginator && this.paginationControls) {
            this.paginator = new OBaseTablePaginator();
            this.paginator.pageSize = this.queryRows;
            this.paginator.pageIndex = this.currentPage;
            this.paginator.showFirstLastButtons = this.showPaginatorFirstLastButtons;
        }
        if (!Util.isDefined(this.selectAllCheckboxVisible)) {
            this._oTableOptions.selectColumn.visible = !!this.state.selectColumnVisible;
        }
        else {
            if (Util.isDefined(this.state.initialConfiguration.selectColumnVisible)
                && this.selectAllCheckboxVisible === this.state.initialConfiguration.selectColumnVisible) {
                this._oTableOptions.selectColumn.visible = !!this.state.selectColumnVisible;
            }
            else {
                this._oTableOptions.selectColumn.visible = this.selectAllCheckboxVisible;
            }
        }
        this.isColumnFiltersActive = this.filterColumnActiveByDefault;
        this.initializeCheckboxColumn();
        if (this.storeState) {
            if (Util.isDefined(this.state.queryRows) && Util.isDefined(this.state.initialConfiguration.queryRows)
                && this.state.initialConfiguration.queryRows === this.originalQueryRows) {
                this.queryRows = this.state.queryRows;
            }
        }
    };
    OTableComponent.prototype.updateStateExpandedColumn = function () {
        if (!this.tableRowExpandable || !this.tableRowExpandable.expandableColumnVisible) {
            return;
        }
        if (this._oTableOptions.visibleColumns[0] === Codes.NAME_COLUMN_SELECT && this._oTableOptions.visibleColumns[1] !== Codes.NAME_COLUMN_EXPANDABLE) {
            this._oTableOptions.visibleColumns = [this._oTableOptions.visibleColumns[0]].concat(Codes.NAME_COLUMN_EXPANDABLE, this._oTableOptions.visibleColumns.splice(1));
        }
        else if (this._oTableOptions.visibleColumns[0] !== Codes.NAME_COLUMN_EXPANDABLE) {
            this._oTableOptions.visibleColumns.unshift(Codes.NAME_COLUMN_EXPANDABLE);
        }
    };
    OTableComponent.prototype.registerTabListener = function () {
        var _this = this;
        this.tabGroupChangeSubscription = this.tabGroupContainer.selectedTabChange.subscribe(function (evt) {
            var interval;
            var timerCallback = function (tab) {
                if (tab && tab.content.isAttached) {
                    clearInterval(interval);
                    if (tab === _this.tabContainer) {
                        _this.insideTabBugWorkaround();
                        if (_this.pendingQuery) {
                            _this.queryData(_this.pendingQueryFilter);
                        }
                        _this.checkViewportSize();
                    }
                }
            };
            interval = setInterval(function () { timerCallback(evt.tab); }, 100);
        });
    };
    OTableComponent.prototype.insideTabBugWorkaround = function () {
        this.refreshSortHeaders();
    };
    OTableComponent.prototype.registerSortListener = function () {
        if (Util.isDefined(this.sort)) {
            this.sortSubscription = this.sort.oSortChange.subscribe(this.onSortChange.bind(this));
            this.sort.setMultipleSort(this.multipleSort);
        }
        if (this.sortColumns && this.staticData) {
            this.updateSortingSubject(true);
        }
    };
    OTableComponent.prototype.updateSortingSubject = function (value) {
        var _this = this;
        var ngZone = this.injector.get(NgZone);
        if (ngZone) {
            ngZone.run(function () { return _this.loadingSortingSubject.next(value); });
        }
        else {
            this.loadingSortingSubject.next(value);
            if (this.cd && !this.cd.destroyed) {
                this.cd.detectChanges();
            }
        }
    };
    OTableComponent.prototype.onSortChange = function (sortArray) {
        var _this = this;
        this.sortColArray = [];
        sortArray.forEach(function (sort) {
            if (sort.direction !== '') {
                _this.sortColArray.push({
                    columnName: sort.id,
                    ascendent: sort.direction === Codes.ASC_SORT
                });
            }
        });
        if (this.pageable) {
            this.reloadData();
        }
        else {
            this.updateSortingSubject(true);
        }
    };
    OTableComponent.prototype.setDatasource = function () {
        delete this.dataSource;
        if (this.onRenderedDataChange)
            this.onRenderedDataChange.unsubscribe();
        var dataSourceService = this.injector.get(OTableDataSourceService);
        this.dataSource = dataSourceService.getInstance(this);
        this.registerDataSourceListeners();
    };
    OTableComponent.prototype.registerDataSourceListeners = function () {
        var _this = this;
        this.onRenderedDataChange = this.dataSource.onRenderedDataChange.subscribe(function () {
            _this.stopEdition();
            _this.checkSelectedItemData();
            if (!_this.pageable) {
                setTimeout(function () {
                    _this.updateSortingSubject(false);
                    if (_this.cd && !_this.cd.destroyed) {
                        _this.cd.detectChanges();
                    }
                }, 500);
            }
        });
    };
    OTableComponent.prototype.getExpandedRowContainerClass = function (rowIndex) {
        return OTableComponent.EXPANDED_ROW_CONTAINER_CLASS + rowIndex;
    };
    OTableComponent.prototype.getExpandableItems = function () {
        return this.expandableItem.selected;
    };
    OTableComponent.prototype.toogleRowExpandable = function (item, rowIndex, event) {
        if (event) {
            event.stopPropagation();
            event.preventDefault();
        }
        this.expandableItem.toggle(item);
        if (this.portalHost[rowIndex]) {
            this.portalHost[rowIndex].detach();
        }
        if (this.getStateExpand(item) === 'collapsed') {
            var eventTableRowExpandableChange = this.emitTableRowExpandableChangeEvent(item, rowIndex);
            this.tableRowExpandable.onCollapsed.emit(eventTableRowExpandableChange);
        }
        else {
            this.portalHost[rowIndex] = new DomPortalOutlet(this.elRef.nativeElement.querySelector('.' + this.getExpandedRowContainerClass(rowIndex)), this._componentFactoryResolver, this.appRef, this.injector);
            var templatePortal = new TemplatePortal(this.tableRowExpandable.templateRef, this._viewContainerRef, { $implicit: item });
            this.portalHost[rowIndex].attachTemplatePortal(templatePortal);
            var eventTableRowExpandableChange = this.emitTableRowExpandableChangeEvent(item, rowIndex);
            this.tableRowExpandable.onExpanded.emit(eventTableRowExpandableChange);
        }
    };
    OTableComponent.prototype.emitTableRowExpandableChangeEvent = function (data, rowIndex) {
        var event = new OTableRowExpandedChange();
        event.rowIndex = rowIndex;
        event.data = data;
        return event;
    };
    OTableComponent.prototype.isExpanded = function (data) {
        return this.expandableItem.isSelected(data);
    };
    OTableComponent.prototype.getStateExpand = function (row) {
        return this.isExpanded(row) ? 'expanded' : 'collapsed';
    };
    OTableComponent.prototype.isColumnExpandable = function () {
        return (Util.isDefined(this.tableRowExpandable) && Util.isDefined(this._oTableOptions.expandableColumn)) ? this._oTableOptions.expandableColumn.visible : false;
    };
    Object.defineProperty(OTableComponent.prototype, "hasExpandedRow", {
        get: function () {
            return Util.isDefined(this.tableRowExpandable);
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.hasInsertableRow = function () {
        return Util.isDefined(this.oTableInsertableRowComponent);
    };
    OTableComponent.prototype.getNumVisibleColumns = function () {
        return this.oTableOptions.visibleColumns.length;
    };
    OTableComponent.prototype.queryData = function (filter, ovrrArgs) {
        var _this = this;
        if (this.isInsideInactiveTab()) {
            this.pendingQuery = true;
            this.pendingQueryFilter = filter;
            return;
        }
        this.pendingQuery = false;
        this.pendingQueryFilter = undefined;
        this.queryCellRenderers().subscribe(function () {
            _super.prototype.queryData.call(_this, filter, ovrrArgs);
        });
    };
    OTableComponent.prototype.isInsideInactiveTab = function () {
        var result = false;
        if (this.tabContainer && this.tabGroupContainer) {
            result = !(this.tabContainer.isActive || (this.tabGroupContainer.selectedIndex === this.tabContainer.position));
        }
        return result;
    };
    OTableComponent.prototype.getComponentFilter = function (existingFilter) {
        if (existingFilter === void 0) { existingFilter = {}; }
        var filter = existingFilter;
        if (this.pageable) {
            if (Object.keys(filter).length > 0) {
                var parentItemExpr = FilterExpressionUtils.buildExpressionFromObject(filter);
                filter = {};
                filter[FilterExpressionUtils.FILTER_EXPRESSION_KEY] = parentItemExpr;
            }
            var beColFilter = this.getColumnFiltersExpression();
            if (beColFilter && !Util.isDefined(filter[FilterExpressionUtils.FILTER_EXPRESSION_KEY])) {
                filter[FilterExpressionUtils.FILTER_EXPRESSION_KEY] = beColFilter;
            }
            else if (beColFilter) {
                filter[FilterExpressionUtils.FILTER_EXPRESSION_KEY] =
                    FilterExpressionUtils.buildComplexExpression(filter[FilterExpressionUtils.FILTER_EXPRESSION_KEY], beColFilter, FilterExpressionUtils.OP_AND);
            }
        }
        return _super.prototype.getComponentFilter.call(this, filter);
    };
    OTableComponent.prototype.getQuickFilterExpression = function () {
        if (Util.isDefined(this.oTableQuickFilterComponent) && this.pageable) {
            return this.oTableQuickFilterComponent.filterExpression;
        }
        return undefined;
    };
    OTableComponent.prototype.getColumnFiltersExpression = function () {
        var columnFilters = this.dataSource.getColumnValueFilters();
        var beColumnFilters = [];
        columnFilters.forEach(function (colFilter) {
            switch (colFilter.operator) {
                case ColumnValueFilterOperator.IN:
                    if (Util.isArray(colFilter.values)) {
                        var besIn = colFilter.values.map(function (value) { return FilterExpressionUtils.buildExpressionEquals(colFilter.attr, value); });
                        var beIn_1 = besIn.pop();
                        besIn.forEach(function (be) {
                            beIn_1 = FilterExpressionUtils.buildComplexExpression(beIn_1, be, FilterExpressionUtils.OP_OR);
                        });
                        beColumnFilters.push(beIn_1);
                    }
                    break;
                case ColumnValueFilterOperator.BETWEEN:
                    if (Util.isArray(colFilter.values) && colFilter.values.length === 2) {
                        var beFrom = FilterExpressionUtils.buildExpressionMoreEqual(colFilter.attr, colFilter.values[0]);
                        var beTo = FilterExpressionUtils.buildExpressionLessEqual(colFilter.attr, colFilter.values[1]);
                        beColumnFilters.push(FilterExpressionUtils.buildComplexExpression(beFrom, beTo, FilterExpressionUtils.OP_AND));
                    }
                    break;
                case ColumnValueFilterOperator.EQUAL:
                    beColumnFilters.push(FilterExpressionUtils.buildExpressionLike(colFilter.attr, colFilter.values));
                    break;
                case ColumnValueFilterOperator.LESS_EQUAL:
                    beColumnFilters.push(FilterExpressionUtils.buildExpressionLessEqual(colFilter.attr, colFilter.values));
                    break;
                case ColumnValueFilterOperator.MORE_EQUAL:
                    beColumnFilters.push(FilterExpressionUtils.buildExpressionMoreEqual(colFilter.attr, colFilter.values));
                    break;
            }
        });
        var beColFilter = beColumnFilters.pop();
        beColumnFilters.forEach(function (be) {
            beColFilter = FilterExpressionUtils.buildComplexExpression(beColFilter, be, FilterExpressionUtils.OP_AND);
        });
        return beColFilter;
    };
    OTableComponent.prototype.updatePaginationInfo = function (queryRes) {
        _super.prototype.updatePaginationInfo.call(this, queryRes);
    };
    OTableComponent.prototype.initViewPort = function (data) {
        if (this.virtualScrollViewport && data) {
            var headerElRef = this.elRef.nativeElement.querySelector(headerSelector);
            var footerElRef = this.elRef.nativeElement.querySelector(footerSelector);
            var rowElRef = this.elRef.nativeElement.querySelector(rowSelector);
            var headerHeight = headerElRef ? headerElRef.offsetHeight : 0;
            var footerHeight = footerElRef ? footerElRef.offsetHeight : 0;
            var rowHeight = rowElRef ? rowElRef.offsetHeight : OTableComponent.DEFAULT_ROW_HEIGHT;
            this.scrollStrategy.setConfig(rowHeight, headerHeight, footerHeight);
            if (this.previousRendererData !== this.dataSource.renderedData) {
                this.scrollStrategy.dataLength = data.length;
            }
        }
    };
    OTableComponent.prototype.setData = function (data, sqlTypes) {
        this.daoTable.sqlTypesChange.next(sqlTypes);
        this.daoTable.setDataArray(data);
        this.updateScrolledState();
        if (this.pageable) {
            ObservableWrapper.callEmit(this.onPaginatedDataLoaded, data);
        }
        ObservableWrapper.callEmit(this.onDataLoaded, this.daoTable.data);
    };
    OTableComponent.prototype.showDialogError = function (error, errorOptional) {
        if (Util.isDefined(error) && !Util.isObject(error)) {
            this.dialogService.alert('ERROR', error);
        }
        else {
            this.dialogService.alert('ERROR', errorOptional);
        }
    };
    OTableComponent.prototype.projectContentChanged = function () {
        var _this = this;
        setTimeout(function () {
            _this.updateSortingSubject(false);
        }, 500);
        this.loadingScrollSubject.next(false);
        this.initViewPort(this.dataSource.renderedData);
        if (this.previousRendererData !== this.dataSource.renderedData) {
            this.previousRendererData = this.dataSource.renderedData;
            ObservableWrapper.callEmit(this.onContentChange, this.dataSource.renderedData);
        }
        if (this.state.selection && this.dataSource.renderedData.length > 0 && this.getSelectedItems().length === 0) {
            this.checkSelectedItemData();
        }
    };
    OTableComponent.prototype.getAttributesValuesToQuery = function () {
        var columns = _super.prototype.getAttributesValuesToQuery.call(this);
        if (this.avoidQueryColumns.length > 0) {
            for (var i = columns.length - 1; i >= 0; i--) {
                var col = columns[i];
                if (this.avoidQueryColumns.indexOf(col) !== -1) {
                    columns.splice(i, 1);
                }
            }
        }
        return columns;
    };
    OTableComponent.prototype.getQueryArguments = function (filter, ovrrArgs) {
        var queryArguments = _super.prototype.getQueryArguments.call(this, filter, ovrrArgs);
        Object.assign(queryArguments[3], this.getSqlTypesForFilter(queryArguments[1]));
        Object.assign(queryArguments[3], ovrrArgs ? ovrrArgs.sqltypes || {} : {});
        if (this.pageable) {
            queryArguments[5] = this.paginator.isShowingAllRows(queryArguments[5]) ? this.state.totalQueryRecordsNumber : queryArguments[5];
            queryArguments[6] = this.sortColArray;
        }
        return queryArguments;
    };
    OTableComponent.prototype.getSqlTypesForFilter = function (filter) {
        var allSqlTypes = this.getSqlTypes();
        var sqlTypes = {};
        Object.keys(allSqlTypes).forEach(function (key) {
            if (filter.indexOf(key) !== -1 && allSqlTypes[key] !== SQLTypes.OTHER) {
                sqlTypes[key] = allSqlTypes[key];
            }
        });
        return sqlTypes;
    };
    OTableComponent.prototype.onExportButtonClicked = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onExportButtonClicked();
        }
    };
    OTableComponent.prototype.onChangeColumnsVisibilityClicked = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onChangeColumnsVisibilityClicked();
        }
    };
    OTableComponent.prototype.onMatTableContentChanged = function () {
    };
    OTableComponent.prototype.add = function () {
        if (!this.checkEnabledActionPermission(PermissionsUtils.ACTION_INSERT)) {
            return;
        }
        _super.prototype.insertDetail.call(this);
    };
    OTableComponent.prototype.remove = function (clearSelectedItems) {
        var _this = this;
        if (clearSelectedItems === void 0) { clearSelectedItems = false; }
        if (!this.checkEnabledActionPermission(PermissionsUtils.ACTION_DELETE)) {
            return;
        }
        var selectedItems = this.getSelectedItems();
        if (selectedItems.length === 0) {
            return;
        }
        this.dialogService.confirm('CONFIRM', 'MESSAGES.CONFIRM_DELETE').then(function (res) {
            if (res === true) {
                if (_this.dataService && (_this.deleteMethod in _this.dataService) && _this.entity && (_this.keysArray.length > 0)) {
                    var filters = ServiceUtils.getArrayProperties(selectedItems, _this.keysArray);
                    var sqlTypesArg = _this.getSqlTypesOfKeys();
                    _this.daoTable.removeQuery(filters, sqlTypesArg).subscribe(function () {
                        ObservableWrapper.callEmit(_this.onRowDeleted, selectedItems);
                    }, function (error) {
                        _this.showDialogError(error, 'MESSAGES.ERROR_DELETE');
                    }, function () {
                        selectedItems.forEach(function (item) {
                            _this.selection.deselect(item);
                        });
                        _this.reloadData();
                    });
                }
                else {
                    _this.deleteLocalItems();
                }
            }
            else if (clearSelectedItems) {
                _this.clearSelection();
            }
        });
    };
    OTableComponent.prototype.refresh = function () {
        this.reloadData();
    };
    OTableComponent.prototype.showAndSelectAllCheckbox = function () {
        if (this.isSelectionModeMultiple()) {
            if (this.selectAllCheckbox) {
                this._oTableOptions.selectColumn.visible = true;
            }
            this.initializeCheckboxColumn();
            this.selectAll();
        }
    };
    OTableComponent.prototype.reloadPaginatedDataFromStart = function (clearSelectedItems) {
        if (clearSelectedItems === void 0) { clearSelectedItems = true; }
        if (this.pageable) {
            this.currentPage = 0;
            this.reloadData(clearSelectedItems);
        }
    };
    OTableComponent.prototype.reloadData = function (clearSelectedItems) {
        if (clearSelectedItems === void 0) { clearSelectedItems = true; }
        if (!this.checkEnabledActionPermission(PermissionsUtils.ACTION_REFRESH)) {
            return;
        }
        this.componentStateService.refreshSelection();
        if (clearSelectedItems) {
            this.clearSelection();
        }
        this.finishQuerySubscription = false;
        this.pendingQuery = true;
        var queryArgs;
        if (this.pageable) {
            queryArgs = {
                offset: this.currentPage * this.queryRows,
                length: this.queryRows
            };
        }
        this.stopEdition(false);
        this.queryData(void 0, queryArgs);
    };
    OTableComponent.prototype.handleClick = function (row, column, rowIndex, cellRef, event) {
        var _this = this;
        this.clickTimer = setTimeout(function () {
            if (!_this.clickPrevent) {
                if (_this.oenabled && column.editor
                    && (_this.detailMode !== Codes.DETAIL_MODE_CLICK)
                    && (_this.editionMode === Codes.EDITION_MODE_CLICK)) {
                    _this.activateColumnEdition(column, row, cellRef);
                }
                else {
                    _this.doHandleClick(row, column.attr, rowIndex, event);
                }
            }
            _this.clickPrevent = false;
        }, this.clickDelay);
    };
    OTableComponent.prototype.doHandleClick = function (row, column, rowIndex, $event) {
        if (!this.oenabled) {
            return;
        }
        if ((this.detailMode === Codes.DETAIL_MODE_CLICK)) {
            this.onClick.emit({ row: row, rowIndex: rowIndex, mouseEvent: $event, columnName: column, cell: row[column] });
            this.saveDataNavigationInLocalStorage();
            this.clearSelection();
            this.selectedRow(row);
            this.viewDetail(row);
            return;
        }
        if (this.isSelectionModeMultiple() && ($event.ctrlKey || $event.metaKey)) {
            this.selectedRow(row);
            this.onClick.emit({ row: row, rowIndex: rowIndex, mouseEvent: $event, columnName: column, cell: row[column] });
        }
        else if (this.isSelectionModeMultiple() && $event.shiftKey) {
            this.handleMultipleSelection(row);
        }
        else if (!this.isSelectionModeNone()) {
            var selectedItems = this.getSelectedItems();
            if (this.isRowSelected(row) && selectedItems.length === 1 && this.editionEnabled) {
                return;
            }
            else {
                this.clearSelectionAndEditing();
            }
            this.selectedRow(row);
            this.onClick.emit({ row: row, rowIndex: rowIndex, mouseEvent: $event, columnName: column, cell: row[column] });
        }
    };
    OTableComponent.prototype.handleMultipleSelection = function (item) {
        var _this = this;
        if (this.selection.selected.length > 0) {
            var first = this.dataSource.renderedData.indexOf(this.selection.selected[0]);
            var last = this.dataSource.renderedData.indexOf(item);
            var indexFrom = Math.min(first, last);
            var indexTo = Math.max(first, last);
            this.clearSelection();
            this.dataSource.renderedData.slice(indexFrom, indexTo + 1).forEach(function (e) { return _this.selectedRow(e); });
            ObservableWrapper.callEmit(this.onClick, this.selection.selected);
        }
    };
    OTableComponent.prototype.saveDataNavigationInLocalStorage = function () {
        _super.prototype.saveDataNavigationInLocalStorage.call(this);
        this.storePaginationState = true;
    };
    OTableComponent.prototype.handleDoubleClick = function (row, column, rowIndex, cellRef, $event) {
        clearTimeout(this.clickTimer);
        this.clickPrevent = true;
        if (this.oenabled && column.editor
            && (!Codes.isDoubleClickMode(this.detailMode))
            && (Codes.isDoubleClickMode(this.editionMode))) {
            this.activateColumnEdition(column, row, cellRef);
        }
        else {
            this.onDoubleClick.emit({ row: row, rowIndex: rowIndex, mouseEvent: $event, columnName: column.attr, cell: row[column.attr] });
            if (this.oenabled && Codes.isDoubleClickMode(this.detailMode)) {
                this.saveDataNavigationInLocalStorage();
                this.viewDetail(row);
            }
        }
    };
    Object.defineProperty(OTableComponent.prototype, "editionEnabled", {
        get: function () {
            return this._oTableOptions.columns.some(function (item) { return item.editing; });
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.handleDOMClick = function (event) {
        if (this._oTableOptions.selectColumn.visible) {
            return;
        }
        if (this.editionEnabled) {
            return;
        }
        var overlayContainer = document.body.getElementsByClassName('cdk-overlay-container')[0];
        if (overlayContainer && overlayContainer.contains(event.target)) {
            return;
        }
        var tableContainer = this.elRef.nativeElement.querySelector('.o-table-container');
        var tableContent = this.elRef.nativeElement.querySelector('.o-table-container table.mat-table');
        if (tableContainer && tableContent && tableContainer.contains(event.target) && !tableContent.contains(event.target)) {
            this.clearSelection();
        }
    };
    OTableComponent.prototype.activateColumnEdition = function (column, row, cellRef) {
        if (cellRef && column.editing && this.editingCell === cellRef) {
            return;
        }
        var columnPermissions = this.getOColumnPermissions(column.attr);
        if (columnPermissions.enabled === false) {
            console.warn(column.attr + " edition not allowed due to permissions");
            return;
        }
        this.clearSelectionAndEditing();
        this.selectedRow(row);
        this.editingCell = cellRef;
        var rowData = {};
        this.keysArray.forEach(function (key) {
            rowData[key] = row[key];
        });
        rowData[column.attr] = row[column.attr];
        this.editingRow = row;
        column.editing = true;
        column.editor.startEdition(rowData);
    };
    OTableComponent.prototype.updateCellData = function (column, data, saveChanges) {
        if (!this.checkEnabledActionPermission(PermissionsUtils.ACTION_UPDATE)) {
            var res = new Observable(function (innerObserver) {
                innerObserver.error();
            });
            return res;
        }
        column.editing = false;
        if (saveChanges && this.editingRow !== undefined) {
            Object.assign(this.editingRow, data);
        }
        this.stopEdition();
        if (saveChanges && column.editor.updateRecordOnEdit) {
            var toUpdate = {};
            toUpdate[column.attr] = data[column.attr];
            var kv = this.extractKeysFromRecord(data);
            return this.updateRecord(kv, toUpdate);
        }
        return undefined;
    };
    OTableComponent.prototype.getKeysValues = function () {
        var _this = this;
        var data = this.getAllValues();
        return data.map(function (row) {
            var obj = {};
            _this.keysArray.forEach(function (key) {
                if (row[key] !== undefined) {
                    obj[key] = row[key];
                }
            });
            return obj;
        });
    };
    OTableComponent.prototype.onShowsSelects = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onShowsSelects();
        }
    };
    OTableComponent.prototype.initializeCheckboxColumn = function () {
        var _this = this;
        if (!this.selectionChangeSubscription && this._oTableOptions.selectColumn.visible) {
            this.selectionChangeSubscription = this.selection.changed.subscribe(function (selectionData) {
                if (_this.triggerSelectionEvents && selectionData) {
                    if (selectionData.added.length > 0) {
                        ObservableWrapper.callEmit(_this.onRowSelected, selectionData.added);
                    }
                    if (selectionData.removed.length > 0) {
                        ObservableWrapper.callEmit(_this.onRowDeselected, selectionData.removed);
                    }
                }
            });
        }
        this.updateSelectionColumnState();
    };
    OTableComponent.prototype.updateSelectionColumnState = function () {
        if (!this._oTableOptions.selectColumn.visible) {
            this.clearSelection();
        }
        if (this._oTableOptions.visibleColumns && this._oTableOptions.selectColumn.visible
            && this._oTableOptions.visibleColumns[0] !== Codes.NAME_COLUMN_SELECT) {
            this._oTableOptions.visibleColumns.unshift(Codes.NAME_COLUMN_SELECT);
        }
        else if (this._oTableOptions.visibleColumns && !this._oTableOptions.selectColumn.visible
            && this._oTableOptions.visibleColumns[0] === Codes.NAME_COLUMN_SELECT) {
            this._oTableOptions.visibleColumns.shift();
        }
        this.updateStateExpandedColumn();
    };
    OTableComponent.prototype.getNumRowSelectedInCurrentData = function () {
        var _this = this;
        return this.dataSource ? this.dataSource.renderedData.filter(function (x) { return !_this.isDisableCheckbox(x) && _this.isRowSelected(x); }).length : 0;
    };
    OTableComponent.prototype.isAllSelected = function () {
        var numSelected = this.getNumRowSelectedInCurrentData();
        var numRows = this.dataSource ? this.dataSource.renderedData.length : 0;
        return numSelected > 0 && numSelected === numRows;
    };
    OTableComponent.prototype.isIndeterminate = function () {
        var numSelected = this.getNumRowSelectedInCurrentData();
        var numRows = this.dataSource ? this.dataSource.renderedData.length : 0;
        return numSelected > 0 && numRows > 0 && numSelected !== numRows;
    };
    OTableComponent.prototype.masterToggle = function (event) {
        event.checked ? this.selectAll() : this.clearSelection();
    };
    OTableComponent.prototype.selectAll = function () {
        var _this = this;
        this.dataSource.renderedData.forEach(function (row) { return _this.setSelected(row); });
    };
    OTableComponent.prototype.selectionCheckboxToggle = function (event, row) {
        if (this.isSelectionModeSingle()) {
            this.clearSelection();
        }
        event.checked ? this.selectedRow(row) : this.selection.deselect(row);
        this.state.selection = this.selection.selected;
    };
    OTableComponent.prototype.selectedRow = function (row) {
        this.setSelected(row);
        this.cd.detectChanges();
    };
    OTableComponent.prototype.clearSelection = function () {
        if (Util.isDefined(this.selection)) {
            this.selection.clear();
        }
        if (Util.isDefined(this.state)) {
            this.state.selection = [];
        }
    };
    OTableComponent.prototype.setSelected = function (item) {
        if (this.isDisableCheckbox(item)) {
            return;
        }
        if (Util.isDefined(item) && !this.isRowSelected(item)) {
            this.selection.select(item);
        }
    };
    Object.defineProperty(OTableComponent.prototype, "showDeleteButton", {
        get: function () {
            return this.deleteButton;
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.getTrackByFunction = function () {
        var _this = this;
        var self = this;
        return function (index, item) {
            var itemId = '';
            if (_this.isGroup(index, item)) {
                itemId += item.column;
            }
            else {
                var keysLenght_1 = self.keysArray.length;
                self.keysArray.forEach(function (key, idx) {
                    var suffix = idx < (keysLenght_1 - 1) ? ';' : '';
                    itemId += item[key] + suffix;
                });
            }
            var hasAsyncAndVisibleCols = _this.asyncLoadColumns.some(function (c) { return _this._oTableOptions.visibleColumns.includes(c); });
            if (self.asyncLoadColumns.length && hasAsyncAndVisibleCols && !self.finishQuerySubscription) {
                self.queryRowAsyncData(index, item);
                if (self.paginator && index === (self.paginator.pageSize - 1)) {
                    self.finishQuerySubscription = true;
                }
                return itemId;
            }
            else {
                return itemId;
            }
        };
    };
    OTableComponent.prototype.queryRowAsyncData = function (rowIndex, rowData) {
        var _this = this;
        var kv = ServiceUtils.getObjectProperties(rowData, this.keysArray);
        var av = this.asyncLoadColumns.filter(function (c) { return _this._oTableOptions.visibleColumns.indexOf(c) !== -1; });
        if (av.length === 0) {
            return;
        }
        var columnQueryArgs = [kv, av, this.entity, undefined, undefined, undefined, undefined];
        var queryMethodName = this.pageable ? this.paginatedQueryMethod : this.queryMethod;
        if (this.dataService && (queryMethodName in this.dataService) && this.entity) {
            if (this.asyncLoadSubscriptions[rowIndex]) {
                this.asyncLoadSubscriptions[rowIndex].unsubscribe();
            }
            this.asyncLoadSubscriptions[rowIndex] = this.dataService[queryMethodName]
                .apply(this.dataService, columnQueryArgs)
                .subscribe(function (res) {
                if (res.isSuccessful()) {
                    var data = void 0;
                    if (Util.isArray(res.data) && res.data.length === 1) {
                        data = res.data[0];
                    }
                    else if (Util.isObject(res.data)) {
                        data = res.data;
                    }
                    _this.daoTable.setAsynchronousColumn(data, rowData);
                    _this.cd.detectChanges();
                }
            });
        }
    };
    OTableComponent.prototype.getValue = function () {
        return this.dataSource.getCurrentData();
    };
    OTableComponent.prototype.getAllValues = function () {
        return this.dataSource.getCurrentAllData();
    };
    OTableComponent.prototype.getAllRenderedValues = function () {
        return this.dataSource.getAllRendererData();
    };
    OTableComponent.prototype.getRenderedValue = function () {
        return this.dataSource.getCurrentRendererData();
    };
    OTableComponent.prototype.getSqlTypes = function () {
        var allSqlTypes = Util.isDefined(this.dataSource.sqlTypes) ? this.dataSource.sqlTypes : {};
        this._oTableOptions.columns.forEach(function (col) {
            if (col.sqlType) {
                allSqlTypes[col.attr] = col.sqlType;
            }
        });
        return allSqlTypes;
    };
    OTableComponent.prototype.setOTableColumnsFilter = function (tableColumnsFilter) {
        this.oTableColumnsFilterComponent = tableColumnsFilter;
    };
    Object.defineProperty(OTableComponent.prototype, "filterColumns", {
        get: function () {
            if (this.state.initialConfiguration.filterColumns === this.originalFilterColumns
                && this.state.filterColumns) {
                return this.state.filterColumns;
            }
            return this.originalFilterColumns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "originalFilterColumns", {
        get: function () {
            var sortColumnsFilter = [];
            if (this.oTableColumnsFilterComponent) {
                sortColumnsFilter = this.oTableColumnsFilterComponent.columnsArray;
            }
            return sortColumnsFilter;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "originalGroupedColumnsArray", {
        get: function () {
            return Util.parseArray(this.groupedColumns, true);
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.getStoredColumnsFilters = function () {
        return this.state.storedFilters;
    };
    OTableComponent.prototype.getStoredGroupedColumns = function () {
        return this.state.storedConfigurations;
    };
    OTableComponent.prototype.onFilterByColumnClicked = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onFilterByColumnClicked();
        }
    };
    OTableComponent.prototype.onStoreFilterClicked = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onStoreFilterClicked();
        }
    };
    OTableComponent.prototype.onLoadFilterClicked = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onLoadFilterClicked();
        }
    };
    OTableComponent.prototype.onClearFilterClicked = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onClearFilterClicked();
        }
    };
    OTableComponent.prototype.clearFilters = function (triggerDatasourceUpdate) {
        if (triggerDatasourceUpdate === void 0) { triggerDatasourceUpdate = true; }
        this.dataSource.clearColumnFilters(triggerDatasourceUpdate);
        if (this.oTableMenu && this.oTableMenu.columnFilterOption) {
            this.oTableMenu.columnFilterOption.setActive(this.isColumnFiltersActive);
        }
        this.onFilterByColumnChange.emit();
        if (this.oTableQuickFilterComponent) {
            this.oTableQuickFilterComponent.setValue(void 0);
        }
        if (this.filterBuilder) {
            this.filterBuilder.clearFilter();
            this.filterBuilder.triggerReload();
        }
    };
    OTableComponent.prototype.clearColumnFilter = function (attr, triggerDatasourceUpdate) {
        if (triggerDatasourceUpdate === void 0) { triggerDatasourceUpdate = true; }
        this.dataSource.clearColumnFilter(attr, triggerDatasourceUpdate);
        this.onFilterByColumnChange.emit();
        this.reloadPaginatedDataFromStart(false);
    };
    OTableComponent.prototype.filterByColumn = function (columnValueFilter) {
        this.dataSource.addColumnFilter(columnValueFilter);
        this.onFilterByColumnChange.emit();
        if (this.pageable) {
            this.reloadPaginatedDataFromStart(false);
        }
    };
    OTableComponent.prototype.clearColumnFilters = function (triggerDatasourceUpdate, columnsAttr) {
        if (triggerDatasourceUpdate === void 0) { triggerDatasourceUpdate = true; }
        this.dataSource.clearColumnFilters(triggerDatasourceUpdate, columnsAttr);
        this.onFilterByColumnChange.emit();
        this.reloadPaginatedDataFromStart(false);
    };
    OTableComponent.prototype.isColumnFilterable = function (column) {
        return Util.isDefined(this.oTableColumnsFilterComponent) ? this.oTableColumnsFilterComponent.isColumnFilterable(column.attr) : this.isSearcheableColumn(column);
    };
    OTableComponent.prototype.isSearcheableColumn = function (column) {
        return this.searcheableColumns.includes(column.attr);
    };
    OTableComponent.prototype.isColumnFilterActive = function (column) {
        return this.isColumnFiltersActive && Util.isDefined(this.dataSource.getColumnValueFilterByAttr(column.attr));
    };
    OTableComponent.prototype.openColumnFilterDialog = function (column, event) {
        var _this = this;
        event.stopPropagation();
        event.preventDefault();
        var dialogRef = this.dialog.open(OTableFilterByColumnDataDialogComponent, {
            data: {
                previousFilter: this.dataSource.getColumnValueFilterByAttr(column.attr),
                column: column,
                activeSortDirection: this.getSortFilterColumn(column),
                tableData: this.dataSource.getCurrentData(),
                preloadValues: this.oTableColumnsFilterComponent ? this.oTableColumnsFilterComponent.preloadValues : true,
                mode: this.oTableColumnsFilterComponent ? this.oTableColumnsFilterComponent.mode : 'default',
                startView: this.getStartViewFilterColumn(column)
            },
            minWidth: '380px',
            disableClose: true,
            panelClass: ['o-dialog-class', 'o-table-dialog']
        });
        dialogRef.afterClosed().subscribe(function (result) {
            switch (result) {
                case TableFilterByColumnDialogResult.ACCEPT:
                    var columnValueFilter = dialogRef.componentInstance.getColumnValuesFilter();
                    _this.filterByColumn(columnValueFilter);
                    break;
                case TableFilterByColumnDialogResult.CLEAR:
                    var col = dialogRef.componentInstance.column;
                    _this.clearColumnFilter(col.attr);
                    break;
            }
        });
        dialogRef.componentInstance.onSortFilterValuesChange.subscribe(function (sortedFilterableColumn) {
            _this.storeFilterColumns(sortedFilterableColumn);
        });
    };
    OTableComponent.prototype.storeFilterColumns = function (sortColumnFilter) {
        if (this.state.filterColumns) {
            var filterColumn = this.filterColumns.find(function (x) { return x.attr === sortColumnFilter.attr; });
            if (Util.isDefined(filterColumn)) {
                filterColumn.sort = sortColumnFilter.sort;
            }
            else {
                this.filterColumns.push(sortColumnFilter);
            }
        }
        this.state.filterColumns = this.filterColumns;
    };
    OTableComponent.prototype.getStartViewFilterColumn = function (column) {
        var startView;
        if (this.state.filterColumns) {
            this.state.filterColumns.forEach(function (element) {
                if (element.attr === column.attr) {
                    startView = element.startView;
                }
            });
        }
        if (!Util.isDefined(startView) && this.oTableColumnsFilterComponent) {
            startView = this.oTableColumnsFilterComponent.getStartViewValueOfFilterColumn(column.attr);
        }
        return startView;
    };
    OTableComponent.prototype.getSortFilterColumn = function (column) {
        var sortColumn;
        if (this.state.filterColumns) {
            var filterCol = this.state.filterColumns.find(function (element) { return element.attr === column.attr; });
            if (filterCol) {
                sortColumn = filterCol.sort;
            }
        }
        if (!Util.isDefined(sortColumn) && this.oTableColumnsFilterComponent) {
            sortColumn = this.oTableColumnsFilterComponent.getSortValueOfFilterColumn(column.attr);
        }
        if (!Util.isDefined(sortColumn) && this.sortColArray.find(function (x) { return x.columnName === column.attr; })) {
            sortColumn = this.isColumnSortActive(column) ? 'asc' : 'desc';
        }
        return sortColumn;
    };
    Object.defineProperty(OTableComponent.prototype, "disableTableMenuButton", {
        get: function () {
            return !!(this.permissions && this.permissions.menu && this.permissions.menu.enabled === false);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableComponent.prototype, "showTableMenuButton", {
        get: function () {
            var permissionHidden = !!(this.permissions && this.permissions.menu && this.permissions.menu.visible === false);
            if (permissionHidden) {
                return false;
            }
            var staticOpt = this.selectAllCheckbox || this.exportButton || this.showConfigurationOption || this.columnsVisibilityButton || (this.showFilterOption && this.oTableColumnsFilterComponent !== undefined);
            return staticOpt || this.tableOptions.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.setOTableInsertableRow = function (tableInsertableRow) {
        var insertPerm = this.getActionPermissions(PermissionsUtils.ACTION_INSERT);
        if (insertPerm.visible) {
            tableInsertableRow.enabled = insertPerm.enabled;
            this.oTableInsertableRowComponent = tableInsertableRow;
            this.showFirstInsertableRow = this.oTableInsertableRowComponent.isFirstRow();
            this.showLastInsertableRow = !this.showFirstInsertableRow;
            this.oTableInsertableRowComponent.initializeEditors();
        }
    };
    OTableComponent.prototype.clearSelectionAndEditing = function (clearSelection) {
        if (clearSelection === void 0) { clearSelection = true; }
        if (clearSelection) {
            this.clearSelection();
        }
        this._oTableOptions.columns
            .filter(function (oColumn) { return oColumn.editing; })
            .forEach(function (oColumn) {
            oColumn.editing = false;
        });
    };
    OTableComponent.prototype.checkSelectedItemData = function () {
        var _this = this;
        this.triggerSelectionEvents = false;
        if (Util.isDefined(this.state.selection) && this.state.selection.length > 0) {
            this.state.selection.forEach(function (selectedItem) {
                var selectedItemKeys = Object.keys(selectedItem);
                var foundItem = _this.dataSource.renderedData.find(function (data) {
                    return selectedItemKeys.every(function (key) { return data[key] === selectedItem[key]; });
                });
                if (foundItem && !_this.isRowSelected(foundItem)) {
                    _this.setSelected(foundItem);
                }
            });
        }
        this.triggerSelectionEvents = true;
    };
    OTableComponent.prototype.useDetailButton = function (column) {
        return column.type === 'editButtonInRow' || column.type === 'detailButtonInRow';
    };
    OTableComponent.prototype.onDetailButtonClick = function (column, row, event) {
        event.preventDefault();
        event.stopPropagation();
        switch (column.type) {
            case 'editButtonInRow':
                this.editDetail(row);
                break;
            case 'detailButtonInRow':
                this.viewDetail(row);
                break;
        }
    };
    OTableComponent.prototype.getDetailButtonIcon = function (column) {
        var result = '';
        switch (column.type) {
            case 'editButtonInRow':
                result = this.editButtonInRowIcon;
                break;
            case 'detailButtonInRow':
                result = this.detailButtonInRowIcon;
                break;
        }
        return result;
    };
    OTableComponent.prototype.usePlainRender = function (column, row) {
        return !this.useDetailButton(column) && !column.renderer && (!column.editor || (!column.editing || !this.isRowSelected(row)));
    };
    OTableComponent.prototype.useCellRenderer = function (column, row) {
        return column.renderer && (!column.editing || column.editing && !this.isRowSelected(row));
    };
    OTableComponent.prototype.useCellEditor = function (column, row) {
        if (column.editor && column.editor.autoCommit) {
            return false;
        }
        return column.editor && column.editing && this.isRowSelected(row);
    };
    OTableComponent.prototype.isSelectionModeMultiple = function () {
        return this.selectionMode === Codes.SELECTION_MODE_MULTIPLE;
    };
    OTableComponent.prototype.isSelectionModeSingle = function () {
        return this.selectionMode === Codes.SELECTION_MODE_SINGLE;
    };
    OTableComponent.prototype.isSelectionModeNone = function () {
        return this.selectionMode === Codes.SELECTION_MODE_NONE;
    };
    OTableComponent.prototype.onChangePage = function (evt) {
        this.finishQuerySubscription = false;
        if (!this.pageable) {
            this.currentPage = evt.pageIndex;
            return;
        }
        var goingBack = evt.pageIndex < this.currentPage;
        this.currentPage = evt.pageIndex;
        var pageSize = this.paginator.isShowingAllRows(evt.pageSize) ? this.state.totalQueryRecordsNumber : evt.pageSize;
        var oldQueryRows = this.queryRows;
        var changingPageSize = (oldQueryRows !== pageSize);
        this.queryRows = pageSize;
        this.paginator.pageSize = pageSize;
        var newStartRecord;
        var queryLength;
        if (goingBack || changingPageSize) {
            newStartRecord = (this.currentPage * this.queryRows);
            queryLength = this.queryRows;
        }
        else {
            newStartRecord = Math.max(this.state.queryRecordOffset, (this.currentPage * this.queryRows));
            var newEndRecord = Math.min(newStartRecord + this.queryRows, this.state.totalQueryRecordsNumber);
            queryLength = Math.min(this.queryRows, newEndRecord - newStartRecord);
        }
        var queryArgs = {
            offset: newStartRecord,
            length: queryLength
        };
        this.finishQuerySubscription = false;
        this.queryData(void 0, queryArgs);
    };
    OTableComponent.prototype.getOColumn = function (attr) {
        return this._oTableOptions ? this._oTableOptions.columns.find(function (item) { return item.name === attr; }) : undefined;
    };
    OTableComponent.prototype.insertRecord = function (recordData, sqlTypes) {
        if (!this.checkEnabledActionPermission(PermissionsUtils.ACTION_INSERT)) {
            return undefined;
        }
        if (!Util.isDefined(sqlTypes)) {
            var allSqlTypes_1 = this.getSqlTypes();
            sqlTypes = {};
            Object.keys(recordData).forEach(function (key) {
                sqlTypes[key] = allSqlTypes_1[key];
            });
        }
        return this.daoTable.insertQuery(recordData, sqlTypes);
    };
    OTableComponent.prototype.updateRecord = function (filter, updateData, sqlTypes) {
        if (!this.checkEnabledActionPermission(PermissionsUtils.ACTION_UPDATE)) {
            return of(this.dataSource.data);
        }
        var sqlTypesArg = sqlTypes || {};
        if (!Util.isDefined(sqlTypes)) {
            var allSqlTypes_2 = this.getSqlTypes();
            Object.keys(filter).forEach(function (key) {
                sqlTypesArg[key] = allSqlTypes_2[key];
            });
            Object.keys(updateData).forEach(function (key) {
                sqlTypesArg[key] = allSqlTypes_2[key];
            });
        }
        return this.daoTable.updateQuery(filter, updateData, sqlTypesArg);
    };
    OTableComponent.prototype.getDataArray = function () {
        return this.daoTable.data;
    };
    OTableComponent.prototype.setDataArray = function (data) {
        if (this.daoTable) {
            this.pageable = false;
            this.staticData = data;
            this.daoTable.usingStaticData = true;
            this.daoTable.setDataArray(this.staticData);
            this.onDataLoaded.emit(this.daoTable.data);
        }
    };
    OTableComponent.prototype.deleteLocalItems = function () {
        var dataArray = this.getDataArray();
        var selectedItems = this.getSelectedItems();
        selectedItems.forEach(function (selectedItem) {
            for (var j = dataArray.length - 1; j >= 0; --j) {
                if (Util.equals(selectedItem, dataArray[j])) {
                    dataArray.splice(j, 1);
                    break;
                }
            }
        });
        this.clearSelection();
        this.setDataArray(dataArray);
    };
    OTableComponent.prototype.isColumnSortActive = function (column) {
        var found = this.sortColArray.find(function (sortC) { return sortC.columnName === column.attr; });
        return found !== undefined;
    };
    OTableComponent.prototype.isColumnDescSortActive = function (column) {
        var found = this.sortColArray.find(function (sortC) { return sortC.columnName === column.attr && !sortC.ascendent; });
        return found !== undefined;
    };
    OTableComponent.prototype.hasTabGroupChangeSubscription = function () {
        return this.tabGroupChangeSubscription !== undefined;
    };
    OTableComponent.prototype.isEmpty = function (value) {
        return !Util.isDefined(value) || ((typeof value === 'string') && !value);
    };
    OTableComponent.prototype.setFiltersConfiguration = function (storage) {
        var _this = this;
        if (storage === void 0) { storage = this.state; }
        if (storage.initialConfiguration.filterCaseSensitive !== this.filterCaseSensitive) {
            this._oTableOptions.filterCaseSensitive = this.filterCaseSensitive;
        }
        else {
            this._oTableOptions.filterCaseSensitive = storage.hasOwnProperty('filter-case-sensitive') ? storage['filter-case-sensitive'] : this.filterCaseSensitive;
        }
        if (storage.initialConfiguration.filterColumnActiveByDefault !== this.filterColumnActiveByDefault) {
            this.isColumnFiltersActive = this.filterColumnActiveByDefault;
        }
        else {
            var confFilterColumnActiveByDefault = storage.hasOwnProperty('filter-column-active') ? storage['filter-column-active'] : this.filterColumnActiveByDefault;
            this.isColumnFiltersActive = confFilterColumnActiveByDefault || this.state.columnValueFilters.length > 0;
        }
        if (Util.isDefined(storage.columnValueFilters)) {
            this.dataSource.initializeColumnsFilters(this.state.columnValueFilters);
            this.onFilterByColumnChange.emit();
        }
        if (this.oTableQuickFilterComponent) {
            this.oTableQuickFilterComponent.setValue(storage['filter']);
            (storage['oColumns'] || []).forEach(function (oColData) {
                var oCol = _this.getOColumn(oColData.attr);
                if (oCol && oColData.hasOwnProperty('searching')) {
                    oCol.searching = oColData.searching;
                }
            });
        }
        if (this.filterBuilder && Util.isDefined(storage.filterBuilderValues)) {
            this.filterBuilder.setFilterValues(storage.filterBuilderValues);
            this.filterBuilder.triggerReload();
        }
    };
    OTableComponent.prototype.onStoreConfigurationClicked = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onStoreConfigurationClicked();
        }
    };
    OTableComponent.prototype.onApplyConfigurationClicked = function () {
        if (this.oTableMenu) {
            this.oTableMenu.onApplyConfigurationClicked();
        }
    };
    OTableComponent.prototype.applyDefaultConfiguration = function () {
        this.initializeParams();
        this.parseVisibleColumns(true);
        this.reinitializateQuickFilterColumns();
        this.resetQueryRows();
        var initialConfigSortColumnsArray = ServiceUtils.parseSortColumns(this.state.initialConfiguration.sortColumns);
        this.reinitializeSortColumns(initialConfigSortColumnsArray);
        this.onReinitialize.emit(null);
        this.clearFilters(false);
        this.reloadData();
    };
    OTableComponent.prototype.resetQueryRows = function () {
        if (Util.isDefined(this.state.initialConfiguration.queryRows)) {
            this.queryRows = this.state.initialConfiguration.queryRows;
        }
    };
    OTableComponent.prototype.applyConfiguration = function (configurationName) {
        var _this = this;
        var storedConfiguration = this.state.getStoredConfiguration(configurationName);
        if (storedConfiguration) {
            var properties = storedConfiguration['stored-properties'] || [];
            properties.forEach(function (property) {
                switch (property) {
                    case 'sort-columns':
                        _this.reinitializeSortColumns();
                        break;
                    case 'oColumns-display':
                        _this.parseVisibleColumns();
                        _this.initializeCheckboxColumn();
                        break;
                    case 'quick-filter':
                    case 'columns-filter':
                    case 'filter-builder':
                        _this.setFiltersConfiguration();
                        break;
                    case 'grouped-columns':
                        _this.parseGroupedColumns();
                        break;
                    case 'page':
                        _this.currentPage = _this.state.currentPage;
                        _this.queryRows = _this.state.queryRows;
                        break;
                }
            });
            this.reloadData();
        }
    };
    OTableComponent.prototype.getTitleAlignClass = function (oCol) {
        var align;
        var hasTitleAlign = Util.isDefined(oCol.definition) && Util.isDefined(oCol.definition.titleAlign);
        var autoAlign = (this.autoAlignTitles && !hasTitleAlign) || (hasTitleAlign && oCol.definition.titleAlign === Codes.COLUMN_TITLE_ALIGN_AUTO);
        if (!autoAlign) {
            return oCol.getTitleAlignClass();
        }
        switch (oCol.type) {
            case 'image':
            case 'date':
            case 'action':
            case 'boolean':
                align = Codes.COLUMN_TITLE_ALIGN_CENTER;
                break;
            case 'currency':
            case 'integer':
            case 'real':
            case 'percentage':
                align = Codes.COLUMN_TITLE_ALIGN_END;
                break;
            case 'service':
            default:
                align = Codes.COLUMN_TITLE_ALIGN_START;
                break;
        }
        return align;
    };
    OTableComponent.prototype.getCellAlignClass = function (column) {
        return Util.isDefined(column.definition) && Util.isDefined(column.definition.contentAlign) ? 'o-' + column.definition.contentAlign : '';
    };
    OTableComponent.prototype.getGroupHeaderCellAlignClass = function (column) {
        var classNameArray = [];
        var oCol = this.getOColumn(column.substr('groupHeader-'.length));
        if (Util.isDefined(oCol)) {
            classNameArray.push(this.getCellAlignClass(oCol));
            classNameArray.push(oCol.className ? oCol.className : '');
        }
        return classNameArray;
    };
    OTableComponent.prototype.addDefaultRowButtons = function () {
        if (this.editButtonInRow) {
            this.addButtonInRow('editButtonInRow');
        }
        if (this.detailButtonInRow) {
            this.addButtonInRow('detailButtonInRow');
        }
    };
    OTableComponent.prototype.addButtonInRow = function (name) {
        var colDef = this.createOColumn(name, this);
        colDef.type = name;
        colDef.visible = true;
        colDef.searchable = false;
        colDef.orderable = false;
        colDef.resizable = false;
        colDef.groupable = false;
        colDef.title = undefined;
        colDef.width = '48px';
        this.pushOColumnDefinition(colDef);
        this._oTableOptions.visibleColumns.push(name);
    };
    Object.defineProperty(OTableComponent.prototype, "headerHeight", {
        get: function () {
            var height = 0;
            if (this.tableHeaderEl && this.tableHeaderEl.nativeElement) {
                height += this.tableHeaderEl.nativeElement.offsetHeight;
            }
            if (this.tableToolbarEl && this.tableToolbarEl.nativeElement) {
                height += this.tableToolbarEl.nativeElement.offsetHeight;
            }
            return height;
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.isDetailMode = function () {
        return this.detailMode !== Codes.DETAIL_MODE_NONE;
    };
    OTableComponent.prototype.copyAll = function () {
        Util.copyToClipboard(JSON.stringify(this.getRenderedValue()));
    };
    OTableComponent.prototype.copySelection = function () {
        var selectedItems = this.dataSource.getRenderedData(this.getSelectedItems());
        Util.copyToClipboard(JSON.stringify(selectedItems));
    };
    OTableComponent.prototype.viewDetail = function (item) {
        if (!this.checkEnabledActionPermission('detail')) {
            return;
        }
        _super.prototype.viewDetail.call(this, item);
    };
    OTableComponent.prototype.editDetail = function (item) {
        if (!this.checkEnabledActionPermission('edit')) {
            return;
        }
        _super.prototype.editDetail.call(this, item);
    };
    OTableComponent.prototype.getOColumnFromTh = function (th) {
        var result;
        var classList = [].slice.call(th.classList);
        var columnClass = classList.find(function (className) { return (className.startsWith('mat-column-')); });
        if (Util.isDefined(columnClass)) {
            result = this.getOColumn(columnClass.substr('mat-column-'.length));
        }
        return result;
    };
    OTableComponent.prototype.getOColumnFromGroupHeaderColumn = function (attr) {
        return this.getOColumn(attr.substr('groupHeader-'.length));
    };
    OTableComponent.prototype.getThWidthFromOColumn = function (oColumn) {
        var widthColumn;
        var thArray = this.tableHeaderEl.nativeElement.children;
        for (var i = 0; i < thArray.length && !Util.isDefined(widthColumn); i++) {
            var th = thArray[i];
            var classList = th.classList;
            for (var j = 0; j < classList.length && !Util.isDefined(widthColumn); j++) {
                if (classList[j] === 'mat-column-' + oColumn.attr) {
                    widthColumn = th.clientWidth;
                }
            }
        }
        return widthColumn;
    };
    OTableComponent.prototype.getColumnInsertable = function (name) {
        return name + this.getSuffixColumnInsertable();
    };
    OTableComponent.prototype.isRowSelected = function (row) {
        var keys = Object.keys(row);
        return !this.isSelectionModeNone() && this.selection.selected.some(function (element) { return keys.every(function (key) { return row[key] === element[key]; }); });
    };
    OTableComponent.prototype.refreshColumnsWidth = function () {
        var _this = this;
        setTimeout(function () {
            _this._oTableOptions.columns.filter(function (c) { return c.visible; }).forEach(function (c) {
                if (Util.isDefined(c.definition) && Util.isDefined(c.definition.width)) {
                    c.width = c.definition.width;
                }
                c.setRenderWidth(_this.horizontalScroll, _this.getClientWidthColumn(c));
            });
            _this.cd.detectChanges();
        }, 0);
    };
    OTableComponent.prototype.createOColumn = function (attr, table, column) {
        var instance = new OColumn();
        if (attr) {
            instance.attr = attr;
        }
        if (table) {
            instance.setDefaultProperties({
                orderable: this.orderable,
                resizable: this.resizable,
                groupable: this.groupable
            });
        }
        if (column) {
            instance.setColumnProperties(column);
        }
        return instance;
    };
    OTableComponent.prototype.registerOTableButtons = function (arg) {
        this.oTableButtons = arg;
        if (this.oTableButtons && this.tableButtons && this.tableButtons.length > 0) {
            this.oTableButtons.registerButtons(this.tableButtons.toArray());
        }
    };
    OTableComponent.prototype.getClientWidthColumn = function (col) {
        return col.DOMWidth || this.getThWidthFromOColumn(col);
    };
    OTableComponent.prototype.getMinWidthColumn = function (col) {
        return Util.extractPixelsValue(col.minWidth, Codes.DEFAULT_COLUMN_MIN_WIDTH) + 'px';
    };
    OTableComponent.prototype.showExpandableRow = function () {
        return Util.isDefined(this.tableRowExpandable);
    };
    Object.defineProperty(OTableComponent.prototype, "enabledVirtualScroll", {
        get: function () {
            return this.virtualScroll && !this.showExpandableRow() && this.groupedColumnsArray.length === 0;
        },
        enumerable: true,
        configurable: true
    });
    OTableComponent.prototype.parseGroupedColumns = function () {
        var _this = this;
        var result = this.state.groupedColumns || this.originalGroupedColumnsArray;
        if (this.state.groupedColumns && this.state.initialConfiguration.groupedColumns) {
            var difference = this.state.initialConfiguration.groupedColumns
                .filter(function (x) { return !_this.originalGroupedColumnsArray.includes(x); });
            if (difference.length > 0) {
                result = this.originalGroupedColumnsArray;
            }
        }
        this.setGroupColumns(result);
    };
    OTableComponent.prototype.parseGroupedColumnTypes = function () {
        this.groupedColumnTypes = this.state.groupedColumnTypes;
    };
    OTableComponent.prototype.groupByColumn = function (column, type) {
        this.checkGroupByColumn(column.attr, true);
        if (type) {
            this.updateGroupedColumnTypes(column.attr, true, type);
        }
        this.dataSource.updateGroupedColumns();
    };
    OTableComponent.prototype.unGroupByColumn = function (column) {
        this.checkGroupByColumn(column.attr, false);
        this.updateGroupedColumnTypes(column.attr, false);
        this.dataSource.updateGroupedColumns();
    };
    OTableComponent.prototype.unGroupByAllColumns = function () {
        this.setGroupColumns([]);
        this.groupedColumnTypes = [];
    };
    OTableComponent.prototype.setGroupColumns = function (value) {
        this.groupedColumnsArray = value;
        this.dataSource.updateGroupedColumns();
    };
    OTableComponent.prototype.setGroupedColumnTypes = function (value) {
        this.groupedColumnTypes = value;
    };
    OTableComponent.prototype.checkGroupByColumn = function (field, add) {
        var e_1, _a;
        var found = null;
        try {
            for (var _b = tslib_1.__values(this.groupedColumnsArray), _c = _b.next(); !_c.done; _c = _b.next()) {
                var column = _c.value;
                if (column === field) {
                    found = this.groupedColumnsArray.indexOf(column, 0);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (found != null && found >= 0) {
            if (!add) {
                this.groupedColumnsArray.splice(found, 1);
            }
        }
        else {
            if (add) {
                this.groupedColumnsArray.push(field);
            }
        }
    };
    OTableComponent.prototype.isGroup = function (index, item) {
        return item instanceof OTableGroupedRow;
    };
    OTableComponent.prototype.isNotGroup = function (index, item) {
        return !(item instanceof OTableGroupedRow);
    };
    OTableComponent.prototype.getLastGroups = function () {
        var scores = this.dataSource.renderedData;
        var maxLevel = scores.reduce(function (acc, curr) { return curr.level > acc ? curr.level : acc; }, 0);
        var maxLevelRenderedData = scores.reduce(function (r, o) { return o.level === maxLevel ? tslib_1.__spread(r, [o]) : r; }, []);
        return maxLevelRenderedData.length;
    };
    OTableComponent.prototype.groupHeaderClick = function (row) {
        this.dataSource.toggleGroupByColumn(row);
    };
    OTableComponent.prototype.isInstanceOfOTableCellRendererServiceComponent = function (renderer) {
        return Util.isDefined(renderer) && renderer.onDataLoaded && renderer.queryAllData;
    };
    OTableComponent.prototype.getColumnDataByAttr = function (attr, row) {
        var operation = null;
        if (this.groupedColumnTypes.length != 0 && this.groupedColumnTypes.findIndex(function (column) { return column.attr == attr; }) != -1) {
            operation = this.groupedColumnTypes[this.groupedColumnTypes.findIndex(function (column) { return column.attr == attr; })].type;
        }
        var oCol = this.getOColumn(attr);
        if (!Util.isDefined(oCol)) {
            return row[attr];
        }
        var useRenderer = oCol.renderer && oCol.renderer.getCellData;
        if (operation == null) {
            return useRenderer ? oCol.renderer.getCellData(row[oCol.attr], row) : row[oCol.attr];
        }
        else {
            var date = moment(row[oCol.attr]);
            var language = this.translateService.getCurrentLang();
            switch (operation) {
                case "YEAR": return date.year();
                case "MONTH": return moment().locale(language).month(date.month()).format("MMMM");
                case "YEAR_MONTH": return moment().locale(language).month(date.month()).year(date.year()).format("MMMM, YYYY");
                case "YEAR_MONTH_DAY": return useRenderer ? oCol.renderer.getCellData(row[oCol.attr], row) : row[oCol.attr];
            }
        }
    };
    OTableComponent.prototype.updateGroupedColumnTypes = function (attr, add, operation) {
        var groupedColumns = [];
        var index = this.groupedColumnTypes.findIndex(function (column) { return column.attr == attr; });
        if (index != -1) {
            if (!add) {
                this.groupedColumnTypes.splice(index, 1);
            }
        }
        else {
            if (add) {
                this.groupedColumnTypes.push({ "attr": attr, "type": operation });
            }
        }
        this.state.groupedColumnTypes = groupedColumns;
    };
    OTableComponent.prototype.getClassNameGroupHeader = function (row) {
        var className = '';
        if (row.level <= 10) {
            className = 'o-table-group-row o-table-group-row-level-' + row.level;
        }
        return className;
    };
    OTableComponent.prototype.stopEdition = function (clearSelection) {
        clearSelection = clearSelection ? clearSelection : false;
        this.editingCell = undefined;
        this.editingRow = undefined;
        this.clearSelectionAndEditing(clearSelection);
    };
    OTableComponent.prototype.storeFilterInState = function (arg) {
        this.componentStateService.storeFilter(arg);
    };
    OTableComponent.prototype.reinitializateQuickFilterColumns = function () {
        this._oTableOptions.columns
            .filter(function (column) { return column.searchable && column.visible; })
            .forEach(function (column) { return column.searching = true; });
    };
    OTableComponent.prototype.reinitializeSortColumns = function (sortColumns) {
        this.tableHeaders = [];
        if (Util.isDefined(sortColumns)) {
            this.sortColArray = sortColumns;
        }
        else {
            this.parseSortColumns();
        }
        this.sort.setSortColumns(this.sortColArray);
        this.refreshSortHeaders();
    };
    OTableComponent.prototype.refreshSortHeaders = function () {
        this.tableHeaders.filter(function (header) { return Util.isDefined(header.matSortHeader); }).forEach(function (header) { return header.matSortHeader.refresh(); });
    };
    OTableComponent.prototype.getQuickFilterValue = function () {
        return Util.isDefined(this.oTableQuickFilterComponent) ?
            this.oTableQuickFilterComponent.value :
            '';
    };
    OTableComponent.prototype.queryCellRenderers = function () {
        var _this = this;
        var quickFilterValue = this.getQuickFilterValue();
        if (Util.isDefined(quickFilterValue) && quickFilterValue.length > 0) {
            var queries = this.oTableOptions.columns
                .filter(function (oCol) { return oCol.searching && _this.isInstanceOfOTableCellRendererServiceComponent(oCol.renderer); })
                .map(function (oCol) { return oCol.renderer.queryAllData(); });
            return queries.length > 0 ? combineLatest(queries) : of(null);
        }
        return of(null);
    };
    OTableComponent.prototype.tableQuickFilterChanged = function (value) {
        var _this = this;
        if (this.pageable) {
            this.queryCellRenderers().subscribe(function () {
                _this.reloadPaginatedDataFromStart(false);
            });
        }
        else {
            if (value && value.length) {
                this.dataSource.quickFilter = value;
            }
        }
    };
    OTableComponent.prototype.filterData = function (value, loadMore) {
    };
    OTableComponent.prototype.setOTableColumnsGrouping = function (value) {
        this.oTableColumnsGroupingComponent = value;
    };
    OTableComponent.prototype.getColumnGroupingComponent = function (columnAttr) {
        var result;
        if (Util.isDefined(this.oTableColumnsGroupingComponent)) {
            result = this.oTableColumnsGroupingComponent.getColumnGrouping(columnAttr);
        }
        return result;
    };
    OTableComponent.prototype.useColumnGroupingAggregate = function (columnAttr) {
        var oCol = this.getOColumn(columnAttr);
        if (!Util.isDefined(oCol)) {
            return false;
        }
        var sqlType = this.getSqlTypes()[columnAttr];
        var hasDefaultAggregate = SQLTypes.isNumericSQLType(sqlType)
            && (!Util.isDefined(oCol.renderer) || OTableComponent.AVAILABLE_GROUPING_COLUMNS_RENDERERS.includes(oCol.type));
        if (!Util.isDefined(this.oTableColumnsGroupingComponent)) {
            return hasDefaultAggregate;
        }
        return this.oTableColumnsGroupingComponent.useColumnAggregate(columnAttr, hasDefaultAggregate);
    };
    OTableComponent.prototype.checkViewportSize = function () {
        if (this.virtualScrollViewport) {
            this.virtualScrollViewport.checkViewportSize();
        }
    };
    OTableComponent.prototype.showExpandableIcon = function (row, rowIndex) {
        return (Util.isDefined(this.showExpandableIconFunction) && this.showExpandableIconFunction instanceof Function) ? Util.wrapIntoObservable(this.showExpandableIconFunction(row, rowIndex)) : of(true);
    };
    OTableComponent.prototype.getColumnsNotIncluded = function () {
        var colsNotIncluded = [];
        colsNotIncluded = this.oTableOptions.columns.filter(function (c) { return void 0 !== c.renderer && c.type === 'image'; }).map(function (c) { return c.attr; });
        colsNotIncluded.push(Codes.NAME_COLUMN_SELECT);
        colsNotIncluded.push(Codes.NAME_COLUMN_EXPANDABLE);
        return colsNotIncluded;
    };
    OTableComponent.prototype.getColumnNames = function (columns) {
        var _this = this;
        var tableColumnNames = {};
        columns.forEach(function (c) {
            var oColumn = _this._oTableOptions.columns.find(function (oc) { return oc.attr === c; });
            tableColumnNames[c] = _this.translateService.get(oColumn.title ? oColumn.title : oColumn.attr);
        });
        return tableColumnNames;
    };
    OTableComponent.prototype.resetColumnsWidth = function () {
        this._oTableOptions.columns.forEach(function (c) {
            if (Util.isDefined(c.definition)) {
                c.width = Util.isDefined(c.definition.width) ? c.definition.width : c.definition.originalWidth;
            }
            else {
                c.width = "auto";
            }
        });
        this.cd.detectChanges();
        this.updateColumnsDOMWidth();
    };
    OTableComponent.prototype.updateColumnsDOMWidth = function () {
        var _this = this;
        this._oTableOptions.columns.forEach(function (c) {
            c.DOMWidth = _this.getThWidthFromOColumn(c);
        });
    };
    OTableComponent.prototype.isDisableCheckbox = function (item) {
        var disable = false;
        if (Util.isDefined(this.disableSelectionFunction)) {
            return this.disableSelectionFunction(item);
        }
        return disable;
    };
    OTableComponent.DEFAULT_BASE_SIZE_SPINNER = 100;
    OTableComponent.FIRST_LAST_CELL_PADDING = 24;
    OTableComponent.EXPANDED_ROW_CONTAINER_CLASS = 'expanded-row-container-';
    OTableComponent.AVAILABLE_GROUPING_COLUMNS_RENDERERS = ['currency', 'integer', 'real'];
    OTableComponent.DEFAULT_ROW_HEIGHT = 36;
    OTableComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table',
                    template: "<div class=\"o-table-container\" fxLayout=\"column\" fxLayoutAlign=\"start stretch\" [style.display]=\"isVisible()? '' : 'none'\"\n  [class.block-events]=\"showLoading | async\">\n\n  <o-data-toolbar #tableToolbar *ngIf=\"hasControls()\" [title]=\"title\" [show-title]=\"showTitle\" class=\"o-table-toolbar\">\n    <ng-container o-data-toolbar-projection-start>\n      <o-table-buttons #tableButtons [insert-button]=\"insertButton\" [refresh-button]=\"refreshButton\" [delete-button]=\"showDeleteButton\">\n        <ng-content select=\"o-table-button\"></ng-content>\n      </o-table-buttons>\n    </ng-container>\n    <ng-content select=\"[o-table-toolbar][position=start]\" ngProjectAs=\"[o-data-toolbar-custom-projection-start]\">\n    </ng-content>\n    <ng-content select=\"[o-table-toolbar][position=end]\" ngProjectAs=\"[o-data-toolbar-custom-projection-end]\">\n    </ng-content>\n    <ng-content select=\"[o-table-toolbar]\" ngProjectAs=\"[o-data-toolbar-custom-projection-start]\">\n    </ng-content>\n    <ng-container o-data-toolbar-projection-end>\n      <ng-container *ngIf=\"quickfilterContentChild; else defaultQuickFilter\">\n        <ng-content select=\"o-table-quickfilter\"></ng-content>\n      </ng-container>\n      <ng-template #defaultQuickFilter>\n        <ng-container *ngIf=\"quickFilter\">\n          <o-table-quickfilter (onChange)=\"tableQuickFilterChanged($event)\">\n          </o-table-quickfilter>\n        </ng-container>\n      </ng-template>\n      <button type=\"button\" *ngIf=\"showTableMenuButton\" mat-icon-button class=\"o-table-menu-button\" [matMenuTriggerFor]=\"tableMenu.matMenu\"\n        (click)=\"$event.stopPropagation()\">\n        <mat-icon svgIcon=\"ontimize:more_vert\"></mat-icon>\n      </button>\n      <o-table-menu #tableMenu [select-all-checkbox]=\"selectAllCheckbox\" [export-button]=\"exportButton\"\n        [columns-visibility-button]=\"columnsVisibilityButton\" [show-configuration-option]=\"showConfigurationOption\"\n        [show-filter-option]=\"showFilterOption\" [show-report-on-demand-option]=\"showReportOnDemandOption\"\n        [show-charts-on-demand-option]=\"showChartsOnDemandOption\" [show-reset-width-option]=\"showResetWidthOption\" [show-group-by-option]=\"groupable\">\n        <ng-content select=\"o-table-option\"></ng-content>\n      </o-table-menu>\n      <ng-template #exportOptsTemplate>\n        <ng-content select=\"o-table-export-button\"></ng-content>\n      </ng-template>\n    </ng-container>\n  </o-data-toolbar>\n\n  <div #tableBody class=\"o-table-body o-scroll\" [class.horizontal-scroll]=\"horizontalScroll\" [class.scrolled]=\"horizontalScrolled\">\n    <ng-container *ngIf=\"!enabledVirtualScroll; else tableWithVirtualScroll\">\n      <div class=\"o-table-overflow o-scroll\">\n        <ng-template *ngTemplateOutlet=\"table\"></ng-template>\n      </div>\n    </ng-container>\n    <ng-template #tableWithVirtualScroll>\n      <cdk-virtual-scroll-viewport #virtualScrollViewPort fxFlex>\n        <ng-template *ngTemplateOutlet=\"table\"></ng-template>\n      </cdk-virtual-scroll-viewport>\n    </ng-template>\n  </div>\n  <!--TABLE PAGINATOR-->\n  <mat-paginator *ngIf=\"paginator\" #matpaginator [length]=\"dataSource?.resultsLength\" [pageIndex]=\"paginator.pageIndex\" [pageSize]=\"queryRows\"\n    [pageSizeOptions]=\"paginator.pageSizeOptions\" (page)=\"onChangePage($event)\" [showFirstLastButtons]=\"paginator.showFirstLastButtons\">\n  </mat-paginator>\n\n  <!--LOADING-->\n  <div #spinnerContainer *ngIf=\"showLoading | async\" fxLayout=\"column\" fxLayoutAlign=\"center center\" [ngStyle]=\"{'top.px': headerHeight}\"\n    class=\"spinner-container\" [class.spinner-container-scrollable]=\"loadingScroll | async\">\n    <mat-progress-spinner mode=\"indeterminate\" strokeWidth=\"3\" [diameter]=\"diameterSpinner\"></mat-progress-spinner>\n  </div>\n\n  <!-- Disable blocker -->\n  <div *ngIf=\"!enabled\" class=\"o-table-disabled-blocker\"></div>\n</div>\n\n<ng-template #table>\n\n  <table mat-table #table [class.autoadjusted]=\"autoAdjust\" [trackBy]=\"getTrackByFunction()\" [dataSource]=\"dataSource\" oMatSort\n    [oMatSortColumns]=\"sortColArray\" [ngClass]=\"rowHeightObservable | async\" (cdkObserveContent)=\"projectContentChanged()\"\n    [oTableExpandedFooter]=\"!(loading | async)\" [oTableExpandedFooterColspan]=\"visibleColArray.length\" [multiTemplateDataRows]=\"showExpandableRow()\"\n    aria-describedby=\"ontimize-web table\">\n\n    <!--Checkbox Column -->\n    <ng-container [matColumnDef]=\"oTableOptions.selectColumn.name\" *ngIf=\"oTableOptions.selectColumn.visible\">\n      <ng-container *ngIf=\"!tableColumnSelectAllContentChild; else customHeaderSelectAllTemplate\">\n        <th mat-header-cell *matHeaderCellDef>\n          <div class=\"content\">\n            <o-table-header-select-all [column]=\"oTableOptions.selectColumn\"></o-table-header-select-all>\n          </div>\n        </th>\n        <td mat-cell *matCellDef=\"let row\">\n          <mat-checkbox name=\"id[]\" (click)=\"$event.stopPropagation()\" [disabled]=\"isDisableCheckbox(row)\"\n            (change)=\"selectionCheckboxToggle($event, row)\" [checked]=\"isRowSelected(row)\">\n          </mat-checkbox>\n        </td>\n      </ng-container>\n      <ng-template #customHeaderSelectAllTemplate>\n        <th mat-header-cell *matHeaderCellDef [class.resizable]=\"resizable\"\n          class=\"mat-header-select-all-with-title o-center\" [style.width]=\"oTableOptions.selectColumn.width\"\n          [style.min-width]=\"getMinWidthColumn(oTableOptions.selectColumn)\" [style.max-width]=\"oTableOptions.selectColumn.maxWidth\">\n          <div class=\"content\">\n            <o-table-header-select-all [column]=\"oTableOptions.selectColumn\"></o-table-header-select-all>\n          </div>\n        </th>\n        <td mat-cell *matCellDef=\"let row\" class=\"o-center\" [style.width]=\"oTableOptions.selectColumn.width\"\n          [style.min-width]=\"getMinWidthColumn(oTableOptions.selectColumn)\" [style.max-width]=\"oTableOptions.selectColumn.maxWidth\">\n          <mat-checkbox name=\"id[]\" (click)=\"$event.stopPropagation()\" [disabled]=\"isDisableCheckbox(row)\"\n            (change)=\"selectionCheckboxToggle($event, row)\" [checked]=\"isRowSelected(row)\">\n          </mat-checkbox>\n        </td>\n      </ng-template>\n\n\n      <td mat-footer-cell *matFooterCellDef></td>\n    </ng-container>\n\n    <!--Expandable Column -->\n    <ng-container [matColumnDef]=\"oTableOptions.expandableColumn.name\" *ngIf=\"isColumnExpandable()\">\n      <th mat-header-cell *matHeaderCellDef>\n        {{ oTableOptions.expandableColumn.title }}\n      </th>\n      <td mat-cell *matCellDef=\"let row;let rowIndex = dataIndex\">\n        <mat-icon *ngIf=\"showExpandableIcon(row, rowIndex) | async\" (click)=\"toogleRowExpandable(row, rowIndex, $event)\">\n          <ng-container *ngIf=\"isExpanded(row)\">{{ tableRowExpandable.iconCollapse }}</ng-container>\n          <ng-container *ngIf=\"!isExpanded(row)\">{{ tableRowExpandable.iconExpand }}</ng-container>\n        </mat-icon>\n      </td>\n    </ng-container>\n\n    <!-- Generic column definition -->\n    <ng-container *ngFor=\"let column of oTableOptions.columns\" [matColumnDef]=\"column.name\">\n      <!--Define header-cell-->\n\n      <th mat-header-cell *matHeaderCellDef [ngClass]=\"getTitleAlignClass(column)\" [class.resizable]=\"resizable\" [style.width]=\"column.width\"\n        [style.min-width]=\"getMinWidthColumn(column)\" [style.max-width]=\"column.maxWidth\">\n\n        <div class=\"content\">\n          <o-table-header [column]=\"column\"></o-table-header>\n        </div>\n      </th>\n\n\n      <!--Define mat-cell-->\n      <ng-container *ngIf=\"!table.multiTemplateDataRows; else cellTemplateMultiTemplateDataRows\">\n        <td #cell mat-cell *matCellDef=\"let row;let rowIndex = index \" [ngClass]=\"[column.className, getCellAlignClass(column)]\"\n          (click)=\"handleClick(row, column, rowIndex, cell, $event)\" (dblclick)=\"handleDoubleClick(row, column, rowIndex, cell, $event)\"\n          [class.empty-cell]=\"isEmpty(row[column.name])\" [matTooltipDisabled]=\"!column.hasTooltip()\" [matTooltip]=\"column.getTooltip(row)\"\n          matTooltipPosition=\"below\" matTooltipShowDelay=\"750\" matTooltipClass=\"o-table-cell-tooltip\"\n          [class.o-mat-cell-multiline]=\"(column.isMultiline | async)\" [oContextMenu]=\"tableContextMenu\"\n          [oContextMenuData]=\"{ cellName:column.name, rowValue:row, rowIndex:rowIndex}\" [style.width]=\"column.width\"\n          [style.min-width]=\"getMinWidthColumn(column)\" [style.max-width]=\"column.maxWidth\"\n          [class.o-table-editing-cell]=\"isRowSelected(row) && column.editing\">\n          <ng-container *ngTemplateOutlet=\"cellRenderer;context:{column:column,row:row}\"></ng-container>\n        </td>\n      </ng-container>\n      <ng-template #cellTemplateMultiTemplateDataRows>\n        <td #cell mat-cell *matCellDef=\"let row;let rowIndex = dataIndex \" [ngClass]=\"[column.className, getCellAlignClass(column)]\"\n          (click)=\"handleClick(row, column, rowIndex, cell, $event)\" (dblclick)=\"handleDoubleClick(row, column, rowIndex, cell, $event)\"\n          [class.empty-cell]=\"isEmpty(row[column.name])\" [matTooltipDisabled]=\"!column.hasTooltip()\" [matTooltip]=\"column.getTooltip(row)\"\n          matTooltipPosition=\"below\" matTooltipShowDelay=\"750\" matTooltipClass=\"o-table-cell-tooltip\"\n          [class.o-mat-cell-multiline]=\"(column.isMultiline | async)\" [oContextMenu]=\"tableContextMenu\"\n          [oContextMenuData]=\"{ cellName:column.name, rowValue:row, rowIndex:rowIndex}\" [style.width]=\"column.width\"\n          [style.min-width]=\"getMinWidthColumn(column)\" [style.max-width]=\"column.maxWidth\"\n          [class.o-table-editing-cell]=\"isRowSelected(row) && column.editing\">\n          <ng-container *ngTemplateOutlet=\"cellRenderer;context:{column:column,row:row}\"></ng-container>\n\n        </td>\n      </ng-template>\n      <!--Define mat-footer-cell-->\n      <ng-container *ngIf=\"showTotals | async\">\n        <td mat-footer-cell *matFooterCellDef [ngClass]=\"column.className\">\n          <div class=\"title\" *ngIf=\"column.aggregate && column.aggregate.title\">\n            {{ column.aggregate.title | oTranslate }}\n          </div>\n          <ng-container *ngIf=\"!column.renderer\">\n            {{ dataSource.getAggregateData(column) }}\n          </ng-container>\n          <ng-template *ngIf=\"column.renderer && column.aggregate\" [ngTemplateOutlet]=\"column.renderer.templateref\"\n            [ngTemplateOutletContext]=\"{cellvalue: dataSource.getAggregateData(column)}\"></ng-template>\n        </td>\n      </ng-container>\n\n    </ng-container>\n\n    <!-- Expanded Content Column - The detail row is made up of this one column that spans across all columns -->\n    <ng-container *ngIf=\"hasExpandedRow\">\n      <ng-container matColumnDef=\"expandedDetail\">\n        <td mat-cell *matCellDef=\"let row;let rowIndex= dataIndex\" [attr.colspan]=\"oTableOptions.visibleColumns.length\">\n          <div [ngClass]=\"getExpandedRowContainerClass(rowIndex)\" [@detailExpand]=\"getStateExpand(row)\">\n          </div>\n        </td>\n      </ng-container>\n    </ng-container>\n\n    <!--FOOTER-INSERTABLE-->\n    <ng-container *ngIf=\"showLastInsertableRow && oTableInsertableRowComponent\">\n      <ng-container [matColumnDef]=\"oTableOptions.selectColumn.name + getSuffixColumnInsertable()\" *ngIf=\"oTableOptions.selectColumn.visible\">\n        <td mat-footer-cell *matFooterCellDef>\n        </td>\n      </ng-container>\n      <ng-container *ngFor=\"let column of oTableOptions.columns\" [matColumnDef]=\"column.name+ getSuffixColumnInsertable()\">\n\n        <td mat-footer-cell *matFooterCellDef [ngClass]=\"column.className\">\n          <ng-container *ngIf=\"oTableInsertableRowComponent.isColumnInsertable(column) && !oTableInsertableRowComponent.useCellEditor(column)\">\n            <mat-form-field class=\"insertable-form-field o-table-cell-editor-text o-table-cell-editor\" [hideRequiredMarker]=\"false\"\n              floatLabel=\"never\">\n              <input matInput type=\"text\" [placeholder]=\"oTableInsertableRowComponent.getPlaceholder(column)\" [id]=\"column.attr\"\n                [formControl]=\"oTableInsertableRowComponent.getControl(column)\" [required]=\"oTableInsertableRowComponent.isColumnRequired(column)\">\n              <mat-error *oMatError=\"oTableInsertableRowComponent.columnHasError(column, 'required')\">\n                {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n              </mat-error>\n            </mat-form-field>\n          </ng-container>\n\n          <ng-container *ngIf=\"oTableInsertableRowComponent.isColumnInsertable(column) && oTableInsertableRowComponent.useCellEditor(column)\">\n            <ng-template [ngTemplateOutlet]=\"oTableInsertableRowComponent.columnEditors[column.attr].templateref\"\n              [ngTemplateOutletContext]=\"{ rowvalue: oTableInsertableRowComponent.rowData }\">\n            </ng-template>\n          </ng-container>\n        </td>\n      </ng-container>\n\n    </ng-container>\n\n    <ng-container *ngIf=\"showFirstInsertableRow && oTableInsertableRowComponent\">\n      <ng-container [matColumnDef]=\"getColumnInsertable(oTableOptions.selectColumn.name)\" *ngIf=\"oTableOptions.selectColumn.visible\">\n        <td mat-header-cell *matHeaderCellDef>\n        </td>\n      </ng-container>\n      <ng-container *ngFor=\"let column of oTableOptions.columns\" [matColumnDef]=\"getColumnInsertable(column.name)\">\n\n        <td mat-header-cell *matHeaderCellDef [ngClass]=\"column.className\">\n          <ng-container *ngIf=\"oTableInsertableRowComponent.isColumnInsertable(column) && !oTableInsertableRowComponent.useCellEditor(column)\">\n            <mat-form-field class=\"insertable-form-field\" [hideRequiredMarker]=\"false\" floatLabel=\"never\">\n              <input matInput type=\"text\" [placeholder]=\"oTableInsertableRowComponent.getPlaceholder(column)\" [id]=\"column.attr\"\n                [formControl]=\"oTableInsertableRowComponent.getControl(column)\" [required]=\"oTableInsertableRowComponent.isColumnRequired(column)\">\n              <mat-error *oMatError=\"oTableInsertableRowComponent.columnHasError(column, 'required')\">\n                {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n              </mat-error>\n            </mat-form-field>\n          </ng-container>\n\n          <ng-container *ngIf=\"oTableInsertableRowComponent.isColumnInsertable(column) && oTableInsertableRowComponent.useCellEditor(column)\">\n            <ng-template [ngTemplateOutlet]=\"oTableInsertableRowComponent.columnEditors[column.attr].templateref\"\n              [ngTemplateOutletContext]=\"{ rowvalue: oTableInsertableRowComponent.rowData }\">\n            </ng-template>\n          </ng-container>\n        </td>\n      </ng-container>\n\n    </ng-container>\n\n    <!-- Definition column group header -->\n    <ng-container *ngFor=\"let column of groupingHeadersRows; let i = index\" [matColumnDef]=\"column\">\n      <td mat-cell *matCellDef=\"let group\" class=\"grouping-row\" [oContextMenu]=\"tableContextMenu\"\n        [oContextMenuData]=\"{ cellName:column, rowValue:group, rowIndex:i}\" [ngClass]=\"getGroupHeaderCellAlignClass(column)\">\n        <div *ngIf=\"i===0\" class=\"grouping-title-wrapper\" [ngStyle]=\"{'padding-left': 20*(group.level-1)+'px'}\">\n          <mat-icon>{{ group.expanded ? 'expand_more' : 'chevron_right' }}</mat-icon>\n          {{ group.title }}\n        </div>\n        <div class=\"grouping-aggregate\" *ngIf=\"group.hasActiveAggregate(visibleColArray[i])\">\n          {{ group.getColumnActiveAggregateTitle(visibleColArray[i]) | oTranslate }} :\n          <ng-container *ngIf=\"!getOColumnFromGroupHeaderColumn(column).renderer\">\n            {{ group.getColumnAggregateValue(visibleColArray[i])}}\n          </ng-container>\n          <ng-container *ngIf=\"getOColumnFromGroupHeaderColumn(column).renderer\">\n            <ng-template\n              *ngTemplateOutlet=\"getOColumnFromGroupHeaderColumn(column).renderer?.templateref; context:{ cellvalue:  group.getColumnAggregateValue(visibleColArray[i]) }\">\n            </ng-template>\n          </ng-container>\n        </div>\n      </td>\n    </ng-container>\n\n    <ng-container matColumnDef=\"position\">\n      <td mat-cell *matCellDef=\"let row\"> {{row}} </td>\n    </ng-container>\n\n    <tr #tableHeader mat-header-row *matHeaderRowDef=\"oTableOptions.visibleColumns; sticky: fixedHeader\"></tr>\n\n    <ng-container *ngIf=\"!table.multiTemplateDataRows; else rowTemplateMultiTemplateDataRows\">\n      <tr mat-row oTableRow *matRowDef=\"let row; columns: oTableOptions.visibleColumns; when:isNotGroup; let rowIndex = index\"\n        [class.selected]=\"isRowSelected(row)\" [ngClass]=\"row | oTableRowClass: rowIndex: rowClass\">\n      </tr>\n    </ng-container>\n    <ng-template #rowTemplateMultiTemplateDataRows>\n      <tr mat-row oTableRow *matRowDef=\"let row; columns: oTableOptions.visibleColumns; when:isNotGroup; let rowIndex = dataIndex\"\n        [class.selected]=\"isRowSelected(row)\" [ngClass]=\"row | oTableRowClass: rowIndex: rowClass\">\n      </tr>\n    </ng-template>\n\n    <!-- Row Group header -->\n    <tr mat-row *matRowDef=\"let row; columns: groupingHeadersRows; when:isGroup\" (click)=\"groupHeaderClick(row)\"\n      [ngClass]=\"getClassNameGroupHeader(row)\">\n    </tr>\n\n    <!-- Expanded detail row-->\n    <ng-container *ngIf=\"hasExpandedRow\">\n      <tr mat-row *matRowDef=\"let row; columns: ['expandedDetail']\" class=\"o-table-row-expanded\"></tr>\n    </ng-container>\n\n    <ng-container *ngIf=\"showLastInsertableRow\">\n      <tr mat-footer-row *matFooterRowDef=\"oTableOptions.columnsInsertables; sticky: true\"\n        (keyup)=\"oTableInsertableRowComponent.handleKeyboardEvent($event)\" class=\"o-table-insertable\"></tr>\n    </ng-container>\n    <ng-container *ngIf=\"showFirstInsertableRow\">\n      <tr mat-header-row *matHeaderRowDef=\"oTableOptions.columnsInsertables; sticky: true\"\n        (keyup)=\"oTableInsertableRowComponent.handleKeyboardEvent($event)\" class=\"o-table-insertable\"> </tr>\n    </ng-container>\n    <ng-container *ngIf=\"showTotals | async\">\n      <tr mat-footer-row *matFooterRowDef=\"oTableOptions.visibleColumns; sticky: true\" class=\"o-table-aggregate\">\n      </tr>\n    </ng-container>\n  </table>\n\n</ng-template>\n\n<ng-container *ngIf=\"!contextMenuContentChild && contextMenu\">\n  <o-table-context-menu [insert]=\"insertButton\" [edit]=\"editionMode !== EDIT_MODE_NONE\" [view-detail]=\"detailMode !== DETAIL_MODE_NONE\"\n    [refresh]=\"refreshButton\" [delete]=\"deleteButton\" [filter]=\"showFilterOption\" [group-by-row]=\"groupable\">\n  </o-table-context-menu>\n</ng-container>\n\n<ng-template #cellRenderer let-row=\"row\" let-column=\"column\">\n  <div class=\"content\">\n\n    <ng-container [ngSwitch]=\"true\">\n      <ng-container *ngSwitchCase=\"column.renderer != null && (!column.editing || column.editing && !isRowSelected(row))\">\n        <ng-template *ngTemplateOutlet=\"column.renderer?.templateref; context:{ cellvalue: row[column.name], rowvalue:row }\">\n        </ng-template>\n      </ng-container>\n      <ng-container *ngSwitchCase=\"isRowSelected(row) && column.editing\">\n        <ng-template *ngTemplateOutlet=\"column.editor?.templateref; context:{ cellvalue: row[column.name], rowvalue:row }\">\n        </ng-template>\n      </ng-container>\n\n      <ng-container *ngSwitchCase=\"column.type === 'editButtonInRow' || column.type === 'detailButtonInRow'\">\n        <div fxLayoutAlign=\"center center\" class=\"o-action-cell-renderer\" (click)=\"onDetailButtonClick(column, row, $event)\">\n          <mat-icon>{{ getDetailButtonIcon(column) }}</mat-icon>\n        </div>\n      </ng-container>\n      <ng-container *ngSwitchDefault>{{ row[column.name] }}</ng-container>\n    </ng-container>\n\n  </div>\n</ng-template>\n",
                    providers: [
                        OntimizeServiceProvider,
                        ComponentStateServiceProvider,
                        OTableDataSourceService,
                        { provide: O_COMPONENT_STATE_SERVICE, useClass: OTableComponentStateService },
                        { provide: VIRTUAL_SCROLL_STRATEGY, useClass: OTableVirtualScrollStrategy }
                    ],
                    animations: [
                        trigger('detailExpand', [
                            state('collapsed', style({ height: '0px', minHeight: '0' })),
                            state('expanded', style({ height: '*' })),
                            transition('expanded <=> collapsed', animate('225ms cubic-bezier(0.4, 0.0, 0.2, 1)')),
                        ])
                    ],
                    inputs: DEFAULT_INPUTS_O_TABLE,
                    outputs: DEFAULT_OUTPUTS_O_TABLE,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.o-table]': 'true',
                        '[class.ontimize-table]': 'true',
                        '[class.o-table-fixed]': 'fixedHeader',
                        '[class.o-table-disabled]': '!enabled',
                        '(document:click)': 'handleDOMClick($event)'
                    },
                    styles: [".o-table{height:100%;max-height:100%;width:100%}.o-table.o-table-disabled{opacity:.4}.o-table .o-table-container{height:100%;display:flex;flex-direction:column;flex-wrap:nowrap;justify-content:flex-start;align-items:flex-start;align-content:stretch;min-width:100%;min-height:400px;position:relative;padding:0 .5%}.o-table .o-table-container .o-table-body .o-table-overflow{overflow-y:auto;overflow-x:hidden;min-width:100%}.o-table .o-table-container .o-table-body.horizontal-scroll .o-table-overflow{overflow-x:auto}.o-table .o-table-container .o-table-body thead .mat-header-row th:last-child .o-table-column-resizer{display:none}.o-table .o-table-container.block-events{pointer-events:none}.o-table .o-table-container.block-events>.o-table-body .mat-header-row,.o-table .o-table-container.block-events>.o-table-toolbar{opacity:.75}.o-table .o-table-container .o-table-toolbar{height:40px}.o-table .o-table-container .o-table-toolbar>div{max-height:100%}.o-table .o-table-container .o-table-toolbar .buttons{margin:0 10px 0 4px}.o-table .o-table-container .o-table-body{display:flex;flex:1 1 auto;max-width:100%;height:100%;overflow:hidden;position:relative}.o-table .o-table-container .o-table-body .spinner-container{position:absolute;top:0;left:0;right:0;bottom:0;z-index:500;visibility:visible;opacity:1;transition:opacity .25s linear}.o-table .o-table-container .o-table-body.horizontal-scroll{overflow-x:auto;padding-bottom:16px}.o-table .o-table-container .o-table-body.horizontal-scroll .mat-header-cell{width:150px}.o-table .o-table-container .o-table-body .o-table-no-results{cursor:default;text-align:center}.o-table .o-table-container .o-table-body .o-table-no-results td{text-align:center}.o-table .o-table-container .mat-table{table-layout:fixed;width:100%}.o-table .o-table-container .mat-table.autoadjusted{table-layout:auto}.o-table .o-table-container .mat-table td .content,.o-table .o-table-container .mat-table th .content{overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.o-table .o-table-container .mat-table.small .column-filter-icon{margin-top:2px}.o-table .o-table-container .mat-table.small .mat-sort-header-arrow{margin-top:4px}.o-table .o-table-container .mat-table.small .mat-header-row .mat-checkbox-inner-container,.o-table .o-table-container .mat-table.small .mat-row .mat-checkbox-inner-container{height:16px;width:16px}.o-table .o-table-container .mat-table.small .mat-header-row .mat-checkbox-inner-container .mat-checkbox-checkmark-path,.o-table .o-table-container .mat-table.small .mat-row .mat-checkbox-inner-container .mat-checkbox-checkmark-path{width:2.13333px}.o-table .o-table-container .mat-table.small .mat-header-row .mat-checkbox-inner-container .mat-checkbox-mixedmark,.o-table .o-table-container .mat-table.small .mat-row .mat-checkbox-inner-container .mat-checkbox-mixedmark{height:2px}.o-table .o-table-container .mat-table.small .mat-header-row .mat-cell .image-avatar,.o-table .o-table-container .mat-table.small .mat-header-row .mat-header-cell .image-avatar,.o-table .o-table-container .mat-table.small .mat-row .mat-cell .image-avatar,.o-table .o-table-container .mat-table.small .mat-row .mat-header-cell .image-avatar{width:24px;height:24px}.o-table .o-table-container .mat-table.medium .column-filter-icon{margin-top:2px}.o-table .o-table-container .mat-table.medium .mat-sort-header-arrow{margin-top:5px}.o-table .o-table-container .mat-table.medium .mat-header-row .mat-checkbox-inner-container,.o-table .o-table-container .mat-table.medium .mat-row .mat-checkbox-inner-container{height:18px;width:18px}.o-table .o-table-container .mat-table.medium .mat-header-row .mat-checkbox-inner-container .mat-checkbox-checkmark-path,.o-table .o-table-container .mat-table.medium .mat-row .mat-checkbox-inner-container .mat-checkbox-checkmark-path{width:2.4px}.o-table .o-table-container .mat-table.medium .mat-header-row .mat-checkbox-inner-container .mat-checkbox-mixedmark,.o-table .o-table-container .mat-table.medium .mat-row .mat-checkbox-inner-container .mat-checkbox-mixedmark{height:2px}.o-table .o-table-container .mat-table.large .column-filter-icon{margin-top:4px}.o-table .o-table-container .mat-table.large .mat-sort-header-arrow{margin-top:7px}.o-table .o-table-container .mat-table tr.mat-row.o-table-row-expanded{height:0}.o-table .o-table-container .mat-table tr.o-table-insertable td{height:1px}.o-table .o-table-container .mat-table .mat-row{box-sizing:border-box;transition:background-color .2s;position:relative;-webkit-touch-callout:none;-webkit-user-select:none;user-select:none}.o-table .o-table-container .mat-table .mat-row .mat-cell{padding:0 12px;overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.o-table .o-table-container .mat-table .mat-row .mat-cell:not(.o-column-image):first-of-type{padding-left:24px}.o-table .o-table-container .mat-table .mat-row .mat-cell:last-of-type{padding-right:24px}.o-table .o-table-container .mat-table .mat-row .mat-cell.grouping-row{padding-top:30px;cursor:pointer}.o-table .o-table-container .mat-table .mat-row .mat-cell.grouping-row .grouping-title-wrapper{position:absolute;width:100%;left:0;top:0;text-align:left}.o-table .o-table-container .mat-table .mat-row .mat-cell.grouping-row .grouping-aggregate{font-weight:700;font-size:14px;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;padding-bottom:8px}.o-table .o-table-container .mat-table .mat-row .mat-cell.empty-cell{min-height:16px}.o-table .o-table-container .mat-table .mat-row .mat-cell .action-cell-renderer{cursor:pointer}.o-table .o-table-container .mat-table .mat-row .mat-cell.o-start{text-align:start}.o-table .o-table-container .mat-table .mat-row .mat-cell.o-center{text-align:center}.o-table .o-table-container .mat-table .mat-row .mat-cell.o-end{text-align:end}.o-table .o-table-container .mat-table .mat-row .mat-cell *{vertical-align:middle}.o-table .o-table-container .mat-table .mat-row .mat-cell.o-mat-cell-multiline:not(.mat-header-cell){padding:6px 12px}.o-table .o-table-container .mat-table .mat-row .mat-cell.o-mat-cell-multiline .content{overflow:initial;white-space:normal;text-overflow:unset}.o-table .o-table-container .mat-table .mat-row .mat-cell .image-avatar{width:32px;height:32px;margin:1px auto;overflow:hidden;border-radius:50%;position:relative;z-index:1}.o-table .o-table-container .mat-table .mat-row .mat-cell .image-avatar img{position:absolute;top:50%;left:50%;transform:translate(-50%,-50%);width:100%;max-width:inherit;max-height:inherit}.o-table .o-table-container .mat-table .o-action-cell-renderer{display:inline-block;cursor:pointer}.o-table .o-table-container .mat-table .mat-header-cell{overflow:hidden;position:relative;box-sizing:border-box;padding:0 12px;vertical-align:middle}.o-table .o-table-container .mat-table .mat-header-cell.resizable{padding-right:24px}.o-table .o-table-container .mat-table .mat-header-cell.mat-column-select.mat-header-select-all-with-title{padding-right:12px}.o-table .o-table-container .mat-table .mat-header-cell:first-of-type{padding-left:0}.o-table .o-table-container .mat-table .mat-header-cell:not(.o-column-image){overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.o-table .o-table-container .mat-table .mat-header-cell .o-table-header-indicator-numbered{font-size:8px;position:absolute;text-align:center;display:inline-block;width:18px;height:18px;line-height:18px;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;pointer-events:none;bottom:-10px;right:-9px}.o-table .o-table-container .mat-table .mat-header-cell .column-filter-icon{cursor:pointer;float:left;font-size:18px;width:18px;height:18px;margin-right:2px;line-height:1}.o-table .o-table-container .mat-table .mat-header-cell .mat-sort-header-button{flex:1;display:block;place-content:center}.o-table .o-table-container .mat-table .mat-header-cell .mat-sort-header-arrow{position:absolute;right:0}.o-table .o-table-container .mat-table .mat-header-cell .header-title-container{cursor:default}.o-table .o-table-container .mat-table .mat-header-cell.resizable .mat-sort-header-arrow{margin-right:12px}.o-table .o-table-container .mat-table .mat-header-cell .header-title-container,.o-table .o-table-container .mat-table .mat-header-cell .mat-sort-header-button{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.o-table .o-table-container .mat-table .mat-header-cell.start,.o-table .o-table-container .mat-table .mat-header-cell.start .mat-sort-header-button{text-align:left}.o-table .o-table-container .mat-table .mat-header-cell.center,.o-table .o-table-container .mat-table .mat-header-cell.center .mat-sort-header-button{text-align:center}.o-table .o-table-container .mat-table .mat-header-cell.center [o-mat-sort-header] .mat-sort-header-button{padding-left:12px}.o-table .o-table-container .mat-table .mat-header-cell.end,.o-table .o-table-container .mat-table .mat-header-cell.end .mat-sort-header-button{text-align:right}.o-table .o-table-container .mat-table .mat-cell.mat-column-select,.o-table .o-table-container .mat-table .mat-header-cell.mat-column-select{box-sizing:content-box;overflow:initial}.o-table .o-table-container .mat-table .mat-cell.mat-column-select:not(.mat-header-select-all-with-title),.o-table .o-table-container .mat-table .mat-header-cell.mat-column-select:not(.mat-header-select-all-with-title){width:18px}.o-table .o-table-container .mat-table .mat-cell.mat-column-select .mat-checkbox-layout,.o-table .o-table-container .mat-table .mat-header-cell.mat-column-select .mat-checkbox-layout{text-overflow:ellipsis;overflow:hidden;display:inline}.o-table .o-table-container .mat-table .mat-cell.mat-column-expandable,.o-table .o-table-container .mat-table .mat-header-cell.mat-column-expandable{width:40px;box-sizing:content-box;padding:0 0 0 24px;overflow:initial}.o-table .o-table-container .mat-table .mat-cell .row-container-expanded{overflow:hidden;display:flex}.o-table .o-table-container .o-table-disabled-blocker{bottom:0;left:0;position:absolute;right:0;top:0;z-index:100}.o-table .spinner-container{position:absolute;top:0;left:0;right:0;bottom:0;z-index:500;visibility:visible;opacity:1;transition:opacity .25s linear}.o-table .spinner-container-scrollable{position:relative}.o-table.o-table-fixed{display:flex}.o-table.o-table-fixed .o-table-container{display:flex;flex-direction:column}.o-table.o-table-fixed .o-table-body{display:flex;flex:1}.o-table.o-table-fixed .o-table-body .o-table-overflow{flex:1;overflow-y:auto}.mat-tooltip.o-table-cell-tooltip{word-wrap:break-word;overflow:hidden;min-width:140px}"]
                }] }
    ];
    OTableComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ElementRef },
        { type: MatDialog },
        { type: ViewContainerRef },
        { type: ApplicationRef },
        { type: ComponentFactoryResolver },
        { type: OFormComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OFormComponent; }),] }] },
        { type: OTableVirtualScrollStrategy, decorators: [{ type: Optional }, { type: Inject, args: [VIRTUAL_SCROLL_STRATEGY,] }] }
    ]; };
    OTableComponent.propDecorators = {
        sort: [{ type: ViewChild, args: [OMatSort, { static: false },] }],
        cdkVirtualScrollViewport: [{ type: ViewChild, args: ['virtualScrollViewPort', { static: false },] }],
        spinnerContainer: [{ type: ViewChild, args: ['spinnerContainer', { read: ElementRef, static: false },] }],
        tableRowExpandable: [{ type: ContentChild, args: [OTableRowExpandableComponent, { static: false },] }],
        tableBodyEl: [{ type: ViewChild, args: ['tableBody', { static: false },] }],
        tableHeaderEl: [{ type: ViewChild, args: ['tableHeader', { read: ElementRef, static: false },] }],
        tableToolbarEl: [{ type: ViewChild, args: ['tableToolbar', { read: ElementRef, static: false },] }],
        oTableMenu: [{ type: ViewChild, args: ['tableMenu', { static: false },] }],
        tableOptions: [{ type: ContentChildren, args: [OTableOptionComponent,] }],
        tableButtons: [{ type: ContentChildren, args: [OTableButtonComponent,] }],
        quickfilterContentChild: [{ type: ContentChild, args: ['o-table-quickfilter', { static: true },] }],
        tableColumnSelectAllContentChild: [{ type: ContentChild, args: [OTableColumnSelectAllDirective, { static: true },] }],
        exportOptsTemplate: [{ type: ViewChild, args: ['exportOptsTemplate', { static: false },] }],
        updateScrolledState: [{ type: HostListener, args: ['window:resize', [],] }],
        contextMenuContentChild: [{ type: ContentChild, args: [OTableContextMenuComponent, { static: true },] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "selectAllCheckbox", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "exportButton", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "showConfigurationOption", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "columnsVisibilityButton", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "showFilterOption", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "showReportOnDemandOption", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "showChartsOnDemandOption", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "showButtonsText", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "filterColumnActiveByDefault", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "showResetWidthOption", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "insertButton", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "refreshButton", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "deleteButton", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "fixedHeader", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "showTitle", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], OTableComponent.prototype, "horizontalScroll", null);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "showPaginatorFirstLastButtons", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "autoAlignTitles", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "multipleSort", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "orderable", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "resizable", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "autoAdjust", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "groupable", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "expandGroupsSameLevel", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "collapseGroupedColumns", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "virtualScroll", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "contextMenu", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableComponent.prototype, "keepSelectedItems", void 0);
    return OTableComponent;
}(AbstractOServiceComponent));
export { OTableComponent };
