// ontimize-web-ngx/esm5/lib/components/table/extensions/header/table-columns-grouping/o-table-columns-grouping.component.js
import { ChangeDetectionStrategy, Component, ContentChildren, forwardRef, Inject, Injector, QueryList } from '@angular/core';
import { Util } from '../../../../../util/util';
import { OTableComponent } from '../../../o-table.component';
import { OTableColumnsGroupingColumnComponent } from './columns/o-table-columns-grouping-column.component';
export var DEFAULT_INPUTS_O_TABLE_COLUMN_GROUPING = [
    'columns',
    'excludedAggregateColumns: excluded-aggregate-columns'
];
export var DEFAULT_OUTPUTS_O_TABLE_COLUMN_GROUPING = [];
var OTableColumnsGroupingComponent = (function () {
    function OTableColumnsGroupingComponent(injector, table) {
        this.injector = injector;
        this.table = table;
        this._columnsArray = [];
        this._excludedAggregateColumnsArray = [];
    }
    Object.defineProperty(OTableColumnsGroupingComponent.prototype, "columns", {
        set: function (arg) {
            this._columnsArray = Util.parseArray(arg, true);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableColumnsGroupingComponent.prototype, "columnsArray", {
        get: function () {
            return this._columnsArray;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableColumnsGroupingComponent.prototype, "excludedAggregateColumns", {
        set: function (arg) {
            this._excludedAggregateColumnsArray = Util.parseArray(arg, true);
        },
        enumerable: true,
        configurable: true
    });
    OTableColumnsGroupingComponent.prototype.ngOnInit = function () {
        if (this._columnsArray.length === 0) {
            this._columnsArray = this.table.groupedColumnsArray;
        }
        this.table.setOTableColumnsGrouping(this);
    };
    OTableColumnsGroupingComponent.prototype.useColumnAggregate = function (columnAttr, hasDefaultAggregate) {
        if (this._excludedAggregateColumnsArray.includes(columnAttr)) {
            return false;
        }
        var columnConf = this.getColumnGrouping(columnAttr);
        return hasDefaultAggregate || Util.isDefined(columnConf);
    };
    OTableColumnsGroupingComponent.prototype.getColumnGrouping = function (columnAttr) {
        return this.groupingColumns.find(function (col) { return col.attr === columnAttr; });
    };
    OTableColumnsGroupingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table-columns-grouping',
                    template: ' ',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_TABLE_COLUMN_GROUPING,
                    outputs: DEFAULT_OUTPUTS_O_TABLE_COLUMN_GROUPING
                }] }
    ];
    OTableColumnsGroupingComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: OTableComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OTableComponent; }),] }] }
    ]; };
    OTableColumnsGroupingComponent.propDecorators = {
        groupingColumns: [{ type: ContentChildren, args: [OTableColumnsGroupingColumnComponent, { descendants: true },] }]
    };
    return OTableColumnsGroupingComponent;
}());
export { OTableColumnsGroupingComponent };
