// ontimize-web-ngx/esm2015/lib/components/input/listpicker/listpicker-renderer/currency/o-list-picker-renderer-currency.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { OCurrencyPipe } from '../../../../../pipes/o-currency.pipe';
import { CurrencyService } from '../../../../../services/currency.service';
import { OListPickerCustomRenderer } from '../o-list-picker-renderer.class';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER_REAL } from '../real/o-list-picker-renderer-real.component';
export const DEFAULT_INPUTS_O_LISTPICKER_RENDERER_CURRENCY = [
    ...DEFAULT_INPUTS_O_LISTPICKER_RENDERER_REAL,
    'currencySymbol: currency-symbol',
    'currencySymbolPosition: currency-symbol-position'
];
export class OListPickerRendererCurrencyComponent extends OListPickerCustomRenderer {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.minDecimalDigits = 2;
        this.maxDecimalDigits = 2;
        this.decimalSeparator = '.';
        this.grouping = true;
        this.thousandSeparator = ',';
        this.currencyService = this.injector.get(CurrencyService);
        this.setComponentPipe();
    }
    setComponentPipe() {
        this.componentPipe = new OCurrencyPipe(this.injector);
    }
    initialize() {
        super.initialize();
        if (typeof this.currencySymbol === 'undefined') {
            this.currencySymbol = this.currencyService.symbol;
        }
        if (typeof this.currencySymbolPosition === 'undefined') {
            this.currencySymbolPosition = this.currencyService.symbolPosition;
        }
        this.pipeArguments = {
            currencySimbol: this.currencySymbol,
            currencySymbolPosition: this.currencySymbolPosition,
            minDecimalDigits: this.minDecimalDigits,
            maxDecimalDigits: this.maxDecimalDigits,
            decimalSeparator: this.decimalSeparator,
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator
        };
    }
}
OListPickerRendererCurrencyComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-list-picker-renderer-currency',
                template: "<ng-template #templateref let-value=\"value\">\n  {{getListPickerValue(value)}}\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_LISTPICKER_RENDERER_CURRENCY
            }] }
];
OListPickerRendererCurrencyComponent.ctorParameters = () => [
    { type: Injector }
];
OListPickerRendererCurrencyComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
};
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Number)
], OListPickerRendererCurrencyComponent.prototype, "minDecimalDigits", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Number)
], OListPickerRendererCurrencyComponent.prototype, "maxDecimalDigits", void 0);
