// ontimize-web-ngx/esm5/lib/components/app-header/o-app-header.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Injector, ViewChild, ViewEncapsulation } from '@angular/core';
import { InputConverter } from '../../decorators/input-converter';
import { AuthService } from '../../services';
import { DialogService } from '../../services/dialog.service';
import { OModulesInfoService } from '../../services/o-modules-info.service';
import { Codes } from '../../util';
import { OUserInfoComponent } from '../user-info/o-user-info.component';
export var DEFAULT_INPUTS_O_APP_HEADER = [
    'showUserInfo: show-user-info',
    'showLanguageSelector: show-language-selector',
    'useFlagIcons: use-flag-icons',
    'color',
    'headerHeight:header-height',
    'showTitle: show-title',
    'staticTitle: static-title',
    'showStaticTitle: show-static-title'
];
export var DEFAULT_OUTPUTS_O_APP_HEADER = [
    'onSidenavToggle'
];
var OAppHeaderComponent = (function () {
    function OAppHeaderComponent(injector) {
        this.injector = injector;
        this.showTitle = false;
        this.showStaticTitle = false;
        this.showUserInfo = true;
        this.showLanguageSelector = true;
        this.useFlagIcons = false;
        this.onSidenavToggle = new EventEmitter();
        this._headerHeight = Codes.DEFAULT_ROW_HEIGHT;
        this.dialogService = this.injector.get(DialogService);
        this.modulesInfoService = this.injector.get(OModulesInfoService);
        this.authService = this.injector.get(AuthService);
    }
    Object.defineProperty(OAppHeaderComponent.prototype, "headerHeight", {
        get: function () {
            return this._headerHeight;
        },
        set: function (value) {
            this._headerHeight = value ? value.toLowerCase() : value;
            if (!Codes.isValidRowHeight(this._headerHeight)) {
                this._headerHeight = Codes.DEFAULT_ROW_HEIGHT;
            }
        },
        enumerable: true,
        configurable: true
    });
    OAppHeaderComponent.prototype.ngOnInit = function () {
        if (!this.showStaticTitle) {
            this.headerTitle$ = this.modulesInfoService.getModuleChangeObservable();
        }
    };
    OAppHeaderComponent.prototype.onLogoutClick = function () {
        this.authService.logoutWithConfirmation();
    };
    Object.defineProperty(OAppHeaderComponent.prototype, "color", {
        get: function () {
            return this._color;
        },
        set: function (newValue) {
            this._color = newValue;
        },
        enumerable: true,
        configurable: true
    });
    OAppHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-app-header',
                    inputs: DEFAULT_INPUTS_O_APP_HEADER,
                    outputs: DEFAULT_OUTPUTS_O_APP_HEADER,
                    template: "<nav fxFlex fxLayout=\"row\" fxLayoutAlign=\"space-between stretch\" [ngClass]=\"color ? 'o-app-header-'+color : ''\">\n  <div fxFlex fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n    <ng-content></ng-content>\n    <ng-content select=\"o-app-layout-header-projection-start\"></ng-content>\n    <span *ngIf=\"showTitle && !showStaticTitle\" class=\"o-app-header-title\" fxLayout=\"row\" fxLayoutAlign=\"center center\">\n      {{ headerTitle$ | async | oTranslate }}\n    </span>\n    <span *ngIf=\"showTitle && showStaticTitle\" class=\"o-app-header-title\" fxLayout=\"row\" fxLayoutAlign=\"center center\">\n      {{ staticTitle | oTranslate }}\n    </span>\n    <div fxLayout=\"row\" fxLayoutAlign=\"end stretch\" class=\"o-app-header-default-actions\">\n      <div class=\"layout-header-projection-end\">\n        <ng-content select=\"o-app-layout-header-projection-end\"></ng-content>\n      </div>\n      <o-user-info #userInfo *ngIf=\"showUserInfo\"></o-user-info>\n      <o-language-selector *ngIf=\"showLanguageSelector\" [use-flag-icons]=\"useFlagIcons\"></o-language-selector>\n    </div>\n  </div>\n</nav>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-app-header]': 'true',
                        '[class.o-app-header-small]': 'headerHeight==="small"',
                        '[class.o-app-header-medium]': 'headerHeight==="medium"',
                        '[class.o-app-header-large]': 'headerHeight==="large"'
                    },
                    styles: [".o-app-header{z-index:2;box-shadow:0 2px 6px #00000029}.o-app-header.o-app-header-large{height:56px}.o-app-header.o-app-header-large+.header-layout{top:56px}.o-app-header.o-app-header-large .o-user-info{font-size:16px}.o-app-header.o-app-header-large .o-user-info .o-user-info-avatar-wrapper{height:40px;width:40px}.o-app-header.o-app-header-medium{height:48px}.o-app-header.o-app-header-medium+.header-layout{top:48px}.o-app-header.o-app-header-medium .o-user-info{font-size:16px}.o-app-header.o-app-header-medium .o-user-info .o-user-info-avatar-wrapper{height:32px;width:32px}.o-app-header.o-app-header-small{height:40px}.o-app-header.o-app-header-small+.header-layout{top:40px}.o-app-header.o-app-header-small .o-user-info{font-size:14px}.o-app-header.o-app-header-small .o-user-info .o-user-info-avatar-wrapper{height:30px;width:30px}.o-app-header nav{padding-right:16px}.o-app-header nav .sidenav-toggle{text-align:center;padding:16px;cursor:pointer}.o-app-header nav .o-app-header-title{padding:0 16px;cursor:default}.o-app-header nav .o-app-header-default-actions{margin-left:auto}.o-app-header nav .o-app-header-default-actions .layout-header-projection-end{margin-right:16px}"]
                }] }
    ];
    OAppHeaderComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OAppHeaderComponent.propDecorators = {
        userInfo: [{ type: ViewChild, args: ['userInfo', { static: false },] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppHeaderComponent.prototype, "showUserInfo", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppHeaderComponent.prototype, "showLanguageSelector", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OAppHeaderComponent.prototype, "useFlagIcons", void 0);
    return OAppHeaderComponent;
}());
export { OAppHeaderComponent };
