// ontimize-web-ngx/esm2015/lib/services/state/o-component-state.service.js
import { Injectable, Injector } from '@angular/core';
import { Util } from '../../util';
import { LocalStorageService } from '../local-storage.service';
import { DefaultComponentStateClass, DefaultServiceComponentStateClass } from './o-component-state.class';
export class AbstractComponentStateService {
    constructor(injector) {
        this.injector = injector;
        this.localStorageService = injector.get(LocalStorageService);
    }
    initialize(comp) {
        this.component = comp;
        if (Util.isDefined(this.state)) {
            this.initializeState(this.state);
        }
    }
    initializeState(state) {
        if (Util.isDefined(this.state) &&
            ((Util.isDefined(this.component.storeState) && this.component.storeState
                ||
                    !Util.isDefined(this.component.storeState)))) {
            state.setData(this.localStorageService.getComponentStorage(this.component, this.component.getRouteKey()));
        }
    }
}
AbstractComponentStateService.decorators = [
    { type: Injectable }
];
AbstractComponentStateService.ctorParameters = () => [
    { type: Injector }
];
export class DefaultComponentStateService extends AbstractComponentStateService {
    initialize(comp) {
        this.state = new DefaultComponentStateClass();
        super.initialize(comp);
    }
}
DefaultComponentStateService.decorators = [
    { type: Injectable }
];
export class DefaultServiceComponentStateService extends AbstractComponentStateService {
    initialize(comp) {
        this.state = new DefaultServiceComponentStateClass();
        super.initialize(comp);
    }
}
DefaultServiceComponentStateService.decorators = [
    { type: Injectable }
];
