// ontimize-web-ngx/esm2015/lib/services/state/o-filter-builder-component-state.class.js
import { Util } from '../../util';
import { DefaultComponentStateClass } from './o-component-state.class';
export class OFilterBuilderComponentStateClass extends DefaultComponentStateClass {
    get storedFilterBuilders() {
        return this['stored-filter-builders'] || [];
    }
    set storedFilterBuilders(value) {
        this['stored-filter-builders'] = value;
    }
    addStoredFilter(filter) {
        if (!Util.isDefined(this['stored-filter-builders'])) {
            this['stored-filter-builders'] = [];
        }
        this.storedFilterBuilders.push(filter);
    }
    deleteStoredFilter(filterName) {
        const index = this.storedFilterBuilders.findIndex((item) => item.name === filterName);
        if (index >= 0) {
            this.storedFilterBuilders.splice(index, 1);
        }
    }
    applyFilter(filterName) {
        const filter = this.getStoredFilter(filterName);
        if (filter) {
            this.filterBuilderValues = filter;
        }
    }
    getStoredFilter(filterName) {
        let result = [];
        const filter = this.storedFilterBuilders.find((item) => item.name === filterName);
        if (Util.isDefined(filter)) {
            result = filter['filter-builder-values'];
        }
        return result;
    }
}
