// ontimize-web-ngx/esm2015/lib/services/state/o-list-component-state.service.js
import { Injectable } from '@angular/core';
import { Codes } from '../../util/codes';
import { Util } from '../../util/util';
import { AbstractComponentStateService } from './o-component-state.service';
import { OListComponentStateClass } from './o-list-component-state.class';
export class OListComponentStateService extends AbstractComponentStateService {
    initialize(component) {
        this.state = new OListComponentStateClass();
        super.initialize(component);
    }
    initializeState(state) {
        super.initializeState(state);
    }
    getDataToStore() {
        let dataToStore = Object.assign({}, this.state);
        dataToStore['query-rows'] = this.component.queryRows;
        if (!this.component.storePaginationState) {
            delete dataToStore['queryRecordOffset'];
        }
        if (this.component.quickFilter && Util.isDefined(this.component.quickFilterComponent)) {
            dataToStore['quickFilterActiveColumns'] = this.component.quickFilterComponent.getActiveColumns().join(Codes.ARRAY_INPUT_SEPARATOR);
        }
        dataToStore['filter-case-sensitive'] = this.component.isFilterCaseSensitive();
        dataToStore.selection = this.state.selection;
        return dataToStore;
    }
    refreshSelection() {
        this.state.selection = this.getSelectionState();
    }
    getSelectionState() {
        const selection = [];
        if (this.component) {
            const keys = this.component.getKeys();
            this.component.getSelectedItems().forEach(item => {
                const data = {};
                keys.forEach(key => {
                    data[key] = item[key];
                });
                selection.push(data);
            });
        }
        return selection;
    }
}
OListComponentStateService.decorators = [
    { type: Injectable }
];
