// ontimize-web-ngx/esm2015/lib/components/container/o-container-component.class.js
import { ElementRef, Inject, Injector, Optional, ViewChild } from '@angular/core';
import { MAT_FORM_FIELD_DEFAULT_OPTIONS } from '@angular/material';
import { Util } from '../../util/util';
export const DEFAULT_INPUTS_O_CONTAINER = [
    'oattr: attr',
    'title',
    'layoutAlign: layout-align',
    'elevation',
    'icon',
    'appearance',
    'layoutGap: layout-gap'
];
export class OContainerComponent {
    constructor(elRef, injector, matFormDefaultOption) {
        this.elRef = elRef;
        this.injector = injector;
        this.matFormDefaultOption = matFormDefaultOption;
        this._elevation = 0;
        this.defaultLayoutAlign = 'start start';
        this._outlineGapCalculationNeededImmediately = false;
        this.titleObserver = new MutationObserver(() => this.updateOutlineGap());
    }
    set containerTitle(elem) {
        this._titleEl = elem;
        if (this._titleEl) {
            this.registerObserver();
            this.updateOutlineGap();
        }
        else {
            this.unRegisterObserver();
        }
    }
    ngAfterViewInit() {
        if (this.elRef) {
            this.elRef.nativeElement.removeAttribute('title');
        }
        this.registerObserver();
    }
    ngAfterContentChecked() {
        if (this._outlineGapCalculationNeededImmediately) {
            this.updateOutlineGap();
        }
    }
    ngOnDestroy() {
        this.unRegisterObserver();
    }
    getAttribute() {
        if (this.oattr) {
            return this.oattr;
        }
        else if (this.elRef && this.elRef.nativeElement.attributes.attr) {
            return this.elRef.nativeElement.attributes.attr.value;
        }
    }
    get appearance() {
        return this._appearance;
    }
    set appearance(value) {
        this._appearance = value;
        setTimeout(() => { this.updateOutlineGap(); }, 0);
    }
    get elevation() {
        return this._elevation;
    }
    set elevation(elevation) {
        this._elevation = elevation;
        this.propagateElevationToDOM();
    }
    get layoutAlign() {
        return this._layoutAlign;
    }
    set layoutAlign(align) {
        if (!align || align.length === 0) {
            align = this.defaultLayoutAlign;
        }
        this._layoutAlign = align;
    }
    get layoutGap() {
        return this._layoutGap;
    }
    set layoutGap(layoutGap) {
        this._layoutGap = layoutGap;
    }
    hasHeader() {
        return !!this.title || !!this.icon;
    }
    isAppearanceOutlineSetted() {
        let isAppearanceSetted = false;
        if (Util.isDefined(this.appearance)) {
            isAppearanceSetted = this.appearance === OContainerComponent.APPEARANCE_OUTLINE;
        }
        return isAppearanceSetted;
    }
    hasHeaderOrAppearanceOutlineSetted() {
        return this.isAppearanceOutlineSetted() || this.hasHeader();
    }
    hasHeaderAndAppearanceOutline() {
        return this.isAppearanceOutline() && this.hasHeader();
    }
    isAppearanceOutline() {
        let isAppearanceOutline = (this.matFormDefaultOption && this.matFormDefaultOption.appearance === OContainerComponent.APPEARANCE_OUTLINE);
        if (Util.isDefined(this.appearance)) {
            isAppearanceOutline = this.appearance === OContainerComponent.APPEARANCE_OUTLINE;
        }
        return isAppearanceOutline;
    }
    hasTitleInAppearanceOutline() {
        return this.isAppearanceOutline() && this.hasHeader();
    }
    propagateElevationToDOM() {
        this.cleanElevationCSSclasses();
        if (this.elevation > 0 && this.elevation <= 12) {
            this.elRef.nativeElement.classList.add('mat-elevation-z' + this.elevation);
        }
    }
    cleanElevationCSSclasses() {
        const classList = [].slice.call(this.elRef.nativeElement.classList);
        if (classList && classList.length) {
            classList.forEach((item) => {
                if (item.startsWith('mat-elevation')) {
                    this.elRef.nativeElement.classList.remove(item);
                }
            });
        }
    }
    updateOutlineGap() {
        if (this.isAppearanceOutline()) {
            const titleEl = this._titleEl ? this._titleEl.nativeElement : null;
            if (!this._containerRef) {
                return;
            }
            if (document.documentElement && !document.documentElement.contains(this.elRef.nativeElement)) {
                this._outlineGapCalculationNeededImmediately = true;
                return;
            }
            const container = this._containerRef.nativeElement;
            const containerRect = container.getBoundingClientRect();
            if (containerRect.width === 0 && containerRect.height === 0) {
                return;
            }
            const containerStart = containerRect.left;
            const labelStart = titleEl.getBoundingClientRect().left;
            const labelWidth = this.hasHeader() ? titleEl.offsetWidth : 0;
            const startWidth = labelStart - containerStart;
            const startEls = container.querySelectorAll('.o-container-outline-start');
            const gapEls = container.querySelectorAll('.o-container-outline-gap');
            gapEls[0].style.width = `${labelWidth}px`;
            startEls[0].style.width = `${startWidth}px`;
            this._outlineGapCalculationNeededImmediately = false;
        }
    }
    registerObserver() {
        if (this._titleEl) {
            this.titleObserver.observe(this._titleEl.nativeElement, {
                childList: true,
                characterData: true,
                subtree: true
            });
        }
    }
    unRegisterObserver() {
        if (this.titleObserver) {
            this.titleObserver.disconnect();
        }
    }
}
OContainerComponent.APPEARANCE_OUTLINE = 'outline';
OContainerComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Injector },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_FORM_FIELD_DEFAULT_OPTIONS,] }] }
];
OContainerComponent.propDecorators = {
    containerTitle: [{ type: ViewChild, args: ['containerTitle', { static: false },] }],
    _containerRef: [{ type: ViewChild, args: ['container', { static: true },] }]
};
