// ontimize-web-ngx/esm5/lib/components/input/listpicker/listpicker-renderer/listpicker-renderer.js
import * as tslib_1 from "tslib";
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER_CURRENCY, OListPickerRendererCurrencyComponent } from './currency/o-list-picker-renderer-currency.component';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER_DATE, OListPickerRendererDateComponent } from './date/o-list-picker-renderer-date.component';
import { OListPickerRendererIntegerComponent } from './integer/o-list-picker-renderer-integer.component';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER_PERCENTAGE, OListPickerRendererPercentageComponent } from './percentage/o-list-picker-renderer-percentage.component';
import { OListPickerRendererRealComponent } from './real/o-list-picker-renderer-real.component';
export var O_LISTPICKER_RENDERERS = [
    OListPickerRendererIntegerComponent,
    OListPickerRendererRealComponent,
    OListPickerRendererCurrencyComponent,
    OListPickerRendererDateComponent,
    OListPickerRendererPercentageComponent
];
export var O_LISTPICKER_RENDERERS_INPUTS = tslib_1.__spread(DEFAULT_INPUTS_O_LISTPICKER_RENDERER_CURRENCY, DEFAULT_INPUTS_O_LISTPICKER_RENDERER_DATE, DEFAULT_INPUTS_O_LISTPICKER_RENDERER_PERCENTAGE);
export var O_LISTPICKER_RENDERERS_OUTPUTS = [];
export var renderersMapping = {
    integer: OListPickerRendererIntegerComponent,
    real: OListPickerRendererRealComponent,
    currency: OListPickerRendererCurrencyComponent,
    date: OListPickerRendererDateComponent,
    percentage: OListPickerRendererPercentageComponent
};
