// ontimize-web-ngx/esm5/lib/components/input/listpicker/listpicker-renderer/o-list-picker-renderer.class.js
import { Util } from '../../../../util';
import { OListPickerComponent } from '../o-list-picker.component';
export var DEFAULT_INPUTS_O_LISTPICKER_RENDERER = [];
var OListPickerCustomRenderer = (function () {
    function OListPickerCustomRenderer(injector) {
        this.injector = injector;
        this.listpickerComponent = this.injector.get(OListPickerComponent);
    }
    OListPickerCustomRenderer.prototype.ngOnInit = function () {
        this.initialize();
    };
    OListPickerCustomRenderer.prototype.initialize = function () {
    };
    OListPickerCustomRenderer.prototype.ngAfterContentInit = function () {
        this.registerRenderer();
    };
    OListPickerCustomRenderer.prototype.registerRenderer = function () {
        this.listpickerComponent.registerRenderer(this);
    };
    OListPickerCustomRenderer.prototype.getListPickerValue = function (record) {
        var _this = this;
        var result = '';
        if (!Util.isDefined(record)) {
            return result;
        }
        this.listpickerComponent.descriptionColArray.forEach(function (col, index) {
            if (Util.isDefined(record[col])) {
                var value = record[col];
                if (_this.componentPipe && _this.pipeArguments != null) {
                    value = _this.componentPipe.transform(value, _this.pipeArguments);
                }
                if (Util.isDefined(value)) {
                    result += value;
                    if (index < _this.listpickerComponent.descriptionColArray.length - 1) {
                        result += _this.listpickerComponent.separator;
                    }
                }
            }
        });
        return result;
    };
    return OListPickerCustomRenderer;
}());
export { OListPickerCustomRenderer };
