// ontimize-web-ngx/esm2015/lib/components/table/extensions/dialog/apply-configuration/o-table-apply-configuration-dialog.component.js
import { SelectionModel } from '@angular/cdk/collections';
import { Component, EventEmitter, Inject, Injector, ViewChild } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef, MatSelectionList } from '@angular/material';
import { DialogService } from '../../../../../services/dialog.service';
export class OTableApplyConfigurationDialogComponent {
    constructor(dialogRef, data, injector) {
        this.dialogRef = dialogRef;
        this.injector = injector;
        this.default_configuration = 'OTableApplyConfigurationDialogComponent-default';
        this.configurations = [];
        this.onDelete = new EventEmitter();
        this.loadConfigurations(data);
        this.dialogService = this.injector.get(DialogService);
    }
    ngOnInit() {
        this.configurationList.selectedOptions = new SelectionModel(false);
    }
    loadConfigurations(configurations) {
        this.configurations = configurations;
    }
    removeConfiguration(configurationName) {
        this.dialogService.confirm('CONFIRM', 'TABLE.DIALOG.CONFIRM_REMOVE_CONFIGURATION').then(result => {
            if (result) {
                this.onDelete.emit(configurationName);
            }
        });
    }
    isDefaultConfigurationSelected() {
        const selected = this.configurationList.selectedOptions.selected;
        const selectedValue = selected.length ? selected[0].value : void 0;
        return selectedValue === this.default_configuration;
    }
    getSelectedConfigurationName() {
        const selected = this.configurationList.selectedOptions.selected;
        return selected.length ? selected[0].value : void 0;
    }
}
OTableApplyConfigurationDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-apply-configuration-dialog',
                template: "<span mat-dialog-title>{{ 'TABLE.BUTTONS.APPLY_CONFIGURATION' | oTranslate }}</span>\n\n<mat-dialog-content fxLayout=\"column\">\n  <div mat-subheader>{{ 'TABLE.DIALOG.APPLY_CONFIGURATION' | oTranslate }}</div>\n  <mat-selection-list #configurationList dense class=\"o-table-apply-configuration-dialog-list\">\n    <mat-list-option checkboxPosition=\"before\" [value]=\"default_configuration\">\n      <span matLine class=\"o-table-apply-configuration-dialog-list-title\">{{ 'TABLE.DIALOG.APPLY_CONFIGURATION_DEFAULT' | oTranslate }}</span>\n      <span matLine>{{ 'TABLE.DIALOG.APPLY_CONFIGURATION_DEFAULT_DESCRIPTION' | oTranslate }}</span>\n    </mat-list-option>\n    <mat-list-option checkboxPosition=\"before\" *ngFor=\"let configuration of configurations; let i = index\" [value]=\"configuration.name\">\n      <span matLine>{{ configuration.name }}</span>\n      <span matLine>{{ configuration.description }}</span>\n    </mat-list-option>\n  </mat-selection-list>\n</mat-dialog-content>\n\n<mat-dialog-actions align=\"end\">\n  <button type=\"button\" mat-stroked-button [disabled]=\"configurationList.selectedOptions.selected.length!==1 || isDefaultConfigurationSelected()\"\n    (click)=\"removeConfiguration(configurationList.selectedOptions.selected[0].value)\" class=\"o-button-danger\">\n    {{ 'DELETE' | oTranslate }}\n  </button>\n  <span fxFlex></span>\n  <button type=\"button\" mat-stroked-button [mat-dialog-close]=\"false\" class=\"o-button-default cancel\">{{ 'CANCEL' | oTranslate }}</button>\n  <button type=\"button\" mat-stroked-button [mat-dialog-close]=\"true\" [disabled]=\"configurationList.selectedOptions.selected.length!==1\"\n    class=\"o-button-primary\">{{ 'OK' | oTranslate }}</button>\n</mat-dialog-actions>\n"
            }] }
];
OTableApplyConfigurationDialogComponent.ctorParameters = () => [
    { type: MatDialogRef },
    { type: Array, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] },
    { type: Injector }
];
OTableApplyConfigurationDialogComponent.propDecorators = {
    configurationList: [{ type: ViewChild, args: [MatSelectionList, { static: true },] }]
};
