// ontimize-web-ngx-report/esm2015/lib/components/report/o-report-detail/o-report-detail.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/report/o-report-detail/o-report-detail.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { MatDialog } from '@angular/material';
import { DialogService, OFormComponent, OTextInputComponent } from 'ontimize-web-ngx';
import { Constants } from '../../../util/constants';
import { OReportViewerComponent } from '../o-report-viewer/o-report-viewer.component';
export class OReportDetailComponent {
    /**
     * @param {?} dialogService
     * @param {?} dialog
     */
    constructor(dialogService, dialog) {
        this.dialogService = dialogService;
        this.dialog = dialog;
        this.hasParams = false;
    }
    /**
     * @private
     * @return {?}
     */
    getValues() {
        for (let i = 0; i < this.parameters.length; i++) {
            /** @type {?} */
            let v = this.paramForm.getFieldValue('value' + i);
            this.values.push(v);
        }
    }
    /**
     * @return {?}
     */
    fillReport() {
        this.values = [];
        this.av = [this.id.getValue()];
        if (this.parameters.length > 0) {
            this.getValues();
            /** @type {?} */
            let size = this.values.length;
            for (let i = 0; i < size; i++)
                this.av.push(this.values.shift());
        }
        this.dialog.open(OReportViewerComponent, {
            height: Constants.DEFAULT_HEIGHT_DIALOG,
            width: Constants.DEFAULT_HEIGHT_DIALOG,
            data: {
                'params': this.av,
                'filter': {}
            }
        });
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onDataLoaded(e) {
        this.parameters = e['PARAMETERS'];
        if (this.parameters.length > 0) {
            this.hasParams = true;
        }
    }
}
OReportDetailComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-report-detail',
                template: "<o-form #form attr=\"report\" service-type=\"report\" entity=\"getReport\" keys=\"UUID\" header-actions=\"R;I;U;D\" show-header=\"yes\"\n  show-header-navigation=\"yes\" fxLayout=\"column\" (onDataLoaded)=\"onDataLoaded($event)\">\n  <o-column attr=\"row1\" title=\"REPORT_DATA\" icon=\"description\" layout-gap=\"14px\" fxFill>\n    <o-text-input #id attr=\"UUID\" label=\"REPORT_ID\" enabled=\"no\" fxFlex=\"100\"></o-text-input>\n    <div fxLayout=\"row wrap\" fxLayoutGap=\"14px\" fxLayoutAlign=\"start start\" fxFlex>\n      <o-text-input attr=\"NAME\" label=\"REPORT_NAME\" required=\"yes\" fxFlex=\"50\"></o-text-input>\n      <o-text-input attr=\"REPORT_TYPE\" label=\"REPORT_TYPE\" required=\"yes\" fxFlex></o-text-input>\n    </div>\n    <o-text-input class=\"form-field\" attr=\"MAIN_REPORT_FILENAME\" label=\"REPORT_FILENAME\" required=\"yes\" fxFlex></o-text-input>\n    <o-textarea-input class=\"form-field\" attr=\"DESCRIPTION\" label=\"REPORT_DESCRIPTION\" required=\"yes\" fxFlex></o-textarea-input>\n  </o-column>\n\n  <div *ngIf=\"hasParams\">\n    <o-column #params title=\"PARAMETERS\" icon=\"info\" class=\"vertical-margin-10\" layout-align=\"start stretch\" appearance=\"outline\">\n      <o-form #paramForm attr=\"report\" service-type=\"report\" entity=\"fillReport\" keys=\"id\" header-mode=\"none\" show-header=\"no\"\n        layout-direction=\"column\">\n        <ng-container *ngFor=\"let elem of parameters; let i = index\">\n          <o-row fxFlex>\n            <ng-template [ngIf]=\"elem.valueClass == 'java.sql.Date' || elem.valueClass == 'java.sql.Timestamp'\" [ngIfElse]=\"textInput\">\n              <o-date-input [attr]=\"'value' + i\" [label]=\"elem.name\" required=\"yes\" value-type=\"string\" format=\"YYYY-MM-DD\" fxFlex></o-date-input>\n            </ng-template>\n            <ng-template #textInput>\n              <o-text-input [attr]=\"'value' + i\" [label]=\"elem.description\" required=\"yes\" fxFlex></o-text-input>\n            </ng-template>\n          </o-row>\n        </ng-container>\n      </o-form>\n    </o-column>\n  </div>\n  <o-button type=\"RAISED\" (click)=\"fillReport()\" label=\"{{ 'FILL_REPORT' | oTranslate }}\" icon=\"download\" layout-padding color=\"accent\"\n    [enabled]=\"form.formGroup?.valid && paramForm?.formGroup?.valid\"></o-button>\n</o-form>\n"
            }] }
];
/** @nocollapse */
OReportDetailComponent.ctorParameters = () => [
    { type: DialogService },
    { type: MatDialog }
];
OReportDetailComponent.propDecorators = {
    id: [{ type: ViewChild, args: ['id', { static: true },] }],
    paramForm: [{ type: ViewChild, args: ['paramForm', { static: false },] }]
};
if (false) {
    /** @type {?} */
    OReportDetailComponent.prototype.id;
    /** @type {?} */
    OReportDetailComponent.prototype.paramForm;
    /**
     * @type {?}
     * @private
     */
    OReportDetailComponent.prototype.values;
    /** @type {?} */
    OReportDetailComponent.prototype.parameters;
    /**
     * @type {?}
     * @private
     */
    OReportDetailComponent.prototype.av;
    /** @type {?} */
    OReportDetailComponent.prototype.hasParams;
    /**
     * @type {?}
     * @protected
     */
    OReportDetailComponent.prototype.dialogService;
    /**
     * @type {?}
     * @protected
     */
    OReportDetailComponent.prototype.dialog;
}
