// ontimize-web-ngx/esm2015/lib/components/app-sidenav/image/o-app-sidenav-image.component.js
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Injector, ViewEncapsulation } from '@angular/core';
import { merge, Subscription } from 'rxjs';
import { OAppSidenavComponent } from '../o-app-sidenav.component';
export const DEFAULT_INPUTS_O_APP_SIDENAV_IMAGE = [
    'openedSrc: opened-src',
    'closedSrc: closed-src'
];
export const DEFAULT_OUTPUTS_O_APP_SIDENAV_IMAGE = [];
export class OAppSidenavImageComponent {
    constructor(injector, cd) {
        this.injector = injector;
        this.cd = cd;
        this.subscription = new Subscription();
        this.sidenav = this.injector.get(OAppSidenavComponent);
    }
    ngOnInit() {
        if (this.sidenav) {
            this.subscription.add(merge(this.sidenav.onSidenavClosedStart, this.sidenav.onSidenavOpenedStart).subscribe(() => this.updateImage()));
        }
        this.updateImage();
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    ngOnChanges(changes) {
        if (changes.openedSrc || changes.closedSrc) {
            this.updateImage();
        }
    }
    updateImage() {
        if (this.sidenav && this.sidenav.sidenav && this.sidenav.sidenav.opened) {
            this.setOpenedImg();
        }
        else {
            this.setClosedImg();
        }
        this.cd.detectChanges();
    }
    set src(val) {
        this._src = val;
    }
    get src() {
        return this._src;
    }
    setOpenedImg() {
        this.src = this.openedSrc;
    }
    setClosedImg() {
        this.src = this.closedSrc;
    }
    get showImage() {
        return (this._src !== undefined && this._src.length > 0);
    }
}
OAppSidenavImageComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-app-sidenav-image',
                inputs: DEFAULT_INPUTS_O_APP_SIDENAV_IMAGE,
                outputs: DEFAULT_OUTPUTS_O_APP_SIDENAV_IMAGE,
                template: "<div class=\"o-app-sidenav-image-container\" *ngIf=\"showImage\">\n  <img class=\"o-app-sidenav-image\" [src]=\"src\" alt=\"application sidenav image\" />\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.o-app-sidenav-image]': 'true'
                },
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".sidenav-toggle-container+.o-app-sidenav-image .o-app-sidenav-image-container{padding-top:0;text-align:center}.mat-drawer-opened .o-app-sidenav-image .o-app-sidenav-image-container{text-align:center}.o-app-sidenav-image .o-app-sidenav-image-container{max-height:75px;margin:24px 0}.o-app-sidenav-image .o-app-sidenav-image-container .o-app-sidenav-image{max-width:100%;max-height:100%}"]
            }] }
];
OAppSidenavImageComponent.ctorParameters = () => [
    { type: Injector },
    { type: ChangeDetectorRef }
];
