// ontimize-web-ngx/esm5/lib/components/table/extensions/header/table-button/o-table-button.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, forwardRef, Inject, Injector, ViewEncapsulation, } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { Codes } from '../../../../../util/codes';
import { Util } from '../../../../../util/util';
import { OTableComponent } from '../../../o-table.component';
export var DEFAULT_INPUTS_O_TABLE_BUTTON = [
    'oattr: attr',
    'enabled',
    'icon',
    'svgIcon: svg-icon',
    'iconPosition: icon-position',
    'olabel: label'
];
export var DEFAULT_OUTPUTS_O_TABLE_BUTTON = [
    'onClick'
];
var OTableButtonComponent = (function () {
    function OTableButtonComponent(injector, elRef, _table) {
        this.injector = injector;
        this.elRef = elRef;
        this._table = _table;
        this.onClick = new EventEmitter();
        this.enabled = true;
    }
    OTableButtonComponent.prototype.ngOnInit = function () {
        if (!Util.isDefined(this.icon) && !Util.isDefined(this.svgIcon)) {
            this.icon = 'priority_high';
        }
        this.iconPosition = Util.parseIconPosition(this.iconPosition);
    };
    OTableButtonComponent.prototype.innerOnClick = function (event) {
        event.stopPropagation();
        this.onClick.emit();
    };
    OTableButtonComponent.prototype.isIconPositionLeft = function () {
        return this.iconPosition === Codes.ICON_POSITION_LEFT;
    };
    Object.defineProperty(OTableButtonComponent.prototype, "table", {
        get: function () {
            return this._table;
        },
        enumerable: true,
        configurable: true
    });
    OTableButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table-button',
                    template: "<ng-container *ngIf=\"table.showButtonsText\">\n  <button type=\"button\" [disabled]=\"!enabled\" [class.disabled]=\"!enabled\" mat-stroked-button (click)=\"innerOnClick($event)\">\n    <mat-icon *ngIf=\"isIconPositionLeft() && svgIcon !== undefined\" [svgIcon]=\"svgIcon\"></mat-icon>\n    <mat-icon *ngIf=\"isIconPositionLeft() && svgIcon === undefined\">{{ icon }}</mat-icon>\n    <span>{{ olabel | oTranslate }}</span>\n    <mat-icon *ngIf=\"!isIconPositionLeft() && svgIcon !== undefined\" [svgIcon]=\"svgIcon\"></mat-icon>\n    <mat-icon *ngIf=\"!isIconPositionLeft() && svgIcon === undefined\">{{ icon }}</mat-icon>\n  </button>\n</ng-container>\n\n<ng-container *ngIf=\"!table.showButtonsText\">\n  <button type=\"button\" [disabled]=\"!enabled\" [class.disabled]=\"!enabled\" mat-icon-button (click)=\"innerOnClick($event)\">\n    <mat-icon *ngIf=\"isIconPositionLeft() && svgIcon !== undefined\" [svgIcon]=\"svgIcon\"></mat-icon>\n    <mat-icon *ngIf=\"isIconPositionLeft() && svgIcon === undefined\">{{ icon }}</mat-icon>\n    <mat-icon *ngIf=\"!isIconPositionLeft() && svgIcon !== undefined\" [svgIcon]=\"svgIcon\"></mat-icon>\n    <mat-icon *ngIf=\"!isIconPositionLeft() && svgIcon === undefined\">{{ icon }}</mat-icon>\n  </button>\n</ng-container>\n",
                    inputs: DEFAULT_INPUTS_O_TABLE_BUTTON,
                    outputs: DEFAULT_OUTPUTS_O_TABLE_BUTTON,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.o-table-button]': 'true',
                        '[class.text]': 'table.showButtonsText',
                        '[class.no-text]': '!table.showButtonsText',
                    }
                }] }
    ];
    OTableButtonComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ElementRef },
        { type: OTableComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OTableComponent; }),] }] }
    ]; };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableButtonComponent.prototype, "enabled", void 0);
    return OTableButtonComponent;
}());
export { OTableButtonComponent };
