// ontimize-web-ngx/esm5/lib/layouts/form-layout/dialog/o-form-layout-dialog.component.js
import { Component, ComponentFactoryResolver, Inject, Injector, ViewChild, ViewEncapsulation } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
import { DialogService } from '../../../services/dialog.service';
import { Util } from '../../../util';
import { OFormLayoutManagerContentDirective } from '../directives/o-form-layout-manager-content.directive';
var OFormLayoutDialogComponent = (function () {
    function OFormLayoutDialogComponent(dialogRef, injector, componentFactoryResolver, data) {
        this.dialogRef = dialogRef;
        this.injector = injector;
        this.componentFactoryResolver = componentFactoryResolver;
        this.dialogService = injector.get(DialogService);
        if (data.title) {
            this.title = data.title;
        }
        if (data.data) {
            this.data = data.data;
            var component = data.data.component;
            this.componentFactory = this.componentFactoryResolver.resolveComponentFactory(component);
            this.params = data.data.params;
            this.queryParams = data.data.queryParams;
            this.urlSegments = data.data.urlSegments;
        }
        if (data.layoutManagerComponent) {
            this.formLayoutManager = data.layoutManagerComponent;
        }
    }
    OFormLayoutDialogComponent.prototype.ngAfterViewInit = function () {
        if (this.contentDirective && this.componentFactory) {
            var viewContainerRef = this.contentDirective.viewContainerRef;
            viewContainerRef.clear();
            viewContainerRef.createComponent(this.componentFactory);
        }
    };
    OFormLayoutDialogComponent.prototype.updateNavigation = function (data) {
        var label = this.formLayoutManager.getLabelFromData(data);
        if (label && label.length) {
            label = ': ' + label;
        }
        this.label = label;
    };
    OFormLayoutDialogComponent.prototype.updateActiveData = function (data) {
        this.data = Object.assign(this.data, data);
    };
    OFormLayoutDialogComponent.prototype.closeDialog = function (options) {
        var _this = this;
        if (Util.isDefined(options) && Util.isDefined(options.exitWithoutConfirmation) && options.exitWithoutConfirmation) {
            this.dialogRef.close();
        }
        else {
            if (this.formLayoutManager.hasToConfirmExit(this.data)) {
                this.dialogService.confirm('CONFIRM', 'MESSAGES.FORM_CHANGES_WILL_BE_LOST').then(function (res) {
                    if (res) {
                        _this.dialogRef.close();
                    }
                });
            }
            else {
                this.dialogRef.close();
            }
        }
    };
    OFormLayoutDialogComponent.prototype.getRouteOfActiveItem = function () {
        var parentRoute = this.formLayoutManager.parentFormLayoutManager.getRouteOfActiveItem();
        var segments = (this.urlSegments || []);
        var route = [];
        segments.forEach(function (segment, index) {
            if (parentRoute[index] !== segment.path) {
                route.push(segment.path);
            }
        });
        return route;
    };
    OFormLayoutDialogComponent.prototype.getParams = function () {
        return this.params;
    };
    OFormLayoutDialogComponent.prototype.getFormCacheData = function () {
        return this.data;
    };
    OFormLayoutDialogComponent.prototype.isMainComponent = function (comp) {
        return !comp.oFormLayoutDialog;
    };
    OFormLayoutDialogComponent.prototype.closeDetail = function (options) {
        this.closeDialog(options);
    };
    OFormLayoutDialogComponent.prototype.getDataToStore = function () {
        return null;
    };
    OFormLayoutDialogComponent.prototype.setModifiedState = function (formAttr, modified, confirmExit) {
        this.data.innerFormsInfo[formAttr] = {
            modified: modified,
            confirmOnExit: confirmExit
        };
    };
    OFormLayoutDialogComponent.prototype.canAddDetailComponent = function () {
        return true;
    };
    OFormLayoutDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-form-layout-dialog',
                    template: "<div fxLayout=\"row\" fxLayoutAlign=\"space-between center\" class=\"title-container\" mat-dialog-title>\n  <span fxFlex fxLayoutAlign=\"center center\" > {{ (title || 'LAYOUT_MANANGER.DIALOG_TITLE') | oTranslate }}{{ label }} </span>\n  <mat-icon (click)=\"closeDialog()\" svgIcon=\"ontimize:close\"></mat-icon>\n</div>\n\n<div mat-dialog-content class=\"form-layout-dialog-container\">\n  <ng-template o-form-layout-manager-content></ng-template>\n</div>",
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-form-layout-dialog]': 'true'
                    },
                    styles: [".o-form-layout-dialog-overlay{width:65%;height:90%}.o-form-layout-dialog-overlay .title-container{cursor:default}.o-form-layout-dialog-overlay .title-container .mat-icon{cursor:pointer}.o-form-layout-dialog-overlay .mat-dialog-container{overflow:hidden}.o-form-layout-dialog-overlay .mat-dialog-container .mat-dialog-content.form-layout-dialog-container{display:block;padding-top:16px;margin:0;max-height:calc(100% - 78px);height:100%}.o-form-layout-dialog-overlay .mat-dialog-container .o-form-layout-dialog o-form-toolbar .mat-toolbar{padding:0}.o-form-layout-dialog-overlay .mat-dialog-container .o-form-layout-dialog .o-form form.inner-form{position:relative;margin-top:0}"]
                }] }
    ];
    OFormLayoutDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    OFormLayoutDialogComponent.propDecorators = {
        contentDirective: [{ type: ViewChild, args: [OFormLayoutManagerContentDirective, { static: false },] }]
    };
    return OFormLayoutDialogComponent;
}());
export { OFormLayoutDialogComponent };
