// ontimize-web-ngx/esm2015/lib/services/auth-guard.service.js
import { Injectable, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { OUserInfoService } from '../services/o-user-info.service';
import { Codes } from '../util/codes';
import { AuthService } from './auth.service';
import { PermissionsService } from './permissions/permissions.service';
import * as i0 from "@angular/core";
export class AuthGuardService {
    constructor(injector) {
        this.injector = injector;
        this.router = this.injector.get(Router);
        this.authService = this.injector.get(AuthService);
        this.oUserInfoService = this.injector.get(OUserInfoService);
        this.permissionsService = this.injector.get(PermissionsService);
    }
    canActivate(next, state) {
        const isLoggedIn = this.authService.isLoggedIn();
        let result = isLoggedIn;
        if (!isLoggedIn) {
            this.permissionsService.restart();
            this.router.navigate([Codes.LOGIN_ROUTE]);
        }
        if (isLoggedIn) {
            this.setUserInformation();
            if (!this.permissionsService.hasPermissions()) {
                result = this.permissionsService.getUserPermissionsAsPromise();
            }
        }
        return result;
    }
    setUserInformation() {
        const sessionInfo = this.authService.getSessionInfo();
        this.oUserInfoService.setUserInfo({
            username: sessionInfo.user,
            avatar: './assets/images/user_profile.png'
        });
    }
}
AuthGuardService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
AuthGuardService.ctorParameters = () => [
    { type: Injector }
];
AuthGuardService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AuthGuardService_Factory() { return new AuthGuardService(i0.ɵɵinject(i0.INJECTOR)); }, token: AuthGuardService, providedIn: "root" });
