// ontimize-web-ngx/esm2015/lib/services/factories.js
import { InjectionToken, Injector } from '@angular/core';
import { AppConfig } from '../config/app-config';
import { Util } from '../util/util';
import { AuthService } from './auth.service';
import { OntimizeAuthService } from './o-auth.service';
import { OntimizeExportDataProviderService3X } from './ontimize-export-data-provider-3x.service';
import { OntimizeExportDataProviderService } from './ontimize-export-data-provider.service';
import { OntimizeEEService } from './ontimize/ontimize-ee.service';
import { OntimizeExportService3X } from './ontimize/ontimize-export-3xx.service';
import { OntimizeExportService } from './ontimize/ontimize-export.service';
import { OntimizeFileService } from './ontimize/ontimize-file.service';
import { OntimizeService } from './ontimize/ontimize.service';
import { OntimizeEEPermissionsService } from './permissions/ontimize-ee-permissions.service';
import { OntimizePermissionsService } from './permissions/ontimize-permissions.service';
import { AbstractComponentStateService, DefaultComponentStateService } from './state/o-component-state.service';
export const O_DATA_SERVICE = new InjectionToken('Ontimize data service');
export const O_TRANSLATE_SERVICE = new InjectionToken('Translate service');
export const O_FILE_SERVICE = new InjectionToken('File uploader service');
export const O_EXPORT_SERVICE = new InjectionToken('Export service');
export const O_PERMISSION_SERVICE = new InjectionToken('Permission service');
export const O_AUTH_SERVICE = new InjectionToken('Authentication service');
export const O_COMPONENT_STATE_SERVICE = new InjectionToken('Component state service');
export const O_CHART_ON_DEMAND_SERVICE = new InjectionToken('Chart on demand service');
export const O_REPORT_SERVICE = new InjectionToken('Report service');
export const O_ERROR_DIALOG_MANAGER = new InjectionToken('Error dialog manager');
export const O_EXPORT_DATA_SERVICE = new InjectionToken('Export data provider');
export const O_MAT_ERROR_OPTIONS = new InjectionToken('o-mat-error-options');
export const O_FORM_MESSAGE_SERVICE = new InjectionToken('Ontimize o-form message service');
export function dataServiceFactory(injector) {
    const serviceClass = _getInjectionTokenValue(O_DATA_SERVICE, injector);
    const service = Util.createServiceInstance(serviceClass, injector);
    if (Util.isDefined(service)) {
        return service;
    }
    const config = injector.get(AppConfig).getConfiguration();
    if (!Util.isDefined(config.serviceType) || 'OntimizeEE' === config.serviceType) {
        return new OntimizeEEService(injector);
    }
    else if ('Ontimize' === config.serviceType) {
        return new OntimizeService(injector);
    }
    return Util.createServiceInstance(config.serviceType, injector);
}
export function fileServiceFactory(injector) {
    const serviceClass = _getInjectionTokenValue(O_FILE_SERVICE, injector);
    const service = Util.createServiceInstance(serviceClass, injector);
    return Util.isDefined(service) ? service : new OntimizeFileService(injector);
}
export function exportServiceFactory(injector) {
    const serviceClass = _getInjectionTokenValue(O_EXPORT_SERVICE, injector);
    const service = Util.createServiceInstance(serviceClass, injector);
    if (Util.isDefined(service)) {
        return service;
    }
    const config = injector.get(AppConfig).getConfiguration();
    if (typeof (config.exportServiceType) === 'undefined') {
        if (config.exportConfiguration) {
            return new OntimizeExportService3X(injector);
        }
        else {
            return new OntimizeExportService(injector);
        }
    }
    return Util.createServiceInstance(config.exportServiceType, injector);
}
export function exportDataFactory(injector) {
    const provider = _getInjectionTokenValue(O_EXPORT_DATA_SERVICE, injector);
    const service = Util.createServiceInstance(provider, injector);
    if (Util.isDefined(service)) {
        return service;
    }
    else {
        const config = injector.get(AppConfig).getConfiguration();
        if (typeof (config.exportConfiguration) === 'undefined') {
            return new OntimizeExportDataProviderService(injector);
        }
        else {
            return new OntimizeExportDataProviderService3X(injector);
        }
    }
}
export function permissionsServiceFactory(injector) {
    const serviceClass = _getInjectionTokenValue(O_PERMISSION_SERVICE, injector);
    const service = Util.createServiceInstance(serviceClass, injector);
    if (Util.isDefined(service)) {
        return service;
    }
    const config = injector.get(AppConfig).getConfiguration();
    if (!Util.isDefined(config.permissionsServiceType) || 'OntimizeEEPermissions' === config.permissionsServiceType) {
        return new OntimizeEEPermissionsService(injector);
    }
    else if ('OntimizePermissions' === config.permissionsServiceType) {
        return new OntimizePermissionsService(injector);
    }
    return Util.createServiceInstance(config.permissionsServiceType, injector);
}
export function authServiceFactory(injector) {
    const serviceClass = _getInjectionTokenValue(O_AUTH_SERVICE, injector);
    const service = Util.createServiceInstance(serviceClass, injector);
    return Util.isDefined(service) ? service : new OntimizeAuthService(injector);
}
export function componentStateFactory(injector) {
    const service = _getInjectionTokenValue(O_COMPONENT_STATE_SERVICE, injector);
    return Util.isDefined(service) ? service : new DefaultComponentStateService(injector);
}
export const OntimizeServiceProvider = { provide: OntimizeService, useFactory: dataServiceFactory, deps: [Injector] };
export const OntimizeExportServiceProvider = { provide: OntimizeExportService, useFactory: exportServiceFactory, deps: [Injector] };
export const OntimizeAuthServiceProvider = { provide: AuthService, useFactory: authServiceFactory, deps: [Injector] };
export const ComponentStateServiceProvider = { provide: AbstractComponentStateService, useFactory: componentStateFactory, deps: [Injector] };
export const ExportDataServiceProvider = { provide: OntimizeExportDataProviderService, useFactory: exportDataFactory, deps: [Injector] };
export function _getInjectionTokenValue(token, injector) {
    let service;
    try {
        service = injector.get(token);
    }
    catch (e) {
    }
    return service;
}
