// ontimize-web-ngx/esm2015/lib/services/login-storage.service.js
import { Injectable, Injector } from '@angular/core';
import { AppConfig } from '../config/app-config';
import { Codes } from '../util/codes';
import * as i0 from "@angular/core";
export class LoginStorageService {
    constructor(injector) {
        this.injector = injector;
        this._config = this.injector.get(AppConfig).getConfiguration();
        this._localStorageKey = this._config.uuid;
    }
    getSessionInfo() {
        const info = localStorage.getItem(this._localStorageKey);
        if (!info) {
            return {};
        }
        const stored = JSON.parse(info);
        return stored[Codes.SESSION_KEY] || {};
    }
    storeSessionInfo(sessionInfo) {
        if (sessionInfo !== undefined) {
            const info = localStorage.getItem(this._localStorageKey);
            let stored = null;
            if (info && info.length > 0) {
                stored = JSON.parse(info);
            }
            else {
                stored = {};
            }
            stored[Codes.SESSION_KEY] = sessionInfo;
            try {
                localStorage.setItem(this._localStorageKey, JSON.stringify(stored));
            }
            catch (e) {
                console.error("Cannot set new item in localStorage. Error: " + e);
            }
        }
    }
    updateSessionId(id) {
        let session = this.getSessionInfo();
        if (session.id !== id) {
            session.id = id;
            this.storeSessionInfo(session);
        }
    }
    sessionExpired() {
        const sessionInfo = this.getSessionInfo();
        delete sessionInfo.id;
        delete sessionInfo.user;
        this.storeSessionInfo(sessionInfo);
    }
    isLoggedIn() {
        const sessionInfo = this.getSessionInfo();
        if (sessionInfo && sessionInfo.id && sessionInfo.user && sessionInfo.user.length > 0) {
            if (typeof sessionInfo.id === 'number' && (isNaN(sessionInfo.id) || sessionInfo.id < 0)) {
                return false;
            }
            return true;
        }
        return false;
    }
}
LoginStorageService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
LoginStorageService.ctorParameters = () => [
    { type: Injector }
];
LoginStorageService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LoginStorageService_Factory() { return new LoginStorageService(i0.ɵɵinject(i0.INJECTOR)); }, token: LoginStorageService, providedIn: "root" });
