// ontimize-web-ngx/esm2015/lib/services/o-auth.service.js
import { Injectable, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { combineLatest, from, Observable } from 'rxjs';
import { AppConfig } from '../config/app-config';
import { Codes } from '../util/codes';
import { AuthService } from './auth.service';
import { LoginStorageService } from './login-storage.service';
import { OntimizeService } from './ontimize/ontimize.service';
import { PermissionsService } from './permissions/permissions.service';
import { ORemoteConfigurationService } from './remote-config.service';
export class OntimizeAuthService extends AuthService {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this._config = this.injector.get(AppConfig).getConfiguration();
        this.router = this.injector.get(Router);
        this.loginStorageService = this.injector.get(LoginStorageService);
        const sessionInfo = this.loginStorageService.getSessionInfo();
        if (sessionInfo && sessionInfo.id && sessionInfo.user && sessionInfo.user.length > 0) {
            this._user = sessionInfo.user;
        }
    }
    get user() {
        return this._user;
    }
    get localStorageKey() {
        return this.loginStorageService._localStorageKey;
    }
    configureOntimizeAuthService(config) {
        this.ontService = this.injector.get(OntimizeService);
        const servConf = {};
        servConf[Codes.SESSION_KEY] = this.loginStorageService.getSessionInfo();
        this.ontService.configureService(servConf);
    }
    retrieveAuthService() {
        return new Promise(resolve => {
            if (this.ontService !== undefined) {
                resolve(this.ontService);
            }
            else {
                this.configureOntimizeAuthService(this._config);
                resolve(this.ontService);
            }
        });
    }
    login(user, password) {
        this._user = user;
        const dataObservable = new Observable(observer => {
            this.retrieveAuthService().then(service => {
                service.startsession(user, password).subscribe(resp => {
                    this.onLoginSuccess(resp);
                    const permissionsService = this.injector.get(PermissionsService);
                    const remoteConfigService = this.injector.get(ORemoteConfigurationService);
                    const pendingArray = [];
                    pendingArray.push(permissionsService.getUserPermissionsAsPromise());
                    pendingArray.push(remoteConfigService.initialize());
                    combineLatest(pendingArray).subscribe(() => {
                        observer.next();
                        observer.complete();
                    });
                }, error => {
                    this.onLoginError(error);
                    observer.error(error);
                });
            });
        });
        return from(dataObservable.toPromise());
    }
    onLoginSuccess(sessionId) {
        const session = {
            user: this._user,
            id: sessionId
        };
        this.loginStorageService.storeSessionInfo(session);
        this.onLogin.next(session);
    }
    onLoginError(error) {
        this.dialogService.alert('ERROR', 'MESSAGES.ERROR_LOGIN');
    }
    logout() {
        this.onLogout.next(null);
        const sessionInfo = this.loginStorageService.getSessionInfo();
        const dataObservable = new Observable(innerObserver => {
            this.retrieveAuthService().then(service => {
                service.endsession(sessionInfo.user, sessionInfo.id).subscribe(resp => {
                    const remoteConfigService = this.injector.get(ORemoteConfigurationService);
                    remoteConfigService.finalize().subscribe(() => {
                        this.onLogoutSuccess(resp);
                        innerObserver.next();
                        innerObserver.complete();
                    });
                }, error => {
                    this.onLogoutError(error);
                    innerObserver.error(error);
                });
            });
        });
        return from(dataObservable.toPromise());
    }
    onLogoutSuccess(sessionId) {
        if (sessionId === 0) {
            this.clearSessionData();
            this.redirectLogin(false);
        }
    }
    onLogoutError(error) {
        console.error('Error on logout');
        this.clearSessionData();
        this.redirectLogin(false);
    }
    clearSessionData() {
        this.loginStorageService.sessionExpired();
    }
    isLoggedIn() {
        return this.loginStorageService.isLoggedIn();
    }
    getSessionInfo() {
        return this.loginStorageService.getSessionInfo();
    }
    storeSessionInfo(info) {
        this.loginStorageService.storeSessionInfo(info);
    }
    redirectLogin(sessionExpired = false) {
        const arg = {};
        arg[Codes.SESSION_EXPIRED_KEY] = sessionExpired;
        const extras = {};
        extras[Codes.QUERY_PARAMS] = arg;
        this.router.navigate([Codes.LOGIN_ROUTE], extras);
    }
}
OntimizeAuthService.decorators = [
    { type: Injectable }
];
OntimizeAuthService.ctorParameters = () => [
    { type: Injector }
];
