// ontimize-web-ngx/esm2015/lib/services/o-error-dialog-manager.service.js
import { Injectable } from '@angular/core';
import { Util } from '../util/util';
import { DialogService } from './dialog.service';
import * as i0 from "@angular/core";
import * as i1 from "./dialog.service";
export class OErrorDialogManager {
    constructor(dialogService) {
        this.dialogService = dialogService;
    }
    restart() {
        this.errorDialogSubscription = null;
    }
    openErrorDialog(err) {
        if (!Util.isDefined(this.errorDialogSubscription)) {
            this.errorDialogSubscription = new Promise((resolve) => {
                const errorMsg = (err && typeof err !== 'object') ? err : 'MESSAGES.ERROR_QUERY';
                this.dialogService.alert('ERROR', errorMsg).then(res => {
                    this.restart();
                    resolve(res);
                });
            });
        }
        return this.errorDialogSubscription;
    }
}
OErrorDialogManager.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
OErrorDialogManager.ctorParameters = () => [
    { type: DialogService }
];
OErrorDialogManager.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OErrorDialogManager_Factory() { return new OErrorDialogManager(i0.ɵɵinject(i1.DialogService)); }, token: OErrorDialogManager, providedIn: "root" });
