// ontimize-web-ngx/esm2015/lib/services/o-modules-info.service.js
import { Injectable, Injector } from '@angular/core';
import { NavigationEnd, Router } from '@angular/router';
import { ReplaySubject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class OModulesInfoService {
    constructor(injector, router) {
        this.injector = injector;
        this.router = router;
        this.subject = new ReplaySubject();
        this.router = this.injector.get(Router);
    }
    getModuleChangeObservable() {
        this.router.events
            .pipe(filter(e => e instanceof NavigationEnd), map(() => {
            let route = this.router.routerState.root;
            let routeTitle = '';
            while (route.firstChild) {
                route = route.firstChild;
            }
            if (route.snapshot.data['oAppHeaderTitle']) {
                routeTitle = route.snapshot.data['oAppHeaderTitle'];
            }
            return routeTitle;
        }))
            .subscribe((title) => {
            if (title) {
                this.subject.next(title);
            }
            else {
                this.subject.next("");
            }
        });
        return this.subject.asObservable();
    }
}
OModulesInfoService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
OModulesInfoService.ctorParameters = () => [
    { type: Injector },
    { type: Router }
];
OModulesInfoService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OModulesInfoService_Factory() { return new OModulesInfoService(i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i1.Router)); }, token: OModulesInfoService, providedIn: "root" });
