// ontimize-web-ngx/esm2015/lib/services/ontimize-export-data-base-provider.service.js
import { Injectable, Injector } from "@angular/core";
export class OntimizeExportDataBaseProviderService {
    constructor(injector) {
        this.injector = injector;
    }
    initializeProvider(table) {
        this.table = table;
        const tableOptions = this.table.oTableOptions;
        this.colsNotIncluded = this.table.getColumnsNotIncluded();
        this.columns = tableOptions.visibleColumns.filter(c => this.colsNotIncluded.indexOf(c) === -1);
        const tableColumnNames = tableOptions.visibleColumns.filter(c => this.colsNotIncluded.indexOf(c) === -1);
        this.columnNames = this.table.getColumnNames(tableColumnNames);
        this.sqlTypes = this.table.getSqlTypes();
        this.entity = this.table.entity;
        this.service = this.table.service;
        this.filter = this.table.getComponentFilter();
    }
}
OntimizeExportDataBaseProviderService.decorators = [
    { type: Injectable }
];
OntimizeExportDataBaseProviderService.ctorParameters = () => [
    { type: Injector }
];
