// ontimize-web-ngx/esm5/lib/components/form/guards/o-form-can-deactivate.guard.js
import { Injectable } from '@angular/core';
import { combineLatest } from 'rxjs';
import { BooleanConverter } from '../../../decorators/input-converter';
import { Codes } from '../../../util/codes';
import { Util } from '../../../util/util';
import { OFormConfirmExitService } from '../navigation/o-form-confirm-exit.service';
var CanDeactivateFormGuard = (function () {
    function CanDeactivateFormGuard(oFormConfirmExitService) {
        this.oFormConfirmExitService = oFormConfirmExitService;
        this.oForms = {};
    }
    CanDeactivateFormGuard.prototype.canDeactivate = function (component, curr, state, future) {
        var _this = this;
        if (BooleanConverter(future.root.queryParams[Codes.IGNORE_CAN_DEACTIVATE])) {
            return true;
        }
        if (Object.keys(this.oForms).length) {
            return new Promise(function (resolve) {
                var arr = Object.keys(_this.oForms).map(function (key) { return Util.wrapIntoObservable(_this.oForms[key].canDeactivate()); });
                combineLatest(arr).subscribe(function (res) {
                    resolve(res.every(function (value) { return value; }));
                });
            });
        }
        return true;
    };
    CanDeactivateFormGuard.prototype.addForm = function (form) {
        this.oForms[form.oattr] = form;
    };
    CanDeactivateFormGuard.prototype.removeForm = function (form) {
        delete this.oForms[form.oattr];
    };
    CanDeactivateFormGuard.CLASSNAME = 'CanDeactivateFormGuard';
    CanDeactivateFormGuard.decorators = [
        { type: Injectable }
    ];
    CanDeactivateFormGuard.ctorParameters = function () { return [
        { type: OFormConfirmExitService }
    ]; };
    return CanDeactivateFormGuard;
}());
export { CanDeactivateFormGuard };
