// ontimize-web-ngx-report/esm5/lib/util/o-reports-translate.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/util/o-reports-translate.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe, Injector } from '@angular/core';
import { OTranslateService } from 'ontimize-web-ngx';
import * as CORE_TRANSLATIONS from '../i18n/i18n';
var OReportsTranslatePipe = /** @class */ (function () {
    function OReportsTranslatePipe(injector) {
        this.injector = injector;
        this.translateService = this.injector.get(OTranslateService);
    }
    /**
     * @param {?} text
     * @return {?}
     */
    OReportsTranslatePipe.prototype.transform = /**
     * @param {?} text
     * @return {?}
     */
    function (text) {
        /** @type {?} */
        var textTranslated = undefined;
        /** @type {?} */
        var bundle = CORE_TRANSLATIONS.MAP[this.translateService.getCurrentLang()];
        if (bundle && bundle[text]) {
            textTranslated = bundle[text];
        }
        else {
            textTranslated = this.translateService.get(text);
        }
        return textTranslated;
    };
    OReportsTranslatePipe.decorators = [
        { type: Pipe, args: [{
                    name: 'oReportsTranslate',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    OReportsTranslatePipe.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OReportsTranslatePipe;
}());
export { OReportsTranslatePipe };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    OReportsTranslatePipe.prototype.translateService;
    /**
     * @type {?}
     * @protected
     */
    OReportsTranslatePipe.prototype.injector;
}
