// ontimize-web-ngx/esm2015/lib/components/input/html-input/o-html-input.component.js
import { animate, state, style, transition, trigger } from '@angular/animations';
import { ChangeDetectorRef, Component, ElementRef, Injector, ViewChild } from '@angular/core';
import { Validators } from '@angular/forms';
import { MatTab, MatTabGroup } from '@angular/material';
import { NumberConverter } from '../../../decorators/input-converter';
import { OFormComponent } from '../../form/o-form.component';
import { CKEditorComponent } from '../../material/ckeditor/ck-editor.component';
import { DEFAULT_INPUTS_O_FORM_DATA_COMPONENT, OFormDataComponent } from '../../o-form-data-component.class';
export const DEFAULT_INPUTS_O_HTML_INPUT = [
    'oattr: attr',
    'data',
    'autoBinding: automatic-binding',
    'autoRegistering: automatic-registering',
    'orequired: required',
    'minLength: min-length',
    'maxLength: max-length',
    'readOnly: read-only',
    'sqlType: sql-type'
];
export const DEFAULT_OUTPUTS_O_HTML_INPUT = [
    ...DEFAULT_INPUTS_O_FORM_DATA_COMPONENT,
    'onFocus',
    'onBlur'
];
export class OHTMLInputComponent extends OFormDataComponent {
    constructor(form, elRef, injector) {
        super(form, elRef, injector);
        this._minLength = -1;
        this._maxLength = -1;
        this._subscriptAnimationState = '';
        this.form = form;
        this.elRef = elRef;
        this._changeDetectorRef = this.injector.get(ChangeDetectorRef);
        try {
            this.tabGroupContainer = this.injector.get(MatTabGroup);
            this.tabContainer = this.injector.get(MatTab);
        }
        catch (error) {
        }
    }
    ngOnInit() {
        super.ngOnInit();
        const self = this;
        if (this.form) {
            this.form.beforeCloseDetail.subscribe(() => this.destroyCKEditor());
            this.form.beforeUpdateMode.subscribe(() => this.destroyCKEditor());
        }
        if (this.tabGroupContainer) {
            this.tabGroupContainer.selectedTabChange.subscribe((evt) => {
                self.destroyCKEditor();
                if (self.isInActiveTab()) {
                    self.ckEditor.initCKEditor(self.oattr);
                }
            });
        }
    }
    ngAfterViewInit() {
        super.ngAfterViewInit();
        this._subscriptAnimationState = 'enter';
        this._changeDetectorRef.detectChanges();
    }
    hasError(error) {
        const result = super.hasError(error);
        this._subscriptAnimationState = result ? 'enter' : 'void';
        return result;
    }
    isInActiveTab() {
        let result = !(this.tabGroupContainer && this.tabContainer);
        if (!result) {
            const self = this;
            this.tabGroupContainer._tabs.forEach((tab, index) => {
                if (tab === self.tabContainer) {
                    result = (self.tabGroupContainer.selectedIndex === index);
                }
            });
        }
        return result;
    }
    resolveValidators() {
        const validators = super.resolveValidators();
        if (this.minLength >= 0) {
            validators.push(Validators.minLength(this.minLength));
        }
        if (this.maxLength >= 0) {
            validators.push(Validators.maxLength(this.maxLength));
        }
        return validators;
    }
    clearValue() {
        super.clearValue();
        this.ckEditor.instance.updateElement();
        this.ckEditor.instance.setData('');
    }
    destroyCKEditor() {
        if (this.ckEditor) {
            this.ckEditor.destroyCKEditor();
        }
    }
    getCKEditor() {
        return this.ckEditor.instance;
    }
    set minLength(val) {
        const old = this._minLength;
        this._minLength = NumberConverter(val);
        if (val !== old) {
            this.updateValidators();
        }
    }
    get minLength() {
        return this._minLength;
    }
    set maxLength(val) {
        const old = this._maxLength;
        this._maxLength = NumberConverter(val);
        if (val !== old) {
            this.updateValidators();
        }
    }
    get maxLength() {
        return this._maxLength;
    }
}
OHTMLInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-html-input',
                template: "<div [formGroup]=\"getFormGroup()\" class=\"mat-form-field mat-form-field\" fxFill>\n  <div class=\"mat-form-field-wrapper mat-form-field-wrapper\">\n    <ck-editor #ckEditor [id]=\"getAttribute()\" [formControlName]=\"getAttribute()\" (focus)=\"innerOnFocus($event)\" (blur)=\"innerOnBlur($event)\" \n    (change)=\"onChangeEvent($event)\"  [required]=\"isRequired\" [readonly]=\"isReadOnly\" class=\"mat-form-field-flex mat-form-field-flex\" fxFill></ck-editor>\n    <div class=\"mat-form-field-subscript-wrapper mat-form-field-subscript-wrapper\">\n      <div class=\"ng-trigger ng-trigger-transitionMessages\" [@transitionMessages]=\"_subscriptAnimationState\">\n        <mat-error *oMatError=\"hasError('required')\">\n          {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n        </mat-error>\n        <mat-error *oMatError=\"hasError('minlength')\">\n          {{ 'FORM_VALIDATION.MIN_LENGTH' | oTranslate }}: {{ getErrorValue('minlength', 'requiredLength') }}\n        </mat-error>\n        <mat-error *oMatError=\"hasError('maxlength')\">\n          {{ 'FORM_VALIDATION.MAX_LENGTH' | oTranslate }}: {{ getErrorValue('maxlength', 'requiredLength') }}\n        </mat-error>\n        <mat-error *ngFor=\"let oError of getActiveOErrors()\"> \n          {{ oError.text | oTranslate }}\n        </mat-error>\n      </div>\n    </div>\n  </div>\n</div>",
                inputs: DEFAULT_INPUTS_O_HTML_INPUT,
                outputs: DEFAULT_OUTPUTS_O_HTML_INPUT,
                animations: [
                    trigger('transitionMessages', [
                        state('enter', style({ opacity: 1, transform: 'translateY(0%)' })),
                        transition('void => enter', [
                            style({ opacity: 0, transform: 'translateY(-100%)' }),
                            animate('300ms cubic-bezier(0.55, 0, 0.55, 0.2)'),
                        ]),
                    ])
                ],
                styles: [".mat-form-field-wrapper{position:relative}.mat-form-field-subscript-wrapper{position:absolute;width:100%;overflow:hidden}"]
            }] }
];
OHTMLInputComponent.ctorParameters = () => [
    { type: OFormComponent },
    { type: ElementRef },
    { type: Injector }
];
OHTMLInputComponent.propDecorators = {
    ckEditor: [{ type: ViewChild, args: ['ckEditor', { static: false },] }]
};
