// ontimize-web-ngx/esm2015/lib/components/table/extensions/header/table-insertable-row/o-table-insertable-row.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ComponentFactoryResolver, EventEmitter, forwardRef, Inject, Injector } from '@angular/core';
import { FormControl, Validators } from '@angular/forms';
import { InputConverter } from '../../../../../decorators/input-converter';
import { SnackBarService } from '../../../../../services/snackbar.service';
import { OTranslateService } from '../../../../../services/translate/o-translate.service';
import { ObservableWrapper } from '../../../../../util/async';
import { Util } from '../../../../../util/util';
import { OTableComponent } from '../../../o-table.component';
export const DEFAULT_INPUTS_O_TABLE_INSERTABLE_ROW = [
    'columns',
    'requiredColumns : required-columns',
    'position',
    'showPlaceHolder: show-placeholder',
    'includeParentKeys: include-parent-keys'
];
export const DEFAULT_OUTPUTS_O_TABLE_INSERTABLE_ROW = [
    'onPostInsertRecord'
];
export class OTableInsertableRowComponent {
    constructor(injector, table, resolver) {
        this.injector = injector;
        this.table = table;
        this.resolver = resolver;
        this.columnsArray = [];
        this.requiredColumnsArray = [];
        this.onPostInsertRecord = new EventEmitter();
        this.columnEditors = {};
        this.position = OTableInsertableRowComponent.DEFAULT_ROW_POSITION;
        this.showPlaceHolder = false;
        this.includeParentKeys = true;
        this.enabled = true;
        this.rowData = {};
        this.controls = {};
        this.translateService = this.injector.get(OTranslateService);
        this.snackBarService = this.injector.get(SnackBarService);
    }
    ngOnInit() {
        this.columnsArray = Util.parseArray(this.columns, true);
        if (this.columnsArray.length === 0) {
            this.columnsArray = this.table.oTableOptions.visibleColumns;
        }
        this.requiredColumnsArray = Util.parseArray(this.requiredColumns, true);
        if (OTableInsertableRowComponent.AVAILABLE_ROW_POSITIONS.indexOf((this.position || '').toLowerCase()) === -1) {
            this.position = OTableInsertableRowComponent.DEFAULT_ROW_POSITION;
        }
        this.table.setOTableInsertableRow(this);
    }
    isFirstRow() {
        return this.position === 'first';
    }
    isColumnInsertable(column) {
        return (this.columnsArray.indexOf(column.attr) !== -1);
    }
    isColumnRequired(column) {
        return (this.requiredColumnsArray.indexOf(column.attr) !== -1);
    }
    initializeEditors() {
        this.table.oTableOptions.columns
            .filter((col) => this.isColumnInsertable(col)).forEach(col => {
            const columnEditorType = col.editor ? col.editor.type : col.type;
            if (col.definition) {
                const editor = col.definition.buildCellEditor(columnEditorType, this.resolver, col.definition.container, col.definition, false);
                this.columnEditors[col.attr] = editor;
                let disabledCol = !this.enabled;
                if (!disabledCol) {
                    const columnPermissions = this.table.getOColumnPermissions(col.attr);
                    disabledCol = columnPermissions.enabled === false;
                }
                editor.enabled = !disabledCol;
                editor.showPlaceHolder = this.showPlaceHolder || editor.showPlaceHolder;
                editor.table = this.table;
                editor.tableColumn = col.editor ? col.editor.tableColumn : col.definition;
                editor.orequired = this.isColumnRequired(col);
                editor.formControl = this.getControl(col, disabledCol);
                editor.formGroup.addControl(editor.cellEditorId, editor.formControl);
                editor.controlArgs = { silent: true };
                editor.rowData = this.rowData;
                editor.startEdition(this.rowData);
                editor.formControl.markAsUntouched();
            }
        });
    }
    useCellEditor(column) {
        return this.isColumnInsertable(column) && Util.isDefined(this.columnEditors[column.attr]);
    }
    getControl(column, disabled = false) {
        if (!this.controls[column.attr]) {
            const validators = this.resolveValidators(column);
            const cfg = {
                value: undefined,
                disabled: disabled
            };
            this.controls[column.attr] = new FormControl(cfg, validators);
        }
        return this.controls[column.attr];
    }
    resolveValidators(column) {
        const validators = [];
        if (column.definition && column.definition.angularValidatorsFn) {
            column.definition.angularValidatorsFn.forEach((fn) => {
                validators.push(fn);
            });
        }
        if (this.isColumnRequired(column)) {
            validators.push(Validators.required);
        }
        return validators;
    }
    getPlaceholder(column) {
        let showPlaceHolder = this.showPlaceHolder;
        const cellEditor = this.columnEditors[column.attr];
        if (cellEditor) {
            showPlaceHolder = cellEditor.showPlaceHolder;
        }
        else if (column.definition) {
            showPlaceHolder = showPlaceHolder || column.definition.showPlaceHolder;
        }
        return showPlaceHolder ? this.translateService.get(column.title) : undefined;
    }
    handleKeyboardEvent(event) {
        if (event.keyCode !== 13) {
            return;
        }
        this.trWrapper = event.currentTarget;
        event.preventDefault();
        event.stopPropagation();
        this.insertRecord();
    }
    insertRecord() {
        if (!this.validateFields()) {
            return;
        }
        const values = this.getAttributesValuesToInsert();
        const insertObservable = this.table.insertRecord(values);
        if (insertObservable) {
            insertObservable.subscribe(res => {
                this.onInsertSuccess(res);
            }, error => {
                this.table.showDialogError(error, 'MESSAGES.ERROR_INSERT');
            });
        }
    }
    validateFields() {
        let valid = true;
        Object.keys(this.controls).forEach((controlKey) => {
            const control = this.controls[controlKey];
            control.markAsTouched();
            valid = valid && control.valid;
        });
        return valid;
    }
    getAttributesValuesToInsert() {
        let attrValues = {};
        if (this.includeParentKeys) {
            attrValues = this.table.getParentKeysValues();
        }
        Object.keys(this.controls).forEach((controlKey) => {
            attrValues[controlKey] = this.controls[controlKey].value;
        });
        return attrValues;
    }
    onInsertSuccess(res) {
        ObservableWrapper.callEmit(this.onPostInsertRecord, res);
        this.snackBarService.open('MESSAGES.INSERTED', { icon: 'check_circle' });
        this.cleanFields();
        if (this.table.daoTable.usingStaticData) {
            this.table.setDataArray(res);
        }
        else {
            this.table.reloadData();
        }
    }
    cleanFields() {
        const controlKeys = Object.keys(this.controls);
        controlKeys.forEach((controlKey) => {
            this.controls[controlKey].setValue(void 0);
        });
        const firstInputEl = this.trWrapper.querySelector('input');
        if (firstInputEl) {
            setTimeout(() => {
                firstInputEl.focus();
            });
        }
    }
    columnHasError(column, error) {
        const control = this.controls[column.attr];
        return control && control.touched && control.hasError(error);
    }
}
OTableInsertableRowComponent.AVAILABLE_ROW_POSITIONS = ['first', 'last'];
OTableInsertableRowComponent.DEFAULT_ROW_POSITION = 'last';
OTableInsertableRowComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-insertable-row',
                template: ' ',
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_TABLE_INSERTABLE_ROW,
                outputs: DEFAULT_OUTPUTS_O_TABLE_INSERTABLE_ROW
            }] }
];
OTableInsertableRowComponent.ctorParameters = () => [
    { type: Injector },
    { type: OTableComponent, decorators: [{ type: Inject, args: [forwardRef(() => OTableComponent),] }] },
    { type: ComponentFactoryResolver }
];
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTableInsertableRowComponent.prototype, "showPlaceHolder", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTableInsertableRowComponent.prototype, "includeParentKeys", void 0);
