// ontimize-web-ngx/esm5/lib/components/input/listpicker/o-list-picker.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, forwardRef, Inject, Injector, Optional, ViewChild } from '@angular/core';
import { FormControl } from '@angular/forms';
import { MatDialog, MatInput } from '@angular/material';
import { InputConverter } from '../../../decorators/input-converter';
import { OntimizeServiceProvider } from '../../../services/factories';
import { Util } from '../../../util/util';
import { OFormComponent } from '../../form/o-form.component';
import { OValueChangeEvent } from '../../o-value-change-event.class';
import { DEFAULT_INPUTS_O_FORM_SERVICE_COMPONENT, DEFAULT_OUTPUTS_O_FORM_SERVICE_COMPONENT, OFormServiceComponent } from '../o-form-service-component.class';
import { OListPickerDialogComponent } from './o-list-picker-dialog.component';
export var DEFAULT_INPUTS_O_LIST_PICKER = tslib_1.__spread(DEFAULT_INPUTS_O_FORM_SERVICE_COMPONENT, [
    'filter',
    'dialogWidth : dialog-width',
    'dialogHeight : dialog-height',
    'queryRows: query-rows',
    'textInputEnabled: text-input-enabled',
    'dialogDisableClose: dialog-disable-close',
    'dialogClass: dialog-class'
]);
export var DEFAULT_OUTPUTS_O_LIST_PICKER = tslib_1.__spread(DEFAULT_OUTPUTS_O_FORM_SERVICE_COMPONENT, [
    'onDialogAccept',
    'onDialogCancel'
]);
var OListPickerComponent = (function (_super) {
    tslib_1.__extends(OListPickerComponent, _super);
    function OListPickerComponent(form, elRef, injector) {
        var _this = _super.call(this, form, elRef, injector) || this;
        _this.onDialogAccept = new EventEmitter();
        _this.onDialogCancel = new EventEmitter();
        _this.textInputEnabled = true;
        _this.dialogDisableClose = false;
        _this.filter = true;
        _this.dialogHeight = '55%';
        _this.blurDelay = 200;
        _this.blurPrevent = false;
        _this.matDialog = _this.injector.get(MatDialog);
        _this.stateCtrl = new FormControl();
        _this.clearButton = true;
        return _this;
    }
    OListPickerComponent.prototype.ngOnInit = function () {
        this.initialize();
        if (!Util.isDefined(this.stateCtrl.value)) {
            this.setStateCtrlValue();
        }
    };
    OListPickerComponent.prototype.ngOnChanges = function (changes) {
        _super.prototype.ngOnChanges.call(this, changes);
        if (typeof (changes.staticData) !== 'undefined') {
            this.cacheQueried = true;
            this.setDataArray(changes.staticData.currentValue);
        }
    };
    OListPickerComponent.prototype.createFormControl = function (cfg, validators) {
        this._fControl = _super.prototype.createFormControl.call(this, cfg, validators);
        this._fControl.fControlChildren = [this.stateCtrl];
        return this._fControl;
    };
    OListPickerComponent.prototype.ensureOFormValue = function (value) {
        _super.prototype.ensureOFormValue.call(this, value);
        this.syncDataIndex(false);
    };
    OListPickerComponent.prototype.setEnabled = function (value) {
        _super.prototype.setEnabled.call(this, value);
        if (this.stateCtrl && this.hasEnabledPermission() || this.hasVisiblePermission()) {
            value ? this.stateCtrl.enable() : this.stateCtrl.disable();
        }
    };
    OListPickerComponent.prototype.ngAfterViewInit = function () {
        _super.prototype.ngAfterViewInit.call(this);
        if (this.queryOnInit) {
            this.queryData();
        }
        else if (this.queryOnBind) {
            this.syncDataIndex();
        }
    };
    OListPickerComponent.prototype.getDescriptionValue = function () {
        if (!Util.isDefined(this.descriptionColArray) || !Util.isDefined(this._currentIndex)) {
            return '';
        }
        if (Util.isDefined(this.renderer)) {
            return this.renderer.getListPickerValue(this.dataArray[this._currentIndex]);
        }
        else {
            return this.getOptionDescriptionValue(this.dataArray[this._currentIndex]);
        }
    };
    OListPickerComponent.prototype.onClickClear = function (e) {
        e.preventDefault();
        e.stopPropagation();
        if (!this.isReadOnly && this.enabled) {
            clearTimeout(this.blurTimer);
            this.blurPrevent = true;
            this.setValue(undefined);
        }
    };
    OListPickerComponent.prototype.onClickInput = function (e) {
        if (!this.textInputEnabled) {
            this.onClickListpicker(e);
        }
    };
    OListPickerComponent.prototype.onClickListpicker = function (e) {
        e.preventDefault();
        e.stopPropagation();
        if (!this.isReadOnly && this.enabled) {
            clearTimeout(this.blurTimer);
            this.openDialog();
        }
    };
    OListPickerComponent.prototype.onDialogClose = function (evt) {
        this.dialogRef = null;
        this.visibleInputValue = undefined;
        if (evt instanceof Object && typeof evt[this.valueColumn] !== 'undefined') {
            var self_1 = this;
            window.setTimeout(function () {
                self_1.setValue(evt[self_1.valueColumn], { changeType: OValueChangeEvent.USER_CHANGE });
                if (self_1._fControl) {
                    self_1._fControl.markAsTouched();
                    self_1._fControl.markAsDirty();
                }
                self_1.onDialogAccept.emit();
            }, 0);
        }
        else {
            this.onDialogCancel.emit();
        }
    };
    OListPickerComponent.prototype.innerOnBlur = function (evt) {
        if (!this.isReadOnly && this.enabled) {
            var self_2 = this;
            this.blurTimer = setTimeout(function () {
                if (!self_2.blurPrevent) {
                    self_2._fControl.markAsTouched();
                    self_2.onBlur.emit(evt);
                    if (self_2.visibleInputValue !== undefined && self_2.visibleInputValue.length > 0) {
                        self_2.openDialog();
                    }
                    else if (self_2.visibleInputValue !== undefined) {
                        self_2.setValue(undefined);
                        self_2.visibleInputValue = undefined;
                    }
                    else {
                        self_2._fControl.markAsTouched();
                    }
                }
                self_2.blurPrevent = false;
            }, this.blurDelay);
        }
    };
    OListPickerComponent.prototype.onVisibleInputChange = function (event) {
        this.visibleInputValue = event.target.value;
    };
    OListPickerComponent.prototype.onKeydownEnter = function (val) {
        clearTimeout(this.blurTimer);
        this.blurPrevent = true;
        this.visibleInputValue = val;
        this.openDialog();
    };
    OListPickerComponent.prototype.setFormValue = function (val, options, setDirty) {
        if (setDirty === void 0) { setDirty = false; }
        _super.prototype.setFormValue.call(this, val, options, setDirty);
        this.setStateCtrlValue();
    };
    OListPickerComponent.prototype.openDialog = function () {
        var _this = this;
        var cfg = {
            role: 'dialog',
            disableClose: this.dialogDisableClose,
            panelClass: ['cdk-overlay-list-picker', 'o-dialog-class', this.dialogClass],
            data: {
                data: this.getDialogDataArray(this.dataArray),
                filter: this.filter,
                searchVal: this.visibleInputValue,
                menuColumns: this.visibleColumns,
                visibleColumns: this.visibleColArray,
                queryRows: this.queryRows,
                renderer: this.renderer
            }
        };
        if (this.dialogWidth !== undefined) {
            cfg.width = this.dialogWidth;
        }
        if (this.dialogHeight !== undefined) {
            cfg.height = this.dialogHeight;
        }
        this.dialogRef = this.matDialog.open(OListPickerDialogComponent, cfg);
        this.dialogRef.afterClosed().subscribe(function (result) { return _this.onDialogClose(result); });
    };
    OListPickerComponent.prototype.getDialogDataArray = function (dataArray) {
        var _this = this;
        var result = [];
        dataArray.forEach(function (item, itemIndex) {
            var newItem = Object.assign({}, item);
            if (!_this.renderer) {
                newItem._parsedVisibleColumnText = _this.getOptionDescriptionValue(item);
            }
            newItem._parsedIndex = itemIndex;
            result.push(newItem);
        });
        return result;
    };
    OListPickerComponent.prototype.registerRenderer = function (renderer) {
        this.renderer = renderer;
        this.renderer.initialize();
    };
    OListPickerComponent.prototype.setStateCtrlValue = function () {
        var descriptionValue = this.getDescriptionValue();
        if (typeof descriptionValue === 'string' && descriptionValue.length === 0) {
            descriptionValue = null;
        }
        this.stateCtrl.setValue(descriptionValue);
    };
    OListPickerComponent.prototype.selectValue = function () {
        var inputEl = document.getElementById('desc_' + this.oattr);
        if (inputEl) {
            inputEl.select();
        }
    };
    OListPickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-list-picker',
                    template: "<div (click)=\"onClickInput($event)\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\" [formGroup]=\"getFormGroup()\"\n  [matTooltip]=\"tooltip\" [matTooltipClass]=\"tooltipClass\" [matTooltipPosition]=\"tooltipPosition\"\n  [matTooltipShowDelay]=\"tooltipShowDelay\" [matTooltipHideDelay]=\"tooltipHideDelay\"\n  class=\"custom-error o-list-picker-content\" [oContextMenu]=\"oContextMenu\">\n  <input #inputModel class=\"input-model\" type=\"text\" [id]=\"getAttribute()\" [formControlName]=\"getAttribute()\">\n\n  <mat-form-field [appearance]=\"appearance\" [floatLabel]=\"floatLabel\" [class.custom-width]=\"hasCustomWidth\"\n    [hideRequiredMarker]=\"hideRequiredMarker\" class=\"icon-field\" #innerInputContainer fxFlexFill>\n    <mat-label *ngIf=\"labelVisible\">{{ olabel | oTranslate }}</mat-label>\n\n    <input #visibleInput matInput type=\"text\" [attr.id]=\"'desc_' + getAttribute()\" [id]=\"'desc_' + getAttribute()\"\n      [formControl]=\"stateCtrl\" [placeholder]=\"placeHolder\" [value]=\"getDescriptionValue()\"\n      (input)=\"onVisibleInputChange($event)\" (focus)=\"innerOnFocus($event)\" (blur)=\"innerOnBlur($event)\"\n      [readonly]=\"isReadOnly || !textInputEnabled\" [required]=\"isRequired\"\n      (keydown.enter)=\"onKeydownEnter(visibleInput.value)\" (change)=\"onChangeEvent($event)\">\n\n    <button type=\"button\" [disabled]=\"isReadOnly || !enabled\" [class.read-only]=\"isReadOnly\" matSuffix mat-icon-button\n      (click)=\"onClickListpicker($event)\">\n      <mat-icon svgIcon=\"ontimize:search\"></mat-icon>\n    </button>\n\n    <button type=\"button\" [disabled]=\"isReadOnly || !enabled\" [class.read-only]=\"isReadOnly\" matSuffix mat-icon-button\n      (click)=\"onClickClearValue($event)\" *ngIf=\"showClearButton\">\n      <mat-icon svgIcon=\"ontimize:clear\"></mat-icon>\n    </button>\n\n    <mat-error *oMatError=\"hasError('required')\">\n      {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n    </mat-error>\n    <mat-error *ngFor=\"let oError of getActiveOErrors()\">\n      {{ oError.text | oTranslate }}\n    </mat-error>\n  </mat-form-field>\n\n  <o-context-menu *ngIf=\"!this.isReadOnly && this.enabled\">\n    <o-context-menu-item attr=\"refresh\" label=\"FORM_SERVICE_COMPONENT.REFRESH\" svg-icon=\"ontimize:autorenew\" (execute)=\"refresh()\">\n    </o-context-menu-item>\n  </o-context-menu>\n</div>",
                    providers: [
                        OntimizeServiceProvider,
                        { provide: OFormServiceComponent, useExisting: forwardRef(function () { return OListPickerComponent; }) }
                    ],
                    inputs: DEFAULT_INPUTS_O_LIST_PICKER,
                    outputs: DEFAULT_OUTPUTS_O_LIST_PICKER,
                    styles: [".o-list-picker-content .input-model{display:none!important}.o-list-picker-content .mat-button.mat-disabled{cursor:default}"]
                }] }
    ];
    OListPickerComponent.ctorParameters = function () { return [
        { type: OFormComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OFormComponent; }),] }] },
        { type: ElementRef },
        { type: Injector }
    ]; };
    OListPickerComponent.propDecorators = {
        inputModel: [{ type: ViewChild, args: ['inputModel', { static: false },] }],
        visibleInput: [{ type: ViewChild, args: ['visibleInput', { static: false },] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OListPickerComponent.prototype, "textInputEnabled", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OListPickerComponent.prototype, "dialogDisableClose", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OListPickerComponent.prototype, "filter", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Number)
    ], OListPickerComponent.prototype, "queryRows", void 0);
    return OListPickerComponent;
}(OFormServiceComponent));
export { OListPickerComponent };
