// ontimize-web-ngx/esm2015/lib/pipes/o-translate.pipe.js
import { ChangeDetectorRef, Injector, NgModule, Pipe } from '@angular/core';
import { OTranslateService } from '../services/translate/o-translate.service';
import { Util } from '../util/util';
export class OTranslatePipe {
    constructor(injector) {
        this.injector = injector;
        this.value = '';
        this._ref = this.injector.get(ChangeDetectorRef);
        this.oTranslateService = this.injector.get(OTranslateService);
    }
    ngOnDestroy() {
        this._dispose();
    }
    transform(text, args) {
        if (!text || text.length === 0) {
            return text;
        }
        if (Util.equals(text, this.lastKey) && Util.equals(args, this.lastParams)) {
            return this.value;
        }
        this.lastKey = text;
        this.lastParams = args;
        this.updateValue(text);
        this._dispose();
        if (!this.onLanguageChanged) {
            this.onLanguageChanged = this.oTranslateService.onLanguageChanged.subscribe(lang => {
                if (this.lastKey) {
                    this.lastKey = null;
                    this.updateValue(text);
                }
            });
        }
        return this.value;
    }
    updateValue(key) {
        const args = Util.isDefined(this.lastParams) ? this.lastParams.values || [] : [];
        const res = this.oTranslateService.get(key, args);
        this.value = res !== undefined ? res : key;
        this.lastKey = key;
        this._ref.markForCheck();
    }
    _dispose() {
        if (typeof this.onLanguageChanged !== 'undefined') {
            this.onLanguageChanged.unsubscribe();
            this.onLanguageChanged = undefined;
        }
    }
}
OTranslatePipe.decorators = [
    { type: Pipe, args: [{
                name: 'oTranslate',
                pure: false
            },] }
];
OTranslatePipe.ctorParameters = () => [
    { type: Injector }
];
export class OTranslateModule {
    static forRoot() {
        return {
            ngModule: OTranslateModule,
            providers: []
        };
    }
}
OTranslateModule.decorators = [
    { type: NgModule, args: [{
                declarations: [OTranslatePipe],
                imports: [],
                exports: [OTranslatePipe]
            },] }
];
