// ontimize-web-ngx/esm5/lib/components/form/services/o-form-message.service.js
import { Injectable } from '@angular/core';
var OFormMessageService = (function () {
    function OFormMessageService() {
    }
    OFormMessageService.prototype.getQueryErrorMessage = function () {
        return 'MESSAGES.ERROR_QUERY';
    };
    OFormMessageService.prototype.getUpdateSuccessMessage = function () {
        return 'MESSAGES.SAVED';
    };
    OFormMessageService.prototype.getUpdateErrorMessage = function () {
        return 'MESSAGES.ERROR_UPDATE';
    };
    OFormMessageService.prototype.getDeleteSuccessMessage = function () {
        return 'MESSAGES.DELETED';
    };
    OFormMessageService.prototype.getDeleteErrorMessage = function () {
        return 'MESSAGES.ERROR_DELETE';
    };
    OFormMessageService.prototype.getDeleteConfirmationMessage = function () {
        return 'MESSAGES.CONFIRM_DELETE';
    };
    OFormMessageService.prototype.getDeleteConfirmationDialogTitle = function () {
        return 'CONFIRM';
    };
    OFormMessageService.prototype.getInsertSuccessMessage = function () {
        return 'MESSAGES.INSERTED';
    };
    OFormMessageService.prototype.getInsertErrorMessage = function () {
        return 'MESSAGES.ERROR_INSERT';
    };
    OFormMessageService.prototype.getValidationError = function () {
        return 'MESSAGES.FORM_VALIDATION_ERROR';
    };
    OFormMessageService.prototype.getValidationErrorDialogTitle = function () {
        return 'ERROR';
    };
    OFormMessageService.prototype.getNothingToUpdateMessage = function () {
        return 'MESSAGES.FORM_NOTHING_TO_UPDATE_INFO';
    };
    OFormMessageService.prototype.getActionPermissionNotEnabledMessage = function () {
        return 'MESSAGES.OPERATION_NOT_ALLOWED_PERMISSION';
    };
    OFormMessageService.prototype.getDiscardChangesConfirmationMessage = function () {
        return 'MESSAGES.FORM_CHANGES_WILL_BE_LOST';
    };
    OFormMessageService.prototype.getDiscardChangesConfirmationDialogTitle = function () {
        return 'CONFIRM';
    };
    OFormMessageService.decorators = [
        { type: Injectable }
    ];
    return OFormMessageService;
}());
export { OFormMessageService };
