// ontimize-web-ngx/esm2015/lib/components/input/combo/combo-renderer/integer/o-combo-renderer-integer.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { OIntegerPipe } from '../../../../../pipes/o-integer.pipe';
import { DEFAULT_INPUTS_O_COMBO_RENDERER, OComboCustomRenderer } from '../o-combo-renderer.class';
export const DEFAULT_INPUTS_O_COMBO_RENDERER_INTEGER = [
    ...DEFAULT_INPUTS_O_COMBO_RENDERER,
    'grouping',
    'thousandSeparator: thousand-separator'
];
export class OComboRendererIntegerComponent extends OComboCustomRenderer {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.grouping = true;
        this.thousandSeparator = ',';
        this.setComponentPipe();
    }
    setComponentPipe() {
        this.componentPipe = new OIntegerPipe(this.injector);
    }
    initialize() {
        super.initialize();
        this.pipeArguments = {
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator
        };
    }
}
OComboRendererIntegerComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-combo-renderer-integer',
                template: "<ng-template #templateref let-value=\"value\">\n  {{getComboData(value)}}\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_COMBO_RENDERER_INTEGER
            }] }
];
OComboRendererIntegerComponent.ctorParameters = () => [
    { type: Injector }
];
OComboRendererIntegerComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
};
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OComboRendererIntegerComponent.prototype, "grouping", void 0);
