// ontimize-web-ngx/esm5/lib/services/login-storage.service.js
import { Injectable, Injector } from '@angular/core';
import { AppConfig } from '../config/app-config';
import { Codes } from '../util/codes';
import * as i0 from "@angular/core";
var LoginStorageService = (function () {
    function LoginStorageService(injector) {
        this.injector = injector;
        this._config = this.injector.get(AppConfig).getConfiguration();
        this._localStorageKey = this._config.uuid;
    }
    LoginStorageService.prototype.getSessionInfo = function () {
        var info = localStorage.getItem(this._localStorageKey);
        if (!info) {
            return {};
        }
        var stored = JSON.parse(info);
        return stored[Codes.SESSION_KEY] || {};
    };
    LoginStorageService.prototype.storeSessionInfo = function (sessionInfo) {
        if (sessionInfo !== undefined) {
            var info = localStorage.getItem(this._localStorageKey);
            var stored = null;
            if (info && info.length > 0) {
                stored = JSON.parse(info);
            }
            else {
                stored = {};
            }
            stored[Codes.SESSION_KEY] = sessionInfo;
            try {
                localStorage.setItem(this._localStorageKey, JSON.stringify(stored));
            }
            catch (e) {
                console.error("Cannot set new item in localStorage. Error: " + e);
            }
        }
    };
    LoginStorageService.prototype.updateSessionId = function (id) {
        var session = this.getSessionInfo();
        if (session.id !== id) {
            session.id = id;
            this.storeSessionInfo(session);
        }
    };
    LoginStorageService.prototype.sessionExpired = function () {
        var sessionInfo = this.getSessionInfo();
        delete sessionInfo.id;
        delete sessionInfo.user;
        this.storeSessionInfo(sessionInfo);
    };
    LoginStorageService.prototype.isLoggedIn = function () {
        var sessionInfo = this.getSessionInfo();
        if (sessionInfo && sessionInfo.id && sessionInfo.user && sessionInfo.user.length > 0) {
            if (typeof sessionInfo.id === 'number' && (isNaN(sessionInfo.id) || sessionInfo.id < 0)) {
                return false;
            }
            return true;
        }
        return false;
    };
    LoginStorageService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    LoginStorageService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    LoginStorageService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LoginStorageService_Factory() { return new LoginStorageService(i0.ɵɵinject(i0.INJECTOR)); }, token: LoginStorageService, providedIn: "root" });
    return LoginStorageService;
}());
export { LoginStorageService };
