// ontimize-web-ngx/esm5/lib/services/o-error-dialog-manager.service.js
import { Injectable } from '@angular/core';
import { Util } from '../util/util';
import { DialogService } from './dialog.service';
import * as i0 from "@angular/core";
import * as i1 from "./dialog.service";
var OErrorDialogManager = (function () {
    function OErrorDialogManager(dialogService) {
        this.dialogService = dialogService;
    }
    OErrorDialogManager.prototype.restart = function () {
        this.errorDialogSubscription = null;
    };
    OErrorDialogManager.prototype.openErrorDialog = function (err) {
        var _this = this;
        if (!Util.isDefined(this.errorDialogSubscription)) {
            this.errorDialogSubscription = new Promise(function (resolve) {
                var errorMsg = (err && typeof err !== 'object') ? err : 'MESSAGES.ERROR_QUERY';
                _this.dialogService.alert('ERROR', errorMsg).then(function (res) {
                    _this.restart();
                    resolve(res);
                });
            });
        }
        return this.errorDialogSubscription;
    };
    OErrorDialogManager.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    OErrorDialogManager.ctorParameters = function () { return [
        { type: DialogService }
    ]; };
    OErrorDialogManager.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OErrorDialogManager_Factory() { return new OErrorDialogManager(i0.ɵɵinject(i1.DialogService)); }, token: OErrorDialogManager, providedIn: "root" });
    return OErrorDialogManager;
}());
export { OErrorDialogManager };
