// ontimize-web-ngx/esm5/lib/services/o-modules-info.service.js
import { Injectable, Injector } from '@angular/core';
import { NavigationEnd, Router } from '@angular/router';
import { ReplaySubject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
var OModulesInfoService = (function () {
    function OModulesInfoService(injector, router) {
        this.injector = injector;
        this.router = router;
        this.subject = new ReplaySubject();
        this.router = this.injector.get(Router);
    }
    OModulesInfoService.prototype.getModuleChangeObservable = function () {
        var _this = this;
        this.router.events
            .pipe(filter(function (e) { return e instanceof NavigationEnd; }), map(function () {
            var route = _this.router.routerState.root;
            var routeTitle = '';
            while (route.firstChild) {
                route = route.firstChild;
            }
            if (route.snapshot.data['oAppHeaderTitle']) {
                routeTitle = route.snapshot.data['oAppHeaderTitle'];
            }
            return routeTitle;
        }))
            .subscribe(function (title) {
            if (title) {
                _this.subject.next(title);
            }
            else {
                _this.subject.next("");
            }
        });
        return this.subject.asObservable();
    };
    OModulesInfoService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    OModulesInfoService.ctorParameters = function () { return [
        { type: Injector },
        { type: Router }
    ]; };
    OModulesInfoService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OModulesInfoService_Factory() { return new OModulesInfoService(i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i1.Router)); }, token: OModulesInfoService, providedIn: "root" });
    return OModulesInfoService;
}());
export { OModulesInfoService };
