// ontimize-web-ngx/esm5/lib/services/ontimize-export-data-provider.service.js
import * as tslib_1 from "tslib";
import { Injectable, Injector } from "@angular/core";
import { OntimizeExportDataBaseProviderService } from "./ontimize-export-data-base-provider.service";
var OntimizeExportDataProviderService = (function (_super) {
    tslib_1.__extends(OntimizeExportDataProviderService, _super);
    function OntimizeExportDataProviderService(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        return _this;
    }
    OntimizeExportDataProviderService.prototype.getExportConfiguration = function () {
        var data = [];
        var exportData = {
            data: data,
            columns: this.columns,
            columnNames: this.columnNames,
            sqlTypes: this.sqlTypes,
            filter: this.filter
        };
        return exportData;
    };
    OntimizeExportDataProviderService.decorators = [
        { type: Injectable }
    ];
    OntimizeExportDataProviderService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OntimizeExportDataProviderService;
}(OntimizeExportDataBaseProviderService));
export { OntimizeExportDataProviderService };
