// ontimize-web-ngx/esm5/lib/services/ontimize-export-data-provider-3x.service.js
import * as tslib_1 from "tslib";
import { Injectable, Injector } from "@angular/core";
import { Util } from "../util/util";
import { OntimizeExportDataBaseProviderService } from "./ontimize-export-data-base-provider.service";
var OntimizeExportDataProviderService3X = (function (_super) {
    tslib_1.__extends(OntimizeExportDataProviderService3X, _super);
    function OntimizeExportDataProviderService3X(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        return _this;
    }
    OntimizeExportDataProviderService3X.prototype.getExportConfiguration = function (param) {
        var currentPage = 0;
        if (this.table.pageable && Util.isDefined(this.table.currentPage)) {
            currentPage = this.table.currentPage;
        }
        var exportData = {
            type: param.format,
            queryParam: {
                columns: this.columns,
                sqltypes: this.sqlTypes,
                offset: this.table.pageable ? currentPage * this.table.queryRows : -1,
                pageSize: this.table.queryRows,
                filter: this.filter,
                orderBy: this.table.sortColArray
            },
            advQuery: (this.table.pageable ? true : false),
            path: param.path,
            dao: this.entity
        };
        if (param.format === 'xlsx' || 'pdf') {
            exportData.columns = this.parseColumns(this.columns);
            exportData.columnTitles = this.columnNames;
            exportData.styles = {};
            exportData.rowStyles = {};
            exportData.columnStyles = {};
            exportData.columnTypes = {};
            exportData.cellStyles = {};
        }
        return exportData;
    };
    OntimizeExportDataProviderService3X.prototype.parseColumns = function (columns) {
        var obj = {};
        columns.forEach(function (column) {
            obj[column] = {};
        });
        return obj;
    };
    OntimizeExportDataProviderService3X.decorators = [
        { type: Injectable }
    ];
    OntimizeExportDataProviderService3X.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OntimizeExportDataProviderService3X;
}(OntimizeExportDataBaseProviderService));
export { OntimizeExportDataProviderService3X };
