// ontimize-web-ngx/esm5/lib/layouts/form-layout/split-pane/options/o-form-layout-split-pane-options.directive.js
import { Directive, Input } from '@angular/core';
import { OFormLayoutManagerComponent } from '../../o-form-layout-manager.component';
var OFormLayoutSplitPaneOptionsDirective = (function () {
    function OFormLayoutSplitPaneOptionsDirective(formLayoutManager) {
        this.formLayoutManager = formLayoutManager;
    }
    OFormLayoutSplitPaneOptionsDirective.prototype.ngOnChanges = function () {
        if (this.formLayoutManager) {
            this.formLayoutManager.addSplitPaneOptions(this.getOptions());
        }
    };
    OFormLayoutSplitPaneOptionsDirective.prototype.getOptions = function () {
        var result = {
            mainWidth: this.mainWidth,
            mainMaxWidth: this.mainMaxWidth,
            mainMinWidth: this.mainMinWidth,
            detailWidth: this.detailWidth,
            detailMaxWidth: this.detailMaxWidth,
            detailMinWidth: this.detailMinWidth
        };
        Object.keys(result).forEach(function (key) { return result[key] == null ? delete result[key] : {}; });
        return result;
    };
    OFormLayoutSplitPaneOptionsDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'o-form-layout-split-pane-options, o-form-layout-manager[mode="split-pane"]'
                },] }
    ];
    OFormLayoutSplitPaneOptionsDirective.ctorParameters = function () { return [
        { type: OFormLayoutManagerComponent }
    ]; };
    OFormLayoutSplitPaneOptionsDirective.propDecorators = {
        mainWidth: [{ type: Input, args: ['main-width',] }],
        mainMaxWidth: [{ type: Input, args: ['main-max-width',] }],
        mainMinWidth: [{ type: Input, args: ['main-min-width',] }],
        detailWidth: [{ type: Input, args: ['detail-width',] }],
        detailMaxWidth: [{ type: Input, args: ['detail-max-width',] }],
        detailMinWidth: [{ type: Input, args: ['detail-min-width',] }]
    };
    return OFormLayoutSplitPaneOptionsDirective;
}());
export { OFormLayoutSplitPaneOptionsDirective };
