// ontimize-web-ngx/esm2015/lib/components/input/checkbox/o-checkbox.component.js
import { Component, ElementRef, forwardRef, Inject, Injector, Optional, ViewEncapsulation } from '@angular/core';
import { OFormComponent } from '../../form/o-form.component';
import { DEFAULT_INPUTS_O_FORM_DATA_COMPONENT, DEFAULT_OUTPUTS_O_FORM_DATA_COMPONENT } from '../../o-form-data-component.class';
import { DEFAULT_INPUTS_O_BOOLEAN_FORM_DATA, OBooleanFormDataComponent } from '../o-boolean-form-data-component.class';
export const DEFAULT_INPUTS_O_CHECKBOX = [
    'color',
    'labelPosition: label-position',
    ...DEFAULT_INPUTS_O_BOOLEAN_FORM_DATA,
    ...DEFAULT_INPUTS_O_FORM_DATA_COMPONENT
];
export const DEFAULT_OUTPUTS_O_CHECKBOX = [
    ...DEFAULT_OUTPUTS_O_FORM_DATA_COMPONENT
];
export class OCheckboxComponent extends OBooleanFormDataComponent {
    constructor(form, elRef, injector) {
        super(form, elRef, injector);
        this.labelPosition = 'after';
    }
    initialize() {
        super.initialize();
        const checkboxCtx = this;
        this.getFormControl().getValue = function () {
            return this.value ? checkboxCtx.trueValue : checkboxCtx.falseValue;
        };
    }
}
OCheckboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-checkbox',
                inputs: DEFAULT_INPUTS_O_CHECKBOX,
                outputs: DEFAULT_OUTPUTS_O_CHECKBOX,
                template: "<div [class.custom-width]=\"hasCustomWidth\" [formGroup]=\"getFormGroup()\" class=\"relative\" [matTooltip]=\"tooltip\"\n  [matTooltipClass]=\"tooltipClass\" [matTooltipPosition]=\"tooltipPosition\" [matTooltipShowDelay]=\"tooltipShowDelay\"\n  [matTooltipHideDelay]=\"tooltipHideDelay\" fxLayout=\"row\" fxLayoutAlign=\"start center\" fxFill>\n  <mat-checkbox [id]=\"getAttribute()\" [formControlName]=\"getAttribute()\" (change)=\"onChangeEvent($event)\"\n    [labelPosition]=\"labelPosition\" [color]=\"color\">\n    {{ olabel | oTranslate }}\n    <mat-error *oMatError=\"hasError('required')\">\n      {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n    </mat-error>\n    <mat-error *ngFor=\"let oError of getActiveOErrors()\">\n      {{ oError.text | oTranslate }}\n    </mat-error>\n  </mat-checkbox>\n  <div *ngIf=\"isReadOnly\" (click)=\"onClickBlocker($event)\" class=\"read-only-blocker\" fxFill></div>\n</div>",
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.o-checkbox]': 'true'
                },
                styles: [".o-checkbox mat-checkbox{z-index:1}.o-checkbox .mat-checkbox-disabled .mat-checkbox-layout .mat-checkbox-label{color:rgba(0,0,0,.36)}.o-checkbox .custom-width{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.o-checkbox .read-only-blocker{z-index:2;position:absolute;top:0;left:0;right:0}"]
            }] }
];
OCheckboxComponent.ctorParameters = () => [
    { type: OFormComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(() => OFormComponent),] }] },
    { type: ElementRef },
    { type: Injector }
];
