// ontimize-web-ngx/esm5/lib/components/table/extensions/row/table-row-expandable/o-table-row-expandable.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ContentChild, EventEmitter, Output, TemplateRef, ViewEncapsulation } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
export var DEFAULT_OUTPUTS_O_TABLE_ROW_EXPANDABLE = [
    'onExpanded',
    'onCollapsed'
];
export var DEFAULT_INPUTS_O_TABLE_ROW_EXPANDABLE = [
    'iconExpand:icon-expand',
    'iconCollapse:icon-collapse',
    'expandableColumnVisible:expandable-column-visible',
    'multiple'
];
var OTableRowExpandedChange = (function () {
    function OTableRowExpandedChange() {
    }
    return OTableRowExpandedChange;
}());
export { OTableRowExpandedChange };
var OTableRowExpandableComponent = (function () {
    function OTableRowExpandableComponent() {
        this.onExpanded = new EventEmitter();
        this.onCollapsed = new EventEmitter();
        this._iconCollapse = 'remove';
        this._iconExpand = 'add';
        this.expandableColumnVisible = true;
        this.multiple = false;
    }
    Object.defineProperty(OTableRowExpandableComponent.prototype, "iconCollapse", {
        get: function () {
            return this._iconCollapse;
        },
        set: function (value) {
            this._iconCollapse = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableRowExpandableComponent.prototype, "iconExpand", {
        get: function () {
            return this._iconExpand;
        },
        set: function (value) {
            this._iconExpand = value;
        },
        enumerable: true,
        configurable: true
    });
    OTableRowExpandableComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table-row-expandable',
                    template: ' ',
                    outputs: DEFAULT_OUTPUTS_O_TABLE_ROW_EXPANDABLE,
                    inputs: DEFAULT_INPUTS_O_TABLE_ROW_EXPANDABLE,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    OTableRowExpandableComponent.propDecorators = {
        templateRef: [{ type: ContentChild, args: [TemplateRef, { static: false },] }],
        onExpanded: [{ type: Output }],
        onCollapsed: [{ type: Output }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableRowExpandableComponent.prototype, "expandableColumnVisible", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTableRowExpandableComponent.prototype, "multiple", void 0);
    return OTableRowExpandableComponent;
}());
export { OTableRowExpandableComponent };
