// ontimize-web-ngx/esm2015/lib/components/input/percent-input/o-percent-input.component.js
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation } from '@angular/core';
import { InputConverter } from '../../../decorators/input-converter';
import { OPercentPipe } from '../../../pipes/o-percentage.pipe';
import { Util } from '../../../util/util';
import { DEFAULT_INPUTS_O_REAL_INPUT, DEFAULT_OUTPUTS_O_REAL_INPUT, ORealInputComponent } from '../real-input/o-real-input.component';
export const DEFAULT_INPUTS_O_PERCENT_INPUT = [
    ...DEFAULT_INPUTS_O_REAL_INPUT,
    'valueBase: value-base'
];
export const DEFAULT_OUTPUTS_O_PERCENT_INPUT = [
    ...DEFAULT_OUTPUTS_O_REAL_INPUT
];
export class OPercentInputComponent extends ORealInputComponent {
    constructor() {
        super(...arguments);
        this.grouping = true;
        this.valueBase = 1;
    }
    ngOnInit() {
        if (!Util.isDefined(this.min)) {
            this.min = 0;
        }
        if (!Util.isDefined(this.max)) {
            this.max = 100;
        }
        super.ngOnInit();
        this.pipeArguments.valueBase = this.valueBase;
    }
    setComponentPipe() {
        this.componentPipe = new OPercentPipe(this.injector);
    }
}
OPercentInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-percent-input',
                template: "<div [formGroup]=\"getFormGroup()\" [matTooltip]=\"tooltip\" [matTooltipClass]=\"tooltipClass\"\n  [matTooltipPosition]=\"tooltipPosition\" [matTooltipShowDelay]=\"tooltipShowDelay\"\n  [matTooltipHideDelay]=\"tooltipHideDelay\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n  <mat-form-field [appearance]=\"appearance\" [floatLabel]=\"floatLabel\" [hideRequiredMarker]=\"hideRequiredMarker\"\n    [class.custom-width]=\"hasCustomWidth\" class=\"icon-field\" fxFlexFill>\n    <mat-label *ngIf=\"labelVisible\">{{ olabel | oTranslate }}</mat-label>\n    <input matInput [type]=\"inputType\" [id]=\"getAttribute()\" [formControlName]=\"getAttribute()\"\n      [placeholder]=\"placeHolder\" (focus)=\"innerOnFocus($event)\" (blur)=\"innerOnBlur($event)\" [readonly]=\"isReadOnly\"\n      (change)=\"onChangeEvent($event)\" [min]=\"min\" [max]=\"max\" [step]=\"step\" [required]=\"isRequired\">\n    <button type=\"button\" *ngIf=\"showClearButton\" matSuffix mat-icon-button (click)=\"onClickClearValue($event)\">\n      <mat-icon svgIcon=\"ontimize:close\"></mat-icon>\n    </button>\n    <mat-icon svgIcon=\"ontimize:PERCENT\" matSuffix class=\"svg-icon\" [class.mat-disabled]=\"!enabled\"></mat-icon>\n    <mat-error *oMatError=\"hasError('required')\">\n      {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n    </mat-error>\n    <mat-error *oMatError=\"hasError('min')\" >\n      {{ 'FORM_VALIDATION.MIN_VALUE' | oTranslate }}: {{ getErrorValue('min', 'requiredMin') }}\n    </mat-error>\n    <mat-error *oMatError=\"hasError('max')\">\n      {{ 'FORM_VALIDATION.MAX_VALUE' | oTranslate }}: {{ getErrorValue('max', 'requiredMax') }}\n    </mat-error>\n    <mat-error *oMatError=\"hasError('minDecimaldigits')\">\n      {{ 'FORM_VALIDATION.MIN_DECIMAL_DIGITS' | oTranslate }}: {{ getErrorValue('minDecimaldigits', 'requiredMinDecimaldigits') }}\n    </mat-error>\n    <mat-error *oMatError=\"hasError('maxDecimaldigits')\">\n      {{ 'FORM_VALIDATION.MAX_DECIMAL_DIGITS' | oTranslate }}: {{ getErrorValue('maxDecimaldigits', 'requiredMaxDecimaldigits') }}\n    </mat-error>\n    <mat-error *ngFor=\"let oError of getActiveOErrors()\">\n      {{ oError.text | oTranslate }}\n    </mat-error>\n  </mat-form-field>\n</div>",
                inputs: DEFAULT_INPUTS_O_PERCENT_INPUT,
                outputs: DEFAULT_OUTPUTS_O_PERCENT_INPUT,
                encapsulation: ViewEncapsulation.None
            }] }
];
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OPercentInputComponent.prototype, "grouping", void 0);
