// ontimize-web-ngx/esm2015/lib/components/input/listpicker/listpicker-renderer/o-list-picker-renderer.class.js
import { Util } from '../../../../util';
import { OListPickerComponent } from '../o-list-picker.component';
export const DEFAULT_INPUTS_O_LISTPICKER_RENDERER = [];
export class OListPickerCustomRenderer {
    constructor(injector) {
        this.injector = injector;
        this.listpickerComponent = this.injector.get(OListPickerComponent);
    }
    ngOnInit() {
        this.initialize();
    }
    initialize() {
    }
    ngAfterContentInit() {
        this.registerRenderer();
    }
    registerRenderer() {
        this.listpickerComponent.registerRenderer(this);
    }
    getListPickerValue(record) {
        let result = '';
        if (!Util.isDefined(record)) {
            return result;
        }
        this.listpickerComponent.descriptionColArray.forEach((col, index) => {
            if (Util.isDefined(record[col])) {
                let value = record[col];
                if (this.componentPipe && this.pipeArguments != null) {
                    value = this.componentPipe.transform(value, this.pipeArguments);
                }
                if (Util.isDefined(value)) {
                    result += value;
                    if (index < this.listpickerComponent.descriptionColArray.length - 1) {
                        result += this.listpickerComponent.separator;
                    }
                }
            }
        });
        return result;
    }
}
