// ontimize-web-ngx/esm5/lib/components/contextmenu/context-menu-separator/o-context-menu-separator.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { OComponentMenuBaseItem } from '../o-content-menu-base-item.class';
export var DEFAULT_CONTEXT_MENU_ITEM_INPUTS = [
    'attr',
    'ovisible: visible'
];
var OContextMenuSeparatorComponent = (function (_super) {
    tslib_1.__extends(OContextMenuSeparatorComponent, _super);
    function OContextMenuSeparatorComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = OComponentMenuBaseItem.TYPE_SEPARATOR_MENU;
        return _this;
    }
    OContextMenuSeparatorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-context-menu-separator',
                    template: ' ',
                    inputs: DEFAULT_CONTEXT_MENU_ITEM_INPUTS,
                    providers: [{ provide: OComponentMenuBaseItem, useExisting: forwardRef(function () { return OContextMenuSeparatorComponent; }) }]
                }] }
    ];
    return OContextMenuSeparatorComponent;
}(OComponentMenuBaseItem));
export { OContextMenuSeparatorComponent };
