// ontimize-web-ngx/esm2015/lib/components/dual-list-selector/o-dual-list-selector.component.js
import { moveItemInArray, transferArrayItem } from '@angular/cdk/drag-drop';
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { Codes } from '../../util';
export const DEFAULT_DUAL_LIST_SELECTOR = [
    'key',
    'display',
    'dataSource:data-source',
    'dataDestination: data-destination',
    'titleListDataSource: title-list-data-source',
    'titleListDataDestination: title-list-data-destination',
    'groupedColumnTypes:grouped-column-types',
    'description'
];
export class ODualListSelectorComponent {
    constructor() {
        this.dataSource = [];
        this.dataDestination = [];
        this.titleListDataSource = '';
        this.titleListDataDestination = '';
        this.description = '';
        this.key = '';
        this.display = '';
        this.groupedColumnTypes = [];
    }
    drop(event) {
        if (event.previousContainer === event.container) {
            moveItemInArray(event.container.data, event.previousIndex, event.currentIndex);
        }
        else {
            transferArrayItem(event.previousContainer.data, event.container.data, event.previousIndex, event.currentIndex);
        }
    }
    addToGroupedColumns(columnsToGrouped) {
        const self = this;
        columnsToGrouped.selectedOptions.selected.forEach(x => {
            self.dataDestination.push(x.value);
            self.dataSource = this.dataSource.filter(groupedColumn => groupedColumn[this.key] != x.value[this.key]);
        });
    }
    removeGroupedColumns(groupedColumns) {
        const self = this;
        groupedColumns.selectedOptions.selected.forEach(x => {
            self.dataSource.push(x.value);
            self.dataDestination = this.dataDestination.filter(groupedColumn => groupedColumn[this.key] != x.value[this.key]);
        });
    }
    onRemoveGroupColumn(column) {
        this.dataDestination = this.dataDestination.filter((oCol) => oCol[this.key] !== column[this.key]);
        this.dataSource.push(column);
    }
    isAllSelected(list) {
        const numSelected = list.selectedOptions.selected.length;
        const numRows = list.options ? list.options.length : 0;
        return numRows !== 0 && numSelected === numRows;
    }
    masterToggle(list) {
        list.selectedOptions.selected.length === list.options.length ?
            list.selectedOptions.clear() :
            list.selectAll();
    }
    hasValue(list) {
        return list.selectedOptions.selected.length > 0;
    }
    isDisabledSortAsc(list) {
        let sortAsc = true;
        if (list.selectedOptions.selected.length !== 0) {
            const index = list.options.toArray().findIndex(x => x.value[this.key] === list.selectedOptions.selected[0].value[this.key]);
            if (index > 0) {
                sortAsc = false;
            }
        }
        return sortAsc;
    }
    isDisabledSortDesc(list) {
        let sortDesc = true;
        if (list.selectedOptions.selected.length !== 0) {
            const index = list.options.toArray().findIndex(x => x.value[this.key] === list.selectedOptions.selected[0].value[this.key]);
            if (index >= 0 && (index + 1) !== list.options.toArray().length) {
                sortDesc = false;
            }
        }
        return sortDesc;
    }
    sortAscSelectedItems(list) {
        const self = this;
        const itemsSelected = list.selectedOptions.selected;
        itemsSelected.forEach(item => {
            const index = list.options.toArray().findIndex(x => x.value[this.key] === item.value[this.key]);
            moveItemInArray(this.dataDestination, index, index - 1);
        });
    }
    sortDescSelectedItems(list) {
        const self = this;
        const itemsSelected = list.selectedOptions.selected;
        itemsSelected.forEach(item => {
            const index = list.options.toArray().findIndex(x => x.value[this.key] === item.value[this.key]);
            moveItemInArray(self.dataDestination, index, index + 1);
        });
    }
    getCheckboxLabel(list) {
        return '(' + list.selectedOptions.selected.length + '/' + list.options.length + ')';
    }
    getSelectedItems() {
        return this.dataDestination;
    }
    setSelectedItems(items) {
        this.dataDestination = items;
    }
    getGroupedColumnTypes() {
        return this.groupedColumnTypes;
    }
    get groupedDateColumns() {
        return this.groupedColumnTypes.filter(groupedColumn => Codes.TYPES_DATE_GROUPS.findIndex(type => type === groupedColumn.type) > -1);
    }
}
ODualListSelectorComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-dual-list-selector',
                template: "<div class=\"mat-subheader\">{{description | oTranslate }} </div>\n\n<div class=\"o-dual-list-selector-wrapper-selector-list\">\n\n  <div class=\"o-dual-list-selector-list\" fxLayout=\"column\" fxLayoutAlign=\"start stretch\" fxFlex=\"50%\">\n    <div fxLayout=\"row\" fxFlex class=\"o-dual-list-selector-source-wrapper\">\n      <mat-selection-list fxFlex dense cdkDropList id=\"item-source-list\" #itemsList [cdkDropListData]=\"dataSource\"\n        [cdkDropListConnectedTo]=\"['item-destination-list']\" (cdkDropListDropped)=\"drop($event)\" cdkDropListSortingDisabled>\n\n        <mat-list-item>\n          <mat-checkbox (change)=\"$event ? masterToggle(itemsList) : null\" [checked]=\"isAllSelected(itemsList)\"\n            [indeterminate]=\"hasValue(itemsList)  && !isAllSelected(itemsList)\" fxFlex>\n            <span fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n              <span>{{ titleListDataSource | oTranslate }}</span>\n              <span>{{ getCheckboxLabel(itemsList) }} </span>\n            </span>\n          </mat-checkbox>\n        </mat-list-item>\n\n        <mat-list-option checkboxPosition=\"before\" *ngFor=\"let item of dataSource\" [value]=\"item\" cdkDrag class=\"o-drag-list-item-box\">\n          <span matLine>{{ (item[display]) | oTranslate }} </span>\n        </mat-list-option>\n      </mat-selection-list>\n\n      <div fxLayout=\"column\" fxLayoutAlign=\"center center\" fxFlex=\"50px\" fxLayoutGap=\"24px\">\n        <button mat-mini-fab (click)=\"addToGroupedColumns(itemsList)\" [disabled]=\"itemsList.selectedOptions.selected.length===0\">\n          <mat-icon>chevron_right</mat-icon>\n        </button>\n        <button mat-mini-fab (click)=\"removeGroupedColumns(selectedItemsList)\" [disabled]=\"selectedItemsList.selectedOptions.selected.length===0\">\n          <mat-icon>chevron_left</mat-icon>\n        </button>\n      </div>\n    </div>\n  </div>\n\n  <div class=\"o-dual-list-selector-list\" fxLayout=\"column\" \u00BAfxLayoutAlign=\"start stretch\" fxFlex=\"50%\">\n    <div fxLayout=\"row\" fxFlex class=\"o-dual-list-selector-destination-wrapper\">\n      <mat-selection-list fxFlex dense cdkDropList id=\"item-destination-list\" #selectedItemsList [cdkDropListData]=\"dataDestination\"\n        [cdkDropListConnectedTo]=\"['item-source-list']\" (cdkDropListDropped)=\"drop($event)\">\n        <mat-list-item>\n          <mat-checkbox (change)=\"$event ? masterToggle(selectedItemsList) : null\" [checked]=\"isAllSelected(selectedItemsList)\"\n            [indeterminate]=\"hasValue(selectedItemsList) && !isAllSelected(selectedItemsList)\" fxFlex>\n            <span fxLayoutAlign=\"space-between center\">\n              <span>{{ titleListDataDestination | oTranslate}}</span>\n              <span>{{ getCheckboxLabel(selectedItemsList) }}</span>\n            </span>\n          </mat-checkbox>\n        </mat-list-item>\n\n        <mat-list-option checkboxPosition=\"before\" *ngFor=\"let itemSelected of dataDestination\" [value]=\"itemSelected\" cdkDrag\n          class=\"o-drag-list-item-box\">\n\n          <ng-container *ngIf=\"itemSelected.type === 'date'; else defaultRenderer\">\n            <o-dual-list-selector-date-item [item]=itemSelected.attr [grouped-date-columns]=\"groupedColumnTypes\"></o-dual-list-selector-date-item>\n          </ng-container>\n\n          <ng-template #defaultRenderer>\n            <span matLine>{{ itemSelected[display] | oTranslate }} </span>\n          </ng-template>\n\n        </mat-list-option>\n\n        <div class=\"mat-caption drop-empty-message\" *ngIf=\"dataDestination.length===0\">\n          {{ 'DUAL_LIST_SELECTOR.DROP_EMPTY_MESSAGE' | oTranslate }}\n        </div>\n      </mat-selection-list>\n\n\n      <div fxLayout=\"column\" fxLayoutAlign=\"center center\" fxFlex=\"50px\" fxLayoutGap=\"24px\">\n        <button mat-mini-fab [disabled]=\"isDisabledSortAsc(selectedItemsList)\" (click)=\"sortAscSelectedItems(selectedItemsList)\">\n          <mat-icon>expand_less</mat-icon>\n        </button>\n        <button mat-mini-fab [disabled]=\"isDisabledSortDesc(selectedItemsList)\" (click)=\"sortDescSelectedItems(selectedItemsList)\">\n          <mat-icon>expand_more</mat-icon>\n        </button>\n      </div>\n    </div>\n\n  </div>\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_DUAL_LIST_SELECTOR,
                host: {
                    '[class.o-dual-list-selector]': 'true'
                },
                styles: [".o-dual-list-selector{width:100%;height:100%;display:flex;flex-direction:column;flex-wrap:nowrap}.o-dual-list-selector .o-dual-list-selector-wrapper-selector-list{flex:1}.o-dual-list-selector .o-dual-list-selector-wrapper-selector-list .o-dual-list-selector-destination-wrapper,.o-dual-list-selector .o-dual-list-selector-wrapper-selector-list .o-dual-list-selector-source-wrapper{overflow:auto}.o-dual-list-selector .o-dual-list-selector-list .mat-mini-fab{line-height:14px;height:18px;width:18px}.o-dual-list-selector .o-dual-list-selector-list .mat-mini-fab .mat-button-wrapper{padding:0;line-height:14px}.o-dual-list-selector .o-dual-list-selector-list .mat-mini-fab .mat-button-wrapper .mat-icon{font-size:18px;height:18px;width:18px}.o-dual-list-selector .o-dual-list-selector-list .mat-selection-list.mat-list-base{overflow-y:scroll;width:100%;position:relative;border-radius:4px;padding:16px}.o-dual-list-selector .o-dual-list-selector-list .mat-selection-list.mat-list-base .mat-list-item.mat-list-item:not(.mat-list-option){margin:0}.o-dual-list-selector .o-dual-list-selector-list .mat-selection-list.mat-list-base .mat-list-item.mat-list-item:not(.mat-list-option) .mat-checkbox-label{font-weight:500}.o-dual-list-selector .o-dual-list-selector-list .mat-selection-list.mat-list-base .mat-list-item.mat-list-item:not(.mat-list-option) .mat-checkbox-label,.o-dual-list-selector .o-dual-list-selector-list .mat-selection-list.mat-list-base .mat-list-item.mat-list-item:not(.mat-list-option) .mat-checkbox-layout{width:100%}.o-dual-list-selector .o-dual-list-selector-list .mat-list-option.cdk-drag.o-drag-list-item-box .mat-list-item-content{cursor:move}.o-dual-list-selector .o-dual-list-selector-list .drop-empty-message{position:absolute;width:calc(100% - 16px);text-align:center;opacity:.5;top:50%}"]
            }] }
];
