// ontimize-web-ngx/esm5/lib/components/user-info/user-info-configuration/o-user-info-configuration.directive.js
import * as tslib_1 from "tslib";
import { ContentChildren, Directive } from "@angular/core";
import { InputConverter } from "../../../decorators";
import { OUserInfoConfigurationItemDirective } from "../user-info-configuration-item/o-user-info-configuration-item.directive";
export var DEFAULT_INPUTS_O_USER_INFO_MENU = [
    'showProfile: show-profile',
    'showSettings: show-settings',
    'showLogout: show-logout',
];
export var DEFAULT_OUTPUTS_O_USER_INFO_MENU = [];
var OUserInfoConfigurationDirective = (function () {
    function OUserInfoConfigurationDirective() {
        this.showProfile = false;
        this.showSettings = true;
        this.showLogout = true;
    }
    OUserInfoConfigurationDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'o-user-info-configuration',
                    inputs: DEFAULT_INPUTS_O_USER_INFO_MENU,
                    outputs: DEFAULT_OUTPUTS_O_USER_INFO_MENU
                },] }
    ];
    OUserInfoConfigurationDirective.ctorParameters = function () { return []; };
    OUserInfoConfigurationDirective.propDecorators = {
        userInfoConfigurationItems: [{ type: ContentChildren, args: [OUserInfoConfigurationItemDirective,] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OUserInfoConfigurationDirective.prototype, "showProfile", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OUserInfoConfigurationDirective.prototype, "showSettings", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OUserInfoConfigurationDirective.prototype, "showLogout", void 0);
    return OUserInfoConfigurationDirective;
}());
export { OUserInfoConfigurationDirective };
