// ontimize-web-ngx/esm5/lib/components/input/listpicker/listpicker-renderer/percentage/o-list-picker-renderer-percentage.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { OPercentPipe } from '../../../../../pipes/o-percentage.pipe';
import { NumberService } from '../../../../../services/number.service';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER_REAL, OListPickerRendererRealComponent } from '../real/o-list-picker-renderer-real.component';
export var DEFAULT_INPUTS_O_LISTPICKER_RENDERER_PERCENTAGE = tslib_1.__spread(DEFAULT_INPUTS_O_LISTPICKER_RENDERER_REAL, [
    'valueBase: value-base'
]);
var OListPickerRendererPercentageComponent = (function (_super) {
    tslib_1.__extends(OListPickerRendererPercentageComponent, _super);
    function OListPickerRendererPercentageComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.decimalSeparator = '.';
        _this.minDecimalDigits = 0;
        _this.maxDecimalDigits = 0;
        _this.valueBase = 1;
        _this.numberService = _this.injector.get(NumberService);
        _this.setComponentPipe();
        return _this;
    }
    OListPickerRendererPercentageComponent.prototype.setComponentPipe = function () {
        this.componentPipe = new OPercentPipe(this.injector);
    };
    OListPickerRendererPercentageComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        this.pipeArguments = {
            minDecimalDigits: this.minDecimalDigits,
            maxDecimalDigits: this.maxDecimalDigits,
            decimalSeparator: this.decimalSeparator,
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator,
            valueBase: this.valueBase
        };
    };
    OListPickerRendererPercentageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-list-picker-renderer-percentage',
                    template: "<ng-template #templateref let-value=\"value\">\n  {{getListPickerValue(value)}}\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_LISTPICKER_RENDERER_PERCENTAGE
                }] }
    ];
    OListPickerRendererPercentageComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OListPickerRendererPercentageComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
    };
    return OListPickerRendererPercentageComponent;
}(OListPickerRendererRealComponent));
export { OListPickerRendererPercentageComponent };
