// ontimize-web-ngx/esm5/lib/components/contextmenu/context-menu-group/o-context-menu-group.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChildren, forwardRef, QueryList } from '@angular/core';
import { Subscription } from 'rxjs';
import { DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS, OComponentMenuBaseItem } from '../o-content-menu-base-item.class';
export var DEFAULT_CONTEXT_MENU_GROUP_INPUTS = tslib_1.__spread(DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS);
var OContextMenuGroupComponent = (function (_super) {
    tslib_1.__extends(OContextMenuGroupComponent, _super);
    function OContextMenuGroupComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = OComponentMenuBaseItem.TYPE_GROUP_MENU;
        _this.children = [];
        _this.subscription = new Subscription();
        return _this;
    }
    OContextMenuGroupComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    OContextMenuGroupComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.subscription.add(this.oContextMenuItems.changes.subscribe(function () {
            _this.updateChildren();
        }));
        this.updateChildren();
    };
    OContextMenuGroupComponent.prototype.updateChildren = function () {
        this.children = this.oContextMenuItems.toArray().slice(1, this.oContextMenuItems.toArray().length);
    };
    OContextMenuGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-context-menu-group',
                    template: ' ',
                    inputs: DEFAULT_CONTEXT_MENU_GROUP_INPUTS,
                    providers: [{ provide: OComponentMenuBaseItem, useExisting: forwardRef(function () { return OContextMenuGroupComponent; }) }]
                }] }
    ];
    OContextMenuGroupComponent.propDecorators = {
        oContextMenuItems: [{ type: ContentChildren, args: [OComponentMenuBaseItem,] }]
    };
    return OContextMenuGroupComponent;
}(OComponentMenuBaseItem));
export { OContextMenuGroupComponent };
