// ontimize-web-ngx/esm2015/lib/components/input/combo/combo-renderer/date/o-combo-renderer-date.component.js
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { OMomentPipe } from '../../../../../pipes/o-moment.pipe';
import { DEFAULT_INPUTS_O_COMBO_RENDERER, OComboCustomRenderer } from '../o-combo-renderer.class';
export const DEFAULT_INPUTS_O_COMBO_RENDERER_DATE = [
    ...DEFAULT_INPUTS_O_COMBO_RENDERER,
    'format'
];
export class OComboRendererDateComponent extends OComboCustomRenderer {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.setComponentPipe();
    }
    setComponentPipe() {
        this.componentPipe = new OMomentPipe(this.injector);
    }
    initialize() {
        super.initialize();
        this.pipeArguments = {
            format: this.format
        };
    }
}
OComboRendererDateComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-combo-renderer-date',
                template: "<ng-template #templateref let-value=\"value\">\n  {{getComboData(value)}}\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_COMBO_RENDERER_DATE
            }] }
];
OComboRendererDateComponent.ctorParameters = () => [
    { type: Injector }
];
OComboRendererDateComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
};
