// ontimize-web-ngx/esm5/lib/components/table/extensions/footer/paginator/o-table-paginator.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, forwardRef, Inject, Injector } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { OTableComponent } from '../../../o-table.component';
import { OBaseTablePaginator } from './o-base-table-paginator.class';
export var DEFAULT_PAGINATOR_TABLE = [
    'pageSize: page-size',
    'pageSizeOptions: page-size-options',
    'showFirstLastButtons: show-first-last-buttons'
];
var OTablePaginatorComponent = (function (_super) {
    tslib_1.__extends(OTablePaginatorComponent, _super);
    function OTablePaginatorComponent(injector, table) {
        var _this = _super.call(this) || this;
        _this.injector = injector;
        _this.table = table;
        _this.showFirstLastButtons = true;
        return _this;
    }
    OTablePaginatorComponent.prototype.ngOnInit = function () {
        this.pageSize = this.table.queryRows;
        this.pageIndex = this.table.currentPage;
        this.showFirstLastButtons = this.table.showPaginatorFirstLastButtons;
        this.table.registerPagination(this);
    };
    Object.defineProperty(OTablePaginatorComponent.prototype, "pageIndex", {
        get: function () {
            return this._pageIndex;
        },
        set: function (value) {
            this._pageIndex = value;
            if (this.table.matpaginator) {
                this.table.matpaginator.pageIndex = this._pageIndex;
            }
        },
        enumerable: true,
        configurable: true
    });
    OTablePaginatorComponent.prototype.isShowingAllRows = function (selectedLength) {
        return false;
    };
    OTablePaginatorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table-paginator',
                    template: ' ',
                    inputs: DEFAULT_PAGINATOR_TABLE,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    OTablePaginatorComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: OTableComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OTableComponent; }),] }] }
    ]; };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OTablePaginatorComponent.prototype, "showFirstLastButtons", void 0);
    return OTablePaginatorComponent;
}(OBaseTablePaginator));
export { OTablePaginatorComponent };
