// ontimize-web-ngx/esm2015/lib/layouts/form-layout/o-form-layout-manager.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, HostListener, Injector, Optional, SkipSelf, ViewChild } from '@angular/core';
import { MatDialog } from '@angular/material';
import { ActivatedRoute, Router } from '@angular/router';
import { Subscription } from 'rxjs';
import { InputConverter } from '../../decorators/input-converter';
import { ComponentStateServiceProvider, O_COMPONENT_STATE_SERVICE } from '../../services/factories';
import { LocalStorageService } from '../../services/local-storage.service';
import { NavigationService } from '../../services/navigation.service';
import { OFormLayoutManagerService } from '../../services/o-form-layout-manager.service';
import { AbstractComponentStateService } from '../../services/state/o-component-state.service';
import { OFormLayoutManagerComponentStateService } from '../../services/state/o-form-layout-manager-component-state.service';
import { OTranslateService } from '../../services/translate/o-translate.service';
import { Codes } from '../../util/codes';
import { Util } from '../../util/util';
import { OFormLayoutDialogComponent } from './dialog/o-form-layout-dialog.component';
import { CanActivateFormLayoutChildGuard } from './guards/o-form-layout-can-activate-child.guard';
export const DEFAULT_INPUTS_O_FORM_LAYOUT_MANAGER = [
    'oattr: attr',
    'mode',
    'storeState: store-state',
    'title',
    'labelColumns: label-columns',
    'separator',
    'titleDataOrigin: title-data-origin',
    'dialogWidth: dialog-width',
    'dialogMinWidth: dialog-min-width',
    'dialogMaxWidth: dialog-max-width',
    'dialogHeight: dialog-height',
    'dialogMinHeight: dialog-min-height',
    'dialogMaxHeight dialog-max-height',
    'dialogClass: dialog-class'
];
export const DEFAULT_OUTPUTS_O_FORM_LAYOUT_MANAGER = [
    'onMainTabSelected',
    'onSelectedTabChange',
    'onCloseTab'
];
export class OFormLayoutManagerComponent {
    constructor(injector, router, actRoute, dialog, elRef, parentFormLayoutManager) {
        this.injector = injector;
        this.router = router;
        this.actRoute = actRoute;
        this.dialog = dialog;
        this.elRef = elRef;
        this.parentFormLayoutManager = parentFormLayoutManager;
        this.OFormLayoutManagerComponent = OFormLayoutManagerComponent;
        this._mode = OFormLayoutManagerComponent.DIALOG_MODE;
        this._separator = ' ';
        this.storeState = true;
        this.dialogClass = '';
        this.onMainTabSelected = new EventEmitter();
        this.onSelectedTabChange = new EventEmitter();
        this.onCloseTab = new EventEmitter();
        this._labelColsArray = [];
        this._tabGroupOptions = {};
        this._dialogOptions = {};
        this._splitPaneOptions = {};
        this.addingGuard = false;
        this._markForUpdate = false;
        this.onTriggerUpdate = new EventEmitter();
        this.subscription = new Subscription();
        this.oFormLayoutManagerService = this.injector.get(OFormLayoutManagerService);
        this.localStorageService = this.injector.get(LocalStorageService);
        this.translateService = this.injector.get(OTranslateService);
        this.navigationService = this.injector.get(NavigationService);
        this.componentStateService = this.injector.get(AbstractComponentStateService);
        if (this.storeState) {
            this.subscription.add(this.localStorageService.onRouteChange.subscribe(res => {
                this.updateStateStorage();
            }));
        }
    }
    get mode() {
        return this._mode;
    }
    set mode(value) {
        const availableModeValues = [OFormLayoutManagerComponent.DIALOG_MODE, OFormLayoutManagerComponent.TAB_MODE, OFormLayoutManagerComponent.SPLIT_PANE_MODE];
        this._mode = (value || '').toLowerCase();
        if (availableModeValues.indexOf(this._mode) === -1) {
            this._mode = OFormLayoutManagerComponent.DIALOG_MODE;
        }
    }
    set separator(value) {
        this._separator = value;
    }
    get separator() {
        return this._separator;
    }
    get labelColumns() {
        return this._labelColumns;
    }
    set labelColumns(value) {
        this._labelColumns = value;
        this._labelColsArray = Util.parseArray(value);
    }
    get labelColsArray() {
        return this._labelColsArray;
    }
    set labelColsArray(value) {
        this._labelColsArray = value;
    }
    get tabGroupOptions() {
        return this._tabGroupOptions;
    }
    addTabGroupOptions(value) {
        Object.assign(this._tabGroupOptions, value);
        if (value.hasOwnProperty('labelColumns')) {
            this.labelColsArray = Util.parseArray(value['labelColumns']);
        }
        if (value.hasOwnProperty('separator')) {
            this.separator = value['separator'];
        }
    }
    get dialogOptions() {
        return this._dialogOptions;
    }
    addDialogOptions(value) {
        Object.assign(this._dialogOptions, value);
        if (value.hasOwnProperty('labelColumns')) {
            this.labelColsArray = Util.parseArray(value['labelColumns']);
        }
        if (value.hasOwnProperty('separator')) {
            this.separator = value['separator'];
        }
    }
    get splitPaneOptions() {
        return this._splitPaneOptions;
    }
    addSplitPaneOptions(value) {
        Object.assign(this._splitPaneOptions, value);
    }
    get state() {
        return this.componentStateService.state;
    }
    ngOnInit() {
        this.addActivateChildGuard();
        if (!Util.isDefined(this.oattr)) {
            this.oattr = (this.title || '') + this.mode;
            console.warn('o-form-layout-manager must have an unique attr');
        }
        this.oFormLayoutManagerService.registerFormLayoutManager(this);
        if (this.storeState) {
            this.componentStateService.initialize(this);
        }
    }
    ngAfterViewInit() {
        setTimeout(() => {
            if (this.elRef) {
                this.elRef.nativeElement.removeAttribute('title');
            }
        });
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
        this.oFormLayoutManagerService.removeFormLayoutManager(this);
        this.destroyActivateChildGuard();
    }
    getAttribute() {
        return this.oattr;
    }
    getComponentKey() {
        return 'OFormLayoutManagerComponent_' + this.oattr;
    }
    getRouteKey() {
        let route = this.router.url;
        this.actRoute.params.subscribe(params => {
            Object.keys(params).forEach(key => {
                route = route.replace(params[key], key);
            });
        });
        return route;
    }
    getDataToStore() {
        const compRef = this.getLayoutModeComponent();
        return Util.isDefined(compRef) ? compRef.getDataToStore() : {};
    }
    beforeunloadHandler() {
        this.updateStateStorage();
    }
    getLabelFromUrlParams(urlParams) {
        let label = '';
        const keys = Object.keys(urlParams);
        keys.forEach((param, i) => {
            label += urlParams[param] + ((i < keys.length - 1) ? this.separator : '');
        });
        return label;
    }
    getFormDataFromLabelColumns(data) {
        const formData = {};
        Object.keys(data).forEach(x => {
            if (this.labelColsArray.indexOf(x) > -1) {
                formData[x] = data[x];
            }
        });
        return formData;
    }
    addActivateChildGuard() {
        const routeConfig = this.getParentActRouteRoute();
        if (Util.isDefined(routeConfig)) {
            const canActivateChildArray = (routeConfig.canActivateChild || []);
            let previouslyAdded = false;
            for (let i = 0, len = canActivateChildArray.length; i < len; i++) {
                const guardChild = this.injector.get(CanActivateFormLayoutChildGuard);
                previouslyAdded = (canActivateChildArray[i] === guardChild.constructor);
                if (previouslyAdded) {
                    break;
                }
            }
            if (!previouslyAdded) {
                this.addingGuard = true;
                canActivateChildArray.push(CanActivateFormLayoutChildGuard);
                routeConfig.canActivateChild = canActivateChildArray;
            }
        }
    }
    destroyActivateChildGuard() {
        if (!this.addingGuard) {
            return;
        }
        this.oFormLayoutManagerService.activeFormLayoutManager = undefined;
        const routeConfig = this.getParentActRouteRoute();
        if (Util.isDefined(routeConfig)) {
            for (let i = (routeConfig.canActivateChild || []).length - 1; i >= 0; i--) {
                const guardChild = this.injector.get(CanActivateFormLayoutChildGuard);
                if (routeConfig.canActivateChild[i] === guardChild.constructor) {
                    routeConfig.canActivateChild.splice(i, 1);
                    break;
                }
            }
        }
    }
    isDialogMode() {
        return this.mode === OFormLayoutManagerComponent.DIALOG_MODE;
    }
    isTabMode() {
        return this.mode === OFormLayoutManagerComponent.TAB_MODE;
    }
    isSplitPaneMode() {
        return this.mode === OFormLayoutManagerComponent.SPLIT_PANE_MODE;
    }
    addDetailComponent(childRoute, url) {
        const newDetailComp = {
            params: childRoute.params,
            queryParams: childRoute.queryParams,
            urlSegments: childRoute.url,
            component: childRoute.routeConfig.component,
            url: url,
            id: Util.randomNumber().toString(),
            label: '',
            innerFormsInfo: {},
            insertionMode: childRoute.queryParams[Codes.INSERTION_MODE] === 'true'
        };
        if (this.isDialogMode()) {
            this.openFormLayoutDialog(newDetailComp);
        }
        else {
            const compRef = this.getLayoutModeComponent();
            if (Util.isDefined(compRef)) {
                compRef.openDetail(newDetailComp);
            }
        }
    }
    closeDetail(options) {
        const compRef = this.getLayoutModeComponent();
        if (Util.isDefined(compRef)) {
            compRef.closeDetail(options);
        }
    }
    openFormLayoutDialog(detailComp) {
        const cssclass = ['o-form-layout-dialog-overlay'];
        if (this.dialogClass) {
            cssclass.push(this.dialogClass);
        }
        const dialogOptions = (this.dialogOptions || {});
        const dialogConfig = {
            data: {
                data: detailComp,
                layoutManagerComponent: this,
                title: dialogOptions.title || this.title,
            },
            width: dialogOptions.width || this.dialogWidth,
            minWidth: dialogOptions.minWidth || this.dialogMinWidth,
            maxWidth: dialogOptions.maxWidth || this.dialogMaxWidth,
            height: dialogOptions.height || this.dialogHeight,
            minHeight: dialogOptions.minHeight || this.dialogMinHeight,
            maxHeight: dialogOptions.maxHeight || this.dialogMaxHeight,
            disableClose: dialogOptions.disableClose || true,
            panelClass: dialogOptions.class || cssclass
        };
        if (this.dialogOptions) {
            dialogConfig.closeOnNavigation = this.dialogOptions.closeOnNavigation;
            dialogConfig.backdropClass = this.dialogOptions.backdropClass;
            dialogConfig.position = this.dialogOptions.position;
            dialogConfig.disableClose = this.dialogOptions.disableClose;
        }
        this.dialogRef = this.dialog.open(OFormLayoutDialogComponent, dialogConfig);
        this.dialogRef.afterClosed().subscribe(() => {
            if (this.markForUpdate) {
                this.updateIfNeeded();
            }
            else {
                this.reloadMainComponents();
            }
        });
    }
    getFormCacheData() {
        const compRef = this.getLayoutModeComponent();
        return Util.isDefined(compRef) ? compRef.getFormCacheData() : undefined;
    }
    setModifiedState(formAttr, modified, confirmExit) {
        const compRef = this.getLayoutModeComponent();
        if (Util.isDefined(compRef)) {
            compRef.setModifiedState(formAttr, modified, confirmExit);
        }
    }
    getLabelFromData(data) {
        let label = '';
        const isDataDefined = Util.isDefined(data);
        if (isDataDefined && data.hasOwnProperty('new_tab_title')) {
            label = this.translateService.get(data.new_tab_title);
        }
        else if (isDataDefined && this.labelColsArray.length !== 0) {
            this.labelColsArray.forEach((col, idx) => {
                if (data[col] !== undefined) {
                    label += data[col] + ((idx < this.labelColsArray.length - 1) ? this.separator : '');
                }
            });
        }
        return label;
    }
    updateNavigation(data, keysValues, insertionMode) {
        const compRef = this.getLayoutModeComponent();
        if (Util.isDefined(compRef)) {
            compRef.updateNavigation(data, keysValues, insertionMode);
        }
    }
    updateActiveData(data) {
        const compRef = this.getLayoutModeComponent();
        if (Util.isDefined(compRef)) {
            compRef.updateActiveData(data);
        }
    }
    getRouteOfActiveItem() {
        const compRef = this.getLayoutModeComponent();
        return Util.isDefined(compRef) ? compRef.getRouteOfActiveItem() : [];
    }
    isMainComponent(comp) {
        if (this.isDialogMode()) {
            return !comp.oFormLayoutDialog;
        }
        const compRef = this.getLayoutModeComponent();
        return Util.isDefined(compRef) && compRef.isMainComponent(comp);
    }
    getRouteForComponent(comp) {
        const result = [];
        if (this.parentFormLayoutManager) {
            const parentRoute = this.parentFormLayoutManager.getRouteForComponent(comp);
            if (parentRoute && parentRoute.length > 0) {
                result.push(...parentRoute);
            }
        }
        if (!this.isMainComponent(comp)) {
            const activeRoute = this.getRouteOfActiveItem();
            if (activeRoute && activeRoute.length > 0) {
                result.push(...activeRoute);
            }
        }
        return result;
    }
    setAsActiveFormLayoutManager() {
        this.oFormLayoutManagerService.activeFormLayoutManager = this;
    }
    reloadMainComponents() {
        this.onTriggerUpdate.emit();
    }
    allowToUpdateNavigation(formAttr) {
        return (this.isTabMode() && Util.isDefined(this.oTabGroup) && Util.isDefined(this.titleDataOrigin)) ?
            this.titleDataOrigin === formAttr :
            true;
    }
    updateStateStorage() {
        if (!this.localStorageService || !this.storeState) {
            return;
        }
        const isTabMode = this.isTabMode() && Util.isDefined(this.oTabGroup);
        const isSplitPaneMode = this.isSplitPaneMode() && Util.isDefined(this.oSplitPane);
        if (isTabMode || isSplitPaneMode) {
            this.localStorageService.updateComponentStorage(this, this.getRouteKey());
        }
    }
    getParentActRouteRoute() {
        let actRoute = this.actRoute;
        while (actRoute.parent !== undefined && actRoute.parent !== null) {
            if (actRoute.routeConfig.children || actRoute.routeConfig.loadChildren) {
                break;
            }
            actRoute = actRoute.parent;
        }
        return actRoute.routeConfig;
    }
    updateIfNeeded() {
        if (this.markForUpdate) {
            this.markForUpdate = false;
            this.onTriggerUpdate.emit();
        }
    }
    getParams() {
        const compRef = this.getLayoutModeComponent();
        return Util.isDefined(compRef) ? compRef.getParams() : undefined;
    }
    set markForUpdate(arg) {
        this._markForUpdate = arg;
        if (this.isSplitPaneMode()) {
            this.updateIfNeeded();
        }
    }
    get markForUpdate() {
        return this._markForUpdate;
    }
    get ignoreCanDeactivate() {
        return !this.isSplitPaneMode();
    }
    getLayoutModeComponent() {
        let compRef;
        if (this.isTabMode() && Util.isDefined(this.oTabGroup)) {
            compRef = this.oTabGroup;
        }
        else if (this.isDialogMode() && Util.isDefined(this.dialogRef)) {
            compRef = this.dialogRef.componentInstance;
        }
        else if (this.isSplitPaneMode() && Util.isDefined(this.oSplitPane)) {
            compRef = this.oSplitPane;
        }
        return compRef;
    }
    allowNavigation() {
        return !this.isTabMode();
    }
    canAddDetailComponent() {
        const compRef = this.getLayoutModeComponent();
        return Util.wrapIntoObservable(Util.isDefined(compRef) ? compRef.canAddDetailComponent() : true);
    }
    hasToConfirmExit(data, options) {
        if (Util.isDefined(options) && options.exitWithoutConfirmation) {
            return false;
        }
        const formsAttr = Object.keys(data.innerFormsInfo);
        let result = false;
        if (formsAttr.length > 0) {
            formsAttr.forEach(formAttr => {
                if (!result) {
                    const formData = data.innerFormsInfo[formAttr];
                    result = formData.confirmOnExit && formData.modified;
                }
            });
        }
        return result;
    }
}
OFormLayoutManagerComponent.DIALOG_MODE = 'dialog';
OFormLayoutManagerComponent.TAB_MODE = 'tab';
OFormLayoutManagerComponent.SPLIT_PANE_MODE = 'split-pane';
OFormLayoutManagerComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-form-layout-manager',
                inputs: DEFAULT_INPUTS_O_FORM_LAYOUT_MANAGER,
                outputs: DEFAULT_OUTPUTS_O_FORM_LAYOUT_MANAGER,
                template: "<ng-container [ngSwitch]=\"mode\">\n  <ng-container *ngSwitchCase=\"OFormLayoutManagerComponent.TAB_MODE\">\n    <o-form-layout-tabgroup #tabGroup [title]=\"title\" (onMainTabSelected)=\"onMainTabSelected.emit($event)\"\n      (onSelectedTabChange)=\"onSelectedTabChange.emit($event)\" (onCloseTab)=\"onCloseTab.emit($event)\"\n      [options]=\"tabGroupOptions\">\n      <ng-content *ngTemplateOutlet=\"contentBlock\"></ng-content>\n    </o-form-layout-tabgroup>\n  </ng-container>\n\n  <ng-container *ngSwitchCase=\"OFormLayoutManagerComponent.SPLIT_PANE_MODE\">\n    <o-form-layout-split-pane #splitPane [options]=\"splitPaneOptions\">\n      <ng-content *ngTemplateOutlet=\"contentBlock\"></ng-content>\n    </o-form-layout-split-pane>\n  </ng-container>\n\n  <ng-container *ngSwitchCase=\"OFormLayoutManagerComponent.DIALOG_MODE\">\n    <ng-content *ngTemplateOutlet=\"contentBlock\"></ng-content>\n  </ng-container>\n</ng-container>\n\n\n<ng-template #contentBlock>\n  <ng-content></ng-content>\n</ng-template>",
                providers: [
                    ComponentStateServiceProvider,
                    { provide: O_COMPONENT_STATE_SERVICE, useClass: OFormLayoutManagerComponentStateService },
                ],
                host: {
                    '[class.o-form-layout-manager]': 'true'
                },
                styles: [".o-form-layout-manager{height:100%}"]
            }] }
];
OFormLayoutManagerComponent.ctorParameters = () => [
    { type: Injector },
    { type: Router },
    { type: ActivatedRoute },
    { type: MatDialog },
    { type: ElementRef },
    { type: OFormLayoutManagerComponent, decorators: [{ type: SkipSelf }, { type: Optional }] }
];
OFormLayoutManagerComponent.propDecorators = {
    oTabGroup: [{ type: ViewChild, args: ['tabGroup', { static: false },] }],
    oSplitPane: [{ type: ViewChild, args: ['splitPane', { static: false },] }],
    beforeunloadHandler: [{ type: HostListener, args: ['window:beforeunload', [],] }]
};
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OFormLayoutManagerComponent.prototype, "storeState", void 0);
