// ontimize-web-ngx/esm2015/lib/components/form/navigation/o-form-confirm-exit.service.js
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { DialogService } from '../../../services/dialog.service';
import { Util } from '../../../util/util';
import * as i0 from "@angular/core";
import * as i1 from "../../../services/dialog.service";
export class OFormConfirmExitService {
    constructor(dialogService) {
        this.dialogService = dialogService;
    }
    subscribeToDiscardChanges(form, ignoreAttrs = []) {
        let subscription;
        if (form.isInitialStateChanged(ignoreAttrs) && this.mustShowConfirmationInForm(form)) {
            subscription = this.getConfirmDialogSubscription(form);
        }
        else {
            const observable = new Observable(observer => {
                observer.next(true);
                observer.complete();
            });
            subscription = observable.toPromise();
        }
        return subscription;
    }
    mustShowConfirmationInForm(form) {
        return form.isInInsertMode() || form.isInUpdateMode();
    }
    restart() {
        this.confirmDialogSubscription = null;
    }
    getConfirmDialogSubscription(form) {
        if (!Util.isDefined(this.confirmDialogSubscription)) {
            this.confirmDialogSubscription = new Promise((resolve) => {
                this.dialogService.confirm(form.messageService.getDiscardChangesConfirmationDialogTitle(), form.messageService.getDiscardChangesConfirmationMessage()).then((res) => {
                    this.restart();
                    resolve(res);
                });
            });
        }
        return this.confirmDialogSubscription;
    }
}
OFormConfirmExitService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
OFormConfirmExitService.ctorParameters = () => [
    { type: DialogService }
];
OFormConfirmExitService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OFormConfirmExitService_Factory() { return new OFormConfirmExitService(i0.ɵɵinject(i1.DialogService)); }, token: OFormConfirmExitService, providedIn: "root" });
