// ontimize-web-ngx/esm2015/lib/components/form/navigation/o-form-navigation.component.js
import { Component, forwardRef, Inject, Injector, ViewEncapsulation } from '@angular/core';
import { Router } from '@angular/router';
import { OntimizeServiceProvider } from '../../../services/factories';
import { NavigationService } from '../../../services/navigation.service';
import { OntimizeService } from '../../../services/ontimize/ontimize.service';
import { Codes } from '../../../util/codes';
import { Util } from '../../../util/util';
import { OFormComponent } from '../o-form.component';
export class OFormNavigationComponent {
    constructor(injector, _form, router) {
        this.injector = injector;
        this._form = _form;
        this.router = router;
        this.navigationData = [];
        this._currentIndex = 0;
        this.formNavigation = this._form.getFormNavigation();
        this.navigationService = this.injector.get(NavigationService);
        this.formLayoutManager = this._form.getFormManager();
        let navData;
        if (this.formLayoutManager && this.formLayoutManager.allowNavigation()) {
            navData = this.navigationService.getLastItem();
        }
        else {
            navData = this.navigationService.getPreviousRouteData();
        }
        if (Util.isDefined(navData)) {
            this.navigationData = navData.keysValues || [];
            this.queryConf = navData.queryConfiguration;
        }
        this.currentIndex = this.getCurrentIndex();
        this.configureService();
    }
    configureService() {
        if (!this.queryConf) {
            return;
        }
        let loadingService = OntimizeService;
        if (this.queryConf.serviceType) {
            loadingService = this.queryConf.serviceType;
        }
        try {
            this.dataService = this.injector.get(loadingService);
            if (Util.isDataService(this.dataService)) {
                const serviceCfg = this.dataService.getDefaultServiceConfiguration(this.queryConf.service);
                if (this.queryConf.entity) {
                    serviceCfg.entity = this.queryConf.entity;
                }
                this.dataService.configureService(serviceCfg);
            }
        }
        catch (e) {
            console.error(e);
        }
    }
    queryNavigationData(offset, length) {
        const self = this;
        return new Promise((resolve, reject) => {
            const conf = self.queryConf;
            const queryArgs = conf.queryArguments;
            queryArgs[1] = self.getKeysArray();
            queryArgs[4] = offset;
            queryArgs[5] = length ? length : conf.queryRows;
            self.querySubscription = self.dataService[conf.queryMethod].apply(self.dataService, queryArgs).subscribe(res => {
                if (res.isSuccessful()) {
                    self.navigationData = res.data;
                    self.queryConf.queryRecordOffset = offset;
                }
                resolve();
            }, () => {
                reject();
            });
        });
    }
    ngOnDestroy() {
        if (this.querySubscription) {
            this.querySubscription.unsubscribe();
        }
    }
    getKeysArray() {
        const navData = this.navigationData ? (this.navigationData[0] || {}) : {};
        const keysArray = [];
        this._form.keysArray.forEach(key => {
            if (navData.hasOwnProperty(key)) {
                keysArray.push(key);
            }
        });
        return keysArray;
    }
    getCurrentIndex() {
        const keysArray = this.getKeysArray();
        const currentKeys = {};
        const currentItem = this.formNavigation.getUrlParams();
        keysArray.forEach(key => {
            currentKeys[key] = currentItem[key];
        });
        const index = (this.navigationData || []).findIndex((item) => {
            const itemKeys = {};
            keysArray.forEach(key => {
                itemKeys[key] = item[key];
            });
            return Util.isEquivalent(itemKeys, currentKeys);
        });
        return index >= 0 ? index : 0;
    }
    next() {
        const total = this.navigationData.length;
        const index = this.currentIndex + 1;
        if (total > index) {
            this.move(index);
        }
        else if (this.queryConf) {
            const offset = (this.queryConf.queryRecordOffset || 0) + this.queryConf.queryRows;
            this.queryNavigationData(offset).then(() => {
                this.move(0);
            });
        }
        else {
            console.error('form-toolbar->next(): total > index');
        }
    }
    previous() {
        const index = this.currentIndex - 1;
        if (index >= 0) {
            this.move(index);
        }
        else if (this.queryConf) {
            const offset = this.queryConf.queryRecordOffset - this.queryConf.queryRows;
            this.queryNavigationData(offset).then(() => {
                this.move(this.navigationData.length - 1);
            });
        }
        else {
            console.error('form-toolbar->next(): index < 0');
        }
    }
    first() {
        if (!this.queryConf || this.queryConf.queryRecordOffset === 0) {
            this.move(0);
        }
        else {
            this.queryNavigationData(0).then(() => {
                this.move(0);
            });
        }
    }
    last() {
        if (!this.queryConf || this.isLast()) {
            const index = this.navigationData.length - 1;
            this.move(index);
        }
        else {
            const offset = this.queryConf.totalRecordsNumber - this.queryConf.queryRows;
            this.queryNavigationData(offset, this.queryConf.queryRows).then(() => {
                this.move(this.navigationData.length - 1);
            });
        }
    }
    isFirst() {
        let result = this.currentIndex === 0;
        if (result && this.queryConf) {
            result = this.queryConf.queryRecordOffset === 0;
        }
        return result;
    }
    isLast() {
        let result = this.currentIndex === (this.navigationData.length - 1);
        if (result && this.queryConf) {
            result = (this.queryConf.queryRecordOffset + this.queryConf.queryRows)
                >= this.queryConf.totalRecordsNumber;
        }
        return result;
    }
    move(index) {
        this._form.showConfirmDiscardChanges().then(res => {
            if (res === true) {
                this.currentIndex = index;
                if (this.formLayoutManager && this.formLayoutManager.allowNavigation()) {
                    this.moveInFormLayoutManager(index);
                }
                else {
                    this.moveWithoutManager(index);
                }
            }
        });
    }
    moveWithoutManager(index) {
        const route = this.getRouteOfSelectedRow(this.navigationData[index]);
        if (route.length > 0) {
            const navData = this.navigationService.getLastItem();
            if (navData) {
                this._form.canDiscardChanges = true;
                const extras = {};
                extras[Codes.QUERY_PARAMS] = Codes.getIsDetailObject();
                const urlArray = navData.url.split(Codes.ROUTE_SEPARATOR);
                const url = urlArray.splice(0, urlArray.length - route.length).join(Codes.ROUTE_SEPARATOR);
                route.unshift(url);
                this.router.navigate(route, extras).then((navigationDone) => {
                    if (navigationDone) {
                        this.currentIndex = index;
                        const url = this.router.routerState.snapshot.url.split('?')[0];
                        if (url !== navData.url) {
                            this.navigationService.removeLastItem();
                        }
                    }
                });
            }
        }
    }
    moveInFormLayoutManager(index) {
        this._form.setUrlParamsAndReload(this.navigationData[index]);
    }
    getRouteOfSelectedRow(item) {
        const route = [];
        if (Util.isObject(item)) {
            this._form.keysArray.forEach(key => {
                if (Util.isDefined(item[key])) {
                    route.push(item[key]);
                }
            });
        }
        return route;
    }
    showNavigation() {
        return (this.navigationData || []).length > 1;
    }
    set currentIndex(arg) {
        this._currentIndex = arg;
    }
    get currentIndex() {
        return this._currentIndex;
    }
    getRecordIndex() {
        let index = this.currentIndex + 1;
        if (this.queryConf) {
            index += this.queryConf.queryRecordOffset;
        }
        return index;
    }
    getTotalRecordsNumber() {
        if (this.queryConf && this.queryConf.totalRecordsNumber) {
            return this.queryConf.totalRecordsNumber;
        }
        return this.navigationData.length;
    }
}
OFormNavigationComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-form-navigation',
                template: "<ng-container *ngIf=\"showNavigation()\">\n  <button mat-icon-button class=\"o-form-toolbar-button\" [disabled]=\"isFirst()\" (click)=\"first()\">\n    <mat-icon aria-label=\"First\" layout-padding svgIcon=\"ontimize:first_page\"></mat-icon>\n  </button>\n  <button mat-icon-button class=\"o-form-toolbar-button\" [disabled]=\"isFirst()\" (click)=\"previous()\">\n    <mat-icon aria-label=\"Previous\" layout-padding svgIcon=\"ontimize:keyboard_arrow_left\"></mat-icon>\n  </button>\n  <span layout-padding>{{ getRecordIndex() }} / {{ getTotalRecordsNumber() }}</span>\n  <button mat-icon-button class=\"o-form-toolbar-button\" [disabled]=\"isLast()\" (click)=\"next()\">\n    <mat-icon aria-label=\"Next\" layout-padding svgIcon=\"ontimize:keyboard_arrow_right\"></mat-icon>\n  </button>\n  <button mat-icon-button class=\"o-form-toolbar-button\" [disabled]=\"isLast()\" (click)=\"last()\">\n    <mat-icon aria-label=\"Last\" layout-padding svgIcon=\"ontimize:last_page\"></mat-icon>\n  </button>\n</ng-container>",
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.o-form-navigation]': 'true'
                },
                providers: [
                    OntimizeServiceProvider
                ],
                styles: [".o-form-navigation .mat-icon{cursor:pointer}.o-form-navigation span{cursor:default}"]
            }] }
];
OFormNavigationComponent.ctorParameters = () => [
    { type: Injector },
    { type: OFormComponent, decorators: [{ type: Inject, args: [forwardRef(() => OFormComponent),] }] },
    { type: Router }
];
