// ontimize-web-ngx/esm5/lib/components/input/listpicker/listpicker-renderer/currency/o-list-picker-renderer-currency.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { OCurrencyPipe } from '../../../../../pipes/o-currency.pipe';
import { CurrencyService } from '../../../../../services/currency.service';
import { OListPickerCustomRenderer } from '../o-list-picker-renderer.class';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER_REAL } from '../real/o-list-picker-renderer-real.component';
export var DEFAULT_INPUTS_O_LISTPICKER_RENDERER_CURRENCY = tslib_1.__spread(DEFAULT_INPUTS_O_LISTPICKER_RENDERER_REAL, [
    'currencySymbol: currency-symbol',
    'currencySymbolPosition: currency-symbol-position'
]);
var OListPickerRendererCurrencyComponent = (function (_super) {
    tslib_1.__extends(OListPickerRendererCurrencyComponent, _super);
    function OListPickerRendererCurrencyComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.minDecimalDigits = 2;
        _this.maxDecimalDigits = 2;
        _this.decimalSeparator = '.';
        _this.grouping = true;
        _this.thousandSeparator = ',';
        _this.currencyService = _this.injector.get(CurrencyService);
        _this.setComponentPipe();
        return _this;
    }
    OListPickerRendererCurrencyComponent.prototype.setComponentPipe = function () {
        this.componentPipe = new OCurrencyPipe(this.injector);
    };
    OListPickerRendererCurrencyComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        if (typeof this.currencySymbol === 'undefined') {
            this.currencySymbol = this.currencyService.symbol;
        }
        if (typeof this.currencySymbolPosition === 'undefined') {
            this.currencySymbolPosition = this.currencyService.symbolPosition;
        }
        this.pipeArguments = {
            currencySimbol: this.currencySymbol,
            currencySymbolPosition: this.currencySymbolPosition,
            minDecimalDigits: this.minDecimalDigits,
            maxDecimalDigits: this.maxDecimalDigits,
            decimalSeparator: this.decimalSeparator,
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator
        };
    };
    OListPickerRendererCurrencyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-list-picker-renderer-currency',
                    template: "<ng-template #templateref let-value=\"value\">\n  {{getListPickerValue(value)}}\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_LISTPICKER_RENDERER_CURRENCY
                }] }
    ];
    OListPickerRendererCurrencyComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OListPickerRendererCurrencyComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Number)
    ], OListPickerRendererCurrencyComponent.prototype, "minDecimalDigits", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Number)
    ], OListPickerRendererCurrencyComponent.prototype, "maxDecimalDigits", void 0);
    return OListPickerRendererCurrencyComponent;
}(OListPickerCustomRenderer));
export { OListPickerRendererCurrencyComponent };
