// ontimize-web-ngx/esm2015/lib/components/input/combo/combo-renderer/o-combo-renderer.class.js
import { Util } from '../../../../util';
import { OComboComponent } from '../o-combo.component';
export const DEFAULT_INPUTS_O_COMBO_RENDERER = [];
export class OComboCustomRenderer {
    constructor(injector) {
        this.injector = injector;
        this.comboComponent = this.injector.get(OComboComponent);
    }
    ngOnInit() {
        this.initialize();
    }
    initialize() {
    }
    ngAfterContentInit() {
        this.registerRenderer();
    }
    registerRenderer() {
        this.comboComponent.registerRenderer(this);
    }
    getComboData(record) {
        let descTxt = '';
        if (!Util.isDefined(record)) {
            return descTxt;
        }
        this.comboComponent.descriptionColArray.forEach((col, index) => {
            if (Util.isDefined(record[col])) {
                let value = record[col];
                if (this.componentPipe && this.pipeArguments != null) {
                    value = this.componentPipe.transform(value, this.pipeArguments);
                }
                if (Util.isDefined(value)) {
                    descTxt += value;
                    if (index < this.comboComponent.descriptionColArray.length - 1) {
                        descTxt += this.comboComponent.separator;
                    }
                }
            }
        });
        return descTxt;
    }
}
