// ontimize-web-ngx/esm2015/lib/components/table/extensions/header/table-header/o-table-header.component.js
import { ChangeDetectionStrategy, Component, forwardRef, Inject, ViewChild, ViewEncapsulation } from "@angular/core";
import { OTableComponent } from "../../../o-table.component";
import { OMatSortHeader } from "../../sort/o-mat-sort-header";
import { OTableHeaderColumnFilterIconComponent } from "../table-header-column-filter-icon/o-table-header-column-filter-icon.component";
export const DEFAULT_INPUTS_O_TABLE_HEADER = [
    'column'
];
export class OTableHeaderComponent {
    constructor(table) {
        this.table = table;
        this.resizable = this.table.resizable;
    }
    set columnFilterIcon(value) {
        this._columnFilterIcon = value;
    }
    isModeColumnFilterable(column) {
        return this.table.isColumnFiltersActive && this.table.isColumnFilterable(column);
    }
    setFilterIconHintVisible(visible) {
        if (this._columnFilterIcon) {
            this._columnFilterIcon.filterIconHintVisible.next(visible);
        }
    }
    ngAfterViewInit() {
        this.table.registerTableHeaders(this);
    }
}
OTableHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-header',
                inputs: DEFAULT_INPUTS_O_TABLE_HEADER,
                template: "<o-table-header-column-filter-icon #columnFilterIcon *ngIf=\"isModeColumnFilterable(column)\" [column]=\"column\">\n</o-table-header-column-filter-icon>\n\n<ng-container *ngIf=\"column.orderable\">\n    <span o-mat-sort-header>{{ column.title | oTranslate }}</span>\n</ng-container>\n<ng-container *ngIf=\"!column.orderable\">\n    <span class=\"header-title-container\">{{ column.title | oTranslate }}</span>\n</ng-container>\n\n<o-table-column-resizer *ngIf=\"resizable\" [column]=\"column\"></o-table-column-resizer>",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[class.o-table-header]': 'true',
                    '(mouseenter)': 'setFilterIconHintVisible(true)',
                    '(mouseleave)': 'setFilterIconHintVisible(false)'
                }
            }] }
];
OTableHeaderComponent.ctorParameters = () => [
    { type: OTableComponent, decorators: [{ type: Inject, args: [forwardRef(() => OTableComponent),] }] }
];
OTableHeaderComponent.propDecorators = {
    columnFilterIcon: [{ type: ViewChild, args: ['columnFilterIcon', { static: false },] }],
    matSortHeader: [{ type: ViewChild, args: [OMatSortHeader, { static: false },] }]
};
