// ontimize-web-ngx/esm2015/lib/components/contextmenu/o-content-menu-base-item.class.js
import { Util } from '../../util/util';
export const DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS = [
    'attr',
    'ovisible: visible',
    'icon',
    'data',
    'label',
    'oenabled: enabled',
    'svgIcon: svg-icon'
];
export class OComponentMenuBaseItem {
    constructor() {
        this.ovisible = true;
        this.type = OComponentMenuBaseItem.TYPE_GROUP_MENU;
        this.enabled = true;
    }
    get disabled() {
        if (this.enabled instanceof Function) {
            return !this.enabled(this.data);
        }
        return !this.enabled;
    }
    get isVisible() {
        if (this.ovisible instanceof Function) {
            return this.ovisible(this.data);
        }
        return this.ovisible;
    }
    set oenabled(value) {
        if (value instanceof Function) {
            this.enabled = value;
        }
        else {
            this.enabled = this.parseInput(value, true);
        }
    }
    parseInput(value, defaultValue) {
        if (value instanceof Function || typeof value === 'boolean') {
            return value;
        }
        return Util.parseBoolean(value, defaultValue);
    }
}
OComponentMenuBaseItem.TYPE_ITEM_MENU = 'item_menu';
OComponentMenuBaseItem.TYPE_GROUP_MENU = 'item_group';
OComponentMenuBaseItem.TYPE_SEPARATOR_MENU = 'item_separator';
