// ontimize-web-ngx/esm2015/lib/components/contextmenu/o-context-menu.component.js
import { Component, ContentChildren, EventEmitter, Injector, QueryList } from '@angular/core';
import { Subscription } from 'rxjs';
import { OComponentMenuBaseItem } from './o-content-menu-base-item.class';
import { OContextMenuService } from './o-context-menu.service';
export const DEFAULT_OUTPUTS_O_CONTEXT_MENU = [
    'onShow',
    'onClose'
];
export class OContextMenuComponent {
    constructor(injector) {
        this.injector = injector;
        this.onShow = new EventEmitter();
        this.onClose = new EventEmitter();
        this.subscription = new Subscription();
        this.oContextMenuService = this.injector.get(OContextMenuService);
    }
    ngOnInit() {
        this.subscription.add(this.oContextMenuService.showContextMenu.subscribe(param => this.showContextMenu(param)));
        this.subscription.add(this.oContextMenuService.closeContextMenu.subscribe(param => this.onClose.emit()));
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    showContextMenu(params) {
        this.origin = params.event.target;
        this.onShow.emit(params);
        if (params.contextMenu !== this) {
            return;
        }
        params.menuItems = this.oContextMenuItems;
        params.externalMenuItems = this.externalContextMenuItems;
        if (params.menuItems.length > 0) {
            this.oContextMenuService.openContextMenu(params);
        }
    }
}
OContextMenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-context-menu',
                template: ' ',
                outputs: DEFAULT_OUTPUTS_O_CONTEXT_MENU,
                providers: [OContextMenuService]
            }] }
];
OContextMenuComponent.ctorParameters = () => [
    { type: Injector }
];
OContextMenuComponent.propDecorators = {
    oContextMenuItems: [{ type: ContentChildren, args: [OComponentMenuBaseItem,] }]
};
