// ontimize-web-ngx/esm5/lib/components/form/navigation/o-form-confirm-exit.service.js
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { DialogService } from '../../../services/dialog.service';
import { Util } from '../../../util/util';
import * as i0 from "@angular/core";
import * as i1 from "../../../services/dialog.service";
var OFormConfirmExitService = (function () {
    function OFormConfirmExitService(dialogService) {
        this.dialogService = dialogService;
    }
    OFormConfirmExitService.prototype.subscribeToDiscardChanges = function (form, ignoreAttrs) {
        if (ignoreAttrs === void 0) { ignoreAttrs = []; }
        var subscription;
        if (form.isInitialStateChanged(ignoreAttrs) && this.mustShowConfirmationInForm(form)) {
            subscription = this.getConfirmDialogSubscription(form);
        }
        else {
            var observable = new Observable(function (observer) {
                observer.next(true);
                observer.complete();
            });
            subscription = observable.toPromise();
        }
        return subscription;
    };
    OFormConfirmExitService.prototype.mustShowConfirmationInForm = function (form) {
        return form.isInInsertMode() || form.isInUpdateMode();
    };
    OFormConfirmExitService.prototype.restart = function () {
        this.confirmDialogSubscription = null;
    };
    OFormConfirmExitService.prototype.getConfirmDialogSubscription = function (form) {
        var _this = this;
        if (!Util.isDefined(this.confirmDialogSubscription)) {
            this.confirmDialogSubscription = new Promise(function (resolve) {
                _this.dialogService.confirm(form.messageService.getDiscardChangesConfirmationDialogTitle(), form.messageService.getDiscardChangesConfirmationMessage()).then(function (res) {
                    _this.restart();
                    resolve(res);
                });
            });
        }
        return this.confirmDialogSubscription;
    };
    OFormConfirmExitService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    OFormConfirmExitService.ctorParameters = function () { return [
        { type: DialogService }
    ]; };
    OFormConfirmExitService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OFormConfirmExitService_Factory() { return new OFormConfirmExitService(i0.ɵɵinject(i1.DialogService)); }, token: OFormConfirmExitService, providedIn: "root" });
    return OFormConfirmExitService;
}());
export { OFormConfirmExitService };
