// ontimize-web-ngx/esm2015/lib/components/table/extensions/footer/paginator/o-base-table-paginator.class.js
import { Codes, Util } from '../../../../../util';
export class OBaseTablePaginator {
    constructor() {
        this._pageIndex = 0;
        this._pageSize = 10;
        this._pageSizeOptions = Codes.PAGE_SIZE_OPTIONS;
        this.showFirstLastButtons = true;
    }
    get pageLenght() {
        return this._pageSize;
    }
    set pageLenght(value) {
        this._pageSize = value;
    }
    get pageIndex() {
        return this._pageIndex;
    }
    set pageIndex(value) {
        this._pageIndex = value;
    }
    get pageSizeOptions() {
        return this._pageSizeOptions;
    }
    set pageSizeOptions(value) {
        if (typeof value === 'string') {
            const opts = Util.parseArray(value, true);
            value = opts.map(o => parseInt(o, 10)).filter(o => !isNaN(o));
        }
        this._pageSizeOptions = value;
    }
    get pageSize() {
        return this._pageSize;
    }
    set pageSize(value) {
        const parsedValue = parseInt(`${value}`, 10);
        if (isNaN(parsedValue) || parsedValue < 0) {
            this._pageSize = this._pageSizeOptions[0];
        }
        else {
            this._pageSize = parsedValue;
        }
        const result = this.pageSizeOptions.find(option => option === this._pageSize);
        if (!result) {
            this._pageSizeOptions.push(this._pageSize);
            this._pageSizeOptions.sort((i, j) => i - j);
        }
    }
    isShowingAllRows(selectedLength) {
        return false;
    }
}
