// ontimize-web-ngx/esm2015/lib/components/table/extensions/footer/paginator/o-table-paginator.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, forwardRef, Inject, Injector } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { OTableComponent } from '../../../o-table.component';
import { OBaseTablePaginator } from './o-base-table-paginator.class';
export const DEFAULT_PAGINATOR_TABLE = [
    'pageSize: page-size',
    'pageSizeOptions: page-size-options',
    'showFirstLastButtons: show-first-last-buttons'
];
export class OTablePaginatorComponent extends OBaseTablePaginator {
    constructor(injector, table) {
        super();
        this.injector = injector;
        this.table = table;
        this.showFirstLastButtons = true;
    }
    ngOnInit() {
        this.pageSize = this.table.queryRows;
        this.pageIndex = this.table.currentPage;
        this.showFirstLastButtons = this.table.showPaginatorFirstLastButtons;
        this.table.registerPagination(this);
    }
    get pageIndex() {
        return this._pageIndex;
    }
    set pageIndex(value) {
        this._pageIndex = value;
        if (this.table.matpaginator) {
            this.table.matpaginator.pageIndex = this._pageIndex;
        }
    }
    isShowingAllRows(selectedLength) {
        return false;
    }
}
OTablePaginatorComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-paginator',
                template: ' ',
                inputs: DEFAULT_PAGINATOR_TABLE,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
OTablePaginatorComponent.ctorParameters = () => [
    { type: Injector },
    { type: OTableComponent, decorators: [{ type: Inject, args: [forwardRef(() => OTableComponent),] }] }
];
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTablePaginatorComponent.prototype, "showFirstLastButtons", void 0);
