// ontimize-web-ngx/esm2015/lib/components/table/column/cell-renderer/image/o-table-cell-renderer-image.component.js
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { Util } from '../../../../../util/util';
import { DEFAULT_INPUTS_O_BASE_TABLE_CELL_RENDERER, OBaseTableCellRenderer } from '../o-base-table-cell-renderer.class';
export const DEFAULT_INPUTS_O_TABLE_CELL_RENDERER_IMAGE = [
    ...DEFAULT_INPUTS_O_BASE_TABLE_CELL_RENDERER,
    'imageType: image-type',
    'emptyImage: empty-image',
    'avatar'
];
export const DEFAULT_OUTPUTS_O_TABLE_CELL_RENDERER_IMAGE = [
    'onClick'
];
export class OTableCellRendererImageComponent extends OBaseTableCellRenderer {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.tableColumn.type = 'image';
        this.tableColumn.orderable = false;
        this.tableColumn.searchable = false;
        this.tableColumn.groupable = false;
    }
    initialize() {
        super.initialize();
        if (this.table) {
            const oCol = this.table.getOColumn(this.tableColumn.attr);
            oCol.title = Util.isDefined(this.tableColumn.title) ? this.tableColumn.title : undefined;
            oCol.definition.contentAlign = oCol.definition.contentAlign ? oCol.definition.contentAlign : 'center';
        }
    }
    getSource(cellData) {
        this._source = '';
        switch (this.imageType) {
            case 'base64':
                this._source = cellData ? ('data:image/png;base64,' + ((typeof (cellData.bytes) !== 'undefined') ? cellData.bytes : cellData)) : this.emptyImage;
                break;
            case 'url':
                this._source = cellData ? cellData : this.emptyImage;
                break;
            default:
                this._source = this.emptyImage;
                break;
        }
        return this._source;
    }
}
OTableCellRendererImageComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-cell-renderer-image',
                template: "<ng-template #templateref let-cellvalue=\"cellvalue\">\n  <div [ngClass]=\"{'image-avatar':avatar,'image-plain':!avatar} \">\n    <img [src]=\"getSource(cellvalue)\" alt=\"renderer cell image\">\n  </div>\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_TABLE_CELL_RENDERER_IMAGE
            }] }
];
OTableCellRendererImageComponent.ctorParameters = () => [
    { type: Injector }
];
OTableCellRendererImageComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
};
