// ontimize-web-ngx-report/esm2015/lib/util/o-reports-translate.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/util/o-reports-translate.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe, Injector } from '@angular/core';
import { OTranslateService } from 'ontimize-web-ngx';
import * as CORE_TRANSLATIONS from '../i18n/i18n';
export class OReportsTranslatePipe {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
        this.translateService = this.injector.get(OTranslateService);
    }
    /**
     * @param {?} text
     * @return {?}
     */
    transform(text) {
        /** @type {?} */
        let textTranslated = undefined;
        /** @type {?} */
        let bundle = CORE_TRANSLATIONS.MAP[this.translateService.getCurrentLang()];
        if (bundle && bundle[text]) {
            textTranslated = bundle[text];
        }
        else {
            textTranslated = this.translateService.get(text);
        }
        return textTranslated;
    }
}
OReportsTranslatePipe.decorators = [
    { type: Pipe, args: [{
                name: 'oReportsTranslate',
                pure: false
            },] }
];
/** @nocollapse */
OReportsTranslatePipe.ctorParameters = () => [
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @protected
     */
    OReportsTranslatePipe.prototype.translateService;
    /**
     * @type {?}
     * @protected
     */
    OReportsTranslatePipe.prototype.injector;
}
