// ontimize-web-ngx-report/esm2015/lib/util/utils.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/util/utils.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Constants } from "./constants";
export class Utils {
    /**
     * @param {?} fullscreen
     * @param {?} dialogRef
     * @return {?}
     */
    static setFullscreenDialog(fullscreen, dialogRef) {
        if (!fullscreen) {
            dialogRef.updateSize("100%", "100%");
        }
        else {
            dialogRef.updateSize(Constants.DEFAULT_WIDTH_DIALOG, Constants.DEFAULT_HEIGHT_DIALOG);
        }
    }
    /**
     * @param {?} data
     * @return {?}
     */
    static cloneObject(data) {
        return JSON.parse(JSON.stringify(data));
    }
    /**
     * @param {?} dialog
     * @param {?} component
     * @param {?} data
     * @return {?}
     */
    static openModalVisor(dialog, component, data) {
        dialog.open(component, {
            maxWidth: '100vw',
            maxHeight: '100vh',
            height: Constants.DEFAULT_HEIGHT_DIALOG,
            width: Constants.DEFAULT_WIDTH_DIALOG,
            panelClass: ['o-dialog-class', 'o-table-dialog', 'o-report'],
            data: data
        });
    }
}
