// ontimize-web-ngx/esm2015/lib/services/ontimize/ontimize-file.service.js
import { HttpEventType, HttpHeaders, HttpRequest } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { share } from 'rxjs/operators';
import { Util } from '../../util';
import { OntimizeBaseService } from './ontimize-base-service.class';
export class OntimizeFileService extends OntimizeBaseService {
    constructor() {
        super(...arguments);
        this.path = '';
    }
    configureService(config) {
        super.configureService(config);
        this.path = config.path;
    }
    upload(files, entity, data) {
        const dataObservable = new Observable(observer => {
            const url = `${this.urlBase}${this.path}/${entity}`;
            const toUpload = new FormData();
            files.forEach(item => {
                item.prepareToUpload();
                item.isUploading = true;
                toUpload.append('name', item.name);
                toUpload.append('file', item.file);
            });
            if (data) {
                toUpload.append('data', JSON.stringify(data));
            }
            const request = new HttpRequest('POST', url, toUpload, {
                headers: this.buildHeaders(),
                reportProgress: true
            });
            this.httpClient.request(request).subscribe(resp => {
                if (HttpEventType.UploadProgress === resp.type) {
                    const progressData = {
                        loaded: resp.loaded,
                        total: resp.total
                    };
                    observer.next(progressData);
                }
                else if (HttpEventType.Response === resp.type) {
                    if (resp.body) {
                        if (resp.body['code'] === 3) {
                            this.authService.logout();
                        }
                        else if (resp.body['code'] === 1) {
                            observer.error(resp.body['message']);
                        }
                        else if (resp.body['code'] === 0) {
                            observer.next(resp.body);
                        }
                        else {
                            observer.error('Service unavailable');
                        }
                    }
                    else {
                        observer.next(resp.body);
                    }
                }
            }, error => {
                console.error(error);
                if (error.status === 401) {
                    this.authService.logout();
                }
                else {
                    observer.error(error);
                }
            }, () => observer.complete());
        });
        return dataObservable.pipe(share());
    }
    buildHeaders() {
        let headers = new HttpHeaders({ 'Access-Control-Allow-Origin': '*' });
        const sessionId = this.authService.getSessionInfo().id;
        if (Util.isDefined(sessionId)) {
            headers = headers.append('Authorization', 'Bearer ' + sessionId);
        }
        return headers;
    }
}
OntimizeFileService.decorators = [
    { type: Injectable }
];
