// ontimize-web-ngx/esm2015/lib/services/ontimize/ontimize-export-3xx.service.js
import { HttpHeaders } from '@angular/common/http';
import { Injectable, Injector } from '@angular/core';
import { Observable } from 'rxjs';
import { share } from 'rxjs/operators';
import { AppConfig } from '../../config/app-config';
import { Util } from '../../util';
import { OntimizeExportDataProviderService } from '../ontimize-export-data-provider.service';
import { OntimizeBaseService } from './ontimize-base-service.class';
export class OntimizeExportService3X extends OntimizeBaseService {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.exportPath = this.injector.get(AppConfig).getExportPath();
        this.exportDataProvider = this.injector.get(OntimizeExportDataProviderService);
    }
    configureService(config) {
        super.configureService(config);
        this.servicePath = config.path;
    }
    buildHeaders() {
        let headers = new HttpHeaders({ 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Headers': 'X-Requested-With,content-type' });
        const sessionId = this.authService.getSessionInfo().id;
        if (Util.isDefined(sessionId)) {
            headers = headers.append('Authorization', 'Bearer ' + sessionId);
        }
        return headers;
    }
    exportData(format) {
        const url = `${this.urlBase}${this.exportPath}/${format}`;
        const options = {
            headers: this.buildHeaders().append('Content-Type', 'application/json;charset=UTF-8'),
            observe: 'response',
            responseType: 'blob'
        };
        let paramExport = {
            format: format,
            path: this.servicePath
        };
        let exportData = this.exportDataProvider.getExportConfiguration(paramExport);
        const body = JSON.stringify(exportData);
        const dataObservable = new Observable(observer => {
            this.httpClient.post(url, body, options).subscribe((resp) => {
                const fileData = resp.body;
                const contentDisposition = resp.headers.get('content-disposition');
                let fileName = 'file.' + format;
                const fileNameRegex = /filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/;
                const matches = fileNameRegex.exec(contentDisposition);
                if (matches != null && matches[1]) {
                    fileName = matches[1].replace(/['"]/g, '');
                }
                const fileURL = URL.createObjectURL(fileData);
                const a = document.createElement('a');
                document.body.appendChild(a);
                a.href = fileURL;
                a.download = fileName;
                a.click();
                document.body.removeChild(a);
                observer.next(fileData);
                URL.revokeObjectURL(fileURL);
            }, error => observer.error(error), () => observer.complete());
        });
        return dataObservable.pipe(share());
    }
}
OntimizeExportService3X.decorators = [
    { type: Injectable }
];
OntimizeExportService3X.ctorParameters = () => [
    { type: Injector }
];
