// ontimize-web-ngx-report/esm5/lib/components/report/o-report-viewer/o-report-viewer.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/report/o-report-viewer/o-report-viewer.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject, ViewEncapsulation } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
import { DialogService } from 'ontimize-web-ngx';
import { OReportService } from '../../../services/o-report.service';
import { Utils } from '../../../util/utils';
var OReportViewerComponent = /** @class */ (function () {
    function OReportViewerComponent(dialogRef, reportService, dialogService, data) {
        var _this = this;
        this.dialogRef = dialogRef;
        this.reportService = reportService;
        this.dialogService = dialogService;
        this.data = data;
        this.pdf = '';
        this.name = '';
        this.fullscreen = false;
        this.reportService.configureService(this.reportService.getDefaultServiceConfiguration());
        this.reportService.configureAdapter();
        this.name = this.data.name;
        this.reportService.fillReport(this.data['params'], 'fillReport', {}, this.data['filter']).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            if (res && res.data.length && res.code === 0) {
                _this.pdf = res.data[0].file;
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            if (_this.dialogService) {
                _this.dialogService.error('ERROR', 'SERVER_ERROR_MESSAGE');
            }
            console.log(err);
        }));
    }
    /**
     * @return {?}
     */
    OReportViewerComponent.prototype.setFullscreenDialog = /**
     * @return {?}
     */
    function () {
        Utils.setFullscreenDialog(this.fullscreen, this.dialogRef);
        this.fullscreen = !this.fullscreen;
    };
    OReportViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-report-viewer',
                    template: "<div mat-dialog-title class=\"dialog-title\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n  {{ name }}\n  <div>\n    <mat-icon aria-label=\"fullscreen dialog\" (click)=\"setFullscreenDialog()\">\n      <ng-container *ngIf=\"fullscreen\">fullscreen_exit</ng-container>\n      <ng-container *ngIf=\"!fullscreen\">fullscreen</ng-container>\n    </mat-icon>\n    <mat-icon svgIcon=\"ontimize:close\" [mat-dialog-close]=\"true\"></mat-icon>\n  </div>\n</div>\n<div #dialog mat-dialog-content class=\"dialog-container\">\n  <ngx-extended-pdf-viewer [base64Src]=\"pdf\" [filenameForDownload]=\"name\" [showBookmarkButton]=\"false\" [showPropertiesButton]=\"false\"\n    [showOpenFileButton]=\"false\" useBrowserLocale=\"true\" delayFirstView=\"1000\" [textLayer]=\"true\"></ngx-extended-pdf-viewer>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-report-viewer]': 'true'
                    },
                    styles: [".o-dialog-class .mat-dialog-container .o-report-viewer .mat-dialog-title .mat-icon{cursor:pointer}.o-dialog-class .mat-dialog-container .o-report-viewer .mat-dialog-content{max-height:calc(100% - 36px - 8px);height:100%;padding-top:8px;padding-bottom:0}.o-dialog-class .mat-dialog-container .o-report-viewer ngx-extended-pdf-viewer .zoom{height:100%}"]
                }] }
    ];
    /** @nocollapse */
    OReportViewerComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: OReportService },
        { type: DialogService },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    return OReportViewerComponent;
}());
export { OReportViewerComponent };
if (false) {
    /** @type {?} */
    OReportViewerComponent.prototype.pdf;
    /** @type {?} */
    OReportViewerComponent.prototype.name;
    /** @type {?} */
    OReportViewerComponent.prototype.fullscreen;
    /** @type {?} */
    OReportViewerComponent.prototype.dialogRef;
    /**
     * @type {?}
     * @private
     */
    OReportViewerComponent.prototype.reportService;
    /**
     * @type {?}
     * @protected
     */
    OReportViewerComponent.prototype.dialogService;
    /** @type {?} */
    OReportViewerComponent.prototype.data;
}
