// ontimize-web-ngx/esm2015/lib/components/contextmenu/context-menu/o-context-menu-content.component.js
import { Component, EventEmitter, HostListener, Injector, ViewChild } from '@angular/core';
import { MatMenuTrigger } from '@angular/material';
import { OContextMenuGroupComponent } from '../context-menu-group/o-context-menu-group.component';
export const DEFAULT_CONTEXT_MENU_CONTENT_INPUTS = [
    'menuItems',
    'externalMenuItems',
    'overlay',
    'data',
    'menuClass'
];
export const DEFAULT_CONTEXT_MENU_CONTENT_OUTPUTS = [
    'execute',
    'close'
];
export class OContextMenuContentComponent {
    constructor(injector) {
        this.injector = injector;
        this.execute = new EventEmitter();
        this.close = new EventEmitter();
    }
    click() {
        this.closeContent();
    }
    ngOnInit() {
        this.initialize();
    }
    ngAfterViewInit() {
        this.trigger.openMenu();
    }
    initialize() {
        const menuItemsArray = this.menuItems ? this.menuItems.toArray() : [];
        const externalItemsArray = this.externalMenuItems ? this.externalMenuItems.toArray() : [];
        this.allMenuItems = menuItemsArray.concat(externalItemsArray);
        this.setData(this.allMenuItems);
    }
    setData(items) {
        if (this.data) {
            (items || []).forEach((menuItem) => {
                menuItem.data = this.data;
                if (menuItem instanceof OContextMenuGroupComponent) {
                    this.setData(menuItem.children);
                }
            });
        }
    }
    onMenuClosed(e) {
        this.closeContent();
    }
    closeContent() {
        this.trigger.closeMenu();
        this.close.emit();
    }
}
OContextMenuContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-context-menu-content',
                template: "<button mat-button [matMenuTriggerFor]=\"menu.childMenu\" (menuClosed)=\"onMenuClosed($event)\"></button>\n<o-wrapper-content-menu #menu [items]=\"allMenuItems\" [class]=\"menuClass\"> </o-wrapper-content-menu>",
                inputs: DEFAULT_CONTEXT_MENU_CONTENT_INPUTS,
                outputs: DEFAULT_CONTEXT_MENU_CONTENT_OUTPUTS,
                host: {
                    '[class.o-context-menu-content]': 'true'
                }
            }] }
];
OContextMenuContentComponent.ctorParameters = () => [
    { type: Injector }
];
OContextMenuContentComponent.propDecorators = {
    trigger: [{ type: ViewChild, args: [MatMenuTrigger, { static: false },] }],
    click: [{ type: HostListener, args: ['document:click',] }]
};
