// ontimize-web-ngx-report/esm5/lib/components/report-on-demand/save-preferences-dialog/save-preferences-dialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/report-on-demand/save-preferences-dialog/save-preferences-dialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Injector } from '@angular/core';
import { FormGroup, FormControl, Validators } from '@angular/forms';
import { MatDialogRef } from '@angular/material';
import { OTableBaseDialogClass } from 'ontimize-web-ngx';
var SavePreferencesDialogComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SavePreferencesDialogComponent, _super);
    function SavePreferencesDialogComponent(dialogo, injector) {
        var _this = _super.call(this, injector) || this;
        _this.dialogo = dialogo;
        _this.injector = injector;
        _this.formGroup = new FormGroup({
            name: new FormControl('', [
                Validators.required
            ]),
            description: new FormControl('')
        });
        _this.setFormControl(_this.formGroup.get('name'));
        return _this;
    }
    /**
     * @return {?}
     */
    SavePreferencesDialogComponent.prototype.save = /**
     * @return {?}
     */
    function () {
        this.dialogo.close({ "name": this.name, "description": this.description });
    };
    SavePreferencesDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-save-preferences-dialog',
                    template: "<span mat-dialog-title>{{ 'TABLE.BUTTONS.SAVE_CONFIGURATION' | oTranslate }}</span>\n\n<mat-dialog-content>\n  <div mat-subheader>{{ 'TABLE.DIALOG.SAVE_CONFIGURATION' | oTranslate }}</div>\n  <form #form [formGroup]=\"formGroup\" fxLayout=\"column\">\n    <mat-form-field>\n      <input [(ngModel)]=\"name\" matInput [matTooltip]=\"tooltipText\" [matTooltipClass]=\"tooltipClass\"\n        placeholder=\"{{ 'TABLE.DIALOG.CONFIGURATION_NAME' | oTranslate }}\" formControlName=\"name\" required>\n      <mat-error *oMatError=\"hasError(formGroup.controls['name'], 'required')\">{{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}</mat-error>\n    </mat-form-field>\n\n    <mat-form-field>\n      <textarea [(ngModel)]=\"description\" matInput placeholder=\"{{ 'TABLE.DIALOG.CONFIGURATION_DESCRIPTION' | oTranslate }}\"\n        formControlName=\"description\" rows=\"4\" cols=\"50\"></textarea>\n    </mat-form-field>\n\n  </form>\n</mat-dialog-content>\n\n<mat-dialog-actions align=\"end\">\n  <button type=\"button\" mat-stroked-button class=\"mat-primary\" [mat-dialog-close]=\"false\" class=\"o-button-default\">\n    {{ 'CANCEL' | oTranslate }}</button>\n  <button type=\"button\" mat-stroked-button class=\"mat-primary\" (click)=\"save()\" [mat-dialog-close]=\"true\" [disabled]=\"formGroup.invalid\">\n    {{ 'SAVE' | oTranslate }}</button>\n</mat-dialog-actions>\n"
                }] }
    ];
    /** @nocollapse */
    SavePreferencesDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: Injector }
    ]; };
    return SavePreferencesDialogComponent;
}(OTableBaseDialogClass));
export { SavePreferencesDialogComponent };
if (false) {
    /** @type {?} */
    SavePreferencesDialogComponent.prototype.name;
    /** @type {?} */
    SavePreferencesDialogComponent.prototype.description;
    /** @type {?} */
    SavePreferencesDialogComponent.prototype.formGroup;
    /** @type {?} */
    SavePreferencesDialogComponent.prototype.dialogo;
    /**
     * @type {?}
     * @protected
     */
    SavePreferencesDialogComponent.prototype.injector;
}
