// ontimize-web-ngx/esm2015/lib/components/input/listpicker/listpicker-renderer/integer/o-list-picker-renderer-integer.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { OIntegerPipe } from '../../../../../pipes/o-integer.pipe';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER, OListPickerCustomRenderer } from '../o-list-picker-renderer.class';
export const DEFAULT_INPUTS_O_LISTPICKER_RENDERER_INTEGER = [
    ...DEFAULT_INPUTS_O_LISTPICKER_RENDERER,
    'grouping',
    'thousandSeparator: thousand-separator'
];
export class OListPickerRendererIntegerComponent extends OListPickerCustomRenderer {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.grouping = true;
        this.thousandSeparator = ',';
        this.setComponentPipe();
    }
    setComponentPipe() {
        this.componentPipe = new OIntegerPipe(this.injector);
    }
    initialize() {
        super.initialize();
        this.pipeArguments = {
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator
        };
    }
}
OListPickerRendererIntegerComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-list-picker-renderer-integer',
                template: "<ng-template #templateref let-value=\"value\">\n  {{getListPickerValue(value)}}\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_LISTPICKER_RENDERER_INTEGER
            }] }
];
OListPickerRendererIntegerComponent.ctorParameters = () => [
    { type: Injector }
];
OListPickerRendererIntegerComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
};
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OListPickerRendererIntegerComponent.prototype, "grouping", void 0);
