// ontimize-web-ngx/esm2015/lib/components/o-data-toolbar/o-data-toolbar.component.js
import * as tslib_1 from "tslib";
import { Component, ChangeDetectionStrategy, ViewEncapsulation } from '@angular/core';
import { InputConverter } from '../../decorators/input-converter';
export const DEFAULT_INPUTS_O_DATA_TOOLBAR = [
    'showTitle: show-title',
    'title',
];
export class ODataToolbarComponent {
    constructor() {
        this.showTitle = false;
    }
}
ODataToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-data-toolbar',
                template: "<ng-content select=\"[o-data-toolbar-projection-start]\"></ng-content>\n<ng-content select=\"[o-data-toolbar-custom-projection-start]\"></ng-content>\n<div fxFlex>\n  <span fxLayoutAlign=\"center center\" *ngIf=\"showTitle\" class=\"title\">{{ title | oTranslate }}</span>\n</div>\n<ng-content select=\"[o-data-toolbar-custom-projection-end]\"></ng-content>\n<ng-content select=\"[o-data-toolbar-projection-end]\"></ng-content>\n",
                inputs: DEFAULT_INPUTS_O_DATA_TOOLBAR,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                host: { class: 'o-data-toolbar' },
                styles: [".o-data-toolbar{padding:0 4px;height:64px;display:flex;align-items:center;align-content:space-between;width:100%;flex-direction:row;white-space:nowrap}.o-data-toolbar.dense{height:48px}.o-data-toolbar .title{font-size:18px;font-weight:400;text-align:center}"]
            }] }
];
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], ODataToolbarComponent.prototype, "showTitle", void 0);
