// ontimize-web-ngx/esm2015/lib/components/list/renderers/avatar/o-list-item-avatar.component.js
import { Component, ElementRef, forwardRef, Inject, Injector, Optional, Renderer2, ViewEncapsulation, } from '@angular/core';
import { OListItemComponent } from '../../list-item/o-list-item.component';
import { DEFAULT_INPUTS_O_TEXT_RENDERER, DEFAULT_OUTPUTS_O_TEXT_RENDERER, OListItemTextRenderer, } from '../o-list-item-text-renderer.class';
export const DEFAULT_INPUTS_O_LIST_ITEM_AVATAR = [
    ...DEFAULT_INPUTS_O_TEXT_RENDERER,
    'avatar',
    'emptyAvatar: empty-avatar',
    'avatarType: avatar-type'
];
export const DEFAULT_OUTPUTS_O_LIST_ITEM_AVATAR = [
    ...DEFAULT_OUTPUTS_O_TEXT_RENDERER
];
export class OListItemAvatarComponent extends OListItemTextRenderer {
    constructor(elRef, _renderer, _injector, _listItem) {
        super(elRef, _renderer, _injector, _listItem);
        this._listItem = _listItem;
    }
    ngAfterViewInit() {
        this.modifyMatListItemElement();
    }
    ngOnInit() {
        let avatarValue = this.avatar;
        if (!this.avatar) {
            avatarValue = this.emptyAvatar;
        }
        else {
            switch (this.avatarType) {
                case 'base64':
                    avatarValue = ('data:image/png;base64,' + ((typeof (avatarValue.bytes) !== 'undefined') ? avatarValue.bytes : avatarValue));
                    break;
                case 'url':
                default:
                    avatarValue = this.avatar;
                    break;
            }
        }
        this.avatarSrc = avatarValue;
    }
}
OListItemAvatarComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-list-item-avatar',
                template: "<div fxLayout=\"row\" fxLayoutAlign=\"center center\">\n  <img matListAvatar class=\"avatar\" [src]=\"avatarSrc | oSafe:'resourceUrl'\" [alt]=\"title\">\n  <div class=\"mat-list-text\">\n    <h3 matLine class=\"primary-text\"> {{ title }} </h3>\n    <h4 *ngIf=\"primaryText !== undefined\" matLine class=\"primary-text\">{{ primaryText }}</h4>\n    <p *ngIf=\"secondaryText !== undefined\" matLine class=\"secondary-text\">{{ secondaryText }}</p>\n  </div>\n  <mat-icon *ngIf=\"icon !== undefined\" class=\"material-icons o-list-item-icon\" (click)=\"onActionIconClick($event)\">{{ icon\n    }}\n  </mat-icon>\n</div>\n",
                inputs: DEFAULT_INPUTS_O_LIST_ITEM_AVATAR,
                outputs: DEFAULT_OUTPUTS_O_LIST_ITEM_AVATAR,
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.o-custom-list-item]': 'true',
                    '[class.o-list-item-avatar]': 'true'
                },
                styles: [".o-list-row-action+.o-list-item-avatar{padding:0 0 0 10px}.mat-list.selectable .o-list-item-avatar .mat-list-text{padding-right:0}.mat-list .mat-list-item .mat-list-item-content .o-list-item-avatar.o-custom-list-item .mat-list-text,.mat-nav-list .mat-list-item .mat-list-item-content .o-list-item-avatar.o-custom-list-item .mat-list-text{padding:0 16px}"]
            }] }
];
OListItemAvatarComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: Injector },
    { type: OListItemComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(() => OListItemComponent),] }] }
];
