// ontimize-web-ngx/esm5/lib/components/table/column/cell-renderer/image/o-table-cell-renderer-image.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { Util } from '../../../../../util/util';
import { DEFAULT_INPUTS_O_BASE_TABLE_CELL_RENDERER, OBaseTableCellRenderer } from '../o-base-table-cell-renderer.class';
export var DEFAULT_INPUTS_O_TABLE_CELL_RENDERER_IMAGE = tslib_1.__spread(DEFAULT_INPUTS_O_BASE_TABLE_CELL_RENDERER, [
    'imageType: image-type',
    'emptyImage: empty-image',
    'avatar'
]);
export var DEFAULT_OUTPUTS_O_TABLE_CELL_RENDERER_IMAGE = [
    'onClick'
];
var OTableCellRendererImageComponent = (function (_super) {
    tslib_1.__extends(OTableCellRendererImageComponent, _super);
    function OTableCellRendererImageComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.tableColumn.type = 'image';
        _this.tableColumn.orderable = false;
        _this.tableColumn.searchable = false;
        _this.tableColumn.groupable = false;
        return _this;
    }
    OTableCellRendererImageComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        if (this.table) {
            var oCol = this.table.getOColumn(this.tableColumn.attr);
            oCol.title = Util.isDefined(this.tableColumn.title) ? this.tableColumn.title : undefined;
            oCol.definition.contentAlign = oCol.definition.contentAlign ? oCol.definition.contentAlign : 'center';
        }
    };
    OTableCellRendererImageComponent.prototype.getSource = function (cellData) {
        this._source = '';
        switch (this.imageType) {
            case 'base64':
                this._source = cellData ? ('data:image/png;base64,' + ((typeof (cellData.bytes) !== 'undefined') ? cellData.bytes : cellData)) : this.emptyImage;
                break;
            case 'url':
                this._source = cellData ? cellData : this.emptyImage;
                break;
            default:
                this._source = this.emptyImage;
                break;
        }
        return this._source;
    };
    OTableCellRendererImageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table-cell-renderer-image',
                    template: "<ng-template #templateref let-cellvalue=\"cellvalue\">\n  <div [ngClass]=\"{'image-avatar':avatar,'image-plain':!avatar} \">\n    <img [src]=\"getSource(cellvalue)\" alt=\"renderer cell image\">\n  </div>\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_TABLE_CELL_RENDERER_IMAGE
                }] }
    ];
    OTableCellRendererImageComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OTableCellRendererImageComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
    };
    return OTableCellRendererImageComponent;
}(OBaseTableCellRenderer));
export { OTableCellRendererImageComponent };
