// ontimize-web-ngx/esm5/lib/components/table/extensions/header/table-header/o-table-header.component.js
import { ChangeDetectionStrategy, Component, forwardRef, Inject, ViewChild, ViewEncapsulation } from "@angular/core";
import { OTableComponent } from "../../../o-table.component";
import { OMatSortHeader } from "../../sort/o-mat-sort-header";
import { OTableHeaderColumnFilterIconComponent } from "../table-header-column-filter-icon/o-table-header-column-filter-icon.component";
export var DEFAULT_INPUTS_O_TABLE_HEADER = [
    'column'
];
var OTableHeaderComponent = (function () {
    function OTableHeaderComponent(table) {
        this.table = table;
        this.resizable = this.table.resizable;
    }
    Object.defineProperty(OTableHeaderComponent.prototype, "columnFilterIcon", {
        set: function (value) {
            this._columnFilterIcon = value;
        },
        enumerable: true,
        configurable: true
    });
    OTableHeaderComponent.prototype.isModeColumnFilterable = function (column) {
        return this.table.isColumnFiltersActive && this.table.isColumnFilterable(column);
    };
    OTableHeaderComponent.prototype.setFilterIconHintVisible = function (visible) {
        if (this._columnFilterIcon) {
            this._columnFilterIcon.filterIconHintVisible.next(visible);
        }
    };
    OTableHeaderComponent.prototype.ngAfterViewInit = function () {
        this.table.registerTableHeaders(this);
    };
    OTableHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table-header',
                    inputs: DEFAULT_INPUTS_O_TABLE_HEADER,
                    template: "<o-table-header-column-filter-icon #columnFilterIcon *ngIf=\"isModeColumnFilterable(column)\" [column]=\"column\">\n</o-table-header-column-filter-icon>\n\n<ng-container *ngIf=\"column.orderable\">\n    <span o-mat-sort-header>{{ column.title | oTranslate }}</span>\n</ng-container>\n<ng-container *ngIf=\"!column.orderable\">\n    <span class=\"header-title-container\">{{ column.title | oTranslate }}</span>\n</ng-container>\n\n<o-table-column-resizer *ngIf=\"resizable\" [column]=\"column\"></o-table-column-resizer>",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class.o-table-header]': 'true',
                        '(mouseenter)': 'setFilterIconHintVisible(true)',
                        '(mouseleave)': 'setFilterIconHintVisible(false)'
                    }
                }] }
    ];
    OTableHeaderComponent.ctorParameters = function () { return [
        { type: OTableComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OTableComponent; }),] }] }
    ]; };
    OTableHeaderComponent.propDecorators = {
        columnFilterIcon: [{ type: ViewChild, args: ['columnFilterIcon', { static: false },] }],
        matSortHeader: [{ type: ViewChild, args: [OMatSortHeader, { static: false },] }]
    };
    return OTableHeaderComponent;
}());
export { OTableHeaderComponent };
