// ontimize-web-ngx/esm5/lib/components/table/column/cell-editor/email/o-table-cell-editor-email.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import { DEFAULT_INPUTS_O_TABLE_CELL_EDITOR, DEFAULT_OUTPUTS_O_TABLE_CELL_EDITOR, OBaseTableCellEditor } from '../o-base-table-cell-editor.class';
export var DEFAULT_INPUTS_O_TABLE_CELL_EDITOR_EMAIL = tslib_1.__spread(DEFAULT_INPUTS_O_TABLE_CELL_EDITOR);
export var DEFAULT_OUTPUTS_O_TABLE_CELL_EDITOR_EMAIL = tslib_1.__spread(DEFAULT_OUTPUTS_O_TABLE_CELL_EDITOR);
var OTableCellEditorEmailComponent = (function (_super) {
    tslib_1.__extends(OTableCellEditorEmailComponent, _super);
    function OTableCellEditorEmailComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        return _this;
    }
    OTableCellEditorEmailComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-table-cell-editor-email',
                    template: "<ng-template #templateref let-cellvalue=\"cellvalue\" let-rowvalue=\"rowvalue\">\n  <div class=\"o-table-cell-editor-email o-table-cell-editor\" [formGroup]=\"formGroup\">\n    <mat-form-field floatLabel=\"never\" cdkFocusInitial>\n      <input #input matInput type=\"email\" [email]=\"true\" [placeholder]=\"getPlaceholder()\" [formControl]=\"formControl\" [id]=\"cellEditorId\"\n        [required]=\"orequired\">\n      <mat-error *oMatError=\"hasError('required')\">\n        {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n      </mat-error>\n      <mat-error *oMatError=\"hasError('email')\">\n        {{ 'FORM_VALIDATION.EMAIL_FORMAT' | oTranslate }}\n      </mat-error>\n      <mat-error *ngFor=\"let oError of getActiveOErrors()\">\n        {{ getErrorText(oError) }}\n      </mat-error>\n    </mat-form-field>\n  </div>\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_TABLE_CELL_EDITOR_EMAIL,
                    outputs: DEFAULT_OUTPUTS_O_TABLE_CELL_EDITOR_EMAIL,
                    encapsulation: ViewEncapsulation.None,
                    styles: [".o-table-cell-editor-text .mat-form-field{width:100%}"]
                }] }
    ];
    OTableCellEditorEmailComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OTableCellEditorEmailComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
    };
    return OTableCellEditorEmailComponent;
}(OBaseTableCellEditor));
export { OTableCellEditorEmailComponent };
