// ontimize-web-ngx/esm2015/lib/components/input/listpicker/listpicker-renderer/date/o-list-picker-renderer-date.component.js
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { OMomentPipe } from '../../../../../pipes/o-moment.pipe';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER, OListPickerCustomRenderer } from '../o-list-picker-renderer.class';
export const DEFAULT_INPUTS_O_LISTPICKER_RENDERER_DATE = [
    ...DEFAULT_INPUTS_O_LISTPICKER_RENDERER,
    'format'
];
export class OListPickerRendererDateComponent extends OListPickerCustomRenderer {
    constructor(injector) {
        super(injector);
        this.injector = injector;
        this.setComponentPipe();
    }
    setComponentPipe() {
        this.componentPipe = new OMomentPipe(this.injector);
    }
    initialize() {
        super.initialize();
        this.pipeArguments = {
            format: this.format
        };
    }
}
OListPickerRendererDateComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-list-picker-renderer-date',
                template: "<ng-template #templateref let-value=\"value\">\n  {{getListPickerValue(value)}}\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_LISTPICKER_RENDERER_DATE
            }] }
];
OListPickerRendererDateComponent.ctorParameters = () => [
    { type: Injector }
];
OListPickerRendererDateComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
};
