// ontimize-web-ngx/esm2015/lib/components/contextmenu/context-menu-group/o-context-menu-group.component.js
import { Component, ContentChildren, forwardRef, QueryList } from '@angular/core';
import { Subscription } from 'rxjs';
import { DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS, OComponentMenuBaseItem } from '../o-content-menu-base-item.class';
export const DEFAULT_CONTEXT_MENU_GROUP_INPUTS = [
    ...DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS
];
export class OContextMenuGroupComponent extends OComponentMenuBaseItem {
    constructor() {
        super(...arguments);
        this.type = OComponentMenuBaseItem.TYPE_GROUP_MENU;
        this.children = [];
        this.subscription = new Subscription();
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    ngAfterContentInit() {
        this.subscription.add(this.oContextMenuItems.changes.subscribe(() => {
            this.updateChildren();
        }));
        this.updateChildren();
    }
    updateChildren() {
        this.children = this.oContextMenuItems.toArray().slice(1, this.oContextMenuItems.toArray().length);
    }
}
OContextMenuGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-context-menu-group',
                template: ' ',
                inputs: DEFAULT_CONTEXT_MENU_GROUP_INPUTS,
                providers: [{ provide: OComponentMenuBaseItem, useExisting: forwardRef(() => OContextMenuGroupComponent) }]
            }] }
];
OContextMenuGroupComponent.propDecorators = {
    oContextMenuItems: [{ type: ContentChildren, args: [OComponentMenuBaseItem,] }]
};
