// ontimize-web-ngx/esm2015/lib/components/table/extensions/dialog/o-table-base-dialog.class.js
import { QueryList, ViewChildren } from '@angular/core';
import { OMatErrorDirective } from '../../../../directives/o-mat-error.directive';
import { O_MAT_ERROR_OPTIONS } from '../../../../services/factories';
import { Codes } from '../../../../util/codes';
import { Util } from '../../../../util/util';
export class OTableBaseDialogClass {
    constructor(injector) {
        this.injector = injector;
        try {
            this.errorOptions = this.injector.get(O_MAT_ERROR_OPTIONS) || {};
        }
        catch (e) {
            this.errorOptions = {};
        }
    }
    setFormControl(formControl) {
        this.formControl = formControl;
    }
    get tooltipClass() {
        let result;
        const liteError = this.errorOptions.type === Codes.O_MAT_ERROR_LITE;
        if (liteError && Util.isDefined(this.formControl) && this.formControlHasErrors()) {
            result = `o-tooltip o-mat-error`;
        }
        return result;
    }
    get tooltipText() {
        let result;
        const liteError = this.errorOptions.type === Codes.O_MAT_ERROR_LITE;
        if (liteError && this.formControlHasErrors() && this.oMatErrorChildren && this.oMatErrorChildren.length > 0) {
            result = this.oMatErrorChildren
                .filter((oMatError) => Util.isDefined(oMatError.text))
                .map((oMatError) => oMatError.text)
                .join('\n');
        }
        return result;
    }
    formControlHasErrors() {
        return Util.isDefined(this.formControl) && this.formControl.touched && Util.isDefined(this.formControl.errors);
    }
    hasError(fControl, error) {
        return fControl && fControl.touched && fControl.hasError(error);
    }
}
OTableBaseDialogClass.propDecorators = {
    oMatErrorChildren: [{ type: ViewChildren, args: [OMatErrorDirective,] }]
};
