// ontimize-web-ngx/esm2015/lib/components/contextmenu/context-menu-item/o-context-menu-item.component.js
import { Component, EventEmitter, forwardRef } from '@angular/core';
import { DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS, OComponentMenuBaseItem } from '../o-content-menu-base-item.class';
export const DEFAULT_CONTEXT_MENU_ITEM_OUTPUTS = [
    'execute'
];
export const DEFAULT_INPUTS_O_CONTEXT_MENU_ITEM = [
    ...DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS
];
export class OContextMenuItemComponent extends OComponentMenuBaseItem {
    constructor() {
        super(...arguments);
        this.execute = new EventEmitter();
        this.type = OComponentMenuBaseItem.TYPE_ITEM_MENU;
    }
    onClick(event) {
        event.preventDefault();
        event.stopPropagation();
        this.triggerExecute(this.data, event);
    }
    triggerExecute(data, $event) {
        if (!this.enabled) {
            return;
        }
        this.execute.emit({ event: $event, data: data });
    }
}
OContextMenuItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-context-menu-item',
                template: ' ',
                inputs: DEFAULT_INPUTS_O_CONTEXT_MENU_ITEM,
                outputs: DEFAULT_CONTEXT_MENU_ITEM_OUTPUTS,
                providers: [{ provide: OComponentMenuBaseItem, useExisting: forwardRef(() => OContextMenuItemComponent) }]
            }] }
];
