// ontimize-web-ngx/esm5/lib/components/contextmenu/context-menu/o-wrapper-content-menu/o-wrapper-content-menu.component.js
import { Component, Injector, Input, ViewChild } from '@angular/core';
import { MatMenu } from '@angular/material/menu';
import { OContextMenuGroupComponent } from '../../context-menu-group/o-context-menu-group.component';
import { OContextMenuItemComponent } from '../../context-menu-item/o-context-menu-item.component';
import { OContextMenuSeparatorComponent } from '../../context-menu-separator/o-context-menu-separator.component';
export var DEFAULT_CONTEXT_MENU_CONTENT_ITEM_INPUTS = [
    'items',
    'class'
];
var OWrapperContentMenuComponent = (function () {
    function OWrapperContentMenuComponent(injector) {
        this.injector = injector;
    }
    OWrapperContentMenuComponent.prototype.onClick = function (item, event) {
        if (item instanceof OContextMenuItemComponent) {
            item.triggerExecute(item.data, event);
        }
    };
    OWrapperContentMenuComponent.prototype.isGroup = function (item) {
        return item instanceof OContextMenuGroupComponent;
    };
    OWrapperContentMenuComponent.prototype.isSeparator = function (item) {
        return item instanceof OContextMenuSeparatorComponent;
    };
    OWrapperContentMenuComponent.prototype.isItem = function (item) {
        return item instanceof OContextMenuItemComponent;
    };
    OWrapperContentMenuComponent.prototype.getChildren = function (item) {
        return item instanceof OContextMenuGroupComponent ? item.children : [];
    };
    OWrapperContentMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-wrapper-content-menu',
                    template: "<mat-menu #childMenu=\"matMenu\" [class]=\"class\" [overlapTrigger]=\"false\" [hasBackdrop]=\"false\">\n  <ng-container *ngFor=\"let child of items\">\n    <!-- Handle branch node menu items -->\n    <span *ngIf=\"isGroup(child) && child.isVisible\">\n      <button mat-menu-item color=\"primary\" [matMenuTriggerFor]=\"menu.childMenu\" [disabled]=\"child.disabled\">\n        <mat-icon *ngIf=\"child.svgIcon != null\" [svgIcon]=\"child.svgIcon\"></mat-icon>\n        <mat-icon *ngIf=\"child.svgIcon == null && child.icon != null\">{{ child.icon }}</mat-icon>\n        <span>{{ child.label | oTranslate}}</span>\n      </button>\n      <o-wrapper-content-menu #menu [items]=\"getChildren(child)\" [class]=\"class\"></o-wrapper-content-menu>\n    </span>\n    <!-- Handle leaf node menu items -->\n    <button mat-menu-item (click)=\"onClick(child, $event)\" [disabled]=\"child.disabled\" *ngIf=\"isItem(child) && child.isVisible\">\n      <mat-icon *ngIf=\"child.svgIcon != null\" [svgIcon]=\"child.svgIcon\"></mat-icon>\n      <mat-icon *ngIf=\"child.svgIcon == null && child.icon != null\">{{ child.icon }}</mat-icon>\n      <span>{{ child.label | oTranslate}} </span>\n    </button>\n    <!-- separator leaf node menu items -->\n    <mat-divider *ngIf=\"isSeparator(child) && child.isVisible\"></mat-divider>\n  </ng-container>\n</mat-menu>\n",
                    inputs: DEFAULT_CONTEXT_MENU_CONTENT_ITEM_INPUTS,
                    styles: ["mat-divider.mat-divider:first-child,mat-divider.mat-divider:last-child{display:none}.mat-icon{display:inline-flex}"]
                }] }
    ];
    OWrapperContentMenuComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OWrapperContentMenuComponent.propDecorators = {
        items: [{ type: Input }],
        childMenu: [{ type: ViewChild, args: ['childMenu', { static: true },] }],
        menu: [{ type: ViewChild, args: [OWrapperContentMenuComponent, { static: true },] }]
    };
    return OWrapperContentMenuComponent;
}());
export { OWrapperContentMenuComponent };
