// ontimize-web-ngx/esm2015/lib/components/user-info/o-user-info.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, Injector, ViewEncapsulation } from '@angular/core';
import { Router } from '@angular/router';
import { InputConverter } from '../../decorators';
import { AuthService } from '../../services/auth.service';
import { DialogService } from '../../services/dialog.service';
import { OUserInfoService } from '../../services/o-user-info.service';
export const DEFAULT_INPUTS_O_USER_INFO = [
    'showProfile: show-profile',
    'showSettings: show-settings',
    'showLogout: show-logout'
];
export const DEFAULT_OUTPUTS_O_USER_INFO = [];
export class OUserInfoComponent {
    constructor(elRef, injector, router) {
        this.elRef = elRef;
        this.injector = injector;
        this.router = router;
        this.showLogout = true;
        this.showSettings = true;
        this.showProfile = false;
        this.dialogService = this.injector.get(DialogService);
        this.authService = this.injector.get(AuthService);
        this.oUserInfoService = this.injector.get(OUserInfoService);
        this.userInfo = this.oUserInfoService.getUserInfo();
        this.userInfoSubscription = this.oUserInfoService.getUserInfoObservable().subscribe(res => {
            this.userInfo = res;
        });
    }
    ngOnDestroy() {
        this.userInfoSubscription.unsubscribe();
    }
    onLogoutClick() {
        this.authService.logoutWithConfirmation();
    }
    onSettingsClick() {
        this.router.navigate(['main/settings']);
    }
    onProfileClick() {
        this.router.navigate(['main/profile']);
    }
    get existsUserInfo() {
        return this.userInfo !== undefined;
    }
    get avatar() {
        return this.userInfo ? this.userInfo.avatar : undefined;
    }
    get username() {
        return this.userInfo ? this.userInfo.username : undefined;
    }
    registerUserInfoConfiguration(userInfoMenu) {
        this.userInfoConfiguration = userInfoMenu;
        this.updateInputsByConfiguration();
    }
    updateInputsByConfiguration() {
        this.showLogout = this.userInfoConfiguration ? this.userInfoConfiguration.showLogout : this.showLogout;
        this.showProfile = this.userInfoConfiguration ? this.userInfoConfiguration.showProfile : this.showProfile;
        this.showSettings = this.userInfoConfiguration ? this.userInfoConfiguration.showSettings : this.showSettings;
    }
}
OUserInfoComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-user-info',
                inputs: DEFAULT_INPUTS_O_USER_INFO,
                outputs: DEFAULT_OUTPUTS_O_USER_INFO,
                template: "<div class=\"o-user-info-container\" fxLayout=\"row\" fxLayoutAlign=\"center\">\n  <ng-container *ngIf=\"existsUserInfo\">\n    <div class=\"o-user-info-wrapper\" fxLayout=\"row\" fxLayoutAlign=\"center center\" [matMenuTriggerFor]=\"menu\"\n      fxLayoutGap=\"16px\">\n      <div fxLayout=\"row\" fxLayoutAlign=\"center center\" class=\"o-user-info-avatar-wrapper\">\n        <img [src]=\"avatar\" *ngIf=\"avatar\" alt=\"avatar image\"/>\n      </div>\n      <span class=\"o-user-info-username\" *ngIf=\"username\"> {{ username }}</span>\n      <mat-icon svgIcon=\"ontimize:keyboard_arrow_down\"></mat-icon>\n    </div>\n    <mat-menu #menu=\"matMenu\" yPosition=\"below\" class=\"o-mat-menu\">\n      <button type=\"button\" mat-menu-item (click)=\"onProfileClick()\" *ngIf=\"showProfile\">\n        <mat-icon>person</mat-icon>\n        <span>{{ 'APP_LAYOUT.USER_PROFILE' | oTranslate }}</span>\n      </button>\n      <button type=\"button\" mat-menu-item (click)=\"onSettingsClick()\" *ngIf=\"showSettings\">\n        <mat-icon svgIcon=\"ontimize:settings\"></mat-icon>\n        <span>{{ 'APP_LAYOUT.SETTINGS' | oTranslate }}</span>\n      </button>\n      <ng-container *ngIf=\"userInfoConfiguration\">\n        <button *ngFor=\"let item of userInfoConfiguration.userInfoConfigurationItems\" type=\"button\" mat-menu-item\n          [matTooltip]=\"item.name | oTranslate\" matTooltipDisabled=\"item.tooltip\" [ngClass]=\"item.class\"\n          (click)=\"item.triggerOnClick($event)\">\n          <mat-icon>{{item.icon}} </mat-icon>\n          <span>{{ item.name | oTranslate }}</span>\n        </button>\n      </ng-container>\n      <button type=\"button\" mat-menu-item (click)=\"onLogoutClick()\" *ngIf=\"showLogout\">\n        <mat-icon svgIcon=\"ontimize:power_settings_new\"></mat-icon>\n        <span>{{ 'LOGOUT' | oTranslate }}</span>\n      </button>\n\n    </mat-menu>\n  </ng-container>\n</div>",
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.o-user-info]': 'true'
                },
                styles: [".o-user-info .o-user-info-container{cursor:pointer;height:100%}.o-user-info .o-user-info-avatar-wrapper{width:30px;height:30px;overflow:hidden;border-radius:100%}.o-user-info .o-user-info-avatar-wrapper img{max-width:100%;height:auto}.o-user-info mat-icon{font-size:16px;line-height:26px;height:100%}"]
            }] }
];
OUserInfoComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Injector },
    { type: Router }
];
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OUserInfoComponent.prototype, "showLogout", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OUserInfoComponent.prototype, "showSettings", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OUserInfoComponent.prototype, "showProfile", void 0);
