// ontimize-web-ngx/esm5/lib/components/dual-list-selector/dual-list-selector-item/date/o-dual-list-selector-date-item.component.js
import { Component, ChangeDetectionStrategy } from "@angular/core";
export var DEFAULT_DUAL_LIST_SELECTOR_DATE_ITEM = [
    'item',
    'groupedDateColumns: grouped-date-columns',
];
var ODualListSelectorDateItemComponent = (function () {
    function ODualListSelectorDateItemComponent() {
        this.dateTypes = [
            { value: 'MONTH', viewValue: 'DUAL_LIST_SELECTOR.GROUP_BY_MONTH' },
            { value: 'YEAR_MONTH', viewValue: 'DUAL_LIST_SELECTOR.GROUP_BY_YEAR_MONTH' },
            { value: 'YEAR', viewValue: 'DUAL_LIST_SELECTOR.GROUP_BY_YEAR' },
            { value: 'YEAR_MONTH_DAY', viewValue: 'DUAL_LIST_SELECTOR.GROUP_BY_YEAR_MONTH_DAY' }
        ];
        this.item = "";
    }
    ODualListSelectorDateItemComponent.prototype.onSelectionChange = function (event, itemSelected) {
        var value = event.value;
        var attr = itemSelected;
        var index = this.groupedDateColumns.findIndex(function (column) { return column.attr == attr; });
        if (index != -1) {
            this.groupedDateColumns.splice(index, 1);
        }
        this.groupedDateColumns.push({ "attr": attr, "type": value });
    };
    ODualListSelectorDateItemComponent.prototype.getSelectValue = function () {
        var _this = this;
        var index = this.groupedDateColumns.findIndex(function (column) { return column.attr == _this.item; });
        return index != -1 ? this.groupedDateColumns[index].type : 'YEAR_MONTH_DAY';
    };
    ODualListSelectorDateItemComponent.prototype.getViewValue = function () {
        var value = this.getSelectValue();
        var indexFindValue = this.dateTypes.findIndex(function (type) { return type.value == value; });
        return indexFindValue > -1 ? this.dateTypes[indexFindValue].viewValue : 'DUAL_LIST_SELECTOR.GROUP_BY_YEAR_MONTH_DAY';
    };
    ODualListSelectorDateItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-dual-list-selector-date-item',
                    template: "<mat-select [value]=\"getSelectValue()\" (selectionChange)=\"onSelectionChange($event,item)\" required>\n  <mat-select-trigger>\n    {{item | oTranslate}} ({{getViewValue() | oTranslate}})\n  </mat-select-trigger>\n  <mat-option *ngFor=\"let type of dateTypes\" [value]=\"type.value\">\n    {{type.viewValue | oTranslate}}\n  </mat-option>\n</mat-select>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_DUAL_LIST_SELECTOR_DATE_ITEM
                }] }
    ];
    return ODualListSelectorDateItemComponent;
}());
export { ODualListSelectorDateItemComponent };
