// ontimize-web-ngx/esm2015/lib/components/table/column/cell-editor/time/o-table-cell-editor-time.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, HostListener, Injector, TemplateRef, ViewChild, ViewEncapsulation, } from '@angular/core';
import { FormControl } from '@angular/forms';
import { DateAdapter, MAT_DATE_LOCALE, MatDatepickerInput } from '@angular/material';
import { MomentDateAdapter } from '@angular/material-moment-adapter';
import moment from 'moment';
import { NgxMaterialTimepickerComponent } from 'ngx-material-timepicker';
import { InputConverter } from '../../../../../decorators/input-converter';
import { MomentService } from '../../../../../services/moment.service';
import { Codes } from '../../../../../util/codes';
import { Util } from '../../../../../util/util';
import { DEFAULT_INPUTS_O_TABLE_CELL_EDITOR, DEFAULT_OUTPUTS_O_TABLE_CELL_EDITOR, OBaseTableCellEditor, } from '../o-base-table-cell-editor.class';
export const DEFAULT_INPUTS_O_TABLE_CELL_EDITOR_TIME = [
    ...DEFAULT_INPUTS_O_TABLE_CELL_EDITOR,
    'oDateFormat: date-format',
    'oDateLocale: date-locale',
    'oDateStartView: date-start-view',
    'oMinDate: date-min',
    'oMaxDate: date-max',
    'oDateTouchUi: date-touch-ui',
    'oDateStartAt: date-start-at',
    'oHourFormat: hour-format',
    'oHourMin: hour-min',
    'oHourMax: hour-max',
    'oHourPlaceholder: hour-placeholder',
    'oDatePlaceholder: date-placeholder'
];
export const DEFAULT_OUTPUTS_O_TABLE_CELL_EDITOR_TIME = [
    ...DEFAULT_OUTPUTS_O_TABLE_CELL_EDITOR
];
export class OTableCellEditorTimeComponent extends OBaseTableCellEditor {
    constructor(injector, adapter) {
        super(injector);
        this.injector = injector;
        this.adapter = adapter;
        this.oStartView = 'month';
        this.oDateFormat = 'L';
        this.oHourFormat = Codes.TWENTY_FOUR_HOUR_FORMAT;
        this.onKeyboardInputDone = false;
        this.enabledCommitOnTabPress = false;
        this.activeKeys = {};
        this.momentSrv = this.injector.get(MomentService);
    }
    onDocumentKeydown(event) {
        this.handleKeydown(event);
    }
    initialize() {
        super.initialize();
        this.createInternalFormControl();
        if (!this._oDateLocale) {
            this.oDateLocale = this.momentSrv.getLocale();
        }
        if (this.oMinDate) {
            const date = new Date(this.oMinDate);
            const momentD = moment(date);
            if (momentD.isValid()) {
                this.minDateString = momentD.format(this.oDateFormat);
            }
        }
        if (this.oMaxDate) {
            const date = new Date(this.oMaxDate);
            const momentD = moment(date);
            if (momentD.isValid()) {
                this.maxDateString = momentD.format(this.oDateFormat);
            }
        }
    }
    createInternalFormControl() {
        if (!this.formControlDate) {
            const validators = this.resolveValidators();
            const cfg = {
                value: undefined,
                disabled: !this.enabled
            };
            this.formControlDate = new FormControl(cfg, validators);
            this.formGroup.addControl('dateInput', this.formControlDate);
        }
        if (!this.formControlHour) {
            const validators = this.resolveValidators();
            const cfg = {
                value: undefined,
                disabled: !this.enabled
            };
            this.formControlHour = new FormControl(cfg, validators);
            this.formGroup.addControl('hourInput', this.formControlHour);
        }
    }
    ngAfterViewChecked() {
        this.modifyPickerMethods();
    }
    setTime(event) {
        this.picker.updateTime(this.formControlHour.value);
    }
    onDateChange(event) {
        const isValid = event.value && event.value.isValid && event.value.isValid();
        const val = isValid ? event.value.valueOf() : moment().startOf('day');
        this.formControlDate.setValue(val, {
            emitModelToViewChange: false,
            emitEvent: false
        });
        this.updateComponentValue();
    }
    updateValeOnInputChange(blurEvent) {
        if (this.onKeyboardInputDone) {
            let value = blurEvent.currentTarget.value;
            value = this.parseHour(value);
            this.formControlHour.setValue(value);
        }
        this.onKeyboardInputDone = false;
    }
    parseHour(value) {
        const strArray = value.split(':');
        let hour = strArray[0];
        if (Codes.TWELVE_FOUR_HOUR_FORMAT === this.oHourFormat) {
            if (hour) {
                hour = parseInt(hour, 10);
                const period = hour <= 12 ? ' AM' : ' PM';
                if (hour > 12) {
                    hour = hour - 12;
                }
                strArray[0] = hour;
                value = strArray.join(':') + period;
            }
        }
        return value;
    }
    onHourChange(event) {
        let value;
        if (event instanceof Event) {
            this.updateValeOnInputChange(event);
        }
        else {
            value = this.convertToFormatString(event);
            this.formControlHour.setValue(value, {
                emitEvent: false,
                emitModelToViewChange: false
            });
        }
        this.updateComponentValue();
    }
    setTimestampValue(value, options) {
        let parsedValue;
        const momentV = Util.isDefined(value) ? moment(value) : value;
        if (momentV && momentV.isValid()) {
            parsedValue = momentV.utcOffset(0).format(this.formatString);
        }
        this.formControlHour.setValue(parsedValue, options);
    }
    convertToFormatString(value) {
        if (value === '00:00' || !Util.isDefined(value)) {
            return value;
        }
        const formatStr = this.oHourFormat === Codes.TWENTY_FOUR_HOUR_FORMAT ? 'HH:mm' : 'hh:mm a';
        let result;
        if (typeof value === 'number') {
            result = moment(value).format(formatStr);
        }
        else {
            result = value ? moment(value, 'h:mm A').format(formatStr) : value;
        }
        return result;
    }
    openDatepicker(d) {
        this.datepicker = d;
        d.open();
    }
    getPlaceholderHour() {
        let placeholder = '';
        if (this.oHourPlaceholder) {
            placeholder = this.translateService.get(this.oHourPlaceholder);
        }
        else {
            placeholder = super.getPlaceholder();
        }
        return placeholder;
    }
    getPlaceholderDate() {
        let placeholder = '';
        if (this.oDatePlaceholder) {
            placeholder = this.translateService.get(this.oDatePlaceholder);
        }
        else {
            placeholder = super.getPlaceholder();
        }
        return placeholder;
    }
    open(e) {
        if (this.picker) {
            this.picker.open();
        }
    }
    handleKeydown(e) {
        this.activeKeys[(e.key || e.keyCode)] = true;
    }
    handleKeyup(e) {
        this.activeKeys[(e.key || e.keyCode)] = false;
        const oColumn = this.table.getOColumn(this.tableColumn.attr);
        if (!oColumn) {
            return;
        }
        if (this.checkKey(e, "Tab", 9) && (this.activeKeys[16] || this.activeKeys["Shift"] || !this.enabledCommitOnTabPress)) {
            return;
        }
        if (!oColumn.editing && this.datepicker && this.datepicker.opened) {
            this.datepicker.close();
        }
        else {
            super.handleKeyup(e);
        }
    }
    updateComponentValue() {
        let timeValue;
        const values = this.formGroup.getRawValue();
        const mDate = (values['dateInput'] ? moment(values['dateInput']) : moment()).startOf('day');
        const mHour = moment(values['hourInput'], this.formatString);
        timeValue = mDate.clone()
            .set('hour', mHour.get('hour'))
            .set('minute', mHour.get('minutes'))
            .valueOf();
        if (this.formControl) {
            this.formControl.setValue(timeValue);
            this.formControl.markAsDirty();
        }
    }
    modifyPickerMethods() {
        if (this.picker && this.picker.inputElement) {
            this.picker.inputElement.addEventListener('change', () => {
                this.onKeyboardInputDone = true;
            });
        }
    }
    hasErrorDate(error) {
        return this.formControlDate && this.formControlDate.touched && this.hasErrorExclusive(error);
    }
    hasErrorExclusive(error) {
        let hasError = false;
        const errorsOrder = ['matDatepickerMax', 'matDatepickerMin', 'matDatepickerFilter', 'matDatepickerParse', 'required'];
        const errors = this.formControlDate.errors;
        if (Util.isDefined(errors)) {
            if (Object.keys(errors).length === 1) {
                return errors.hasOwnProperty(error);
            }
            else {
                for (let i = 0, len = errorsOrder.length; i < len; i++) {
                    hasError = errors.hasOwnProperty(errorsOrder[i]);
                    if (hasError) {
                        hasError = (errorsOrder[i] === error);
                        break;
                    }
                }
            }
        }
        return hasError;
    }
    hasErrorHour(error) {
        return this.formControlHour && this.formControlHour.touched;
    }
    getCellDataDate() {
        const value = super.getCellData();
        if (Util.isDefined(value)) {
            const m = moment(value);
            let result = value;
            if (Util.isDefined(m)) {
                result = m.toDate();
            }
            return result;
        }
        return value;
    }
    getCellDataHour() {
        const value = super.getCellData();
        if (Util.isDefined(value)) {
            const m = moment(value);
            let result = value;
            if (Util.isDefined(m)) {
                result = m.format(Codes.formatString(this.oHourFormat));
            }
            return result;
        }
        return value;
    }
    startEdition(data) {
        super.startEdition(data);
        const cellDataDate = this.getCellDataDate();
        this.formControlDate.setValue(cellDataDate);
        const cellDataHour = this.getCellDataHour();
        this.formControlHour.setValue(cellDataHour);
        this.formGroup.markAsTouched();
    }
    get formatString() {
        return Codes.formatString(this.oHourFormat);
    }
    get minDateString() {
        return this._minDateString;
    }
    set minDateString(val) {
        this._minDateString = val;
    }
    get maxDateString() {
        return this._maxDateString;
    }
    set maxDateString(val) {
        this._maxDateString = val;
    }
    set oDateLocale(value) {
        this._oDateLocale = value;
        if (Util.isDefined(this._oDateLocale)) {
            this.adapter.setLocale(value);
        }
    }
    get minDate() {
        return new Date(this.oMinDate);
    }
    get maxDate() {
        return new Date(this.oMaxDate);
    }
    onDatepickerClosed() {
        this.dateInput.nativeElement.focus();
    }
    onTimepickerClosed() {
        this.hourInput.nativeElement.focus();
    }
    commitEdition() {
        if (!this.formGroup.invalid) {
            super.commitEdition();
        }
    }
    onKeyDown(e) {
        if (!Codes.isHourInputAllowed(e)) {
            e.preventDefault();
        }
    }
}
OTableCellEditorTimeComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-cell-editor-time',
                template: "<ng-template #templateref let-cellvalue=\"cellvalue\" let-rowvalue=\"rowvalue\">\n  <div [formGroup]=\"formGroup\" class=\"o-table-cell-editor-time o-table-cell-editor\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\"\n    fxLayoutGap=\"8px\">\n    <mat-form-field floatLabel=\"never\">\n\n      <input #dateInput matInput [placeholder]=\"getPlaceholderDate()\" [formControl]=\"formControlDate\"\n        [required]=\"orequired\" [matDatepicker]=\"d\" (dateChange)=\"onDateChange($event)\" [min]=\"minDate\" [max]=\"maxDate\"\n        (focus)=\"enabledCommitOnTabPress = false\">\n\n      <mat-datepicker #d [startView]=\"oStartView\" [startAt]=\"oDateStartAt\" [touchUi]=\"oDateTouchUi\"\n        (closed)=\"onDatepickerClosed()\">\n      </mat-datepicker>\n\n      <span class=\"icon-btn\" (click)=\"openDatepicker(d)\" matSuffix>\n        <mat-icon svgIcon=\"ontimize:today\"></mat-icon>\n      </span>\n\n      <mat-error *oMatError=\"hasErrorDate('required') || hasErrorHour('required')\">\n        {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n      </mat-error>\n      <mat-error *oMatError=\"hasErrorDate('matDatepickerParse')\">\n        {{ 'FORM_VALIDATION.DATE_PARSE' | oTranslate }} {{ oHourFormat }}\n      </mat-error>\n      <mat-error *oMatError=\"hasErrorDate('matDatepickerFilter')\">\n        {{ 'FORM_VALIDATION.DATE_FILTER' | oTranslate }}\n      </mat-error>\n      <mat-error *oMatError=\"hasErrorDate('matDatepickerMin')\">\n        {{ 'FORM_VALIDATION.DATE_MIN' | oTranslate }} {{ minDateString }}\n      </mat-error>\n      <mat-error *oMatError=\"hasErrorDate('matDatepickerMax')\">\n        {{ 'FORM_VALIDATION.DATE_MAX' | oTranslate }} {{ maxDateString }}\n      </mat-error>\n      <mat-error *ngFor=\"let oError of getActiveOErrors()\">\n        {{ getErrorText(oError) }}\n      </mat-error>\n\n    </mat-form-field>\n\n    <span class=\"separator\">&ndash;</span>\n\n    <mat-form-field floatLabel=\"never\">\n\n      <input #hourInput matInput [ngxTimepicker]=\"picker\" [placeholder]=\"getPlaceholderHour()\"\n        [formControl]=\"formControlHour\" [required]=\"orequired\" (change)=\"onHourChange($event)\" [min]=\"oHourMin\"\n        [max]=\"oHourMax\" (keydown)=\"onKeyDown($event)\" [disableClick]=\"true\" [format]=\"oHourFormat\"\n        (blur)=\"enabledCommitOnTabPress = true\">\n\n      <button type=\"button\" matSuffix mat-icon-button (click)=\"open($event)\">\n        <mat-icon ngxMaterialTimepickerToggleIcon svgIcon=\"ontimize:clock\"></mat-icon>\n      </button>\n\n      <mat-error *oMatError=\"hasErrorHour('invalidFormatHour')\">\n        {{ 'FORM_VALIDATION.HOUR_FORMAT' | oTranslate }} {{ formatString }}\n      </mat-error>\n      <mat-error *ngFor=\"let oError of getActiveOErrors()\">\n        {{ oError.text | oTranslate }}\n      </mat-error>\n\n    </mat-form-field>\n\n    <ngx-material-timepicker #picker (timeSet)=\"onHourChange($event)\" [confirmBtnTmpl]=\"confirmBtn\"\n      [cancelBtnTmpl]=\"cancelBtn\" (closed)=\"onTimepickerClosed()\"></ngx-material-timepicker>\n    <ng-template #confirmBtn>\n      <button mat-stroked-button type=\"button\"><span>{{'OK' | oTranslate}}</span></button>\n    </ng-template>\n    <ng-template #cancelBtn>\n      <button mat-stroked-button type=\"button\"><span>{{'CANCEL' | oTranslate}}</span></button>\n    </ng-template>\n  </div>\n\n</ng-template>",
                inputs: DEFAULT_INPUTS_O_TABLE_CELL_EDITOR_TIME,
                outputs: DEFAULT_OUTPUTS_O_TABLE_CELL_EDITOR_TIME,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [
                    { provide: DateAdapter, useClass: MomentDateAdapter, deps: [MAT_DATE_LOCALE] }
                ],
                styles: [".separator{cursor:default}.mat-form-field:not(.custom-width) .mat-form-field-infix{width:84px}button.mat-stroked-button{margin:0 6px}"]
            }] }
];
OTableCellEditorTimeComponent.ctorParameters = () => [
    { type: Injector },
    { type: DateAdapter }
];
OTableCellEditorTimeComponent.propDecorators = {
    templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }],
    dateInput: [{ type: ViewChild, args: ['dateInput', { static: false },] }],
    hourInput: [{ type: ViewChild, args: ['hourInput', { static: false },] }],
    picker: [{ type: ViewChild, args: ['picker', { static: false },] }],
    datepickerInput: [{ type: ViewChild, args: [MatDatepickerInput, { static: false },] }],
    onDocumentKeydown: [{ type: HostListener, args: ['document:keydown', ['$event'],] }]
};
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTableCellEditorTimeComponent.prototype, "oDateTouchUi", void 0);
