// ontimize-web-ngx/esm2015/lib/components/form/guards/o-form-can-deactivate.guard.js
import { Injectable } from '@angular/core';
import { combineLatest } from 'rxjs';
import { BooleanConverter } from '../../../decorators/input-converter';
import { Codes } from '../../../util/codes';
import { Util } from '../../../util/util';
import { OFormConfirmExitService } from '../navigation/o-form-confirm-exit.service';
export class CanDeactivateFormGuard {
    constructor(oFormConfirmExitService) {
        this.oFormConfirmExitService = oFormConfirmExitService;
        this.oForms = {};
    }
    canDeactivate(component, curr, state, future) {
        if (BooleanConverter(future.root.queryParams[Codes.IGNORE_CAN_DEACTIVATE])) {
            return true;
        }
        if (Object.keys(this.oForms).length) {
            return new Promise((resolve) => {
                const arr = Object.keys(this.oForms).map((key) => Util.wrapIntoObservable(this.oForms[key].canDeactivate()));
                combineLatest(arr).subscribe(res => {
                    resolve(res.every(value => value));
                });
            });
        }
        return true;
    }
    addForm(form) {
        this.oForms[form.oattr] = form;
    }
    removeForm(form) {
        delete this.oForms[form.oattr];
    }
}
CanDeactivateFormGuard.CLASSNAME = 'CanDeactivateFormGuard';
CanDeactivateFormGuard.decorators = [
    { type: Injectable }
];
CanDeactivateFormGuard.ctorParameters = () => [
    { type: OFormConfirmExitService }
];
