// ontimize-web-ngx/esm2015/lib/shared/components/filter/store-filter/o-store-filter-dialog.component.js
import { ChangeDetectionStrategy, Component, Inject, Injector } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
import { OTableBaseDialogClass } from '../../../../components/table/extensions/dialog/o-table-base-dialog.class';
export class OStoreFilterDialogComponent extends OTableBaseDialogClass {
    constructor(dialogRef, injector, data) {
        super(injector);
        this.dialogRef = dialogRef;
        this.injector = injector;
        this.filterNames = [];
        this.formGroup = new FormGroup({
            name: new FormControl('', [
                Validators.required,
                this.filterNameValidator.bind(this)
            ]),
            description: new FormControl('')
        });
        this.setFormControl(this.formGroup.get('name'));
        this.loadFilterNames(data);
    }
    loadFilterNames(filterNames) {
        this.filterNames = filterNames;
    }
    getFilterAttributes() {
        return this.formGroup.value;
    }
    filterNameValidator(control) {
        const ctrlValue = control.value;
        if (this.filterNames.indexOf(ctrlValue) !== -1) {
            return { filterNameAlreadyExists: true };
        }
        return {};
    }
}
OStoreFilterDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-store-filter-dialog',
                template: "<span mat-dialog-title>{{ 'TABLE.BUTTONS.SAVE_FILTER' | oTranslate }}</span>\n\n<mat-dialog-content>\n  <div mat-subheader>{{ 'TABLE.DIALOG.SAVE_FILTER' | oTranslate }}</div>\n  <form #form [formGroup]=\"formGroup\" fxLayout=\"column\">\n    <mat-form-field>\n      <input matInput [matTooltip]=\"tooltipText\" [matTooltipClass]=\"tooltipClass\" placeholder=\"{{ 'TABLE.DIALOG.FILTER_NAME' | oTranslate }}\"\n        formControlName=\"name\" required>\n      <mat-error *oMatError=\"hasError(formGroup.controls['name'], 'filterNameAlreadyExists')\">\n        {{ 'TABLE.DIALOG.FILTER_NAME_ALREADY_EXISTS' | oTranslate }}\n      </mat-error>\n      <mat-error *oMatError=\"hasError(formGroup.controls['name'], 'required')\">\n        {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n      </mat-error>\n    </mat-form-field>\n    <mat-form-field class=\"example-full-width\">\n      <textarea matInput placeholder=\"{{ 'TABLE.DIALOG.FILTER_DESCRIPTION' | oTranslate }}\" formControlName=\"description\" rows=\"4\"\n        cols=\"50\"></textarea>\n    </mat-form-field>\n  </form>\n</mat-dialog-content>\n\n<mat-dialog-actions align=\"end\">\n  <button type=\"button\" mat-stroked-button [mat-dialog-close]=\"false\" class=\"o-button-default cancel\">{{ 'CANCEL' | oTranslate }}</button>\n  <button type=\"button\" mat-stroked-button [mat-dialog-close]=\"true\" [disabled]=\"!formGroup.valid\" class=\"o-button-primary\">\n    {{'SAVE' | oTranslate }}\n  </button>\n</mat-dialog-actions>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
OStoreFilterDialogComponent.ctorParameters = () => [
    { type: MatDialogRef },
    { type: Injector },
    { type: Array, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
];
