// ontimize-web-ngx/esm2015/lib/components/table/extensions/header/table-columns-grouping/columns/o-table-columns-grouping-column.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { InputConverter } from '../../../../../../decorators/input-converter';
import { Util } from '../../../../../../util';
export const DEFAULT_INPUTS_O_TABLE_COLUMN_GROUPING_COLUMN = [
    'attr',
    'title',
    'aggregateName: aggregate-name',
    'aggregate',
    'aggregateFunction: aggregate-function',
    'expandGroupsSameLevel: expand-groups-same-level',
    'changeAggregateSameLevel: change-aggregate-same-level',
];
export class OTableColumnsGroupingColumnComponent {
    constructor() {
        this.expandGroupsSameLevel = true;
        this.changeAggregateSameLevel = true;
    }
    set aggregate(value) {
        this._aggregate = value;
    }
    get aggregate() {
        if (Util.isDefined(this.aggregateFunction) && Util.isDefined(this.aggregateName)) {
            return this.aggregateName;
        }
        return this._aggregate;
    }
    getAggregateConfiguration() {
        return {
            attr: this.attr,
            title: this.title,
            aggregateName: this.aggregateName,
            aggregate: this.aggregate,
            aggregateFunction: this.aggregateFunction
        };
    }
}
OTableColumnsGroupingColumnComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-table-columns-grouping-column',
                template: ' ',
                changeDetection: ChangeDetectionStrategy.OnPush,
                inputs: DEFAULT_INPUTS_O_TABLE_COLUMN_GROUPING_COLUMN
            }] }
];
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTableColumnsGroupingColumnComponent.prototype, "expandGroupsSameLevel", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OTableColumnsGroupingColumnComponent.prototype, "changeAggregateSameLevel", void 0);
