// ontimize-web-ngx-report/esm5/lib/services/o-report.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/o-report.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector } from "@angular/core";
import { Observable, OErrorDialogManager, OntimizeEEService, Util } from 'ontimize-web-ngx';
import { HttpEventType, HttpHeaders, HttpRequest } from "@angular/common/http";
import { share } from 'rxjs/operators';
import * as i0 from "@angular/core";
var OReportService = /** @class */ (function (_super) {
    tslib_1.__extends(OReportService, _super);
    function OReportService(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _super.prototype.configureService.call(_this, _this.getDefaultServiceConfiguration());
        _this.oErrorDialogManager = injector.get(OErrorDialogManager);
        return _this;
    }
    /**
     * @param {?} reportparams
     * @return {?}
     */
    OReportService.prototype.createReport = /**
     * @param {?} reportparams
     * @return {?}
     */
    function (reportparams) {
        /** @type {?} */
        var body = JSON.stringify(reportparams);
        /** @type {?} */
        var url = this.urlBase + '/dynamicjasper/report';
        return this.doRequest({
            method: 'POST',
            url: url,
            body: body
        });
    };
    /**
     * @param {?=} preferencesparams
     * @return {?}
     */
    OReportService.prototype.saveAsPreferences = /**
     * @param {?=} preferencesparams
     * @return {?}
     */
    function (preferencesparams) {
        /** @type {?} */
        var body = JSON.stringify(preferencesparams);
        /** @type {?} */
        var url = this.urlBase + '/preferences/save';
        return this.doRequest({
            method: 'POST',
            url: url,
            body: body
        });
    };
    /**
     * @param {?} id
     * @param {?=} preferencesparams
     * @return {?}
     */
    OReportService.prototype.savePreferences = /**
     * @param {?} id
     * @param {?=} preferencesparams
     * @return {?}
     */
    function (id, preferencesparams) {
        /** @type {?} */
        var body = JSON.stringify(preferencesparams);
        /** @type {?} */
        var url = this.urlBase + '/preferences/update/' + id;
        return this.doRequest({
            method: 'PUT',
            url: url,
            body: body
        });
    };
    /**
     * @param {?=} entity
     * @param {?=} service
     * @return {?}
     */
    OReportService.prototype.getPreferences = /**
     * @param {?=} entity
     * @param {?=} service
     * @return {?}
     */
    function (entity, service) {
        /** @type {?} */
        var url = this.urlBase + '/preferences/preferences?entity=' + entity + '&service=' + service + "&type=REPORT";
        return this.doRequest({
            method: 'GET',
            url: url
        });
    };
    /**
     * @param {?=} functionparams
     * @return {?}
     */
    OReportService.prototype.getFunctions = /**
     * @param {?=} functionparams
     * @return {?}
     */
    function (functionparams) {
        /** @type {?} */
        var body = JSON.stringify(functionparams);
        /** @type {?} */
        var url = this.urlBase + '/dynamicjasper/functionsName';
        return this.doRequest({
            method: 'POST',
            url: url,
            body: body
        });
    };
    /**
     * @param {?=} id
     * @return {?}
     */
    OReportService.prototype.deletePreferences = /**
     * @param {?=} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var url = this.urlBase + '/preferences/remove/' + id;
        return this.doRequest({
            method: 'DELETE',
            url: url
        });
    };
    /** overridden method to add error callback for all requests */
    /**
     * overridden method to add error callback for all requests
     * @param {?} param
     * @return {?}
     */
    OReportService.prototype.doRequest = /**
     * overridden method to add error callback for all requests
     * @param {?} param
     * @return {?}
     */
    function (param) {
        return _super.prototype.doRequest.call(this, {
            method: param.method,
            url: param.url,
            body: param.body,
            errorCallBack: this.errorCallBack
        });
    };
    /**
     * @param {?} httpErrorResponse
     * @return {?}
     */
    OReportService.prototype.errorCallBack = /**
     * @param {?} httpErrorResponse
     * @return {?}
     */
    function (httpErrorResponse) {
        /** @type {?} */
        var error = httpErrorResponse.error;
        if (Util.isObject(error)) {
            if (error['code'] === 1 && Util.isDefined(error['message'])) {
                this.showNotificationError(error['message']);
                return;
            }
        }
        this.showNotificationError('MESSAGES.ERROR_QUERY');
    };
    /**
     * @param {?} error
     * @return {?}
     */
    OReportService.prototype.showNotificationError = /**
     * @param {?} error
     * @return {?}
     */
    function (error) {
        this.oErrorDialogManager.openErrorDialog(error);
    };
    /**
     * @param {?} files
     * @param {?} entity
     * @param {?=} data
     * @return {?}
     */
    OReportService.prototype.upload = /**
     * @param {?} files
     * @param {?} entity
     * @param {?=} data
     * @return {?}
     */
    function (files, entity, data) {
        var _this = this;
        /** @type {?} */
        var dataObservable = new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        function (observer) {
            /** @type {?} */
            var url = _this.urlBase + "/reportstore/" + entity;
            /** @type {?} */
            var toUpload = new FormData();
            files.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                item.prepareToUpload();
                item.isUploading = true;
                toUpload.append('name', item.name);
                toUpload.append('file', item.file);
            }));
            if (data) {
                toUpload.append('data', JSON.stringify(data));
            }
            /** @type {?} */
            var request = new HttpRequest('POST', url, toUpload, {
                headers: _this.buildHeadersReport(),
                reportProgress: true
            });
            _this.httpClient.request(request).subscribe((/**
             * @param {?} resp
             * @return {?}
             */
            function (resp) {
                if (HttpEventType.UploadProgress === resp.type) {
                    // Upload progress event received
                    /** @type {?} */
                    var progressData = {
                        loaded: resp.loaded,
                        total: resp.total
                    };
                    observer.next(progressData);
                }
                else if (HttpEventType.Response === resp.type) {
                    // Full response received
                    if (resp.body) {
                        _this.bodyCode(resp, observer);
                    }
                    else {
                        observer.next(resp.body);
                    }
                }
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                console.error(error);
                if (error.status === 401) {
                    _this.authService.logout();
                }
                else {
                    observer.error(error);
                }
            }), (/**
             * @return {?}
             */
            function () { return observer.complete(); }));
        }));
        return dataObservable.pipe(share());
    };
    /**
     * @protected
     * @param {?} resp
     * @param {?} observer
     * @return {?}
     */
    OReportService.prototype.bodyCode = /**
     * @protected
     * @param {?} resp
     * @param {?} observer
     * @return {?}
     */
    function (resp, observer) {
        if (resp.body['code'] === 3) {
            this.authService.logout();
        }
        else if (resp.body['code'] === 1) {
            observer.error(resp.body['message']);
        }
        else if (resp.body['code'] === 0) {
            // RESPONSE
            observer.next(resp.body);
        }
        else {
            // Unknow state -> error
            observer.error('Service unavailable');
        }
    };
    /**
     * @param {?=} _kv
     * @param {?=} _av
     * @param {?=} entity
     * @param {?=} _sqltypes
     * @param {?=} offset
     * @param {?=} _pagesize
     * @param {?=} _orderby
     * @return {?}
     */
    OReportService.prototype.advancedQuery = /**
     * @param {?=} _kv
     * @param {?=} _av
     * @param {?=} entity
     * @param {?=} _sqltypes
     * @param {?=} offset
     * @param {?=} _pagesize
     * @param {?=} _orderby
     * @return {?}
     */
    function (_kv, _av, entity, _sqltypes, offset, _pagesize, _orderby) {
        offset = (Util.isDefined(offset)) ? offset : this.offset;
        // Calculate page
        /** @type {?} */
        var page = 0;
        if (Util.isDefined(offset)) {
            page = Math.trunc(offset / 10) + 1;
        }
        /** @type {?} */
        var url = this.urlBase + '/reportstore/' + entity + '/?format=json' + '&page=' + page;
        return this.doRequest({
            method: 'GET',
            url: url
        });
    };
    /**
     * @param {?=} kv
     * @param {?=} _av
     * @param {?=} entity
     * @param {?=} _sqltypes
     * @return {?}
     */
    OReportService.prototype.query = /**
     * @param {?=} kv
     * @param {?=} _av
     * @param {?=} entity
     * @param {?=} _sqltypes
     * @return {?}
     */
    function (kv, _av, entity, _sqltypes) {
        /** @type {?} */
        var identifier = kv['UUID'];
        /** @type {?} */
        var url = '';
        if (Object.keys(kv).length === 0) {
            url = this.urlBase + "/reportstore/" + entity;
        }
        else {
            url = this.urlBase + "/reportstore/" + entity + "/" + identifier;
        }
        return this.doRequest({
            method: 'GET',
            url: url
        });
    };
    /**
     * @param {?=} av
     * @param {?=} entity
     * @param {?=} _sqltypes
     * @param {?=} filter
     * @return {?}
     */
    OReportService.prototype.fillReport = /**
     * @param {?=} av
     * @param {?=} entity
     * @param {?=} _sqltypes
     * @param {?=} filter
     * @return {?}
     */
    function (av, entity, _sqltypes, filter) {
        /** @type {?} */
        var identifier = av[0];
        /** @type {?} */
        var params = '';
        for (var i = 1; i < av.length; i++)
            params = params + av[i].toString() + ',';
        /** @type {?} */
        var body = JSON.stringify({
            params: params,
            filter: filter
        });
        /** @type {?} */
        var url = this.urlBase + "/reportstore/" + entity + "/" + identifier;
        return this.doRequest({
            method: 'POST',
            url: url,
            body: body
        });
    };
    /**
     * @param {?=} kv
     * @param {?=} _entity
     * @param {?=} _sqltypes
     * @return {?}
     */
    OReportService.prototype.delete = /**
     * @param {?=} kv
     * @param {?=} _entity
     * @param {?=} _sqltypes
     * @return {?}
     */
    function (kv, _entity, _sqltypes) {
        /** @type {?} */
        var identifier = kv.valueOf()[Object.keys(kv)[0]];
        /** @type {?} */
        var url = this.urlBase + "/reportstore/removeReport/" + identifier;
        return this.doRequest({
            method: 'DELETE',
            url: url
        });
    };
    /**
     * @param {?=} kv
     * @param {?=} av
     * @param {?=} _entity
     * @param {?=} _sqltypes
     * @return {?}
     */
    OReportService.prototype.update = /**
     * @param {?=} kv
     * @param {?=} av
     * @param {?=} _entity
     * @param {?=} _sqltypes
     * @return {?}
     */
    function (kv, av, _entity, _sqltypes) {
        /** @type {?} */
        var identifier = kv.valueOf()[Object.keys(kv)[0]];
        /** @type {?} */
        var url = this.urlBase + "/reportstore/updateReport/" + identifier;
        return this.doRequest({
            method: 'PUT',
            url: url,
            body: av
        });
    };
    /**
     * @protected
     * @return {?}
     */
    OReportService.prototype.buildHeadersReport = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = new HttpHeaders({ 'Access-Control-Allow-Origin': '*' });
        /** @type {?} */
        var sessionId = this.authService.getSessionInfo().id;
        if (Util.isDefined(sessionId)) {
            headers = headers.append('Authorization', 'Bearer ' + sessionId);
        }
        return headers;
    };
    OReportService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    OReportService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    /** @nocollapse */ OReportService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OReportService_Factory() { return new OReportService(i0.ɵɵinject(i0.INJECTOR)); }, token: OReportService, providedIn: "root" });
    return OReportService;
}(OntimizeEEService));
export { OReportService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    OReportService.prototype.oErrorDialogManager;
    /**
     * @type {?}
     * @protected
     */
    OReportService.prototype.injector;
}
