// ontimize-web-ngx-report/esm5/lib/services/o-report-store.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/o-report-store.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector } from "@angular/core";
import { MatDialog } from "@angular/material";
import { DialogService } from "ontimize-web-ngx";
import { OReportViewerComponent } from "../components/report/o-report-viewer/o-report-viewer.component";
import { Utils } from "../util/utils";
import { OReportService } from "./o-report.service";
var OReportStoreService = /** @class */ (function () {
    function OReportStoreService(injector) {
        this.injector = injector;
        this.dialogService = this.injector.get(DialogService);
        this.dialog = this.injector.get(MatDialog);
        this.reportService = this.injector.get(OReportService);
    }
    /**
     * @param {?} reportId
     * @param {?} parametersValues
     * @param {?} filter
     * @return {?}
     */
    OReportStoreService.prototype.openFillReport = /**
     * @param {?} reportId
     * @param {?} parametersValues
     * @param {?} filter
     * @return {?}
     */
    function (reportId, parametersValues, filter) {
        var _this = this;
        this.reportService.configureService(this.reportService.getDefaultServiceConfiguration());
        this.reportService.configureAdapter();
        /** @type {?} */
        var kv = { 'UUID': reportId };
        this.reportService.query(kv, null, 'getReport', {}).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            var e_1, _a;
            if (res && res.data.length && res.code === 0) {
                /** @type {?} */
                var parameters = res.data[0].PARAMETERS;
                /** @type {?} */
                var name_1 = res.data[0].NAME;
                /** @type {?} */
                var av = [];
                if (parameters.length > 0) {
                    av = [reportId];
                    /** @type {?} */
                    var values = Object.values(parametersValues);
                    try {
                        for (var values_1 = tslib_1.__values(values), values_1_1 = values_1.next(); !values_1_1.done; values_1_1 = values_1.next()) {
                            var value = values_1_1.value;
                            av.push(value);
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (values_1_1 && !values_1_1.done && (_a = values_1.return)) _a.call(values_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
                else {
                    av = [reportId];
                }
                /** @type {?} */
                var data = { 'params': av, 'filter': filter, 'name': name_1 };
                Utils.openModalVisor(_this.dialog, OReportViewerComponent, data);
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            if (_this.dialogService) {
                _this.dialogService.error('ERROR', 'SERVER_ERROR_MESSAGE');
            }
            console.log(err);
        }));
    };
    OReportStoreService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    OReportStoreService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OReportStoreService;
}());
export { OReportStoreService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    OReportStoreService.prototype.reportService;
    /**
     * @type {?}
     * @protected
     */
    OReportStoreService.prototype.dialogService;
    /**
     * @type {?}
     * @protected
     */
    OReportStoreService.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    OReportStoreService.prototype.injector;
}
