// ontimize-web-ngx/esm5/lib/components/user-info/user-info-configuration-item/o-user-info-configuration-item.directive.js
import * as tslib_1 from "tslib";
import { Directive, Injector } from "@angular/core";
import { Router } from "@angular/router";
import { InputConverter } from "../../../decorators";
import { DialogService } from "../../../services";
export var DEFAULT_INPUTS_O_USER_INFO_MENU_ITEM = [
    'name',
    'route',
    'icon',
    'action',
    'confirm',
    'confirmText: confirm-text',
    'tooltip',
    'class'
];
var OUserInfoConfigurationItemDirective = (function () {
    function OUserInfoConfigurationItemDirective(injector) {
        this.injector = injector;
        this.toolTip = false;
        this.confirm = false;
        this.router = this.injector.get(Router);
        this.dialogService = this.injector.get(DialogService);
    }
    OUserInfoConfigurationItemDirective.prototype.triggerOnClick = function (e) {
        if (this.route) {
            this.navigate();
        }
        else {
            this.executeItemAction();
        }
    };
    OUserInfoConfigurationItemDirective.prototype.navigate = function () {
        if (this.router.url !== this.route) {
            this.router.navigate([this.route]);
        }
    };
    OUserInfoConfigurationItemDirective.prototype.executeItemAction = function () {
        var _this = this;
        if (this.confirm) {
            this.dialogService.confirm('CONFIRM', this.confirmText || 'MESSAGES.CONFIRM_ACTION').then(function (result) { return result ? _this.action() : null; });
        }
        else {
            this.action();
        }
    };
    OUserInfoConfigurationItemDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'o-user-info-configuration-item',
                    inputs: DEFAULT_INPUTS_O_USER_INFO_MENU_ITEM
                },] }
    ];
    OUserInfoConfigurationItemDirective.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OUserInfoConfigurationItemDirective.prototype, "toolTip", void 0);
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OUserInfoConfigurationItemDirective.prototype, "confirm", void 0);
    return OUserInfoConfigurationItemDirective;
}());
export { OUserInfoConfigurationItemDirective };
