// ontimize-web-ngx/esm5/lib/components/container/row-collapsible/o-row-collapsible.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, Inject, Injector, Optional, ViewEncapsulation } from '@angular/core';
import { MAT_FORM_FIELD_DEFAULT_OPTIONS } from '@angular/material';
import { DEFAULT_INPUTS_O_CONTAINER_COLLAPSIBLE, DEFAULT_OUTPUTS_O_CONTAINER_COLLAPSIBLE, OContainerCollapsibleComponent } from '../o-container-collapsible-component.class';
export var DEFAULT_INPUTS_O_ROW_COLLAPSIBLE = tslib_1.__spread(DEFAULT_INPUTS_O_CONTAINER_COLLAPSIBLE);
export var DEFAULT_OUTPUTS_O_ROW_COLLAPSIBLE = tslib_1.__spread(DEFAULT_OUTPUTS_O_CONTAINER_COLLAPSIBLE);
var ORowCollapsibleComponent = (function (_super) {
    tslib_1.__extends(ORowCollapsibleComponent, _super);
    function ORowCollapsibleComponent(elRef, injector, matFormDefaultOption) {
        var _this = _super.call(this, elRef, injector, matFormDefaultOption) || this;
        _this.elRef = elRef;
        _this.injector = injector;
        _this.matFormDefaultOption = matFormDefaultOption;
        return _this;
    }
    ORowCollapsibleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-row-collapsible',
                    template: "<mat-expansion-panel #expPanel [expanded]=\"expanded\" class=\"o-container\" (opened)=\"updateInnerHeight(null)\" (closed)=\"updateInnerHeight(0)\">\n  <mat-expansion-panel-header #containerTitle [expandedHeight]=\"expandedHeight\" [collapsedHeight]=\"collapsedHeight\">\n    <mat-panel-title fxLayout=\"row\" fxLayoutAlign=\"start center\" class=\"o-container-title\">\n      <mat-icon *ngIf=\"icon\">{{ icon }}</mat-icon>\n      <span *ngIf=\"title\">{{ title | oTranslate }}</span>\n    </mat-panel-title>\n    <mat-panel-description fxLayout=\"row\" fxLayoutAlign=\"start center\">\n      <span>{{ description | oTranslate }}</span>\n    </mat-panel-description>\n  </mat-expansion-panel-header>\n  <div #containerContent id=\"innerRow\" fxLayout=\"row\" fxLayoutAlign=\"{{ layoutAlign }}\" fxLayoutGap=\"{{ layoutGap }}\"\n    class=\"o-container-scroll o-scroll\" fxFlex=\"grow\">\n    <ng-content></ng-content>\n  </div>\n</mat-expansion-panel>\n<div #oContainerOutline *ngIf=\"isAppearanceOutline()\" [class.o-container-outline-expanded]=\"expPanel.expanded\" class=\"o-container-outline\">\n  <div class=\"o-container-outline-start\"></div>\n  <div class=\"o-container-outline-gap-title\"></div>\n  <div class=\"o-container-outline-gap-empty1\"></div>\n  <div class=\"o-container-outline-gap-description\"></div>\n  <div class=\"o-container-outline-gap-empty2\"></div>\n  <div class=\"o-container-outline-gap-icon\"></div>\n  <div class=\"o-container-outline-end\"></div>\n</div>\n",
                    inputs: DEFAULT_INPUTS_O_ROW_COLLAPSIBLE,
                    outputs: DEFAULT_OUTPUTS_O_ROW_COLLAPSIBLE,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-row-collapsible]': 'true',
                        '[class.o-appearance-outline]': 'isAppearanceOutline()',
                        '[class.o-appearance-outline-title]': 'hasTitleInAppearanceOutline()'
                    }
                }] }
    ];
    ORowCollapsibleComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_FORM_FIELD_DEFAULT_OPTIONS,] }] }
    ]; };
    return ORowCollapsibleComponent;
}(OContainerCollapsibleComponent));
export { ORowCollapsibleComponent };
