// ontimize-web-ngx/esm5/lib/components/contextmenu/o-context-menu.directive.js
import { Directive, HostListener, Injector } from '@angular/core';
import { Util } from '../../util/util';
export var DEFAULT_CONTEXT_MENU_DIRECTIVE_INPUTS = [
    'oContextMenu',
    'oContextMenuData'
];
var OContextMenuDirective = (function () {
    function OContextMenuDirective(injector) {
        this.injector = injector;
    }
    OContextMenuDirective.prototype.onRightClick = function (event) {
        event.preventDefault();
        event.stopPropagation();
        if (Util.isDefined(this.oContextMenu)) {
            this.oContextMenu.oContextMenuService.showContextMenu.next({
                contextMenu: this.oContextMenu,
                event: event,
                data: this.oContextMenuData
            });
        }
    };
    OContextMenuDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[oContextMenu]',
                    inputs: DEFAULT_CONTEXT_MENU_DIRECTIVE_INPUTS
                },] }
    ];
    OContextMenuDirective.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OContextMenuDirective.propDecorators = {
        onRightClick: [{ type: HostListener, args: ['contextmenu', ['$event'],] }]
    };
    return OContextMenuDirective;
}());
export { OContextMenuDirective };
