// ontimize-web-ngx/esm5/lib/components/contextmenu/o-context-menu-overlay.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var OContextMenuOverlayService = (function () {
    function OContextMenuOverlayService() {
        this.overlays = [];
    }
    OContextMenuOverlayService.prototype.addOverlay = function (value) {
        this.overlays.push(value);
    };
    OContextMenuOverlayService.prototype.destroyOverlays = function () {
        if (this.overlays) {
            this.overlays.forEach(function (overlay) {
                overlay.detach();
                overlay.dispose();
            });
        }
        this.overlays = [];
    };
    OContextMenuOverlayService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    OContextMenuOverlayService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OContextMenuOverlayService_Factory() { return new OContextMenuOverlayService(); }, token: OContextMenuOverlayService, providedIn: "root" });
    return OContextMenuOverlayService;
}());
export { OContextMenuOverlayService };
