// ontimize-web-ngx/esm5/lib/components/contextmenu/o-context-menu.component.js
import { Component, ContentChildren, EventEmitter, Injector, QueryList } from '@angular/core';
import { Subscription } from 'rxjs';
import { OComponentMenuBaseItem } from './o-content-menu-base-item.class';
import { OContextMenuService } from './o-context-menu.service';
export var DEFAULT_OUTPUTS_O_CONTEXT_MENU = [
    'onShow',
    'onClose'
];
var OContextMenuComponent = (function () {
    function OContextMenuComponent(injector) {
        this.injector = injector;
        this.onShow = new EventEmitter();
        this.onClose = new EventEmitter();
        this.subscription = new Subscription();
        this.oContextMenuService = this.injector.get(OContextMenuService);
    }
    OContextMenuComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription.add(this.oContextMenuService.showContextMenu.subscribe(function (param) { return _this.showContextMenu(param); }));
        this.subscription.add(this.oContextMenuService.closeContextMenu.subscribe(function (param) { return _this.onClose.emit(); }));
    };
    OContextMenuComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    OContextMenuComponent.prototype.showContextMenu = function (params) {
        this.origin = params.event.target;
        this.onShow.emit(params);
        if (params.contextMenu !== this) {
            return;
        }
        params.menuItems = this.oContextMenuItems;
        params.externalMenuItems = this.externalContextMenuItems;
        if (params.menuItems.length > 0) {
            this.oContextMenuService.openContextMenu(params);
        }
    };
    OContextMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-context-menu',
                    template: ' ',
                    outputs: DEFAULT_OUTPUTS_O_CONTEXT_MENU,
                    providers: [OContextMenuService]
                }] }
    ];
    OContextMenuComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OContextMenuComponent.propDecorators = {
        oContextMenuItems: [{ type: ContentChildren, args: [OComponentMenuBaseItem,] }]
    };
    return OContextMenuComponent;
}());
export { OContextMenuComponent };
