// ontimize-web-ngx/esm5/lib/components/contextmenu/o-content-menu-base-item.class.js
import { Util } from '../../util/util';
export var DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS = [
    'attr',
    'ovisible: visible',
    'icon',
    'data',
    'label',
    'oenabled: enabled',
    'svgIcon: svg-icon'
];
var OComponentMenuBaseItem = (function () {
    function OComponentMenuBaseItem() {
        this.ovisible = true;
        this.type = OComponentMenuBaseItem.TYPE_GROUP_MENU;
        this.enabled = true;
    }
    Object.defineProperty(OComponentMenuBaseItem.prototype, "disabled", {
        get: function () {
            if (this.enabled instanceof Function) {
                return !this.enabled(this.data);
            }
            return !this.enabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OComponentMenuBaseItem.prototype, "isVisible", {
        get: function () {
            if (this.ovisible instanceof Function) {
                return this.ovisible(this.data);
            }
            return this.ovisible;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OComponentMenuBaseItem.prototype, "oenabled", {
        set: function (value) {
            if (value instanceof Function) {
                this.enabled = value;
            }
            else {
                this.enabled = this.parseInput(value, true);
            }
        },
        enumerable: true,
        configurable: true
    });
    OComponentMenuBaseItem.prototype.parseInput = function (value, defaultValue) {
        if (value instanceof Function || typeof value === 'boolean') {
            return value;
        }
        return Util.parseBoolean(value, defaultValue);
    };
    OComponentMenuBaseItem.TYPE_ITEM_MENU = 'item_menu';
    OComponentMenuBaseItem.TYPE_GROUP_MENU = 'item_group';
    OComponentMenuBaseItem.TYPE_SEPARATOR_MENU = 'item_separator';
    return OComponentMenuBaseItem;
}());
export { OComponentMenuBaseItem };
