// ontimize-web-ngx/esm5/lib/components/input/listpicker/listpicker-renderer/integer/o-list-picker-renderer-integer.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Injector, TemplateRef, ViewChild } from '@angular/core';
import { InputConverter } from '../../../../../decorators/input-converter';
import { OIntegerPipe } from '../../../../../pipes/o-integer.pipe';
import { DEFAULT_INPUTS_O_LISTPICKER_RENDERER, OListPickerCustomRenderer } from '../o-list-picker-renderer.class';
export var DEFAULT_INPUTS_O_LISTPICKER_RENDERER_INTEGER = tslib_1.__spread(DEFAULT_INPUTS_O_LISTPICKER_RENDERER, [
    'grouping',
    'thousandSeparator: thousand-separator'
]);
var OListPickerRendererIntegerComponent = (function (_super) {
    tslib_1.__extends(OListPickerRendererIntegerComponent, _super);
    function OListPickerRendererIntegerComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.grouping = true;
        _this.thousandSeparator = ',';
        _this.setComponentPipe();
        return _this;
    }
    OListPickerRendererIntegerComponent.prototype.setComponentPipe = function () {
        this.componentPipe = new OIntegerPipe(this.injector);
    };
    OListPickerRendererIntegerComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        this.pipeArguments = {
            grouping: this.grouping,
            thousandSeparator: this.thousandSeparator
        };
    };
    OListPickerRendererIntegerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-list-picker-renderer-integer',
                    template: "<ng-template #templateref let-value=\"value\">\n  {{getListPickerValue(value)}}\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_LISTPICKER_RENDERER_INTEGER
                }] }
    ];
    OListPickerRendererIntegerComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OListPickerRendererIntegerComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef, static: true },] }]
    };
    tslib_1.__decorate([
        InputConverter(),
        tslib_1.__metadata("design:type", Boolean)
    ], OListPickerRendererIntegerComponent.prototype, "grouping", void 0);
    return OListPickerRendererIntegerComponent;
}(OListPickerCustomRenderer));
export { OListPickerRendererIntegerComponent };
