// ontimize-web-ngx/esm2015/lib/components/input/listpicker/o-list-picker.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, forwardRef, Inject, Injector, Optional, ViewChild } from '@angular/core';
import { FormControl } from '@angular/forms';
import { MatDialog, MatInput } from '@angular/material';
import { InputConverter } from '../../../decorators/input-converter';
import { OntimizeServiceProvider } from '../../../services/factories';
import { Util } from '../../../util/util';
import { OFormComponent } from '../../form/o-form.component';
import { OValueChangeEvent } from '../../o-value-change-event.class';
import { DEFAULT_INPUTS_O_FORM_SERVICE_COMPONENT, DEFAULT_OUTPUTS_O_FORM_SERVICE_COMPONENT, OFormServiceComponent } from '../o-form-service-component.class';
import { OListPickerDialogComponent } from './o-list-picker-dialog.component';
export const DEFAULT_INPUTS_O_LIST_PICKER = [
    ...DEFAULT_INPUTS_O_FORM_SERVICE_COMPONENT,
    'filter',
    'dialogWidth : dialog-width',
    'dialogHeight : dialog-height',
    'queryRows: query-rows',
    'textInputEnabled: text-input-enabled',
    'dialogDisableClose: dialog-disable-close',
    'dialogClass: dialog-class'
];
export const DEFAULT_OUTPUTS_O_LIST_PICKER = [
    ...DEFAULT_OUTPUTS_O_FORM_SERVICE_COMPONENT,
    'onDialogAccept',
    'onDialogCancel'
];
export class OListPickerComponent extends OFormServiceComponent {
    constructor(form, elRef, injector) {
        super(form, elRef, injector);
        this.onDialogAccept = new EventEmitter();
        this.onDialogCancel = new EventEmitter();
        this.textInputEnabled = true;
        this.dialogDisableClose = false;
        this.filter = true;
        this.dialogHeight = '55%';
        this.blurDelay = 200;
        this.blurPrevent = false;
        this.matDialog = this.injector.get(MatDialog);
        this.stateCtrl = new FormControl();
        this.clearButton = true;
    }
    ngOnInit() {
        this.initialize();
        if (!Util.isDefined(this.stateCtrl.value)) {
            this.setStateCtrlValue();
        }
    }
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        if (typeof (changes.staticData) !== 'undefined') {
            this.cacheQueried = true;
            this.setDataArray(changes.staticData.currentValue);
        }
    }
    createFormControl(cfg, validators) {
        this._fControl = super.createFormControl(cfg, validators);
        this._fControl.fControlChildren = [this.stateCtrl];
        return this._fControl;
    }
    ensureOFormValue(value) {
        super.ensureOFormValue(value);
        this.syncDataIndex(false);
    }
    setEnabled(value) {
        super.setEnabled(value);
        if (this.stateCtrl && this.hasEnabledPermission() || this.hasVisiblePermission()) {
            value ? this.stateCtrl.enable() : this.stateCtrl.disable();
        }
    }
    ngAfterViewInit() {
        super.ngAfterViewInit();
        if (this.queryOnInit) {
            this.queryData();
        }
        else if (this.queryOnBind) {
            this.syncDataIndex();
        }
    }
    getDescriptionValue() {
        if (!Util.isDefined(this.descriptionColArray) || !Util.isDefined(this._currentIndex)) {
            return '';
        }
        if (Util.isDefined(this.renderer)) {
            return this.renderer.getListPickerValue(this.dataArray[this._currentIndex]);
        }
        else {
            return this.getOptionDescriptionValue(this.dataArray[this._currentIndex]);
        }
    }
    onClickClear(e) {
        e.preventDefault();
        e.stopPropagation();
        if (!this.isReadOnly && this.enabled) {
            clearTimeout(this.blurTimer);
            this.blurPrevent = true;
            this.setValue(undefined);
        }
    }
    onClickInput(e) {
        if (!this.textInputEnabled) {
            this.onClickListpicker(e);
        }
    }
    onClickListpicker(e) {
        e.preventDefault();
        e.stopPropagation();
        if (!this.isReadOnly && this.enabled) {
            clearTimeout(this.blurTimer);
            this.openDialog();
        }
    }
    onDialogClose(evt) {
        this.dialogRef = null;
        this.visibleInputValue = undefined;
        if (evt instanceof Object && typeof evt[this.valueColumn] !== 'undefined') {
            const self = this;
            window.setTimeout(() => {
                self.setValue(evt[self.valueColumn], { changeType: OValueChangeEvent.USER_CHANGE });
                if (self._fControl) {
                    self._fControl.markAsTouched();
                    self._fControl.markAsDirty();
                }
                self.onDialogAccept.emit();
            }, 0);
        }
        else {
            this.onDialogCancel.emit();
        }
    }
    innerOnBlur(evt) {
        if (!this.isReadOnly && this.enabled) {
            const self = this;
            this.blurTimer = setTimeout(() => {
                if (!self.blurPrevent) {
                    self._fControl.markAsTouched();
                    self.onBlur.emit(evt);
                    if (self.visibleInputValue !== undefined && self.visibleInputValue.length > 0) {
                        self.openDialog();
                    }
                    else if (self.visibleInputValue !== undefined) {
                        self.setValue(undefined);
                        self.visibleInputValue = undefined;
                    }
                    else {
                        self._fControl.markAsTouched();
                    }
                }
                self.blurPrevent = false;
            }, this.blurDelay);
        }
    }
    onVisibleInputChange(event) {
        this.visibleInputValue = event.target.value;
    }
    onKeydownEnter(val) {
        clearTimeout(this.blurTimer);
        this.blurPrevent = true;
        this.visibleInputValue = val;
        this.openDialog();
    }
    setFormValue(val, options, setDirty = false) {
        super.setFormValue(val, options, setDirty);
        this.setStateCtrlValue();
    }
    openDialog() {
        const cfg = {
            role: 'dialog',
            disableClose: this.dialogDisableClose,
            panelClass: ['cdk-overlay-list-picker', 'o-dialog-class', this.dialogClass],
            data: {
                data: this.getDialogDataArray(this.dataArray),
                filter: this.filter,
                searchVal: this.visibleInputValue,
                menuColumns: this.visibleColumns,
                visibleColumns: this.visibleColArray,
                queryRows: this.queryRows,
                renderer: this.renderer
            }
        };
        if (this.dialogWidth !== undefined) {
            cfg.width = this.dialogWidth;
        }
        if (this.dialogHeight !== undefined) {
            cfg.height = this.dialogHeight;
        }
        this.dialogRef = this.matDialog.open(OListPickerDialogComponent, cfg);
        this.dialogRef.afterClosed().subscribe(result => this.onDialogClose(result));
    }
    getDialogDataArray(dataArray) {
        const result = [];
        dataArray.forEach((item, itemIndex) => {
            const newItem = Object.assign({}, item);
            if (!this.renderer) {
                newItem._parsedVisibleColumnText = this.getOptionDescriptionValue(item);
            }
            newItem._parsedIndex = itemIndex;
            result.push(newItem);
        });
        return result;
    }
    registerRenderer(renderer) {
        this.renderer = renderer;
        this.renderer.initialize();
    }
    setStateCtrlValue() {
        let descriptionValue = this.getDescriptionValue();
        if (typeof descriptionValue === 'string' && descriptionValue.length === 0) {
            descriptionValue = null;
        }
        this.stateCtrl.setValue(descriptionValue);
    }
    selectValue() {
        const inputEl = document.getElementById('desc_' + this.oattr);
        if (inputEl) {
            inputEl.select();
        }
    }
}
OListPickerComponent.decorators = [
    { type: Component, args: [{
                selector: 'o-list-picker',
                template: "<div (click)=\"onClickInput($event)\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\" [formGroup]=\"getFormGroup()\"\n  [matTooltip]=\"tooltip\" [matTooltipClass]=\"tooltipClass\" [matTooltipPosition]=\"tooltipPosition\"\n  [matTooltipShowDelay]=\"tooltipShowDelay\" [matTooltipHideDelay]=\"tooltipHideDelay\"\n  class=\"custom-error o-list-picker-content\" [oContextMenu]=\"oContextMenu\">\n  <input #inputModel class=\"input-model\" type=\"text\" [id]=\"getAttribute()\" [formControlName]=\"getAttribute()\">\n\n  <mat-form-field [appearance]=\"appearance\" [floatLabel]=\"floatLabel\" [class.custom-width]=\"hasCustomWidth\"\n    [hideRequiredMarker]=\"hideRequiredMarker\" class=\"icon-field\" #innerInputContainer fxFlexFill>\n    <mat-label *ngIf=\"labelVisible\">{{ olabel | oTranslate }}</mat-label>\n\n    <input #visibleInput matInput type=\"text\" [attr.id]=\"'desc_' + getAttribute()\" [id]=\"'desc_' + getAttribute()\"\n      [formControl]=\"stateCtrl\" [placeholder]=\"placeHolder\" [value]=\"getDescriptionValue()\"\n      (input)=\"onVisibleInputChange($event)\" (focus)=\"innerOnFocus($event)\" (blur)=\"innerOnBlur($event)\"\n      [readonly]=\"isReadOnly || !textInputEnabled\" [required]=\"isRequired\"\n      (keydown.enter)=\"onKeydownEnter(visibleInput.value)\" (change)=\"onChangeEvent($event)\">\n\n    <button type=\"button\" [disabled]=\"isReadOnly || !enabled\" [class.read-only]=\"isReadOnly\" matSuffix mat-icon-button\n      (click)=\"onClickListpicker($event)\">\n      <mat-icon svgIcon=\"ontimize:search\"></mat-icon>\n    </button>\n\n    <button type=\"button\" [disabled]=\"isReadOnly || !enabled\" [class.read-only]=\"isReadOnly\" matSuffix mat-icon-button\n      (click)=\"onClickClearValue($event)\" *ngIf=\"showClearButton\">\n      <mat-icon svgIcon=\"ontimize:clear\"></mat-icon>\n    </button>\n\n    <mat-error *oMatError=\"hasError('required')\">\n      {{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\n    </mat-error>\n    <mat-error *ngFor=\"let oError of getActiveOErrors()\">\n      {{ oError.text | oTranslate }}\n    </mat-error>\n  </mat-form-field>\n\n  <o-context-menu *ngIf=\"!this.isReadOnly && this.enabled\">\n    <o-context-menu-item attr=\"refresh\" label=\"FORM_SERVICE_COMPONENT.REFRESH\" svg-icon=\"ontimize:autorenew\" (execute)=\"refresh()\">\n    </o-context-menu-item>\n  </o-context-menu>\n</div>",
                providers: [
                    OntimizeServiceProvider,
                    { provide: OFormServiceComponent, useExisting: forwardRef(() => OListPickerComponent) }
                ],
                inputs: DEFAULT_INPUTS_O_LIST_PICKER,
                outputs: DEFAULT_OUTPUTS_O_LIST_PICKER,
                styles: [".o-list-picker-content .input-model{display:none!important}.o-list-picker-content .mat-button.mat-disabled{cursor:default}"]
            }] }
];
OListPickerComponent.ctorParameters = () => [
    { type: OFormComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(() => OFormComponent),] }] },
    { type: ElementRef },
    { type: Injector }
];
OListPickerComponent.propDecorators = {
    inputModel: [{ type: ViewChild, args: ['inputModel', { static: false },] }],
    visibleInput: [{ type: ViewChild, args: ['visibleInput', { static: false },] }]
};
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OListPickerComponent.prototype, "textInputEnabled", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OListPickerComponent.prototype, "dialogDisableClose", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Boolean)
], OListPickerComponent.prototype, "filter", void 0);
tslib_1.__decorate([
    InputConverter(),
    tslib_1.__metadata("design:type", Number)
], OListPickerComponent.prototype, "queryRows", void 0);
