// ontimize-web-ngx/esm5/lib/components/contextmenu/context-menu-item/o-context-menu-item.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, forwardRef } from '@angular/core';
import { DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS, OComponentMenuBaseItem } from '../o-content-menu-base-item.class';
export var DEFAULT_CONTEXT_MENU_ITEM_OUTPUTS = [
    'execute'
];
export var DEFAULT_INPUTS_O_CONTEXT_MENU_ITEM = tslib_1.__spread(DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS);
var OContextMenuItemComponent = (function (_super) {
    tslib_1.__extends(OContextMenuItemComponent, _super);
    function OContextMenuItemComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.execute = new EventEmitter();
        _this.type = OComponentMenuBaseItem.TYPE_ITEM_MENU;
        return _this;
    }
    OContextMenuItemComponent.prototype.onClick = function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.triggerExecute(this.data, event);
    };
    OContextMenuItemComponent.prototype.triggerExecute = function (data, $event) {
        if (!this.enabled) {
            return;
        }
        this.execute.emit({ event: $event, data: data });
    };
    OContextMenuItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-context-menu-item',
                    template: ' ',
                    inputs: DEFAULT_INPUTS_O_CONTEXT_MENU_ITEM,
                    outputs: DEFAULT_CONTEXT_MENU_ITEM_OUTPUTS,
                    providers: [{ provide: OComponentMenuBaseItem, useExisting: forwardRef(function () { return OContextMenuItemComponent; }) }]
                }] }
    ];
    return OContextMenuItemComponent;
}(OComponentMenuBaseItem));
export { OContextMenuItemComponent };
