// ontimize-web-ngx/esm5/lib/components/app-sidenav/image/o-app-sidenav-image.component.js
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Injector, ViewEncapsulation } from '@angular/core';
import { merge, Subscription } from 'rxjs';
import { OAppSidenavComponent } from '../o-app-sidenav.component';
export var DEFAULT_INPUTS_O_APP_SIDENAV_IMAGE = [
    'openedSrc: opened-src',
    'closedSrc: closed-src'
];
export var DEFAULT_OUTPUTS_O_APP_SIDENAV_IMAGE = [];
var OAppSidenavImageComponent = (function () {
    function OAppSidenavImageComponent(injector, cd) {
        this.injector = injector;
        this.cd = cd;
        this.subscription = new Subscription();
        this.sidenav = this.injector.get(OAppSidenavComponent);
    }
    OAppSidenavImageComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (this.sidenav) {
            this.subscription.add(merge(this.sidenav.onSidenavClosedStart, this.sidenav.onSidenavOpenedStart).subscribe(function () { return _this.updateImage(); }));
        }
        this.updateImage();
    };
    OAppSidenavImageComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    OAppSidenavImageComponent.prototype.ngOnChanges = function (changes) {
        if (changes.openedSrc || changes.closedSrc) {
            this.updateImage();
        }
    };
    OAppSidenavImageComponent.prototype.updateImage = function () {
        if (this.sidenav && this.sidenav.sidenav && this.sidenav.sidenav.opened) {
            this.setOpenedImg();
        }
        else {
            this.setClosedImg();
        }
        this.cd.detectChanges();
    };
    Object.defineProperty(OAppSidenavImageComponent.prototype, "src", {
        get: function () {
            return this._src;
        },
        set: function (val) {
            this._src = val;
        },
        enumerable: true,
        configurable: true
    });
    OAppSidenavImageComponent.prototype.setOpenedImg = function () {
        this.src = this.openedSrc;
    };
    OAppSidenavImageComponent.prototype.setClosedImg = function () {
        this.src = this.closedSrc;
    };
    Object.defineProperty(OAppSidenavImageComponent.prototype, "showImage", {
        get: function () {
            return (this._src !== undefined && this._src.length > 0);
        },
        enumerable: true,
        configurable: true
    });
    OAppSidenavImageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'o-app-sidenav-image',
                    inputs: DEFAULT_INPUTS_O_APP_SIDENAV_IMAGE,
                    outputs: DEFAULT_OUTPUTS_O_APP_SIDENAV_IMAGE,
                    template: "<div class=\"o-app-sidenav-image-container\" *ngIf=\"showImage\">\n  <img class=\"o-app-sidenav-image\" [src]=\"src\" alt=\"application sidenav image\" />\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-app-sidenav-image]': 'true'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".sidenav-toggle-container+.o-app-sidenav-image .o-app-sidenav-image-container{padding-top:0;text-align:center}.mat-drawer-opened .o-app-sidenav-image .o-app-sidenav-image-container{text-align:center}.o-app-sidenav-image .o-app-sidenav-image-container{max-height:75px;margin:24px 0}.o-app-sidenav-image .o-app-sidenav-image-container .o-app-sidenav-image{max-width:100%;max-height:100%}"]
                }] }
    ];
    OAppSidenavImageComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    return OAppSidenavImageComponent;
}());
export { OAppSidenavImageComponent };
