// ontimize-web-ngx-report/esm5/lib/types/report-preferences.type.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/types/report-preferences.type.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var DefaultOReportPreferences = /** @class */ (function () {
    function DefaultOReportPreferences() {
        this.title = '';
        this.subtitle = '';
        this.vertical = true;
        this.columns = [];
        this.groups = [];
        this.functions = [];
        this.style = {
            grid: false, rowNumber: false, columnName: true, backgroundOnOddRows: false, hideGroupDetails: false, groupNewPage: false, firstGroupNewPage: false
        };
        this.orderBy = [];
        this.entity = '';
        this.service = '';
    }
    return DefaultOReportPreferences;
}());
export { DefaultOReportPreferences };
if (false) {
    /** @type {?} */
    DefaultOReportPreferences.prototype.title;
    /** @type {?} */
    DefaultOReportPreferences.prototype.subtitle;
    /** @type {?} */
    DefaultOReportPreferences.prototype.vertical;
    /** @type {?} */
    DefaultOReportPreferences.prototype.columns;
    /** @type {?} */
    DefaultOReportPreferences.prototype.groups;
    /** @type {?} */
    DefaultOReportPreferences.prototype.functions;
    /** @type {?} */
    DefaultOReportPreferences.prototype.style;
    /** @type {?} */
    DefaultOReportPreferences.prototype.orderBy;
    /** @type {?} */
    DefaultOReportPreferences.prototype.entity;
    /** @type {?} */
    DefaultOReportPreferences.prototype.service;
}
